---
title: "Figure5-Figure6-Revisions"
author: "Anthony"
date: "7/18/2021"
output: pdf_document
---

# {.tabset}  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup {.tabset}

### Helper Functions
- helper functions come from the original figure 6 methods
- may need to import figure 2 methods as well.
```{r helperFunction, eval=TRUE}

 zScorePatientExpression<-function(full.dn4=NULL,markers=NULL,ROIID.column=NULL){
  X3<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(sd))%>%data.frame
  X2<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(mean))%>%data.frame
   ###
   z<-full.dn4
  for(cl in unique(full.dn4$ROIID)){
   for(mark in colnames(X2)[-1]){
  z[which(z$ROIID==cl),which(colnames(z)==mark)]<- ( z[which(z$ROIID==cl),which(colnames(z)==mark)]-X2[which(X2$ROIID==cl),which(colnames(X2)==mark)])/(X3[which(X3$ROIID==cl),which(colnames(X3)==mark)])
   }
 }
 return(z)
 }


 makeJoyPlot<-function(z,marker=NULL){
   ltlist<-list()
     for(roi in unique(z$ROIID)){
      lt = data.frame(density(z[which(z$ROIID==roi),marker])[c("x","y")])
    ltlist[[(length(ltlist)+1)]]<-lt
     }
    marker1= rowAnnotation(Z= anno_joyplot(ltlist, width = unit(4, "cm"), 
    gp = gpar(fill = color_clusters),
	 transparency = 0.35)) 
  marker1@anno_list[[1]]@name <-gsub("Cell_","",marker)
  names(marker1@anno_list)<-gsub("Cell_","",marker)
    return(marker1)
   }


 ## this function performs Chi-square on the direct standardized counts per case
## direct standardization, standardizes the number of cells per case.
## direct standardization takes the given sub-community relative proportion, and then multiplies by the average number of cells across all cases (19,919). Each sub-community cluster has a different relative abundance.  By multiplying the cluster proportion by the standardized case cell average, it returns the expected number of cells in a sub-cluster assuming that all cases had the same number of cells.  It balances the expectation if all cases captured the same number of cells. so it scales the relative case proportions under the expectation that all cases had the same average cells. 

## The parameter 'conditionOnPrimary' can compare enrichment across COO conditioned on a primary component ###(TRUE), or across all components (FALSE).  default is to leave FALSE, more conservative.
 
chiHelperMeanScale<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn="unsup.subcommunity",conditionOnPrimary=FALSE){
 ### the features are average to scale the data.
	##only works for HANS COO
	##phenotype column is the column of phenoytpe levels.
    message(pheno)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
    ta$COO<-full.dn4$COO[match(rownames(ta),full.dn4$case)]
    ta2<-ta[,c(pheno,"COO")]
    x11<-round(mean(ta2[which(ta2$COO=="GCB"),1]))+1
    x12<-round(mean(ta2[which(ta2$COO!="GCB"),1]))+1
	##conditioned on the same phenotype family.
  if(conditionOnPrimary==TRUE){
    ta3<-ta[,grepl(primary,colnames(ta))]
  }else{
      ta3<-ta[,which(colnames(ta)!=pheno)]
  }
    ta3<-ta3[,which(colnames(ta3)!=pheno)]
    ta3$COO<-ta2$COO[match(rownames(ta3),rownames(ta2))]
   ta3<-ta3%>%group_by(COO)%>%summarise_all(funs(mean))%>%data.frame()
    x21<-round(mean(as.numeric(ta3[which(ta3$COO=="GCB"),-1])))+1
    x22<-round(mean(as.numeric(ta3[which(ta3$COO=="NGCB"),-1])))+1
    mat<-matrix(c(x11,x21,x12,x22),nrow=2,ncol=2)
	rownames(mat)<-c(paste0(pheno,"+"),paste0(pheno,"-"))
	colnames(mat)<-c("GCB","NGCB")
 ### need OR with confidence intervals
     require(epitools)
     dd<-oddsratio(mat,rev="both")
	return(dd)
}

geeGLMHANS<-function(pheno=NULL,cluster.counts=NULL,COO="HANS.binary",conditionOnPrimary=FALSE){
 ### the features are percentages.
	##only works for HANS COO binary cluster feature must be made
	##phenotype column is the column of phenoytpe levels.
    message(pheno)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])

 library(gee)
 ### need OR with confidence intervals
    form<-formula(paste0(COO,"~",pheno,"+ipi"))
     dd<- geeglm(form, id = Case_Number, data = cluster.counts,family = binomial(link = "logit"), corstr = "exchangeable", scale.fix = TRUE)
     dd<-coef(summary(dd))
	return(dd)
}







chiHelperMeanScaleLabeled<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn="unsup.subcommunity"){
 ### the features are average to scale the data.
	##only works for HANS COO
	##phenotype column is the column of phenoytpe levels.
    message(pheno)
    primary<-pheno
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
    ta$COO<-full.dn4$COO[match(rownames(ta),full.dn4$case)]
    ta2<-ta[,c(pheno,"COO")]
    x11<-round(mean(ta2[which(ta2$COO=="GCB"),1]))+1
    x12<-round(mean(ta2[which(ta2$COO!="GCB"),1]))+1
   
    ta3<-ta[,which(colnames(ta)!=pheno)]
    ta3$COO<-ta2$COO[match(rownames(ta3),rownames(ta2))]
   ta3<-ta3%>%group_by(COO)%>%summarise_all(funs(mean))%>%data.frame()
    x21<-round(mean(as.numeric(ta3[which(ta3$COO=="GCB"),-1])))+1
    x22<-round(mean(as.numeric(ta3[which(ta3$COO=="NGCB"),-1])))+1
    mat<-matrix(c(x11,x21,x12,x22),nrow=2,ncol=2)
	rownames(mat)<-c(paste0(pheno,"+"),paste0(pheno,"-"))
	colnames(mat)<-c("GCB","NGCB")
 ### need OR with confidence intervals
     require(epitools)
     dd<-oddsratio(mat,rev="both")
	return(dd)
  }

 ### logistic regression using direct standardization.
 ##logistic regression standard.
 logisticRegressionClinicalOdds<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn=NULL,mycFeature=NULL,groupA=1){
    message(pheno)
   #this function returns the OR (exp the coefficients and CI in the model.)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
	ta<-as.data.frame(scale(ta))
    ta$feature<-full.dn4[match(rownames(ta),full.dn4$case),mycFeature]
    ta$feature<-ifelse(ta$feature==groupA,1,0)
  colnames(ta)[which(colnames(ta)=="feature")]<-mycFeature
    myForm<-formula(paste0(mycFeature,"~",pheno))
	fit<-glm(myForm,ta,family='binomial')
	ci<-exp(confint(fit))
  dd<-exp(coef(glm(myForm,ta,family='binomial')))
   dd<-cbind(ci,dd)%>%data.frame
  colnames(dd)<-c("lower.2.5%","upper.97.5%",paste0("OddsRatio.",groupA))
  pv<-summary(fit)%>%coef%>%data.frame
  dd$p<-pv[,4]
	return(dd[pheno,])
  } 

  logisticRegressionClinicalLogOdds<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn=NULL,mycFeature=NULL,groupA=1){
    message(pheno)
   #this function returns the OR (exp the coefficients and CI in the model.)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
	ta<-as.data.frame(scale(ta))
    ta$feature<-full.dn4[match(rownames(ta),full.dn4$case),mycFeature]
    ta$feature<-ifelse(ta$feature==groupA,1,0)
  colnames(ta)[which(colnames(ta)=="feature")]<-mycFeature
    myForm<-formula(paste0(mycFeature,"~",pheno))
	fit<-glm(myForm,ta,family='binomial')
	ci<-(confint(fit))
  dd<-(coef(glm(myForm,ta,family='binomial')))
   dd<-cbind(ci,dd)%>%data.frame
  colnames(dd)<-c("lower.2.5%","upper.97.5%",paste0("LogOddsRatio.",groupA))
  pv<-summary(fit)%>%coef%>%data.frame
  dd$p<-pv[,4]
	return(dd[pheno,])
  } 
  
  ## REVISION 10-15 adding a chisquare test for EBER in general way.
  generalChiHelperMeanScale<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn="unsup.subcommunity",conditionOnPrimary=FALSE,responseColumn='eber'){
 ### the features are average to scale the data.
	##only works for HANS COO
	##phenotype column is the column of phenoytpe levels.
    ## the responseColumn is the response binary variable y~ phenotypeColumn_i, where y is binary
    message(pheno)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
    ta$COO<-full.dn4[match(rownames(ta),full.dn4$case),responseColumn]
    ta2<-ta[,c(pheno,"COO")]
    x11<-round(mean(ta2[which(ta2$COO==1),1]))+1
    x12<-round(mean(ta2[which(ta2$COO!=1),1]))+1
	##conditioned on the same phenotype family.
  if(conditionOnPrimary==TRUE){
    ta3<-ta[,grepl(primary,colnames(ta))]
  }else{
      ta3<-ta[,which(colnames(ta)!=pheno)]
  }
    ta3<-ta3[,which(colnames(ta3)!=pheno)]
    ta3$COO<-ta2$COO[match(rownames(ta3),rownames(ta2))]
   ta3<-ta3%>%group_by(COO)%>%summarise_all(funs(mean))%>%data.frame()
    x21<-round(mean(as.numeric(ta3[which(ta3$COO==1),-1])))+1
    x22<-round(mean(as.numeric(ta3[which(ta3$COO==0),-1])))+1
    mat<-matrix(c(x11,x21,x12,x22),nrow=2,ncol=2)
	rownames(mat)<-c(paste0(pheno,"+"),paste0(pheno,"-"))
	colnames(mat)<-c("1","0")
 ### need OR with confidence intervals
     require(epitools)
     dd<-oddsratio(mat,rev="both")
	return(dd)
  }

color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")

 getPhenoColors<-function(full.dn4){
 require(circlize)
  tsn<-full.dn4[,c("Cell_CD20",
	"Cell_CD3",
	"Cell_CD4",
	"Cell_CD31",
	"Cell_CD4",
	"Cell_CD68",
	"Cell_CD8",
	"Cell_FOXP3",
	"Cell_BCL2",
	"uniqueLabel","response","ROIID",
	"unsup.subcommunity")]
  ##balanced response representation.
   table(tsn$response)
  tsn$primary.phenotype<-tsn$unsup.subcommunity
 
   cd8s<-colorRampPalette(c("darkseagreen","green"),20)
  cd4s<-colorRampPalette(c("lightsteelblue1","blue"),20)
  macs<-colorRampPalette(c("magenta","maroon4"),20)
   tregs<-colorRampPalette(c("plum2","purple"),20)
   tumrs<-colorRampPalette(c("grey24","grey64"),20)
   endos<-colorRampPalette(c("rosybrown1","red"),20)

   phenoColors<-c(cd4s(14)[seq(1,14,2)],
	cd8s(14)[seq(1,14,2)],macs(12)[seq(1,12,2)],tregs(12)[seq(1,12,2)],tumrs(11),endos(6))

 names(phenoColors)<-levels(factor(tsn$primary.phenotype))
   ##manual color change code.
   phenoColors["Macrophage_2"]<-"darkblue"
   phenoColors["Macrophage_5"]<-"yellow"
   phenoColors["CD8_7"]<-"black"

   return(phenoColors)
 }



mutationClinicalSheetAssembly<-function(mutationsToGather=c("TP53","BCL2","BCL6","CD79B","SGK1","MYD88","MYC","NOTCH1","NOTCH2","EZH2","CREBBP"),full.dn4=NULL){
 ## the mutation of MYD88 is specifically controlled for L265P region.
 ## the column header for the non-syn SNP annotation for MYD88 is AA.Mutation.Cosmic_v70

   mutations<-read.csv("~/IMC-Ranalysis/DLBCL/mutational-genomic-table/TMA CGI genomic data.csv",header=T)
   colnames(mutations)<-gsub("USC076","USC76",colnames(mutations))
  #  clinic<-read.csv("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/patient.clinical.table.csv",header=T)
    # library(XLConnect) 
  # wb<-XLConnect::loadWorkbook("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/clinicalDataRevisedFinal_final.xlsx")
  # clinic <- readWorksheet(wb, 1);  
   clinic<-read.csv("~/IMC-Ranalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/clinicalDataRevisedFinal_finalSheet.csv",header=T)

  clinic$MB<-gsub(" ","",clinic$End..MB)
 mut.tag<-colnames(mutations)[9:67]
  mutt<-data.frame(mut.tag,id=sapply(strsplit(gsub("\\.","_",mut.tag),"_"),function(x) x[1])
) 

  mutt$id[which(mutt[,"id"]=="USC076")]<-"USC76"
  tag<- as.character(clinic[,"MB"])
  tag<-gsub(" ","",tag)
  tag<-data.frame(tag=tag,stringsAsFactors=FALSE)
  inters<-intersect(tag[,1],mutt$id)
   key<-data.frame(MB=tag[match(inters,tag[,1]),],stringsAsFactors=FALSE)
  key$id<-mutt[match(inters,mutt[,2]),1]
 ##fix a column header
 ## USC76 has a funcy header.
 ##
  casemuts<-mutations[,match(key$id,colnames(mutations))]
  casemuts$gene.name<-mutations[,"gene_name..Homo_sapiens_ensembl_v74_mRNA."]
  casemuts$chromosome<-as.character(mutations$Chromosome)
 casemuts$aminoAcid<-as.character(mutations$AA.Mutation.Cosmic_v70)
 casemuts$aminoAcid<-gsub( "\"","",mutations$AA.Mutation.Cosmic_v70)
  casemuts$gene.name<-as.character(casemuts$gene.name)
 ##for MYD88, the gene name will have the aminoAcid appended.
  casemuts$gene.name[which(casemuts$gene.name=="MYD88" & casemuts$aminoAcid=="L265P, L265P")]<-"MYD88.L265P"
  casemuts<-casemuts[which(casemuts$aminoAcid!="S219C, S219C, S219C, S219C" & casemuts$aminoAcid!="M232T, M232T" & casemuts$aminoAcid!="0.999, 0.999" & casemuts$aminoAcid!="S251N, S243N, S251N, S243N"),]
  casemuts[which(casemuts$gene.name=="MYD88" & casemuts$chromosome==""),"gene.name"]<-"MYD88.L265P"
 casemuts[which(casemuts$gene.name=="MYD88.L265P" & casemuts$chromosome==""),1:20]<-ifelse(casemuts[which(casemuts$gene.name=="MYD88.L265P" & casemuts$chromosome!=""),1:20]!=0,1,0)
  casem<- casemuts[which(casemuts$chromosome==""),]
 mutationsToGather[which(mutationsToGather=="MYD88")]<-"MYD88.L265P"
  dat<-casem[casem$gene.name%in%mutationsToGather,]
   ### 
   for(i in mutationsToGather){
  key[,i]<-0
 key[match(colnames(dat)[which(dat[dat$gene.name==i,]==1)],key$id),i]<-1
   }
  #write.csv(key,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/Mutational.CNV.categorical.table.csv")
write.csv(key,file="~/IMC-Ranalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/Mutational.CNV.categorical.table.csv")

  ## append to the clinical sheet
  clinic2<-clinic[,!colnames(clinic)%in%mutationsToGather]
 clinic2<-left_join(clinic2,key[,c("MB",mutationsToGather)],by="MB")
  for(i in mutationsToGather){
  clinic2[which(is.na(clinic2[,i])), i]<-0
  }
 ##add the factor to control for replicates
  replicate<-model.matrix(~0+clinic2$MB)
  replicate<-replicate[,which(colSums(replicate)>1)]  
 colnames(replicate)<-substring(colnames(replicate),11)
  clinic2<-data.frame(clinic2,replicate)
 clinic2$WES.avail<-FALSE
  clinic2$WES.avail[match(key$MB,clinic2$MB)]<-TRUE
 clinic2$Case_ID<-paste0("Case_",clinic2$X)
  return(clinic2)

}


 chiHelperStatus<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn=NULL,mycFeature="MYC.Rearrange",groupA=1){
 ## sums the phenotype totals per patient across a clinical feature
  ## we do not use averages, this is penalized for missing phenotypes.
  ## generalized for other clinical factors.
  ## Patient level data
###phenotype column is the pheno cluster column
 ### mycFetaure is patient clinical var
 ## groupA is the level of the mycFeature for positive.
	##use the means to scale the data.
    message(pheno)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/(rowSums(ta)+1)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
    ta$MYC<-full.dn4[match(rownames(ta),full.dn4$case),mycFeature]
    ta2<-ta[,c(pheno,"MYC")]
    x11<-round(sum(ta2[which(ta2$MYC==groupA),1]))+100
    x12<-round(sum(ta2[which(ta2$MYC!=groupA),1]))+100
    ta3<-ta[,grepl(primary,colnames(ta))]
    ta3<-ta3[,which(colnames(ta3)!=pheno)]
    ta3$MYC<-ta2$MYC[match(rownames(ta3),rownames(ta2))]
   ta3<-ta3%>%group_by(MYC)%>%summarise_all(funs(sum))%>%data.frame()
    x21<-round(sum(as.numeric(ta3[which(ta3$MYC==groupA),-1])))+100
    x22<-round(sum(as.numeric(ta3[which(ta3$MYC!=groupA),-1])))+100
    mat<-matrix(c(x11,x21,x12,x22),nrow=2,ncol=2)
	rownames(mat)<-c(paste0(pheno,"+"),paste0(pheno,"-"))
	colnames(mat)<-c(paste0(mycFeature,"+"),paste0(mycFeature,"-"))
 ### need OR with confidence intervals
     require(epitools)
     dd<-oddsratio(mat,rev="both")
	return(dd)
  }



 ### individual mutations at the class level
 ## ignores 'Tumor_11' cluster that cluster was not studied in phenotype analysis.
  patientMutationOddsRatio<-function(mutation=NULL){
   ### odds ratio of mutations
	print(mutation)
 ta<-as.data.frame.matrix(table(full.dn4[which(full.dn4$WES.available==TRUE & full.dn4$dimCD20.filtered=='NO'),mutation],full.dn4$COO[which(full.dn4$WES.available==TRUE& full.dn4$dimCD20.filtered=='NO')]))+400
  dd<-epitools::oddsratio(as.matrix(ta))
   return(dd$measure[2,])  
  }


 zScorePatientExpression<-function(full.dn4=NULL,markers=NULL,ROIID.column=NULL){
  X3<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(sd))%>%data.frame
  X2<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(mean))%>%data.frame
   ###
   z<-full.dn4
  for(cl in unique(full.dn4$ROIID)){
   for(mark in colnames(X2)[-1]){
  z[which(z$ROIID==cl),which(colnames(z)==mark)]<- ( z[which(z$ROIID==cl),which(colnames(z)==mark)]-X2[which(X2$ROIID==cl),which(colnames(X2)==mark)])/(X3[which(X3$ROIID==cl),which(colnames(X3)==mark)])
   }
 }
 return(z)
 }


 empiricalFitMutationCaseLevel<-function(full.dn4=NULL,design=NULL,toDrop=NULL,phenotypeCommunity=NULL){
 ## fit the model at the case level, and not ROI level
   patient<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeCommunity]))
  ### patient/rowSums(patient) is the rate of phenotype_j per patient
 ##  direct standardization multiples all the rates by a constant cohort average across all patients and all phenotypes.  this models equal number of phenotypes per patient.
   ##modeled as a proportion 
  y<-100*patient/rowSums(patient)
 # y<-mean(table(full.dn4$ROIID))*y
  #mean(rowMeans(patient))
	y2<-as.data.frame(t(y))
   y2$annotation<-sapply(strsplit(rownames(y2),"_"),function(x) x[1])
   pat.pheno<-y2
    pat.pheno<-pat.pheno[,which(colnames(pat.pheno)!="annotation")]
    design<-design[match(colnames(pat.pheno),rownames(design)),]  
   stopifnot(all( colnames(pat.pheno)==rownames(design)))
    require(limma)
  if(!is.null(toDrop)==TRUE){
  pat.pheno<-pat.pheno[!rownames(pat.pheno)%in%toDrop,]
  }
  fit <- lmFit(pat.pheno,design)
 #  Moderated t-statistic
  fit <- eBayes(fit)
 print( topTable(fit,coef=2))
  return(fit)
 }
 
  empiricalFitMutationROIIDLevel<-function(full.dn4=NULL,design=NULL,toDrop=NULL,phenotypeCommunity=NULL){
 ## fit the model at the case level, and not ROI level
   patient<-as.data.frame.matrix(table(full.dn4$ROIID,full.dn4[,phenotypeCommunity]))
  ### patient/rowSums(patient) is the rate of phenotype_j per patient
 ##  direct standardization multiples all the rates by a constant cohort average across all patients and all phenotypes.  this models equal number of phenotypes per patient.
   ##modeled as a proportion 
  y<-100*patient/rowSums(patient)
 # y<-mean(table(full.dn4$ROIID))*y
  #mean(rowMeans(patient))
	y2<-as.data.frame(t(y))
   y2$annotation<-sapply(strsplit(rownames(y2),"_"),function(x) x[1])
   pat.pheno<-y2
    pat.pheno<-pat.pheno[,which(colnames(pat.pheno)!="annotation")]
    design<-design[match(colnames(pat.pheno),rownames(design)),]  
   stopifnot(all( colnames(pat.pheno)==rownames(design)))
    require(limma)
  if(!is.null(toDrop)==TRUE){
  pat.pheno<-pat.pheno[!rownames(pat.pheno)%in%toDrop,]
  }
  fit <- lmFit(pat.pheno,design)
 #  Moderated t-statistic
  fit <- eBayes(fit)
 print( topTable(fit,coef=2))
  return(fit)
 }


getMutationColors<-function(){

 # mutCols<-c("thistle1",
#	"darkorchid2",
#	"purple3",
#	"khaki",
#	"white",
#	"dodgerblue",
#	"cyan2",
#	"cadetblue4",
#	"hotpink",
#	"lightblue1",
#	"seagreen2",
#	 "black")
  mutCols<-c(
	#BCL2+	
	"thistle1",
	#BCL2+BCL6+EZH2"
	"grey79",
	#BCL2+EZH2+
	"aquamarine4",
	#BCL6+CD79b+
	"purple3",
	#CD79b
	"khaki",
	#EZH2
	"black",
	#MYD88
	"hotpink",
	##TP53"
 	"dodgerblue",
	##TP53 CD79b
	"cyan2",
	#TP53+CD79B+MYD88
	"darkorchid2",
	#TP53+SGK+
	"cadetblue2",
	"lightblue1",
	"seagreen2",
	"none")


  names(mutCols)<-c("BCL2+",
	"BCL6+BCL2+EZH2+","BCL2+EZH2+","BCL6+CD79B+","CD79B+","EZH2+","MYD88.L265P+",
	"TP53+","TP53+CD79B+","TP53+CD79B+MYD88.L265P+",
	"TP53+SGK1+","NOTCH1+","NOTCH2+","none")
 return(mutCols)
}


local_diffcyt<-
function (d_input, experiment_info = NULL, marker_info = NULL, 
    design = NULL, formula = NULL, contrast, analysis_type = c("DA", 
        "DS"), method_DA = c("diffcyt-DA-edgeR", 
        "diffcyt-DA-voom", "diffcyt-DA-GLMM","diffcyt-DA-GLMM-Local"), method_DS = c("diffcyt-DS-limma", 
        "diffcyt-DS-LMM"), markers_to_test = NULL, clustering_to_use = NULL, 
    cols_to_include = NULL, subsampling = FALSE, n_sub = NULL, 
    seed_sub = NULL, transform = TRUE, cofactor = 5, cols_clustering = NULL, 
    xdim = 10, ydim = 10, meta_clustering = FALSE, meta_k = 40, 
    seed_clustering = NULL, min_cells = 3, min_samples = NULL, 
    normalize = FALSE, norm_factors = "TMM", trend_method = "none", 
    block_id = NULL, trend = TRUE, weights = TRUE, plot = FALSE, 
    path = ".", verbose = TRUE) 
{
    analysis_type <- match.arg(analysis_type)
    method_DA <- match.arg(method_DA)
    method_DS <- match.arg(method_DS)
    if (!is(d_input, "SingleCellExperiment")) {
        if (is.null(experiment_info) | is.null(marker_info)) {
            stop("'experiment_info' and 'marker_info' must be provided (unless using a SingleCellExperiment ", 
                "object from CATALYST as input)")
        }
        if (verbose) 
            message("preparing data...")
        d_se <- prepareData(d_input, experiment_info, marker_info, 
            cols_to_include, subsampling, n_sub, seed_sub)
        if (transform) {
            if (verbose) 
                message("transforming data...")
            d_se <- transformData(d_se, cofactor)
        }
        if (verbose) 
            message("generating clusters...")
        d_se <- generateClusters(d_se, cols_clustering, xdim, 
            ydim, meta_clustering, meta_k, seed_clustering)
    }
    else if (is(d_input, "SingleCellExperiment")) {
        if (verbose) 
            message("using SingleCellExperiment object from CATALYST as input")
        if (is.null(clustering_to_use)) {
            stopifnot("cluster_id" %in% colnames(colData(d_input)))
            if (verbose) 
                message("using cluster IDs stored in column named 'cluster_id' in 'colData' of ", 
                  "SingleCellExperiment object from CATALYST")
            clustering_name <- colnames(metadata(d_input)$cluster_codes)[1]
        }
        else if (!is.null(clustering_to_use)) {
            stopifnot(as.character(clustering_to_use) %in% colnames(metadata(d_input)$cluster_codes))
            stopifnot("cluster_id" %in% colnames(colData(d_input)))
            if (verbose) 
                message("using cluster IDs from clustering stored in column '", 
                  clustering_to_use, "' of 'cluster_codes' data frame in 'metadata' of SingleCellExperiment object from CATALYST")
            code_id <- colData(d_input)$cluster_id
            cluster_id <- metadata(d_input)$cluster_codes[, clustering_to_use][code_id]
            stopifnot(length(cluster_id) == nrow(colData(d_input)), 
                length(code_id) == nrow(colData(d_input)))
            colData(d_input)$code_id <- code_id
            colData(d_input)$cluster_id <- cluster_id
            clustering_name <- clustering_to_use
        }
        stopifnot("sample_id" %in% colnames(colData(d_input)))
        stopifnot("cluster_id" %in% colnames(colData(d_input)))
        stopifnot("cluster_codes" %in% names(metadata(d_input)))
        if (!("experiment_info" %in% names(metadata(d_input)))) {
            if (verbose) 
                message("generating 'experiment_info' from input object")
            m <- match(levels(droplevels(factor(d_input$sample_id))), 
                d_input$sample_id)
            experiment_info <- data.frame(colData(d_input)[m, 
                ], check.names = FALSE, row.names = NULL)
            metadata <- as.list(c(metadata(d_input), experiment_info))
        }
        else {
            experiment_info <- metadata(d_input)$experiment_info
            metadata <- metadata(d_input)
        }
        cs_by_s <- split(seq_len(ncol(d_input)), colData(d_input)$sample_id)
        cs <- unlist(cs_by_s[as.character(experiment_info$sample_id)])
        es <- t(assays(d_input)[["exprs"]])[cs, , drop = FALSE]
        d_se <- SummarizedExperiment(assays = list(exprs = es), 
            rowData = colData(d_input)[cs, ], colData = rowData(d_input), 
            metadata = metadata)
    }
    if (verbose) 
        message("calculating features...")
    d_counts <- calcCounts(d_se)
    if (analysis_type == "DS") {
        d_medians <- calcMedians(d_se)
        d_medians_by_cluster_marker <- calcMediansByClusterMarker(d_se)
        d_medians_by_sample_marker <- calcMediansBySampleMarker(d_se)
    }
    if (analysis_type == "DA" && method_DA == "diffcyt-DA-edgeR") {
        if (verbose) 
            message("calculating DA tests using method 'diffcyt-DA-edgeR'...")
        res <- testDA_edgeR(d_counts, design, contrast, trend_method, 
            min_cells, min_samples, normalize, norm_factors)
    }
    if (analysis_type == "DA" && method_DA == "diffcyt-DA-voom") {
        if (verbose) 
            message("calculating DA tests using method 'diffcyt-DA-voom'...")
        res <- testDA_voom(d_counts, design, contrast, block_id, 
            min_cells, min_samples, normalize, norm_factors, 
            plot, path)
    }
    if (analysis_type == "DA" && method_DA == "diffcyt-DA-GLMM") {
        if (verbose) 
            message("calculating DA tests using method 'diffcyt-DA-GLMM'...")
        res <- testDA_GLMM(d_counts, formula, contrast, min_cells, 
            min_samples, normalize, norm_factors)
    }
    
    if (analysis_type == "DA" && method_DA == "diffcyt-DA-GLMM-Local") {
        if (verbose) 
            message("calculating DA tests using method 'diffcyt-DA-GLMM'...")
        res <- testDA_GLMM_Local(d_counts, formula, contrast, min_cells, 
            min_samples, normalize, norm_factors)
    }
    
    if (analysis_type == "DS" && method_DS == "diffcyt-DS-limma") {
        if (verbose) 
            message("calculating DS tests using method 'diffcyt-DS-limma'...")
        res <- testDS_limma(d_counts, d_medians, design, contrast, 
            block_id, trend, weights, markers_to_test, min_cells, 
            min_samples, plot, path)
    }
    if (analysis_type == "DS" && method_DS == "diffcyt-DS-LMM") {
        if (verbose) 
            message("calculating DS tests using method 'diffcyt-DS-LMM'...")
        res <- testDS_LMM(d_counts, d_medians, formula, contrast, 
            weights, markers_to_test, min_cells, min_samples)
    }
    if (!is(d_input, "SingleCellExperiment")) {
        if (analysis_type == "DA") {
            return(list(res = res, d_se = d_se, d_counts = d_counts))
        }
        else if (analysis_type == "DS") {
            return(list(res = res, d_se = d_se, d_counts = d_counts, 
                d_medians = d_medians, d_medians_by_cluster_marker = d_medians_by_cluster_marker, 
                d_medians_by_sample_marker = d_medians_by_sample_marker))
        }
    }
    else if (is(d_input, "SingleCellExperiment")) {
        metadata(res) <- as.list(c(metadata(res), clustering_name = clustering_name))
        if (analysis_type == "DA") {
            return(list(res = res, d_counts = d_counts))
        }
        else if (analysis_type == "DS") {
            return(list(res = res, d_counts = d_counts, d_medians = d_medians, 
                d_medians_by_cluster_marker = d_medians_by_cluster_marker, 
                d_medians_by_sample_marker = d_medians_by_sample_marker))
        }
    }
}



library(multcomp)
    
testDA_GLMM_Local<-function (d_counts, formula, contrast=createContrast(c(0,0,0,0,0,0,0,0,1,0,0,0,-1/2,0,-1/2,0,0,0,0)), min_cells = 3, min_samples = NULL, 
    normalize = FALSE, norm_factors = "TMM") 
{
    if (is.null(min_samples)) {
        min_samples <- ncol(d_counts)/2
    }
    counts <- assays(d_counts)[["counts"]]
    cluster_id <- rowData(d_counts)$cluster_id
    tf <- counts >= min_cells
    ix_keep <- apply(tf, 1, function(r) sum(r) >= min_samples)
    counts <- counts[ix_keep, , drop = FALSE]
    cluster_id <- cluster_id[ix_keep]
    if (normalize & norm_factors == "TMM") {
        norm_factors <- edgeR::calcNormFactors(counts, method = "TMM")
    }
    if (normalize) {
        n_cells_smp <- colSums(counts)/norm_factors
    }
    else {
        n_cells_smp <- colSums(counts)
    }
    if (ncol(contrast) == 1 & nrow(contrast) > 1) {
        contrast <- t(contrast)
    }
    p_vals <- rep(NA, length(cluster_id))
    estimatedEff<-rep(NA,length(cluster_id))
    for (i in seq_along(cluster_id)) {
        p_vals[i] <- tryCatch({
            y <- counts[i, ]/n_cells_smp
            data_i <- cbind(y, n_cells_smp, formula$data)
            fit <- lme4::glmer(formula$formula, data = data_i, family = "binomial", 
                weights = n_cells_smp)
            test <- multcomp::glht(fit, contrast)
            summary(test)$test$pvalues
        }, error = function(e) NA)
    }
    
    ###extracting the LogFC.
    for (i in seq_along(cluster_id)) {
    estimatedEff[i] <- tryCatch({
      y <- counts[i, ]/n_cells_smp
      data_i <- cbind(y, n_cells_smp, formula$data)
      fit <- lme4::glmer(formula$formula, data = data_i, family = "binomial", 
        weights = n_cells_smp)
       # fitOut<-msigetSummary.merMod(fit)$coef
      #estimates<-estimates<-fitOut[,,]
      #estimatedEff[i]<-sum(estimates[,1]*contrast)
      test <- multcomp::glht(fit, contrast)
      summary(test)$test$coefficients
          }, error = function(e) NA)
 }
    
    p_adj <- p.adjust(p_vals, method = "fdr")
    stopifnot(length(p_vals) == length(p_adj))
    out <- data.frame(p_val = p_vals, p_adj = p_adj, estimate=estimatedEff, stringsAsFactors = FALSE)
    row_data <- as.data.frame(matrix(as.numeric(NA), nrow = nlevels(cluster_id), 
        ncol = ncol(out)))
    colnames(row_data) <- colnames(out)
    cluster_id_nm <- as.numeric(cluster_id)
    row_data[cluster_id_nm, ] <- out
    row_data <- cbind(cluster_id = rowData(d_counts)$cluster_id, 
        row_data)
    res <- d_counts
    rowData(res) <- row_data
    if (normalize) {
        metadata(res)$norm_factors <- norm_factors
    }
    res<-res[order(rowData(res)$p_adj),]
}


###states
testDS_LMM_Local<-function (d_counts, d_medians, formula, contrast, weights = TRUE, 
    markers_to_test = NULL, min_cells = 3, min_samples = NULL) 
{
    if (is.null(min_samples)) {
        min_samples <- ncol(d_counts)/2
    }
    if (!is.null(markers_to_test)) {
        markers_to_test <- markers_to_test
    }
    else {
        markers_to_test <- metadata(d_medians)$id_state_markers
    }
    counts <- assays(d_counts)[["counts"]]
    cluster_id <- rowData(d_counts)$cluster_id
    tf <- counts >= min_cells
    ix_keep <- apply(tf, 1, function(r) sum(r) >= min_samples)
    counts <- counts[ix_keep, , drop = FALSE]
    cluster_id <- cluster_id[ix_keep]
    if (is.logical(weights)) {
        if (weights) {
            weights <- colSums(counts)
        }
        else {
            weights <- NULL
        }
    }
    else if (is.numeric(weights)) {
        stopifnot(length(weights) == ncol(d_counts))
    }
    if (ncol(contrast) == 1 & nrow(contrast) > 1) {
        contrast <- t(contrast)
    }
    state_names <- names(assays(d_medians))[markers_to_test]
    meds <- do.call("rbind", {
        lapply(as.list(assays(d_medians)[state_names]), function(a) a[as.character(cluster_id), 
            , drop = FALSE])
    })
    meds_all <- do.call("rbind", as.list(assays(d_medians)[state_names]))
    p_vals <- rep(NA, nrow(meds))
    estimatedEff<-rep(NA,length(cluster_id))

    for (i in seq_len(nrow(meds))) {
        p_vals[i] <- tryCatch({
            y <- meds[i, ]
            data_i <- cbind(cbind(y, weights), formula$data)
            if (formula$random_terms) {
                fit <- lme4::lmer(formula$formula, data = data_i, weights = weights)
            }
            else {
                fit <- lm(formula$formula, data = data_i, weights = weights)
            }
            test <- multcomp::glht(fit, contrast)
            summary(test)$test$pvalues
        }, error = function(e) NA)
    }
    
    ###extracting the LogFC.
    for (i in seq_along(cluster_id)) {
    estimatedEff[i] <- tryCatch({
      y <- meds[i, ]
      data_i <- cbind(cbind(y, weights), formula$data)
     if (formula$random_terms) {
                fit <- lme4::lmer(formula$formula, data = data_i, weights = weights)
            }
            else {
                fit <- lm(formula$formula, data = data_i, weights = weights)
            }
       
      test <- multcomp::glht(fit, contrast)
      summary(test)$test$coefficients
          }, error = function(e) NA)
 }
    
    p_adj <- p.adjust(p_vals, method = "fdr")
    stopifnot(length(p_vals) == length(p_adj))
    out <- data.frame(p_val = p_vals, p_adj = p_adj,estimate=estimatedEff, stringsAsFactors = FALSE)
    row_data <- as.data.frame(matrix(as.numeric(NA), nrow = nlevels(cluster_id) * 
        length(state_names), ncol = ncol(out)))
    colnames(row_data) <- colnames(out)
    cluster_id_nm <- as.numeric(cluster_id)
    s <- seq(0, nlevels(cluster_id) * (length(state_names) - 
        1), by = nlevels(cluster_id))
    r1 <- rep(cluster_id_nm, length(state_names))
    r2 <- rep(s, each = length(cluster_id_nm))
    stopifnot(length(s) == length(state_names))
    stopifnot(length(r1) == length(r2))
    rows <- r1 + r2
    row_data[rows, ] <- out
    clus <- factor(rep(levels(cluster_id), length(state_names)), 
        levels = levels(cluster_id))
    stat <- factor(rep(state_names, each = length(levels(cluster_id))), 
        levels = state_names)
    stopifnot(length(clus) == nrow(row_data), length(stat) == 
        nrow(row_data))
    row_data <- cbind(data.frame(cluster_id = clus, marker_id = stat, 
        stringsAsFactors = FALSE), row_data)
    col_data <- colData(d_medians)
    res <- SummarizedExperiment(meds_all, rowData = row_data, 
        colData = col_data)
    res<-res[order(rowData(res)$p_adj),]
}
 


testDS_limma_local<-function (d_counts, d_medians, design, contrast, block_id = NULL, 
    trend = TRUE, weights = TRUE, markers_to_test = NULL, min_cells = 3, 
    min_samples = NULL, plot = FALSE, path = ".") 
{
    if (!is.null(block_id) & !is.factor(block_id)) {
        block_id <- factor(block_id, levels = unique(block_id))
    }
    if (is.null(min_samples)) {
        min_samples <- ncol(d_counts)/2
    }
    if (!is.null(markers_to_test)) {
        markers_to_test <- markers_to_test
    }
    else {
        markers_to_test <- metadata(d_medians)$id_state_markers
    }
    counts <- assays(d_counts)[["counts"]]
    cluster_id <- rowData(d_counts)$cluster_id
    tf <- counts >= min_cells
    ix_keep <- apply(tf, 1, function(r) sum(r) >= min_samples)
    counts <- counts[ix_keep, , drop = FALSE]
    cluster_id <- cluster_id[ix_keep]
    state_names <- names(assays(d_medians))[markers_to_test]
    meds <- do.call("rbind", {
        lapply(as.list(assays(d_medians)[state_names]), function(a) a[as.character(cluster_id), 
            , drop = FALSE])
    })
    meds_all <- do.call("rbind", as.list(assays(d_medians)[state_names]))
    if (!is.null(block_id)) {
        dupcor <- duplicateCorrelation(meds, design, block = block_id)
    }
    if (weights) {
        weights <- counts[as.character(rep(cluster_id, length(state_names))), 
            ]
        stopifnot(nrow(weights) == nrow(meds))
    }
    else {
        weights <- NULL
    }
    if (!is.null(block_id)) {
        message("Fitting linear models with random effects term for 'block_id'.")
        fit <- lmFit(meds, design, weights = weights, block = block_id, 
            correlation = dupcor$consensus.correlation)
    }
    else {
        fit <- lmFit(meds, design, weights = weights)
    }
    fit <- contrasts.fit(fit, contrast)
    efit <- eBayes(fit, trend = trend)
    if (plot) {
        pdf(file.path(path, "SA_plot.pdf"), width = 6, 
            height = 6)
        plotSA(efit)
        dev.off()
    }
    top <- limma::topTable(efit, coef = 1, number = Inf, adjust.method = "BH", 
        sort.by = "none")
    if (!all(top$ID %in% cluster_id)) {
        stop("cluster labels do not match")
    }
    row_data <- as.data.frame(matrix(as.numeric(NA), nrow = nlevels(cluster_id) * 
        length(state_names), ncol = ncol(top)))
    colnames(row_data) <- colnames(top)
    cluster_id_nm <- as.numeric(cluster_id)
    s <- seq(0, nlevels(cluster_id) * (length(state_names) - 
        1), by = nlevels(cluster_id))
    r1 <- rep(cluster_id_nm, length(state_names))
    r2 <- rep(s, each = length(cluster_id_nm))
    stopifnot(length(s) == length(state_names))
    stopifnot(length(r1) == length(r2))
    rows <- r1 + r2
    row_data[rows, ] <- top
    clus <- factor(rep(levels(cluster_id), length(state_names)), 
        levels = levels(cluster_id))
    stat <- factor(rep(state_names, each = length(levels(cluster_id))), 
        levels = state_names)
    stopifnot(length(clus) == nrow(row_data), length(stat) == 
        nrow(row_data))
    row_data <- cbind(data.frame(cluster_id = clus, marker_id = stat, 
        stringsAsFactors = FALSE), row_data)
    colnames(row_data)[colnames(row_data) == "P.Value"] <- "p_val"
    colnames(row_data)[colnames(row_data) == "adj.P.Val"] <- "p_adj"
    col_data <- colData(d_medians)
    res <- SummarizedExperiment(meds_all, rowData = row_data, 
        colData = col_data)
   res<- res[order(rowData(res)$p_adj),]
}


    
plotMedianExpressionBoxplot<-function(d_medians=NULL,cluster=NULL,marker=NULL,clinicalFactor=NULL){
  ex<-assays(d_medians)[[as.character(marker)]]
  ex<-as.data.frame(t(ex))
    ex$highIPI<-factor(exper_info2$highIPI[match(rownames(ex),exper_info2$sample_id)])
    ex$REF<-factor(exper_info2$group_id[match(rownames(ex),exper_info2$sample_id)])
  ex$NGCB<-factor(exper_info2$NGCB[match(rownames(ex),exper_info2$sample_id)])
  ex$C1<-factor(exper_info2$C1[match(rownames(ex),exper_info2$sample_id)])
  ex$C2<-factor(exper_info2$C2[match(rownames(ex),exper_info2$sample_id)])
  ex$C3<-factor(exper_info2$C3[match(rownames(ex),exper_info2$sample_id)])
  ex$C4<-factor(exper_info2$C4[match(rownames(ex),exper_info2$sample_id)])
  ex$C5<-factor(exper_info2$C5[match(rownames(ex),exper_info2$sample_id)])
  ex$doubleExpressor<-factor(exper_info2$doubleExpressor[match(rownames(ex),exper_info2$sample_id)])
   gg<-ggplot(ex,aes_string(x=clinicalFactor,y=cluster))+geom_boxplot()+ylab(paste0(marker," Normalized Median expression"))+ggtitle(cluster)+geom_jitter(color="grey")
  return(gg)
}


     
plotAbundanceCPMDiffCyt<-function(d_counts=NULL,cluster=NULL,marker=NULL,clinicalFactor=NULL){
   
  counts<-assays(d_counts)[["counts"]]
  cluster_id <- rowData(d_counts)$cluster_id
    tf <- counts >= 50
    ix_keep <- apply(tf, 1, function(r) sum(r) >=4)
    counts <- counts[ix_keep, , drop = FALSE]
    cluster_id <- cluster_id[ix_keep]
    
   norm.factors<-edgeR::calcNormFactors(counts,method="TMM")
    y <- edgeR::DGEList(counts,norm.factors = norm.factors)
  
  ###generate CPM counts of abundance  
  y <- edgeR::estimateDisp(y, design, trend.method = "none")
  logCPM<-asinh(edgeR::cpm(y))
  logCPM<-as.data.frame(t(logCPM))
  logCPM$highIPI<-factor(exper_info2$highIPI[match(rownames(logCPM),exper_info2$sample_id)])
    logCPM$REF<-factor(exper_info2$group_id[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$NGCB<-factor(exper_info2$NGCB[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$C1<-factor(exper_info2$C1[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$C2<-factor(exper_info2$C2[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$C3<-factor(exper_info2$C3[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$C4<-factor(exper_info2$C4[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$C5<-factor(exper_info2$C5[match(rownames(logCPM),exper_info2$sample_id)])
  logCPM$doubleExpressor<-factor(exper_info2$doubleExpressor[match(rownames(logCPM),exper_info2$sample_id)])
 gg<-ggplot(logCPM,aes_string(x=clinicalFactor,y=cluster))+geom_boxplot()+ylab(paste0(cluster," Log-CPM"))+ggtitle(cluster)+geom_jitter(color="grey")
 return(gg)
} 


createAnnotation<-function(DS=NULL,full.dn4,X=NULL,phenotype="unsup.subcommunity"){
  ddf<-diffcyt::topTable(DS,format_vals = TRUE,top_n=10, show_logFC = TRUE)%>%as.data.frame
    print(ddf)
  key<-data.frame(cluster_id=NA,marker_id=NA,logFC=0,p_adj=NA,symbol=NA,id=rownames(X))
  key<-key[match(rownames(X),key$id),]
  rez<-ddf[which(ddf$p_adj<0.1),]
  rez$id<-as.character(full.dn4[match(rez$cluster_id,full.dn4[,phenotype]),phenotype])
  rez$symbol<-NA
  rez$symbol[which(rez$p_adj<=0.01)]<-7
  rez$symbol[which(rez$p_adj>0.01 & rez$p_adj<=0.05)]<-3
    rez$symbol[which(rez$p_adj>0.05 & rez$p_adj<0.1)]<-17
key[match(rez$id,key$id),c("cluster_id","logFC","p_adj","symbol")]<-rez[,c("cluster_id","logFC","p_adj","symbol")]  
     return(key)
}   


 
findChapuyProportion<-function(sce=NULL,cluster="C1",phenotype="unsup.subcommunity",tmeOnly=TRUE,modelResults=NULL){
  #computes the conditional prob, and Z-score from mean 
  modelRes<-as.data.frame(topTable(modelResults,top_n=60,show_logFC = TRUE))
    c1size<-as.data.frame.matrix(table(colData(sce)$cluster_id,colData(sce)[,cluster]))
 clusterTotal<- as.data.frame(table(full.dn4[,phenotype]))
  c1size$clusterTotal<-0
  c1size$clusterTotal[match(clusterTotal$Var1, rownames(c1size))]<-clusterTotal$Freq
  c1size$chapuyProp<-0
  c1size$condProb<-0
  c1size$Z<-0
  tmeID<-!grepl("Tumor",rownames(c1size))
   c1size$chapuyProp<-100*c1size[,"1"]/sum(c1size[,"1"])
 
 if(any(c1size$chapuyProp>13)){
  c1size$chapuyProp[which(c1size$chapuyProp>13)]<-8.5
 }
   
   
 if(any(c1size$chapuyProp<1)){
  c1size$chapuyProp[which(c1size$chapuyProp<2)]<-1
  }
  
   Tot<-ncol(sce)
  if(tmeOnly==FALSE){
  c1size$condProb<-100*(c1size[,"1"]/Tot)/(sum(c1size[,"1"])/Tot)
     c1size$Z<-(c1size$condProb-mean(c1size$condProb))/sd(c1size$condProb)
  }else{
   c1size$condProb[tmeID]<-100*(c1size[tmeID,"1"]/Tot)/(sum(c1size[tmeID,"1"])/Tot)
   c1size$Z[tmeID]<-(c1size$condProb[tmeID]-mean(c1size$condProb[tmeID]))/sd(c1size$condProb[tmeID])
  }
   c1size$logFC<-modelRes$logFC[match(rownames(c1size),rownames(modelRes))]
   c1size$logFC[is.na(c1size$logFC)]<-0
   
     c1size$adj_p<-modelRes$p_adj[match(rownames(c1size),rownames(modelRes))]

   return(c1size)
}


grabMedianNormalized<-function(d_medians=NULL,cluster=NULL,marker=NULL,clinicalFactor=NULL){
  ex<-assays(d_medians)[[as.character(marker)]]
  ex<-as.data.frame(t(ex))
   #ex<-ex[,cluster]
    ex$highIPI<-factor(exper_info2$highIPI[match(rownames(ex),exper_info2$sample_id)])
    ex$REF<-factor(exper_info2$group_id[match(rownames(ex),exper_info2$sample_id)])
  ex$NGCB<-factor(exper_info2$NGCB[match(rownames(ex),exper_info2$sample_id)])
  ex$C1<-factor(exper_info2$C1[match(rownames(ex),exper_info2$sample_id)])
  ex$C2<-factor(exper_info2$C2[match(rownames(ex),exper_info2$sample_id)])
  ex$C3<-factor(exper_info2$C3[match(rownames(ex),exper_info2$sample_id)])
  ex$C4<-factor(exper_info2$C4[match(rownames(ex),exper_info2$sample_id)])
  ex$C5<-factor(exper_info2$C5[match(rownames(ex),exper_info2$sample_id)])
  ex$doubleExpressor<-factor(exper_info2$doubleExpressor[match(rownames(ex),exper_info2$sample_id)])
  return(ex[,c(cluster,"highIPI","REF","NGCB","C1","C2","C3","C4","C5")])
}






## phenograph to spatial point patter (PPP) object
 phenographToPPP<-function(case=NULL,marksCase=NULL,shift=FALSE){
 suppressPackageStartupMessages( require(spatstat))
  ##case: the case is the phenographed CSV 
  ##returns the PPP object, with marks as the original data frame.
  ##note that the window minimum is set to 0. This **SHOULD** not hopefully interfere with the spatial coordinates.
 ##this change of min 0 was necessary for clusterset function which conflicted with the window range.
  if(shift==TRUE){
  minX<-min(case[,"X_position"])
  minY<-min(case[,"Y_position"])
 case[,"X_position"]<-case[,"X_position"]-minX
 case[,"Y_position"]<-case[,"Y_position"]-minY
    mypat<-ppp(case[,"X_position"],case[,"Y_position"],c(min(case[,"X_position"]),summary(case[,"X_position"])[6]),c(min(case[,"Y_position"]),summary(case[,"Y_position"])[6]),marks=marksCase)
  ##shifting via spatstat causes non-stationary errors. FIXME: use manual shifting of coordinates prior   
 #mypat<-spatstat::shift(mypat,c(-minX,-minY))
  }else{
    #mypat<-ppp(case[,"X_position"],case[,"Y_position"],c(0,summary(case[,"X_position"])[6]),c(0,summary(case[,"Y_position"])[6]),marks=marksCase)
 mypat<-ppp(case[,"X_position"],case[,"Y_position"],c(min(case[,"X_position"]),max(case[,"X_position"])),c(min(case[,"Y_position"]),max(case[,"Y_position"])),marks=marksCase) 
 } 

 ## scale in  micrometers
   unitname(mypat)<-"micrometer"

  return(mypat)
 }

 #### borrowing from Vito's pipeline.
  findSecondObjectNumber<-function(pp=NULL,minDistance=13.5){
  ##FIX ME : use apply nbhd
   x<-pairdist(pp)
    id<-apply(x,1,function(x) which(x<=minDistance))
  ##stack the list into long format
  names(id)<-as.character(seq(1,npoints(pp)))
  datRel<-stack(id)
 ##the second object near the first fixed object.
  colnames(datRel)<-c("Second Object Number","First Object Number")
   ##exclude self
   datRel<-datRel[which(datRel[,"First Object Number"]!=datRel[,"Second Object Number"]),]
   datRel[,1]<-as.numeric(datRel[,1])
  datRel[,2]<-as.numeric(datRel[,2])
   return(datRel)
  }




 createRelationTable<-function(dn5=NULL,myPheno=NULL,minDistance=12){
  ##for each ROI, create PPP
  dat_relation2<-c()
  for(roi in unique(dn5$ROIID)){
   message(roi)
   dats<-subset(dn5,dn5$ROIID==roi)
   synth3.cells<-phenographToPPP(case=dats,marks=factor(dats[,myPheno]))
   dat_relation<-findSecondObjectNumber(pp=synth3.cells,minDistance=minDistance)
  dat_relation$Relationship<-"Neighbors"
  dat_relation[,"First Object Name"]<-"cell"
  dat_relation[,"First Image Number"]<-roi
  dat_relation[,"Second Object Name"]<-"cell"
  dat_relation[,"Second Image Number"]<-roi
  dat_relation2<-rbind(dat_relation2,dat_relation)
  }
 return(data.table(dat_relation2))
}



makePermutationNBHD<-function(dat_cells=NULL,dat_relation=NULL,n_perm=100,pthreshold=0.05){
  d = prepare_tables(dat_cells, dat_relation)
  dat_baseline = apply_labels(d[[1]], d[[2]]) %>%
  aggregate_histo()


 set.seed(12312)
   dat_perm = rbindlist(mclapply(1:n_perm, function(x){
  dat_labels = shuffle_labels(d[[1]])
  apply_labels(dat_labels, d[[2]]) %>%
    aggregate_histo()
  },mc.cores = 1
  ), idcol = 'run') 
 
   dat_p <- calc_p_vals(dat_baseline, dat_perm, n_perm = n_perm, p_tresh = pthreshold) 
 ### plotting heatmap.
  
   pmat = dcast(dat_p, 'FirstLabel ~ SecondLabel', value.var = 'sigval', fun.aggregate = sum,
             fill=0, drop=F)
   rname = pmat$FirstLabel
   pmat<-as.data.table(pmat)
   pmat = pmat[,which(colnames(pmat)!="FirstLabel"),with=FALSE] %>%
    as.matrix()
row.names(pmat) <- rname
 ### he structures his probabilities greater than and less than.
  return(list(pmat=pmat,dat_p=dat_p))
 }

## Vito Methodology.
 performVitoInteraction<-function(full.dn4.5nn2=NULL,phenotypeColumn=NULL,saveName=NULL,minDistance=15){
  zz<-full.dn4.5nn2
 table(zz[,phenotypeColumn])
   ### Vito's spatial analysis subtypes.
  dat_cells<-data.frame(ImageNumber=zz$ROIID,
	ObjectNumber=zz$CellId,
	label=as.character(zz[,phenotypeColumn]),
	group=zz$ROIID)
  dat_cells<-data.table(dat_cells)
  dat_relation<-createRelationTable(dn5=zz,myPheno=phenotypeColumn,minDistance=minDistance)
 save(dat_relation,file=paste0("dat_relation_",minDistance,"_",saveName,".RData"))

  p1<-makePermutationNBHD(dat_cells=dat_cells,dat_relation=dat_relation,n_perm=1000)
  save(p1,file=paste0("p1_",minDistance,"_",saveName,".RData"))
   require(pheatmap)
  cols = rev(brewer.pal(11,'Spectral'))
  cmap = colorRampPalette(cols)
  hr <- hclust(dist(p1$pmat), method="ward.D")
     pheatmap(p1$pmat, 
	color = cmap(25), 
    cluster_cols = TRUE, 
    cluster_rows = TRUE, 
    labels_col = colnames(p1),
    labels_row = colnames(p1), 
    display_numbers = TRUE,  
    fontsize = 14,
    cutree_rows=2,
	cutree_cols=2)

  return(p1)
 }
 
 
 
 getPhenoColorsAnnotatedLabel<-function(full.dn4){
 ##this requires the T-Cell names to not be included.
 require(circlize)
  tsn<-full.dn4[which(!is.na(full.dn4$tSNE1) & !is.na(full.dn4$tSNE2)),c("tSNE1",
	"tSNE2",
	"Cell_CD20",
	"Cell_CD3",
	"Cell_CD4",
	"Cell_CD31",
	"Cell_CD4",
	"Cell_CD68",
	"Cell_CD8",
	"Cell_FOXP3",
	"Cell_BCL2",
	"uniqueLabel","response","ROIID",
	"annotated.label")]
  ##balanced response representation.
   table(tsn$response)
  tsn$primary.phenotype<-(tsn$annotated.label)
   cd8s<-colorRampPalette(c("darkseagreen","green"),5)
  cd4s<-colorRampPalette(c("lightsteelblue1","blue"),5)
  macs<-colorRampPalette(c("magenta","maroon4"),5)
   tregs<-colorRampPalette(c("plum2","purple"),5)
   tumrs<-colorRampPalette(c("grey24","grey64"),20)
   phenoColors<-c(cd4s(5),
	cd8s(3),"red",macs(4),
	tregs(3),
	tumrs(2),'bisque')

 names(phenoColors)<-c(levels(factor(tsn$primary.phenotype)),'Tumor_11')
   ##manual color change code.
   phenoColors["Memory CD4"]<-"darkblue"
    phenoColors["CD45RA- exhausted CD4"]<-"turquoise2"
   phenoColors["Ki67+ memory CD4"]<-"steelblue4"

   phenoColors["Exhausted CD8"]<-"springgreen4"

 ##
    phenoColors["M1-Macrophage"]<-"magenta4"
    phenoColors["M2-Macrophage"]<-"hotpink2"
     phenoColors["PD-L1+ M2-Macrophage"]<-"lightpink"
 
 ## tregs
      phenoColors["T-reg"]<-"mediumorchid4"
       phenoColors["Highly suppressive T-reg"]<-"yellow"

 # full<-ggplot(tsn,  aes(x = tSNE1, y = tSNE2, color = primary.phenotype)) +
 # geom_point(size = 0.8) +
 # theme_bw() +
 # scale_color_manual(values = phenoColors) +
 # guides(color = guide_legend(override.aes = list(size = 4), ncol = 2))
 # full

   return(phenoColors)
 }
 
 

 getPhenoColorsAkilAnnotatedLabel<-function(full.dn4){
 ##this requires the T-Cell names to not be included.
 require(circlize)
  tsn<-full.dn4[,c("tSNE1",
	"tSNE2",
	"Cell_CD20",
	"Cell_CD3",
	"Cell_CD4",
	"Cell_CD31",
	"Cell_CD4",
	"Cell_CD68",
	"Cell_CD8",
	"Cell_FOXP3",
	"Cell_BCL2",
	"uniqueLabel","response","ROIID",
	"annotated.label.back_akil")]
  ##balanced response representation.
   table(tsn$response)
  tsn$primary.phenotype<-(tsn$annotated.label.back_akil)
   cd8s<-colorRampPalette(c("darkseagreen","green"),5)
  cd4s<-colorRampPalette(c("lightsteelblue1","blue"),5)
  macs<-colorRampPalette(c("magenta","maroon4"),5)
   tregs<-colorRampPalette(c("plum2","purple"),5)
   tumrs<-colorRampPalette(c("grey24","grey64"),20)
      endos<-colorRampPalette(c("rosybrown1","red"),20)

   phenoColors<-c(cd4s(4),
	cd8s(4),
   macs(3),
	tregs(3),
	tumrs(2),
	 endos(5),
	"bisque1")

 names(phenoColors)<-c(levels(factor(tsn$primary.phenotype)))
   ##manual color change code.
   phenoColors["Late Exhausted/Inflam. CD4(1,5)"]<-"darkblue"
    phenoColors["Activated/Early/Exhausted CD4(2,3,7)"]<-"turquoise2"
   phenoColors["Baseline CD4(4)"]<-"steelblue4"
    phenoColors["Activated/Prolif. CD4(6)"]<-"lightpink4"
    
   phenoColors["Baseline CD8(7)"]<-"springgreen4"
 ##
    phenoColors["M1-Macrophage(3,4)"]<-"magenta4"
    phenoColors["M2-Macrophage(1,5,6)"]<-"hotpink2"
     phenoColors["PD-L1+ M2-Macrophage(2)"]<-"lightpink" 
 ## tregs
      phenoColors["Baseline T-reg(3,4)"]<-"mediumorchid4"
       phenoColors["Highly Suppress. T-reg(1,2,6)"]<-"yellow"

 ## tumor groups
        phenoColors["Non-Inflam. B-Cell tumor(1-5,7-9)"]<-"khaki3"
        phenoColors["Inflammatory B-Cell tumor(6,10)"]<-"violetred4"


   return(phenoColors)
 }
    
    
distanceCalculationROI<-function(synth=NULL,zz=NULL,cluster.column="cd8.case.cluster"){
  require(Rphenograph)
  nnd1<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=1)
    nnd2<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=2)
  nnd3<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=3)
  nnd4<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=4)
  nnd5<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=5)
  #nnd6<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=6)
  #nnd7<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=7)
  #nnd8<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=8)
  
  ## assemble
  nnd<-cbind(nnd1[,grepl("k_",colnames(nnd1))],
	nnd2[,grepl("k_",colnames(nnd2))],
	nnd3[,grepl("k_",colnames(nnd3))],
	nnd4[,grepl("k_",colnames(nnd4))],
	nnd5[,grepl("k_",colnames(nnd5))])
	#nnd6[,grepl("k_",colnames(nnd6))],
	#nnd7[,grepl("k_",colnames(nnd7))],
	#nnd8[,grepl("k_",colnames(nnd8))])
    nnd$uniqueLabel<-zz$uniqueLabel
    nnd$marks<-marks(synth)
   stopifnot(all(nrow(nnd)==nrow(zz)))
     nnd$ROIID<-zz$ROIID
    nnd$cluster<-zz[,cluster.column]
  return(nnd)
}




 grabDistanceRange<-function(synth=NULL,uniqueLabel=NULL,k=1,from=NULL,to=NULL){
  nnd<-as.data.frame(nndist(synth,k=k,by=marks(synth)))
  colnames(nnd)<-paste0("k_",k,"_",colnames(nnd))
  for(i in 1:ncol(nnd)){
  nnd[which(is.infinite(nnd[,i])),i] <-300
 }##
   nnd$marks<-marks(synth)
   nnd$K=1
   nnd$uniqueLabel<-uniqueLabel 
  return(nnd)
 }
 
 
 

 createExpressionFormat<-function(ex=NULL,exper_info2){
  ex<-as.data.frame(t(ex))
    ex$highIPI<-factor(exper_info2$highIPI[match(rownames(ex),exper_info2$sample_id)])
    ex$REF<-factor(exper_info2$group_id[match(rownames(ex),exper_info2$sample_id)])
  ex$NGCB<-factor(exper_info2$NGCB[match(rownames(ex),exper_info2$sample_id)])
  ex$C1<-factor(exper_info2$C1[match(rownames(ex),exper_info2$sample_id)])
  ex$C2<-factor(exper_info2$C2[match(rownames(ex),exper_info2$sample_id)])
  ex$C3<-factor(exper_info2$C3[match(rownames(ex),exper_info2$sample_id)])
  ex$C4<-factor(exper_info2$C4[match(rownames(ex),exper_info2$sample_id)])
  ex$C5<-factor(exper_info2$C5[match(rownames(ex),exper_info2$sample_id)])
  ex$doubleExpressor<-factor(exper_info2$doubleExpressor[match(rownames(ex),exper_info2$sample_id)])
  ex$REF<-as.character(ex$REF)
  ex$REF[which(ex$REF=="Primary refract")]<-"Primary refractory"
  return(ex)
 }
 
 
 getSpatialPhenoColors<-function(full.dn4=NULL){
	##this uses the color coding from the Tumor spatial classification
  phenoColors<-phenoColors<-getPhenoColors(full.dn4)

   cols<-data.frame(phenotype= unique(full.dn4$tumor.5nn.class.unsup),
	colors="black",
	stringsAsFactors=FALSE)
   for(i in 1:nrow(cols)){
   id<-which(names(phenoColors)[i]==cols$phenotype)
   cols$colors[id]<-phenoColors[i]
  }
        
 	cols[which(cols$phenotype=="Tumor_b"),"colors"]<-"lightslategrey"
     cols[which(cols$phenotype=="Tumor_g"),"colors"]<-"gray89"
       cols[which(cols$phenotype=="Tumor_c"),"colors"]<-"black"
 	 cols[which(cols$phenotype=="Tumor_f"),"colors"]<-"khaki1"
   cols[which(cols$phenotype=="Tumor_i"),"colors"]<-"tan3"
 ###REF
	       cols[which(cols$phenotype=="Tumor_e"),"colors"]<-"orange"
 	       cols[which(cols$phenotype=="Tumor_a"),"colors"]<-"cyan"
  	       cols[which(cols$phenotype=="Tumor_h"),"colors"]<-"lightpink"

    	cols[which(cols$phenotype=="Tumor_d"),"colors"]<-color_clusters[6]


   return(cols)
 }
 
 
 

 getPhenoColorsAnnotatedLabelUMAP<-function(full.dn4){
 ##this requires the T-Cell names to not be included.
 require(circlize)
  tsn<-full.dn4[which(!is.na(full.dn4$tSNE1) & !is.na(full.dn4$tSNE2)),c("tSNE1",
	"tSNE2",
	"Cell_CD20",
	"Cell_CD3",
	"Cell_CD4",
	"Cell_CD31",
	"Cell_CD4",
	"Cell_CD68",
	"Cell_CD8",
	"Cell_FOXP3",
	"Cell_BCL2",
	"uniqueLabel","response","ROIID",
	"annotated.umap")]
  ##balanced response representation.
   table(tsn$response)
  tsn$primary.phenotype<-(tsn$annotated.umap)
   cd8s<-colorRampPalette(c("darkseagreen","green"),5)
  cd4s<-colorRampPalette(c("lightsteelblue1","blue"),5)
  macs<-colorRampPalette(c("magenta","maroon4"),5)
   tregs<-colorRampPalette(c("plum2","purple"),5)
   tumrs<-colorRampPalette(c("grey24","grey64"),20)
   phenoColors<-c(cd4s(6),
	cd8s(3),
	"red",
	macs(4),
	tregs(3),
	tumrs(6),'bisque')

 names(phenoColors)<-c(levels(factor(tsn$primary.phenotype)),'Tumor_11')
   ##manual color change code.
   phenoColors["Memory CD4"]<-"darkblue"
    phenoColors["CD45RA- exhausted CD4"]<-"turquoise2"
   phenoColors["Ki67+ memory CD4"]<-"steelblue4"
    phenoColors["CD45RA+TIM3+ CD4"]<-"lightpink4"
   phenoColors["Exhausted CD8"]<-"springgreen4"
 ##
    phenoColors["M1-Macrophage"]<-"magenta4"
    phenoColors["M2-Macrophage"]<-"hotpink2"
     phenoColors["PD-L1+ M2-Macrophage"]<-"lightpink" 
 ## tregs
      phenoColors["T-reg"]<-"mediumorchid4"
       phenoColors["Highly suppressive T-reg"]<-"yellow"

 ## tumor groups
        phenoColors["Ki67+ B-Cell tumor"]<-"gray79"
        phenoColors["Ki67- B-Cell tumor"]<-"khaki3"
        phenoColors["Ki67+CCR4+ B-Cell tumor"]<-"violetred4"
        phenoColors["cMYC+Ki67+ B-Cell tumor"]<-"darkorange"

  #full<-ggplot(tsn,  aes(x = tSNE1, y = tSNE2, color = primary.phenotype)) +
  #geom_point(size = 0.8) +
  #theme_bw() +
  #scale_color_manual(values = phenoColors) +
  #guides(color = guide_legend(override.aes = list(size = 4), ncol = 2))
  #full

   return(phenoColors)
 }

distanceClusterROI.5NN<-function(synth=NULL,zz=NULL){
  require(Rphenograph)
  nnd1<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=1)
    nnd2<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=2)
  nnd3<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=3)
  nnd4<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=4)
  nnd5<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=5)
  ## need to average them.
 
  ## assemble
  nnd<-cbind(nnd1[,grepl("k_",colnames(nnd1))],
	nnd2[,grepl("k_",colnames(nnd2))],
	nnd3[,grepl("k_",colnames(nnd3))],
	nnd4[,grepl("k_",colnames(nnd4))],
	nnd5[,grepl("k_",colnames(nnd5))])
	#nnd9[,1:3],
	#nnd10[,1:3])
   nndImmune<-rowMeans(nnd[,grepl("_immune",colnames(nnd))])
   nndTumor<-rowMeans(nnd[,grepl("_Tumor",colnames(nnd))])
   nnd.ti<-data.frame(nndImmune,nndTumor)
  ###use the 5NN distances to immune and tumor.
    pheno<-Rphenograph(nnd.ti,k=40)
 ##########
    nnd$cluster<-factor(membership(pheno[[2]]))
     nnd$uniqueLabel<-zz$uniqueLabel
     nnd$marks<-marks(synth)
   stopifnot(all(nrow(nnd)==nrow(zz)))
   zz$tumor.immune.5nn<-nnd$cluster
 zz$nndImmune.5nn<-nnd.ti$nndImmune
 zz$nndTumor.5nn<-nnd.ti$nndTumor
  return(zz)
}

  plotShape<-function(path="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-shape-objects-revised/final-shapes/",ROI="3964",cols=NULL,full.dn4,compartment=NULL,phenotype=NULL,phenotype.column.toShape=NULL){
   sh<-dir(path)
  n<-sh[grepl(ROI,sh)]
  load(paste0(path,n))
  spdf$cellType2<-sapply(strsplit(as.character(spdf$cellType),"_"),function(x) x[1])
  mycase<-full.dn4[which(full.dn4$ROIID==ROI),]
  tumor.id<-mycase$uniqueLabel[which(mycase[,compartment]==phenotype) ]
  spdf$uniqueLabel<-spdf$id
  spdf<-left_join(spdf,mycase[,c("uniqueLabel",phenotype.column.toShape)],by="uniqueLabel")
   spdf$cellType3<-factor(spdf[,phenotype.column.toShape],levels=unique(spdf[,phenotype.column.toShape]))
 test<- ggplot(spdf,aes(x=long,y=lat,group=id,fill=cellType3)) +
      geom_polygon( size=0.1,colour="black") +scale_fill_manual(values=cols[match(levels(spdf$cellType3),cols$phenotype),"colors"])+ggtitle(ROI)+theme_bw()
 
  # print(test)
  return(test)
  }

```

### Load the data
- load the unfiltered complete data 692,463 cells
- include annotations from Dr. Merchant
- include the endothelial clusters.
```{r loadData}

  require(ggplot2)
 suppressPackageStartupMessages( require(ComplexHeatmap))
 suppressPackageStartupMessages( require(knitr))
 suppressPackageStartupMessages( require(kableExtra))
suppressPackageStartupMessages( require(spatstat))
 suppressPackageStartupMessages(library(dplyr))
 suppressPackageStartupMessages(library(magrittr))
 suppressPackageStartupMessages(library(survival))
  suppressPackageStartupMessages(library(Hmisc))

 library(data.table)
library(dplyr)
library(magrittr)
library(dtplyr)
library(ggplot2)
library(parallel)
library(neighbouRhood)
library(gplots)
library(RColorBrewer)
 library(hrbrthemes)
    library(igraph)
library(imcExperiment)
  
   library(CATALYST)
library(cowplot)
library(flowCore)
library(ggplot2)
library(SingleCellExperiment)
library(diffcyt)
library(imcExperiment)
 library(gridExtra)
 library(magrittr);library(dplyr)
  
  load(file="~/IMC-Ranalysis/DLBCL/monirath/myData/revision-comments-11-6/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")

   full.dn4$unsup.subcommunity<-as.character(full.dn4$unsup.subcommunity)
     full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="Endothelial")]<-paste0("Endothelial_",full.dn4$Endothelial_subphenograph_cluster_meta[which(full.dn4$subType.correction=="Endothelial")])
     dim(full.dn4)

  full.dn4$unsup.subcommunity<-factor(full.dn4$unsup.subcommunity,
	levels=c("CD4_1",
	       "CD4_2",
		"CD4_3",
		"CD4_4",
		"CD4_5","CD4_6","CD4_7",
	        "CD8_1","CD8_2","CD8_3","CD8_4","CD8_5","CD8_6","CD8_7",
 	       "Macrophage_1","Macrophage_2","Macrophage_3","Macrophage_4","Macrophage_5",
	"Macrophage_6",
	"TREG_1","TREG_2","TREG_3","TREG_4","TREG_5",     
	 "TREG_6",
	 "Tumor_1","Tumor_2","Tumor_3", "Tumor_4", "Tumor_5","Tumor_6","Tumor_7","Tumor_8", "Tumor_9",  "Tumor_10" ,"Tumor_11",
	"Endothelial_1",	"Endothelial_2",	"Endothelial_3", 	"Endothelial_4", 	"Endothelial_5", 	"Endothelial_6"  ))

 lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
	"Area.norm",
	"Eccentricity.norm",
	"Solidity.norm",
	"Perimeter.norm",
	"Percent_Touching.norm",
	"Number_Neighbors.norm")

```  
  
### Adding the annotations from Figure 2A based on Akil's suggestion 5-2- revision
- add custom row labels to each phenotype showing cluster and labeled name.
- REVISION 5-2: we add the cluster labels to the names for back-tracking the names.
- FIX ME: the objects need to be re-cast to match the renamed labels.  the order is not right.
- added annotations for endothelial
- 1-8: tumor_8 was updated to be an inflammatory tumor during the writing.  this should reflect the other scripts changes as well.
```{r}

### REVISION 5-2: Annotate cluster to cell populations 


 ## need to add labels.
  label<-full.dn4$unsup.subcommunity%>%levels%>%data.frame
 colnames(label)<-'cluster'
  label$label<-'none'
 ## CD4 labels.
  label$label[which(label$cluster=='CD4_1')]<-'Late Exhausted/Inflam. CD4(1,5)'
  label$label[which(label$cluster=='CD4_2')]<-"Activated/Early/Exhausted CD4(2,3,7)"
  label$label[which(label$cluster=='CD4_3')]<-'Activated/Early/Exhausted CD4(2,3,7)'
   label$label[which(label$cluster=='CD4_4')]<-'Baseline CD4(4)'
   label$label[which(label$cluster=='CD4_5')]<-'Late Exhausted/Inflam. CD4(1,5)'
   label$label[which(label$cluster=='CD4_6')]<-'Activated/Prolif. CD4(6)'
    label$label[which(label$cluster=='CD4_7')]<-'Activated/Early/Exhausted CD4(2,3,7)'

 ## CD8 labels.
      label$label[which(label$cluster=='CD8_1')]<-'Activated/Prolif. CD8(1,4)'
      label$label[which(label$cluster=='CD8_2')]<-'Exhausted/Inflam. CD8(2,6)'
      label$label[which(label$cluster=='CD8_3')]<-'Terminally Exhaust. CD8(3,5)'
      label$label[which(label$cluster=='CD8_4')]<-'Activated/Prolif. CD8(1,4)'
      label$label[which(label$cluster=='CD8_5')]<-'Terminally Exhaust. CD8(3,5)'
      label$label[which(label$cluster=='CD8_6')]<-'Exhausted/Inflam. CD8(2,6)'
      label$label[which(label$cluster=='CD8_7')]<-'Baseline CD8(7)'
 ##exhausted
 
 ## MAC
	 label$label[which(label$cluster=='Macrophage_1')]<-'M2-Macrophage(1,5,6)'
        label$label[which(label$cluster=='Macrophage_2')]<-'PD-L1+ M2-Macrophage(2)'
        label$label[which(label$cluster=='Macrophage_3')]<-'M1-Macrophage(3,4)'	
	 label$label[which(label$cluster=='Macrophage_4')]<-'M1-Macrophage(3,4)'	
        label$label[which(label$cluster=='Macrophage_5')]<-'M2-Macrophage(1,5,6)'
	 label$label[which(label$cluster=='Macrophage_6')]<-'M2-Macrophage(1,5,6)'	

 ## TREG.
 	 label$label[which(label$cluster=='TREG_1')]<-'Highly Suppress. T-reg(1,2,6)'
 	 label$label[which(label$cluster=='TREG_2')]<-'Highly Suppress. T-reg(1,2,6)'
 	 label$label[which(label$cluster=='TREG_3')]<-'Baseline T-reg(3,4)'
 	 label$label[which(label$cluster=='TREG_4')]<-'Baseline T-reg(3,4)'
	 label$label[which(label$cluster=='TREG_5')]<-'Activated/Prolif. T-reg(5)'
 	 label$label[which(label$cluster=='TREG_6')]<-'Highly Suppress. T-reg(1,2,6)'

 ##ENDO
	label$label[which(label$cluster=='Endothelial_1')]<-'EphrinB2+Endothelial(1)'
	label$label[which(label$cluster=='Endothelial_2')]<-'Proliferative Endothelial(2)'
	label$label[which(label$cluster=='Endothelial_3')]<-'Activated Endothelial(3,6)'
	label$label[which(label$cluster=='Endothelial_4')]<-'PD-L1+Endothelial(4)'
	label$label[which(label$cluster=='Endothelial_5')]<-'Baseline Endothelial(5)'
	label$label[which(label$cluster=='Endothelial_6')]<-'Activated Endothelial(3,6)'

 

 ## Tumor
 ## t1:CXCR3-mid, BCL2+
 ##t2: CD20+ B-cell
 #t3: BCL2+
 # t4: Ki67+, BCL6+
 # t5: Ki67+, BCL6+
 
 #t6: PD-1-moderate, Ki67+, BCL6+, BCL2+
 #t7:CD20+ B-cell
 #t8: CD20+ B-cell
 #t9: Ki67+
 #t10: Ki67+, BCL2+,BCL6+
 ## labels
   label$label[which(label$cluster=='Tumor_1')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_2')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_3')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_4')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_5')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_6')]<-'Inflammatory B-Cell tumor(6,8,10)'
   label$label[which(label$cluster=='Tumor_7')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_8')]<-'Inflammatory B-Cell tumor(6,8,10)'
   label$label[which(label$cluster=='Tumor_9')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_10')]<-'Inflammatory B-Cell tumor(6,8,10)'
      label$label[which(label$cluster=='Tumor_11')]<-'none(11)'



 label$easyLabel<-c("Late_Exhausted_CD4",
	"Act_Exh_early_CD4",
	"Act_Exh_early_CD4",
	"Baseline_CD4",
	"Late_Exhausted_CD4",
	"Act_prol_CD4",
	"Act_Exh_early_CD4",
	
	"Act_prol_CD8",
	"Exh_Inflam_CD8",
	"Term_Exh_CD8",
	"Act_prol_CD8",
	"Term_Exh_CD8",
	"Exh_Inflam_CD8",
	"Baseline_CD8",
	
	"M2MAC",
	"PDL1M2MAC",
	"M1MAC",
	"M1MAC",
	"M2MAC",
	"M2MAC",
	
	"Hi_Supp_TREG",
	"Hi_Supp_TREG",
	"Baseline_TREG",
	"Baseline_TREG",
	"Act_prol_TREG",
	"Hi_Supp_TREG",
	
	"Tumor",
	"Tumor",
	"Tumor",
	"Tumor",
	"Tumor",
	"InflamTumor",
	"Tumor",
	"InflamTumor",
	"Tumor",
	"InflamTumor",
	'none',
	"EphrinB2_Endothelial",
	"Ki67_Endothelial",
	"Activ_Endothelial",
	"PDL1_Endothelial",
	"Baseline_Endothelial",
	"Activ_Endothelial"	)
	
 full.dn4$annotated.label.simple_akil<-NA
 full.dn4$annotated.label.simple_akil<-label$easyLabel[match(full.dn4$unsup.subcommunity,label$cluster)]


 ## REVISION: 5-2 we add the 'mapped-back' label with cluster names to the full.dn4 data-set.
 full.dn4$annotated.label.back_akil<-NA
 full.dn4$annotated.label.back_akil<-label$label[match(full.dn4$unsup.subcommunity,label$cluster)]
 
 
 full.dn4$annotated.label.back_akil<-factor(full.dn4$annotated.label.back_akil,levels=c("Late Exhausted/Inflam. CD4(1,5)",
                                                                         "Activated/Early/Exhausted CD4(2,3,7)",
                                                                         "Baseline CD4(4)",
                                                                         "Activated/Prolif. CD4(6)",
                                                                         
                                                                         "Baseline CD8(7)",
                                                                         "Activated/Prolif. CD8(1,4)",
                                                                         "Exhausted/Inflam. CD8(2,6)",
                                                                         "Terminally Exhaust. CD8(3,5)",
                                                                         
                                                                         "M1-Macrophage(3,4)",
                                                                         "M2-Macrophage(1,5,6)",
                                                                         "PD-L1+ M2-Macrophage(2)",
                                                                         
                                                                         "Highly Suppress. T-reg(1,2,6)",
                                                                         "Activated/Prolif. T-reg(5)",
                                                                         "Baseline T-reg(3,4)",
                                                                         
                                                                         "Non-Inflam. B-Cell tumor(1-5,7,9)",
                                                                         "Inflammatory B-Cell tumor(6,8,10)",
                                                                         
                                                                          "EphrinB2+Endothelial(1)",
                                                                         "Proliferative Endothelial(2)",
                                                                         "Activated Endothelial(3,6)",
                                                                         "PD-L1+Endothelial(4)",
                                                                         "Baseline Endothelial(5)",
                                                                         'none(11)'))
 
 print(label)
 table(full.dn4$annotated.label.back_akil)
##labels are updated to match Dr. Merchant's labeling clusters.
 
 
 ##need to create a heatmap with the labels appended 5-2.
 full.dn4$subcluster_annotated<-'none'
 full.dn4$subcluster_annotated<-paste0(full.dn4$unsup.subcommunity," (",as.character(full.dn4$annotated.label.back_akil),")")
 
 
 
 full.dn4$subcluster_annotated<-factor(full.dn4$subcluster_annotated,levels= 
 c( "CD4_1 (Late Exhausted/Inflam. CD4(1,5))" ,
    "CD4_2 (Activated/Early/Exhausted CD4(2,3,7))",
    "CD4_3 (Activated/Early/Exhausted CD4(2,3,7))" ,
    "CD4_4 (Baseline CD4(4))" ,
    
    "CD4_5 (Late Exhausted/Inflam. CD4(1,5))",
    "CD4_6 (Activated/Prolif. CD4(6))",
    "CD4_7 (Activated/Early/Exhausted CD4(2,3,7))",
    
     "CD8_1 (Activated/Prolif. CD8(1,4))",
     "CD8_2 (Exhausted/Inflam. CD8(2,6))",
    "CD8_3 (Terminally Exhaust. CD8(3,5))" ,
    "CD8_4 (Activated/Prolif. CD8(1,4))",
     "CD8_5 (Terminally Exhaust. CD8(3,5))",
    "CD8_6 (Exhausted/Inflam. CD8(2,6))",
     "CD8_7 (Baseline CD8(7))",
    
    "Macrophage_1 (M2-Macrophage(1,5,6))",
     "Macrophage_2 (PD-L1+ M2-Macrophage(2))",
    "Macrophage_3 (M1-Macrophage(3,4))",
    "Macrophage_4 (M1-Macrophage(3,4))"  ,
    "Macrophage_5 (M2-Macrophage(1,5,6))" ,
    "Macrophage_6 (M2-Macrophage(1,5,6))",
    
    "TREG_1 (Highly Suppress. T-reg(1,2,6))",
    "TREG_2 (Highly Suppress. T-reg(1,2,6))",
      "TREG_3 (Baseline T-reg(3,4))" ,
    "TREG_4 (Baseline T-reg(3,4))"   ,
     "TREG_5 (Activated/Prolif. T-reg(5))",
    "TREG_6 (Highly Suppress. T-reg(1,2,6))",
    
    
    "Tumor_1 (Non-Inflam. B-Cell tumor(1-5,7,9))",
    "Tumor_2 (Non-Inflam. B-Cell tumor(1-5,7,9))",
     "Tumor_3 (Non-Inflam. B-Cell tumor(1-5,7,9))" ,
     "Tumor_4 (Non-Inflam. B-Cell tumor(1-5,7,9))",
      "Tumor_5 (Non-Inflam. B-Cell tumor(1-5,7,9))" ,
     "Tumor_6 (Inflammatory B-Cell tumor(6,8,10))" ,
      "Tumor_7 (Non-Inflam. B-Cell tumor(1-5,7,9))",
    "Tumor_8 (Inflammatory B-Cell tumor(6,8,10))",
    "Tumor_9 (Non-Inflam. B-Cell tumor(1-5,7,9))",
     "Tumor_10 (Inflammatory B-Cell tumor(6,8,10))",
      "Tumor_11 (none(11))" ,
    
            "Endothelial_1 (EphrinB2+Endothelial(1))",
      "Endothelial_2 (Proliferative Endothelial(2))",
    "Endothelial_3 (Activated Endothelial(3,6))",
    "Endothelial_4 (PD-L1+Endothelial(4))",
    "Endothelial_5 (Baseline Endothelial(5))",
    "Endothelial_6 (Activated Endothelial(3,6))" ))
 

```
 
 
## Append the spatial phenotypes and identify neighborhoods of the labeled sets.
- Originally used label tag based on the UMAP analysis, revised 7-18 to use Merchant annotation.
- incorporates Dr. Merchant's annotation labels and simple label with the tumor topology labels.
- creates a cluster based label tag for TME at the cluster level, and tumor topology level.
```{r}
 full.dn4$tumor.5nn.umap.simple.label<-as.character(full.dn4$annotated.label.simple_akil)


 full.dn4$tumor.5nn.umap.simple.label[which(full.dn4$subType.correction=="Tumor")]<-full.dn4$tumor.5nn.class.unsup[which(full.dn4$subType.correction=="Tumor")]

 table(full.dn4$annotated.label.simple_akil,full.dn4$tumor.5nn.umap.simple.label)
 
 
  ## account for the tumor_11 none cluster
 full.dn4$tumor.5nn.umap.simple.label[which(is.na(full.dn4$tumor.5nn.umap.simple.label))]<-'none'

  table(full.dn4$tumor.5nn.umap.simple.label)

 table(full.dn4$tumor.5nn.umap.simple.label,full.dn4$annotated.label.simple_akil)

 
 
 ### unsup community with topology
 table(full.dn4$tumor.5nn.class.unsup)
 #full.dn4$tumor.5nn.class.unsup<-as.character(full.dn4$unsup.subcommunity)
 full.dn4$tumor.5nn.class.unsup[which(is.na(full.dn4$tumor.5nn.class.unsup))]<-'none'


  ##append the endo sub-clusters.
 full.dn4$tumor.5nn.class.unsup[which(full.dn4$unsup.subcommunity=="Endothelial_1")]<-"Endothelial_1"
 full.dn4$tumor.5nn.class.unsup[which(full.dn4$unsup.subcommunity=="Endothelial_2")]<-"Endothelial_2"
 full.dn4$tumor.5nn.class.unsup[which(full.dn4$unsup.subcommunity=="Endothelial_3")]<-"Endothelial_3"
 full.dn4$tumor.5nn.class.unsup[which(full.dn4$unsup.subcommunity=="Endothelial_4")]<-"Endothelial_4"
 full.dn4$tumor.5nn.class.unsup[which(full.dn4$unsup.subcommunity=="Endothelial_5")]<-"Endothelial_5"
 full.dn4$tumor.5nn.class.unsup[which(full.dn4$unsup.subcommunity=="Endothelial_6")]<-"Endothelial_6"

  table(full.dn4$unsup.subcommunity,full.dn4$tumor.5nn.class.unsup)%>%tail(30)
table(full.dn4$tumor.5nn.class.unsup)


```


### perform the spatial interactions family level and topology
- Perform the spatial analysis on labeled expression using 1,000 permutations and p<0.025 as the significance threshold.  can be slow.
- REVISION 10-4: this section is deprecated, the next section below is using p=0.01 threshold.
```{r, eval=FALSE}
 
  ##perform's Vito's interaction/permutations.
 # can be slow.
  dat_cells<-data.frame(ImageNumber=full.dn4$ROIID,
	ObjectNumber=full.dn4$CellId,
	label=as.character(full.dn4$tumor.5nn.umap.simple.label),
	group=full.dn4$ROIID)
  dat_cells<-data.table(dat_cells)
 
  
   dat_relation_15<-createRelationTable(dn5=full.dn4,
	myPheno="tumor.5nn.umap.simple.label",
	minDistance=15)
  
  
  save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/dat_relation_15_tumor.5nn.umap.simple.label.15distance.RData")
  
  
 #load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure6-final/dat_relation_15_tumor.5nn.umap.simple.label.15distance.RData")
## windows local
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure6-final/dat_relation_15_tumor.5nn.umap.simple.label.15distance.RData")
  
 ## using the p=0.025 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

 # save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_annotatedLabel.tumor.5nn.umap.simple.label.15distance.025.RData")
  
   save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_annotatedLabel.tumor.5nn.umap.simple.label.15distance.01.RData")
  
 ## compare with original.
```



### perform the spatial interactions cluster level and topology
- Perform the spatial analysis on labeled expression using 1,000 permutations and p<0.01 as the significance threshold.  can be slow.
- uses the topology and the unsupervised clusters of TME.
- REVISION 10-4: this has the topology with p=0.01 threshold.
```{r, eval=FALSE}
 
  ##perform's Vito's interaction/permutations.
 # can be slow.


  dat_cells<-data.frame(ImageNumber=full.dn4$ROIID,
	ObjectNumber=full.dn4$CellId,
	label=as.character(full.dn4$tumor.5nn.class.unsup),
	group=full.dn4$ROIID)
  dat_cells<-data.table(dat_cells)
  dat_relation_15<-createRelationTable(dn5=full.dn4,
	myPheno="tumor.5nn.class.unsup",
	minDistance=15)
  
  
  save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/dat_relation_15_tumor.5nn.unsup.15distance.RData")
  
  

   load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure6-final/dat_relation_15_tumor.5nn.unsup.15distance.RData")
  
 ## using the p=0.025 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

  save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")
  
 load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")

  
  
 ## compare with original.
```


## Interaction with pooled label at p=0.01 update
- 1-19 creates a heatmap with immune/tumor interactions at 0.01 alpha level.
```{r}

 dat_cells<-data.frame(ImageNumber=full.dn4$ROIID,
	ObjectNumber=full.dn4$CellId,
	label=as.character(full.dn4$annotated.label.back_akil),
	group=full.dn4$ROIID)
  dat_cells<-data.table(dat_cells)
 
  
   dat_relation_15<-createRelationTable(dn5=full.dn4,
	myPheno="annotated.label.back_akil",
	minDistance=15)
  
  
  save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/dat_relation_15_annotated.label.back_akil.15distance.RData")
  
  
 #load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure6-final/dat_relation_15_tumor.5nn.umap.simple.label.15distance.RData")
## windows local
load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure6-final/dat_relation_15_annotated.label.back_akil.15distance.RData")
  
 ## using the p=0.025 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

   save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_annotatedLabel.annotated.label.back_akil.15distance.01.RData")
  


## load the full.dn4 object with back labeling.
 full.dn4$tumor.5nn.label.back<-as.character(full.dn4$annotated.label.back_akil)
 full.dn4$tumor.5nn.label.back[which(full.dn4$subType.correction=="Tumor")]<-full.dn4$tumor.5nn.class.unsup[which(full.dn4$subType.correction=="Tumor")]

  
  
   
  

 ## load the interactions
load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_annotatedLabel.annotated.label.back_akil.15distance.01.RData")

 ##change the names.
 
 
 ## heatmap 
  interact<-p1$pmat
    interact<-interact[which(rownames(interact)!='none'),which(colnames(interact)!='none')]

 
 rownames(interact)<-full.dn4$tumor.5nn.label.back[match(rownames(interact),full.dn4$tumor.5nn.umap.simple.label)]
 colnames(interact)<-full.dn4$tumor.5nn.label.back[match(colnames(interact),full.dn4$tumor.5nn.umap.simple.label)]

 ###shorten names
 rownames(interact)<-gsub("Exhausted","Exh.",rownames(interact))
  colnames(interact)<-gsub("Exhausted","Exh.",colnames(interact))
  
  
###### supplementary 11b
  require(dendsort);require(pvclust)
  rpt.pv<-pvclust(interact[,!grepl("Tumor",colnames(interact))],nboot=100,method.hclust="average")
  rpt.dend<-dendsort(hclust(dist(interact[,!grepl("Tumor",colnames(interact))])),isReverse=TRUE)
  library(ComplexHeatmap);library(dendextend)
  
  inth<-Heatmap(interact[,!grepl("Tumor",colnames(interact))],
                  cluster_rows=rpt.dend,
                  cluster_columns=rpt.pv$hclust,
		name="Immune interaction Z-score",
		rect_gp = gpar(col = "white", lwd = 1))

   tum.pv<-pvclust(interact[,grepl("Tumor",colnames(interact))],nboot=100,method.hclust="average")
  tum.dend<-dendsort(hclust(dist(interact[,!grepl("Tumor",colnames(interact))])),isReverse=TRUE)
  library(ComplexHeatmap);library(dendextend)

 
   ### row annotate the native expression.

   immune.interact<-interact[,!grepl("Tumor",colnames(interact))]
  tumor.interact<-interact[,grepl("Tumor",rownames(interact))]

 
  col_fun <- circlize::colorRamp2(breaks = c((-3), 0, 3),
                  colors = c("black", "purple", "yellow"),
                  transparency = 0.02)



    txb.df<-data.frame(phenotype=rownames(interact))
  
    phenoColors<-getSpatialPhenoColors(full.dn4=full.dn4)
 
    umapColors<-getPhenoColorsAkilAnnotatedLabel(full.dn4)
    ##shorten names
    names(umapColors)<-gsub("Exhausted","Exh.",names(umapColors))
    
    umapColors<-data.frame(phenotype=names(umapColors),colors=umapColors,label=names(umapColors))
   umapColors<-umapColors[!grepl("tumor",rownames(umapColors)),]
  umapColors<-umapColors[!grepl("none",rownames(umapColors)),]

  spatialphenoColors<-phenoColors[grepl("Tumor_",phenoColors$phenotype),]
  ## assign spatial colors
  spatialphenoColors$label<-spatialphenoColors$phenotype
  phenoColors<-rbind(umapColors,spatialphenoColors)
  
  ## REVISION 12-12: adjust phenoColors labels to match the back labeling
 
  rownames(phenoColors)<-as.character(phenoColors$label)
  phenoColors$phenotype<-as.character(phenoColors$label)


 ## order to match interaction
 phenoColors<-phenoColors[match(rownames(immune.interact),phenoColors$label),]

   pColors<-phenoColors$colors
  names(pColors)<-phenoColors$phenotype

 txb.sel.col=list(phenotype=pColors[match(rownames(interact),names(pColors))])
  
 rowAnnLabeled <- HeatmapAnnotation(df=txb.df,
                                    which="row", 
	col=txb.sel.col,
	show_annotation_name=FALSE,
	 annotation_width=unit(c(1, 5), "cm"), 
	show_legend = FALSE,
	gap=unit(1, "mm"),
	annotation_legend_param=list(phenotype=list(ncol=3,title_position="topcenter")))
	#link = anno_mark(at =seq(1,nrow(txb.df)) ,
      #  labels = as.character(txb.df[,1]),labels_gp=gpar(fontsize=9)))

  inth.tumor<-Heatmap(interact[,grepl("Tumor",rownames(interact))],
                  cluster_rows=tum.dend,
                  cluster_columns=tum.pv$hclust,
                  right_annotation = rowAnnLabeled,
                  show_heatmap_legend=TRUE,
		name="Tumor Interaction Z-score",
		rect_gp = gpar(col = "white", lwd = 1))
 
  
  inth+inth.tumor
```


### DLBCL spatial topology analysis
- We perform the topology classification in DLBCL and classify B-cell phenotypes 
- 12-13: this is added as a reference only, this is the orignal performance/computatoin of clustering.  this is a record of the script
- 1-8: reproduce this on windows system for regenerating figure 3/4 results.
```{r, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 14}

 # load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/revision-comments-11-6/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")


 ### use the 5NN analysis approach.

 ## classify Tumor 5-NN to immune.  
 ## omit endo.
  ## use the synthetic immune/tumor classification. 
    full.dn4$compartment<-"immune" 
    full.dn4$compartment[which(full.dn4$subType.correction=="Tumor")]<-"Tumor" 
    full.dn4$compartment[which(full.dn4$subType.correction=="Endothelial")]<-"Endothelial"

   full.dn4.5nn2<-full.dn4
   full.dn4.5nn2$nndImmune.5nn<-0
   full.dn4.5nn2$nndTumor.5nn<-0
  full.dn4.5nn2$tumor.immune.5nn<-NA
  ## clusters distance neighbors for each ROI using k=40, and looks at 8-NN distances from tumor to immune
  for(roi in unique(full.dn4.5nn2$ROIID)){
  myroi<-subset(full.dn4.5nn2,full.dn4.5nn2$ROIID==roi)
   p<-phenographToPPP(case=myroi,marksCase=factor(myroi$compartment),shift=FALSE)
    myroi<-distanceClusterROI.5NN(synth=p,zz=myroi)
    full.dn4.5nn2$tumor.immune.5nn[match(myroi$uniqueLabel,full.dn4.5nn2$uniqueLabel)]<-myroi$tumor.immune.5nn
   full.dn4.5nn2$nndImmune.5nn[match(myroi$uniqueLabel,full.dn4$uniqueLabel)]<-myroi$nndImmune.5nn
  full.dn4.5nn2$nndTumor.5nn[match(myroi$uniqueLabel,full.dn4$uniqueLabel)]<-myroi$nndTumor.5nn
  }



  ErikMeta=full.dn4.5nn2[,
	c("nndImmune.5nn","nndTumor.5nn",
	"ROIID",
	"tumor.immune.5nn")] %>% group_by(ROIID,tumor.immune.5nn) %>% 
       dplyr::summarise_all(funs(mean)) %>%data.frame
     
          ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c("nndImmune.5nn","nndTumor.5nn")], k = 55)
     distance.meta.5nn.cluster<- factor(membership(erikPheno[[2]]))
     ErikMeta=cbind(ErikMeta,distance.meta.5nn.cluster)

   full.dn4.5nn2=left_join(full.dn4.5nn2, ErikMeta[,c("ROIID",
	"tumor.immune.5nn","distance.meta.5nn.cluster")], by = c("ROIID", "tumor.immune.5nn"))

  library(hrbrthemes)
 full.dn4.5nn2%>%group_by(ROIID,subType.correction,distance.meta.5nn.cluster)%>%summarise(immune=mean(nndImmune.5nn),tum=mean(nndTumor.5nn))%>%ggplot(aes(x=distance.meta.5nn.cluster,y=immune))+geom_boxplot()+facet_wrap(~subType.correction)+theme_ipsum(base_family="Arial")+ylab("Average distance to 5-NN immune cells (microns)")+xlab("distance meta cluster")
  
   #ggsave("Tumor.immune.metacluster.5NN.png")

 ##assign letters to tumor.
  full.dn4.5nn2$tumor.5nn.class<-as.character(full.dn4.5nn2$subType.correction)
 full.dn4.5nn2$tumor.5nn.class[which(full.dn4.5nn2$tumor.5nn.class=="Tumor")]<-paste0("Tumor_",letters[full.dn4.5nn2$distance.meta.5nn.cluster[which(full.dn4.5nn2$subType.correction=='Tumor')]])

 ## add the unsupervised phenotypes.
 full.dn4.5nn2$tumor.5nn.class.unsup<-as.character(full.dn4.5nn2$unsup.subcommunity)
 full.dn4.5nn2$tumor.5nn.class.unsup[which(full.dn4.5nn2$subType.correction=="Tumor")]<-full.dn4.5nn2$tumor.5nn.class[which(full.dn4.5nn2$subType.correction=="Tumor")]

 color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")
    spatCol<-getSpatialPhenoColors(full.dn4.5nn2)


```


# Clinical Sheet gather
- 1-10: identical to blood-revision paper, this section is required for the clinical regression
```{r}
 ndROI<-full.dn4$ROIID[which(full.dn4$treatment.status=='ND')]%>%unique()
treatROI<-full.dn4$ROIID[which(full.dn4$treatment.status!='ND')]%>%unique()


 c1<-c("BCL6","BCL10","TNFAIP3","UBE2A","CD70","B2M","NOTCH2","TMEM30A","FAS","TP63","ZEB2","HLAB","SPEN","PDL1")
  c1.rank<-c(1,2,3,4,5,6,7,8,9,11,12,13,14,15)
  c1.dataFrame<-data.frame(gene=c1,Rank=c1.rank,cluster='C1',contained=FALSE)
  c2<-c("TP53")
  c2.rank<-1
  c2.dataFrame<-data.frame(gene=c2,Rank=c2.rank,cluster='C2',contained=FALSE)

  c3<-c("BCL2","CREBBP","EZH2","KMT2D","TNFRSF14","HVCN1","IRF8","GNA13","MEF2B","PTEN")
  c3.rank<-c(seq(1,length(c3)))
  c3.dataFrame<-data.frame(gene=c3,Rank=c3.rank,cluster='C3',contained=FALSE)

  c4<-c("SGK1","HIST1H1E","NFKBIE","BRAF","CD83","NFKBIA","CD58","HIST1H2BC","STAT3","HIST1H1C","ZFP36L1","KLHL6","HIST1H1D","HIST1H1B","ETS1","TOX","HIST1H2AM","HIST1H2BK","RHOA","ACTB","LTB","SF3B1","CARD11","HIST1H2AC")
  c4.rank<-seq(1,length(c4))
    c4.dataFrame<-data.frame(gene=c4,Rank=c4.rank,cluster='C4',contained=FALSE)

  c5<-c("CD79B","MYD88","ETV6","PIM1","TBL1XR1","GRHPR","ZC3H12A","HLAA","PRDM1","BTG1")
  c5.rank<-c(3,5,6,8,10,12,13,16,17,18)
    c5.dataFrame<-data.frame(gene=c5,Rank=c5.rank,cluster='C5',contained=FALSE)

    
##read in the mutational data.    
     clinic<-mutationClinicalSheetAssembly(mutationsToGather=c(c1,c2,c3,c4,c5),full.dn4=full.dn4)
      any(clinic$Run.ID%in%ndROI)

      
       mutations<-read.csv("~/IMC-Ranalysis/DLBCL/monirath/myData/mutational-genomic-table/TMA CGI genomic data.csv")
  
c1.dataFrame[c1.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c2.dataFrame[c2.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c3.dataFrame[c3.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c4.dataFrame[c4.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c5.dataFrame[c5.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
clusterData<-rbind(c1.dataFrame,c2.dataFrame,c3.dataFrame,c4.dataFrame,c5.dataFrame)

   ## REVISION 10-20 hans mutation association with C1-C5.
 ### HANS association with each mutation class logistic regression. 
 clinic$C1.count<-rowSums(clinic[,colnames(clinic)%in%c1])
 clinic$C2.count<-(clinic[,colnames(clinic)%in%c2])
 clinic$C3.count<-rowSums(clinic[,colnames(clinic)%in%c3])
 clinic$C4.count<-rowSums(clinic[,colnames(clinic)%in%c4])
 clinic$C5.count<-rowSums(clinic[,colnames(clinic)%in%c5])
 ##ngcb
 clinic$NGCB.bin<-ifelse(clinic$Cell.of.origin..HANS.=='NGCB',1,0)
 clinic$Case_Number<-as.numeric(substr(clinic$Case_ID,6,7))
##chapuy binary.
  clinic$C1<-ifelse(clinic[,'C1.count']>0,1,0)
  clinic$C2<-ifelse(clinic[,'C2.count']>0,1,0)
  clinic$C3<-ifelse(clinic[,'C3.count']>0,1,0)
  clinic$C4<-ifelse(clinic[,'C4.count']>0,1,0)
  clinic$C5<-ifelse(clinic[,'C5.count']>0,1,0)

  clinic$Treatment.Category[which(clinic$Treatment.Category=="sur+chemo")]<-"SUR+CHEMO"
clinic$X9p24.gain.status[which(is.na(clinic$X9p24.gain.status))]<-0

 #clinic<-clinic[match(pData(fsimc)$sample_id,clinic$Run.ID),]
#  all(clinic$Run.ID==pData(fsimc)$sample_id)
 

```

### Figure 4B, 4C
- 12-13: final figures for the revised figure 4B, 4C.
- Revision 3-23-2022: 
```{r, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 14}

  

   absoluteTMEProportion<-function(full.dn4=NULL,tumorsToExclude=NULL,phenoType=NULL){
     ## the phenotype should be a column vector of "immune/tumor" only binary##
      
   immune<-full.dn4[!full.dn4[,phenoType]%in%tumorsToExclude,]
   roiSum<-immune%>%group_by_("case",phenoType)%>%summarise(count=n())%>%mutate(freq=count/sum(count))%>%data.frame
  roiSum<-roiSum[which(roiSum[,2]=="immune"),]
   roiSum<-roiSum[order(roiSum$freq,decreasing=FALSE),]
    return(roiSum)
   }

    
#load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/Figure5/DLBCL-RData/full.dn4.5nn2.CD8.MC.RData")
load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/full.dn4.5nn2.CD8.MC.RData")
   spatCol<-getSpatialPhenoColors(full.dn4.5nn2)
   cols<-spatCol
   cols[which(cols$phenotype=="Tumor_d"),"colors"]<-"plum4"
   
   
    full.dn4$compartment<-"immune" 
    full.dn4$compartment[which(full.dn4$subType.correction=="Tumor")]<-"Tumor" 
    full.dn4$compartment[which(full.dn4$subType.correction=="Endothelial")]<-"Endothelial"

  ## for record keeping here is the script to generate 5nn
   #library(igraph)
   #full.dn4.5nn2<-full.dn4
   #full.dn4.5nn2$nndImmune.5nn<-0
   #full.dn4.5nn2$nndTumor.5nn<-0
  #full.dn4.5nn2$tumor.immune.5nn<-NA
  ## clusters distance neighbors for each ROI using k=40, and looks at 8-NN distances from tumor to immune
  #for(roi in unique(full.dn4.5nn2$ROIID)){
  #myroi<-subset(full.dn4.5nn2,full.dn4.5nn2$ROIID==roi)
  # p<-phenographToPPP(case=myroi,marksCase=factor(myroi$compartment),shift=FALSE)
  #  myroi<-distanceClusterROI.5NN(synth=p,zz=myroi)
  #  full.dn4.5nn2$tumor.immune.5nn[match(myroi$uniqueLabel,full.dn4.5nn2$uniqueLabel)]<-myroi$tumor.immune.5nn
  # full.dn4.5nn2$nndImmune.5nn[match(myroi$uniqueLabel,full.dn4$uniqueLabel)]<-myroi$nndImmune.5nn
  #full.dn4.5nn2$nndTumor.5nn[match(myroi$uniqueLabel,full.dn4$uniqueLabel)]<-myroi$nndTumor.5nn
  #}
  

 full.dn4$immune.prop<-full.dn4.5nn2[match(full.dn4$uniqueLabel,full.dn4.5nn2$uniqueLabel),"immune.prop"]
 full.dn4$tumor.5nn.class<-full.dn4$tumor.5nn.class.unsup
  full.dn4$nndImmune.5nn<-full.dn4.5nn2[match(full.dn4$uniqueLabel,full.dn4.5nn2$uniqueLabel),"nndImmune.5nn"]

 # full.dn4.5nn2$tumor.5nn.class<-"none"
  #full.dn4.5nn2$tumor.5nn.class[grepl("Tumor",full.dn4.5nn2$tumor.5nn.class)]<-full.dn4.5nn2$
 
 
    tme<-absoluteTMEProportion(full.dn4=full.dn4.5nn2,phenoType="immune.prop")
   roiSum<-full.dn4.5nn2[which(full.dn4.5nn2$subType.correction=="Tumor"& full.dn4.5nn2$tumor.5nn.class!="none"),]%>%group_by(case,tumor.5nn.class)%>%summarise(count=n())%>%mutate(freq=count/sum(count))%>%data.frame

    ##must have the absolute ratio plot created.
    colnames(roiSum)<-c("case","Spatial.Class","count","proportion")
      roiSum$Spatial.Class<-factor(roiSum$Spatial.Class)
   roiSum$case<-factor(roiSum$case,levels=rev(as.character(tme$case)))
  cols<-getSpatialPhenoColors(full.dn4.5nn2)
  roiSum<-roiSum[which(roiSum$case!="Case_20"),]
  finalStack<-ggplot(roiSum,aes(y=proportion,x=case,fill=Spatial.Class))+geom_bar(stat='identity')+coord_flip()+scale_fill_manual(values=cols[match(levels(roiSum$Spatial.Class),cols$phenotype),"colors"])+theme_ipsum(base_family="Arial")
  ##FIGURE 4B
  print(finalStack)
   ## saving
 save(finalStack,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/finalStack_Figure4BTop.RData")

 
 
 ## FIGURE 4C
   centroidSD<- full.dn4.5nn2[which(full.dn4.5nn2$subType.correction=="Tumor"),]%>%group_by(tumor.5nn.class.unsup)%>%summarise(nndImmune.5nn=sd(nndImmune.5nn))%>%data.frame()

 di<- full.dn4.5nn2[which(full.dn4.5nn2$subType.correction=="Tumor" & full.dn4.5nn2$tumor.5nn.class!="none"),]%>%group_by(tumor.5nn.class.unsup)%>%summarise(nndImmune.5nn=mean(nndImmune.5nn))%>%data.frame()
   di[,1]<-factor(di[,1],levels=c(di[order(di[,2]),1]))
   cols[9,2]<-color_clusters[5]
  di%>%reshape2::melt()%>%ggplot(aes(x=tumor.5nn.class.unsup,y=value,fill=tumor.5nn.class.unsup))+geom_bar(stat='identity')+facet_wrap(~variable)+theme_ipsum(base_family="Arial")+theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_fill_manual(values=cols[match(levels(di[,1]),cols[,1]),"colors"])

  centroidSD<-centroidSD[which(centroidSD$tumor.5nn.class.unsup!="none"),]

    phenoColors<-getSpatialPhenoColors(full.dn4=full.dn4.5nn2)
 rownames(centroidSD)<-centroidSD$tumor.5nn.class.unsup[which(centroidSD$tumor.5nn.class.unsup!="none")]
 sdPrim<-centroidSD[,-1]
  ## confidence interval. 95%
 sdPrim<- sdPrim/sqrt(33)
  error<- qt(0.975,df=9-1)*sdPrim
   di$lower<-di[,2]-error
   di$upper<-di[,2]+error
  
 

 
  distancePlot<-ggplot(di,aes(x=tumor.5nn.class.unsup,y=nndImmune.5nn,fill=tumor.5nn.class.unsup))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.5,alpha=0.5)+theme_ipsum(base_family="Arial")+theme(axis.text.x = element_text(angle = 45, hjust = 1))+scale_fill_manual(values=phenoColors[match(levels(di$tumor.5nn.class.unsup),phenoColors$phenotype),'colors'])+xlab("Tumor spatial cluster")+ylab("Average 5-NN distance (microns)")
   ## saving for Final Figure
  # save(distancePlot,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/distancePlot_Figure4C.RData")

 #ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/Nearest.5NN.Immune.spatial.phenotype.distances.png")
 ####

  
  ### cartoon with topology colors.
   full.dn4$tumor.5nn.class.unsup<-full.dn4.5nn2$tumor.5nn.class.unsup[match(full.dn4$uniqueLabel,full.dn4.5nn2$uniqueLabel)]
   full.dn4$topology.TME<-as.character(full.dn4$rename)
   full.dn4$topology.TME[full.dn4$rename=="Tumor"]<-full.dn4$tumor.5nn.class.unsup[full.dn4$rename=="Tumor"]
   
   
    spatcols<-getSpatialPhenoColors(full.dn4=full.dn4.5nn2)
   spatcols2<-data.frame(phenotype=unique(full.dn4$topology.TME))
     spatcols2$colors<-spatcols$colors[match(spatcols2$phenotype,spatcols$phenotype)]   
     spatcols2$colors[spatcols2$phenotype=="T cytotoxic"]<-"green"
      spatcols2$colors[spatcols2$phenotype=="Endothelial"]<-"red"
      spatcols2$colors[spatcols2$phenotype=="T Helper"]<-"blue"
      spatcols2$colors[spatcols2$phenotype=="CD68 Macrophage"]<-"magenta"
            spatcols2$colors[spatcols2$phenotype=="FOXP3 T REG"]<-"purple"
      spatcols2$colors[spatcols2$phenotype=="none"]<-"azure"
      spatcols2$colors[spatcols2$phenotype=="Tumor_d"]<-"plum4"


  
	c26.top<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/cartoons/cartoon-shapeFiles/final-shapes/",
          ROI='3975',
	cols=spatcols2,
	full.dn4,
	compartment="topology.TME",
	phenotype="CD4_1",
	phenotype.column.toShape="topology.TME")
  print(c26.top)
  
  #Save for Tucker
   # save(c26.top,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/c26.top_Figure4B_Bottom.RData")

  
  

 ## Figure 4D multivariate regression.
   absol.prop<-as.data.frame.matrix(table(full.dn4.5nn2$case,full.dn4.5nn2$tumor.5nn.class.unsup))
  absol.prop<-100*absol.prop/rowSums(absol.prop)
  absol.prop$HANS<-full.dn4.5nn2$COO[match(rownames(absol.prop),full.dn4.5nn2$case)]
 absol.prop$NGCB<-ifelse(absol.prop$HANS=="NGCB",1,0)
  absol.prop$IPI<-full.dn4.5nn2$IPI[match(rownames(absol.prop),full.dn4.5nn2$case)]
  absol.prop$immune<-tme$freq[match(rownames(absol.prop),tme$case)]

  spatCol<-getSpatialPhenoColors(full.dn4.5nn2)

  tum.lr<-glm(NGCB~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)
  mlr.lr<-tum.lr%>%confint%>%data.frame
  mlr.lr$estimate<-coef(tum.lr)
   mlr.lr<-mlr.lr[order(mlr.lr$estimate),]
  mlr.lr<-mlr.lr[grepl("Tumor_",rownames(mlr.lr)),]
  mlr.lr$phenotype<-rownames(mlr.lr)
  mlr.lr$phenotype<-factor(mlr.lr$phenotype,levels=rownames(mlr.lr))

  ggplot(mlr.lr,aes(x=phenotype,y=estimate))+geom_point(size=4,aes(color=phenotype))+geom_errorbar(aes(ymin=X2.5..,ymax=X97.5..),width=0.2,color='grey39')+scale_color_manual(values=spatCol$colors[match(levels(mlr.lr$phenotype),spatCol$phenotype)])+theme_ipsum(base_family='Arial')+theme(axis.text.x = element_text(angle = 90, hjust = 1))+xlab("Spatial phenotype")+ylab("Non-GC log-odds (IPI adjusted)")+ggtitle("Spatial Phenotype Non-GC association")+geom_hline(yintercept=0,linetype='dashed',color='red')
 
#ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Tumor-analysis-5NN/mult.LR.IPI.adjusted.spatial.png")

  write.csv(mlr.lr,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-model-worksheets/logisticRegression/mlr.topology.cluster.csv")
  
### model of NGCB/GCB over  dispersed/ mantle/crust /core regions (pooled)
  
  
### we don't see strong evidence for Chapuy  
  ## model of C1-C5 over dispersed/mantle/crust/core regions  (pooled)
absol.prop$C1<-clinic$C1[match(rownames(absol.prop),clinic$Case_ID)]
absol.prop$C2<-clinic$C2[match(rownames(absol.prop),clinic$Case_ID)]
absol.prop$C3<-clinic$C3[match(rownames(absol.prop),clinic$Case_ID)]
absol.prop$C4<-clinic$C4[match(rownames(absol.prop),clinic$Case_ID)]
  absol.prop$C5<-clinic$C5[match(rownames(absol.prop),clinic$Case_ID)]
  absol.prop$CD79b<-clinic$CD79B[match(rownames(absol.prop),clinic$Case_ID)]

absol.prop$C124<-ifelse(absol.prop$C1==1 | absol.prop$C2==1| absol.prop$C4==1, 1,0)
absol.prop$C14<-ifelse(absol.prop$C1==1 |  absol.prop$C4==1, 1,0)
absol.prop$C35<-ifelse(absol.prop$C5==1 | absol.prop$C3==1, 1,0)
 c35.lr<-glm(C35~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)
 c124.lr<-glm(C124~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)
 c14.lr<-glm(C14~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)
 c5.lr<-glm(C5~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)

  mlr.lr<-c35.lr%>%confint%>%data.frame
  mlr.lr$estimate<-coef(c35.lr)
   mlr.lr<-mlr.lr[order(mlr.lr$estimate),]
  mlr.lr<-mlr.lr[grepl("Tumor_",rownames(mlr.lr)),]
  mlr.lr$phenotype<-rownames(mlr.lr)
  mlr.lr$phenotype<-factor(mlr.lr$phenotype,levels=rownames(mlr.lr))

   ggplot(mlr.lr,aes(x=phenotype,y=estimate))+geom_point(size=4,aes(color=phenotype))+geom_errorbar(aes(ymin=X2.5..,ymax=X97.5..),width=0.2,color='grey39')+scale_color_manual(values=spatCol$colors[match(levels(mlr.lr$phenotype),spatCol$phenotype)])+theme_ipsum(base_family='Arial')+theme(axis.text.x = element_text(angle = 90, hjust = 1))+xlab("Spatial phenotype")+ylab("Non-GC log-odds (IPI adjusted)")+ggtitle("Spatial Phenotype Non-GC association")+geom_hline(yintercept=0,linetype='dashed',color='red')
 
   ## null
 library(glmnet)
   x<-absol.prop[,grepl("Tumor_",colnames(absol.prop))]%>%as.matrix
   fit<-cv.glmnet(x,absol.prop$C1,family="binomial")
   
   absol.prop$Crust.Mantle<-rowMeans(absol.prop[,c("Tumor_c","Tumor_h","Tumor_i","Tumor_f")])
   absol.prop$Dispersed<-rowMeans(absol.prop[,c("Tumor_b","Tumor_a","Tumor_e","Tumor_g")])
aov(Tumor_d~C1+C2+C3+C4+C5+IPI+NGCB,absol.prop)%>%summary
aov(Dispersed~C1+C2+C3+C4+C5+IPI+NGCB,absol.prop)%>%summary
aov(Crust.Mantle~C1+C2+C3+C4+C5+IPI+NGCB,absol.prop)%>%summary

```

## Chisquare Test of CD4/MAC/CD8 and topology zones
- 1-9: counts the total signed interaction values and tests using Chisquare test  (3x2) 
- 1-9: performs a CHisquare test, and a GLM logistic model for outer/crust vs. mantle/core of interactions from the tumor perpsective (Figure 4E) OR.
- 1-9: performs a chapuy association of interactions.
```{r}


### pooling logistic regression into Dispersed/Mantle/Crust/Core ~ CD4/CD8/Mac association
   load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")
  ## use TP for chi-square test 

 
  
   ## first label is the TME toward tumor topology
###  Cd4/Cd8/MAC  x  core/mantle vs crust/outer
  
  cd4.core.mantle<- p1$pmat[grepl("CD4",rownames(p1$pmat)),c(c("Tumor_d","Tumor_i","Tumor_f"))]%>%sum   
   # CD4 (outer/crust)
  cd4.crust.outer<- p1$pmat[grepl("CD4",rownames(p1$pmat)),c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h"))]%>%sum   

  cd8.crust.outer<- p1$pmat[grepl("CD8",rownames(p1$pmat)),c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h"))]%>%sum
   cd8.core.mantle<-   p1$pmat[grepl("CD8",rownames(p1$pmat)),c(c("Tumor_d","Tumor_i","Tumor_f"))]%>%sum 

  mac.crust.outer<-p1$pmat[grepl("Macrophage",rownames(p1$pmat)),c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h"))]%>%sum 
  mac.core.mantle<-p1$pmat[grepl("Macrophage",rownames(p1$pmat)),c(c("Tumor_d","Tumor_i","Tumor_f"))]%>%sum 
 
  t.crust.outer<-p1$pmat[grepl("TREG",rownames(p1$pmat)),c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h"))]%>%sum 
  t.core.mantle<-p1$pmat[grepl("TREG",rownames(p1$pmat)),c(c("Tumor_d","Tumor_i","Tumor_f"))]%>%sum 
 
 # for epitools  
####      -   +  
####   -
###    +
 
 cd4Mat<-matrix(c(cd8.crust.outer,mac.crust.outer,cd4.crust.outer,t.crust.outer,
                  cd8.core.mantle,mac.core.mantle,cd4.core.mantle,t.core.mantle),nrow=3,ncol=2)
 
 rownames(cd4Mat)<-c("CD8","MAC","CD4","TREG")
 colnames(cd4Mat)<-c("Crust.outer","Core.Mantle")
  chisq.test(cd4Mat)

  library(epitools)
   epitools::oddsratio(cd4Mat)
   
   
   ### flipping directions.
   # topology  toward TME
    core.mantle.cd4<- p1$pmat[c(c("Tumor_d","Tumor_i","Tumor_f")),grepl("CD4",colnames(p1$pmat))]%>%sum   
   # CD4 (outer/crust)
  crust.outer.cd4<- p1$pmat[c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h")),grepl("CD4",colnames(p1$pmat))]%>%sum   

  crust.outer.cd8<- p1$pmat[c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h")),grepl("CD8",colnames(p1$pmat))]%>%sum
  core.mantle.cd8<-p1$pmat[c(c("Tumor_d","Tumor_i","Tumor_f")),grepl("CD8",colnames(p1$pmat))]%>%sum 

  crust.outer.mac<-p1$pmat[c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h")),grepl("Macrophage",colnames(p1$pmat))]%>%sum 
  core.mantle.mac<-p1$pmat[c(c("Tumor_d","Tumor_i","Tumor_f")),grepl("Macrophage",colnames(p1$pmat))]%>%sum 
 
    crust.outer.t<-p1$pmat[c(c("Tumor_a","Tumor_b","Tumor_e","Tumor_g","Tumor_c","Tumor_h")),grepl("TREG",colnames(p1$pmat))]%>%sum 
  core.mantle.t<-p1$pmat[c(c("Tumor_d","Tumor_i","Tumor_f")),grepl("TREG",colnames(p1$pmat))]%>%sum 
 
 ## ambi
 ambi.core.mantle.cd4<-sum(cd4.core.mantle,core.mantle.cd4)
 ambi.core.mantle.cd8<-sum(cd8.core.mantle,core.mantle.cd8)
 ambi.core.mantle.mac<-sum(mac.core.mantle,core.mantle.mac)
 ambi.core.mantle.t<-sum(t.core.mantle,core.mantle.t)

 ambi.crust.outer.cd4<-sum(cd4.crust.outer,crust.outer.cd4)
 ambi.crust.outer.cd8<-sum(cd8.crust.outer,crust.outer.cd8)
 ambi.crust.outer.mac<-sum(mac.crust.outer,crust.outer.mac)
 ambi.crust.outer.t<-sum(t.crust.outer,crust.outer.t)

 
 
 cd4Mat2<-matrix(c(ambi.crust.outer.cd8,ambi.crust.outer.mac,ambi.crust.outer.cd4,ambi.crust.outer.t,
                  ambi.core.mantle.cd8,ambi.core.mantle.mac,ambi.core.mantle.cd4,ambi.core.mantle.t),nrow=4,ncol=2)
 
 rownames(cd4Mat2)<-c("CD8","MAC","CD4","TREG")
 colnames(cd4Mat2)<-c("Crust.outer","Core.Mantle")
 
 ### finalized statistics that we report.
  chisq.test(cd4Mat2)

  library(epitools)
   epitools::oddsratio(cd4Mat2)
   
   
   
   

   
   
### patient regression MLR
   ## from first label as tumor is how the Figure 4E is generated.
    pat.int<-p1$dat_p[grepl("Tumor",p1$dat_p$FirstLabel),]
   pat.int$firstFam<-NA
   pat.int$firstFam[grepl("CD4",pat.int$FirstLabel)]<-"CD4"
   pat.int$firstFam[grepl("CD8",pat.int$FirstLabel)]<-"CD8"
   pat.int$firstFam[grepl("Macrophage",pat.int$FirstLabel)]<-"Macrophage"
      pat.int$firstFam[grepl("TREG",pat.int$FirstLabel)]<-"TREG"
       pat.int$firstFam[pat.int$FirstLabel%in%c("Tumor_d")]<-"Core"
      pat.int$firstFam[pat.int$FirstLabel%in%c("Tumor_i","Tumor_f")]<-"Mantle"
      pat.int$firstFam[pat.int$FirstLabel%in%c("Tumor_a","Tumor_b","Tumor_e","Tumor_g")]<-"Outer"
      pat.int$firstFam[pat.int$FirstLabel%in%c("Tumor_c","Tumor_h")]<-"Crust"


   pat.int$secondFam<-NA
   pat.int$secondFam[grepl("CD4",pat.int$SecondLabel)]<-"CD4"
   pat.int$secondFam[grepl("CD8",pat.int$SecondLabel)]<-"CD8"
   pat.int$secondFam[grepl("Macrophage",pat.int$SecondLabel)]<-"Macrophage"
      pat.int$secondFam[grepl("TREG",pat.int$SecondLabel)]<-"TREG"
   pat.int$secondFam[pat.int$SecondLabel%in%c("Tumor_d")]<-"Core"
      pat.int$secondFam[pat.int$SecondLabel%in%c("Tumor_i","Tumor_f")]<-"Mantle"
      pat.int$secondFam[pat.int$SecondLabel%in%c("Tumor_a","Tumor_b","Tumor_e","Tumor_g")]<-"Outer"
      pat.int$secondFam[pat.int$SecondLabel%in%c("Tumor_c","Tumor_h")]<-"Crust"

      
  test<-pat.int%>%group_by(group,firstFam,secondFam)%>%summarise(sigval=sum(sigval))%>%data.frame    
  test<-test[which(!is.na(test$firstFam)),]
  test$C1<-clinic$C1[match(test$group,clinic$Run.ID)]
  test$C2<-clinic$C2[match(test$group,clinic$Run.ID)]
  test$C3<-clinic$C3[match(test$group,clinic$Run.ID)]
  test$C4<-clinic$C4[match(test$group,clinic$Run.ID)]
  test$C5<-clinic$C5[match(test$group,clinic$Run.ID)]
  test$IPI<-clinic$ipi[match(test$group,clinic$Run.ID)]
  test$NGCB<-clinic$NGCB.bin[match(test$group,clinic$Run.ID)]

  test$IPI<-ifelse(test$IPI>=3,1,0)
  library(tidyverse)
  test2<-test%>%spread(secondFam,sigval)
   test2$CoreMantle<-ifelse(test2$firstFam%in%c("Core","Mantle"),1,0)

    test2$OuterCrust<-ifelse(test2$firstFam%in%c("Crust","Outer"),1,0)

  glm(CoreMantle~CD8+CD4+Macrophage+TREG+IPI,family='binomial',test2)%>%summary
  glm(CoreMantle~CD8+Macrophage,family='binomial',test2)%>%summary
confint(glm(CoreMantle~CD8+Macrophage,family='binomial',test2))%>%exp()
  
  glm(OuterCrust~CD8+CD4+Macrophage+TREG+IPI,family='binomial',test2)%>%summary
  
  glm(Mantle~CD8+CD4+Macrophage+TREG+IPI,family='binomial',test2)%>%summary

  ### Chapuy model for interactions  (we see the NULL)
   library(biostat3)
glm(C1~CD8+CD4+Macrophage+TREG+IPI+CoreMantle+NGCB,family='binomial',test2)%>%summary
 c1fit<-glm(C1~CD4+TREG+CoreMantle+NGCB+IPI,family='binomial',test2)
 lincom(c1fit,c("CD4+CoreMantle","TREG+CoreMantle"),eform=TRUE)
  

  glm(C2~CD8+CD4+Macrophage+TREG+IPI+CoreMantle+NGCB,family='binomial',test2)%>%summary
  glm(C3~CD8+CD4+Macrophage+TREG+IPI+CoreMantle+NGCB,family='binomial',test2)%>%summary
  glm(C4~CD8+CD4+Macrophage+TREG+IPI+CoreMantle+NGCB,family='binomial',test2)%>%summary
  glm(C5~CD8+CD4+Macrophage+TREG+IPI+CoreMantle+NGCB,family='binomial',test2)%>%summary
  c5fit<-glm(C5~CD4+CoreMantle+NGCB,family='binomial',test2)
 lincom(c5fit,c("CD4+CoreMantle"),eform=TRUE)
  summary(c5fit)

  

 
#### pooled proportions counts of cells OR for chapuy.
 full.dn4.5nn2$tumor.5nn.pool.unsup<-full.dn4.5nn2$tumor.5nn.class.unsup
  full.dn4.5nn2$tumor.5nn.pool.unsup[full.dn4.5nn2$tumor.5nn.pool.unsup%in%c("Tumor_i","Tumor_f")]<-"Mantle"
    full.dn4.5nn2$tumor.5nn.pool.unsup[full.dn4.5nn2$tumor.5nn.pool.unsup%in%c("Tumor_a","Tumor_b","Tumor_e","Tumor_g")]<-"Dispersed"
  full.dn4.5nn2$tumor.5nn.pool.unsup[full.dn4.5nn2$tumor.5nn.pool.unsup%in%c("Tumor_c","Tumor_h")]<-"Crust"
    full.dn4.5nn2$tumor.5nn.pool.unsup[full.dn4.5nn2$tumor.5nn.pool.unsup%in%c("Tumor_d")]<-"Core"

 absol.prop2<-as.data.frame.matrix(table(full.dn4.5nn2$case,full.dn4.5nn2$tumor.5nn.pool.unsup))

 
   absol.prop2<-100*absol.prop2/rowSums(absol.prop2)
  absol.prop2$HANS<-full.dn4.5nn2$COO[match(rownames(absol.prop2),full.dn4.5nn2$case)]
 absol.prop2$NGCB<-ifelse(absol.prop2$HANS=="NGCB",1,0)
  absol.prop2$IPI<-full.dn4.5nn2$IPI[match(rownames(absol.prop2),full.dn4.5nn2$case)]
  absol.prop2$immune<-tme$freq[match(rownames(absol.prop2),tme$case)]

  absol.prop2$C1<-clinic$C1[match(rownames(absol.prop2),clinic$Case_ID)]
  absol.prop2$C2<-clinic$C2[match(rownames(absol.prop2),clinic$Case_ID)]
  absol.prop2$C3<-clinic$C3[match(rownames(absol.prop2),clinic$Case_ID)]
  absol.prop2$C4<-clinic$C4[match(rownames(absol.prop2),clinic$Case_ID)]
  absol.prop2$C5<-clinic$C5[match(rownames(absol.prop2),clinic$Case_ID)]
  
  glm(C1~Mantle+Dispersed+Crust+Core+NGCB+IPI,family='binomial',absol.prop2)%>%summary()
   glm(C2~Mantle+Dispersed+Crust+Core+NGCB+IPI,family='binomial',absol.prop2)%>%summary()
  glm(C3~Mantle+Dispersed+Crust+Core+IPI+NGCB,family='binomial',absol.prop2)%>%summary()
  glm(C4~Mantle+Dispersed+Crust+Core+IPI+NGCB,family='binomial',absol.prop2)%>%summary()
  glm(C5~Mantle+Dispersed+Crust+Core+IPI+NGCB,family='binomial',absol.prop2)%>%summary()

 

  ### chapuy and topology clusters.
  ## we do not drop terms unless we check for effect modification.  do not subset out terms .  
     absol.prop<-as.data.frame.matrix(table(full.dn4.5nn2$case,full.dn4.5nn2$tumor.5nn.class.unsup))
 absol.prop<-100*absol.prop/rowSums(absol.prop)
  absol.prop$HANS<-full.dn4.5nn2$COO[match(rownames(absol.prop),full.dn4.5nn2$case)]
 absol.prop$NGCB<-ifelse(absol.prop$HANS=="NGCB",1,0)
  absol.prop$IPI<-full.dn4.5nn2$IPI[match(rownames(absol.prop),full.dn4.5nn2$case)]
  absol.prop$immune<-tme$freq[match(rownames(absol.prop),tme$case)]

  absol.prop$C1<-clinic$C1[match(rownames(absol.prop),clinic$Case_ID)]
  absol.prop$C2<-clinic$C2[match(rownames(absol.prop),clinic$Case_ID)]
  absol.prop$C3<-clinic$C3[match(rownames(absol.prop),clinic$Case_ID)]
  absol.prop$C4<-clinic$C4[match(rownames(absol.prop),clinic$Case_ID)]
  absol.prop$C5<-clinic$C5[match(rownames(absol.prop),clinic$Case_ID)]
  
  
  #### tests for proportion variation of each topology classes
  ### proportion of each topology class, association with immune proportion.
  test<-melt(absol.prop[,c("immune","Tumor_a","Tumor_b","Tumor_c","Tumor_d","Tumor_e","Tumor_f","Tumor_g","Tumor_h","Tumor_i")],id.vars="immune")
  aov(value~variable+immune,test)%>%summary
  #############################################################
  
  
  ## Null
  glm(C1~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%summary()
 
  ### Tumor_i (mantle) 
  glm(C2~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%summary()
   glm(C2~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%coef%>%exp() 
   confint(glm(C2~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop))%>%exp()
   
  #############################################################
  #############################################################
  #############################################################
  #############################################################
   ## revision to drop COO into the multiple model.
###  revision 3-23 
     glm(C2~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%summary()
     glm(C2~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%coef%>%exp()
     confint(glm(C2~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop))%>%exp()

### HANS association model (Tumor e associated with NGCB)
  glm(NGCB~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%summary()
   glm(NGCB~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%coef%>%exp()
  confint(glm(NGCB~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop))%>%exp()

#####       
    ## C3: Tumor_e (disper)
  glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%summary()
  
#####################################
#####################################  
#####################################  
  ### 3-23-2022 revision to drop NGCB
  ## revision reviewer asked to drop NGCB
  glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%summary()
  glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%coef%>%exp()
 confint(glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop))%>%exp()

  ## best model
  glm(C3~Tumor_a+Tumor_b+Tumor_g+Tumor_e+IPI,family='binomial',absol.prop)%>%summary()
  
 

  ###############  
  
  ## drop IPI ,  it is not an effect modifier for tumor_e
glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB,family='binomial',absol.prop)%>%summary()
glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB,family='binomial',absol.prop)%>%coef%>%exp
 confint(glm(C3~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB,family='binomial',absol.prop))%>%exp
 
 ## C4 Tumor_e (disp)   (no manipulations)
  glm(C4~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%summary()
    glm(C4~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%coef%>%exp()

confint(  glm(C4~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop))%>%exp()


#####################################
#####################################  
##################################### 
#### revision 3-23-2022
 ### C4 association
   glm(C4~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%summary()
   glm(C4~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)%>%coef%>%exp()
   confint(glm(C4~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop))%>%exp()

  
  glm(C5~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+NGCB+IPI,family='binomial',absol.prop)%>%summary()

  
  ### interaction summaries per cluster (descriptive)
     load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")
p1$pmat["Tumor_e",]

p1$pmat["Tumor_c",]

p1$pmat["Tumor_i",]


p1$pmat["Tumor_d",]


```



## Figure 4E revision proportion (%) interactions within each zone (final 4E)
 This corresponds to Figure 5G.  From the context of each tumor zone, it shows the total sum of signed interactions (p<0.05) in each tumor zone.  The tumor zones in the figure are ordered by distance to TME.  From the perspective of the tumor domains, we counted the ROIs that had significant spatial interactions (p<0.05) within 15 microns from each tumor domain.  The y-axis denotes the number of cases with significant interactions between the phenotypes, and the x-axis denotes the primary TME that was significantly attracted/repulsed to the tumor domain.
- REVISION 10-4:  include an anova to examine change across factors.
- 12-13: this is the final figure 4E version.
- 1-20:  includes beta regression on the spatial attraction neighborhood frequencies comparing act/supp/base to disper/mantle/crust/core regions.
```{r,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 14}

  # load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/tumor-spatial-class5NN-unsupCommunity/p1_15_Tumor.5nn.unsup.RData")
  load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")



   ints<- as.data.frame(p1$pmat)
 ##this counts the signed interactions from the reference of the tumor zones, and summarizes major TME level.
 ##Figure 6D
  x<-ints[grepl("Tumor_",rownames(ints)),]%>%t()%>%data.frame
  x<-x[which(rownames(x)!="none"),]
  ints<-x
  ints$primary<-sapply(strsplit(rownames(ints),"_"),function(x) x[1])
  
   tme.interact<-melt(ints[!grepl("Tumor",rownames(ints)),])
  colnames(tme.interact)<-c("TME","Topology","Interaction")

  lm(Interaction~TME*Topology,tme.interact)%>%anova

  sink("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-model-worksheets/Figure2-9-20/anova.TME.interactions.topology.txt")
  lm(Interaction~TME*Topology,tme.interact)%>%anova
  sink()
  
  
  avgPrim<-ints%>%group_by(primary)%>%summarise_all(mean)%>%data.frame
 rownames(avgPrim)<-avgPrim$primary
 avgPrim<-avgPrim[which(avgPrim$primary!="Endothelial"),grepl("Tumor",colnames(avgPrim))]
 avgPrim<-avgPrim[which(rownames(avgPrim)!="Tumor"),]
  
 
 sdPrim<-ints%>%group_by(primary)%>%summarise_all(funs(sd))%>%data.frame
 rownames(sdPrim)<-sdPrim$primary
 sdPrim<-sdPrim[which(sdPrim$primary!="Endothelial"),grepl("Tumor",colnames(sdPrim))]
 sdPrim<-sdPrim[which(rownames(sdPrim)!="Tumor"),]
  ## confidence interval. 95%
 sdPrim<- sdPrim/sqrt(ncol(sdPrim))
 error<- qt(0.975,df=ncol(sdPrim)-1)*sdPrim

 ## this is identical to firstLabel sums with Tumor spatial domain SecondLabel.
  colSums(ints[grepl("CD4",rownames(ints)),grepl("Tumor",colnames(ints))])
  colSums(ints[grepl("CD8",rownames(ints)),grepl("Tumor",colnames(ints))])
  colSums(ints[grepl("Macrophage",rownames(ints)),grepl("Tumor",colnames(ints))])
  
 ## the total CD4 (firstLabel) interactions with each tumor
 totalPrimary<-ints%>%group_by(primary)%>%summarise_all(funs(sum))%>%data.frame
 rownames(totalPrimary)<-totalPrimary$primary
 totalPrimary<-totalPrimary[which(rownames(totalPrimary)!="Tumor"),grepl("Tumor",colnames(totalPrimary))]
 
# totalPrimary<-100*totalPrimary/rowSums(totalPrimary)
 totalPrimary<-totalPrimary[which(rownames(totalPrimary)!="Endothelial"),]
 totalPrimary$primary<-rownames(totalPrimary)
  
 lower<-totalPrimary[,colnames(error)]-error
  lower$primary<-rownames(lower)
 lower<-reshape2::melt(lower)
 upper<-totalPrimary[,colnames(error)]+error
   upper$primary<-rownames(upper)
 upper<-reshape2::melt(upper)
  tp<-melt(totalPrimary)
  tp$lower<-lower$value
  tp$upper<-upper$value

 orderTop<-c("Tumor_b","Tumor_a","Tumor_e","Tumor_g","Tumor_c","Tumor_h","Tumor_i","Tumor_f","Tumor_d")
 topID<-c(which(tp$variable=='Tumor_b'),
	which(tp$variable=='Tumor_a'),
	which(tp$variable=='Tumor_e'),
	which(tp$variable=='Tumor_g'),
	which(tp$variable=='Tumor_c'),
	which(tp$variable=='Tumor_h'),
	which(tp$variable=='Tumor_i'),
	which(tp$variable=='Tumor_f'),
	which(tp$variable=='Tumor_d'))

  tp<-tp[topID,]
  tp$variable<-factor(as.character(tp$variable),levels=orderTop)


 
 ## this is the proportion averages.
 final4E<-ggplot(tp,aes(x=primary,y=value,fill=primary))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.5)+facet_wrap(~variable)+scale_fill_manual(values=c("blue","green","magenta","purple"))+theme_ipsum(base_family="Arial")+theme(axis.text.x=element_text(angle=90),axis.text=element_text(color="black"),text=element_text(size=6))+ylab("Total number of significant interactions (p<0.01)")+geom_hline(yintercept = 0,linetype='dashed',color='red',size=1)
 

 #ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/Major.phenotype.interactions.topologyReference.png")
# save(final4E,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/final4E.RData")


 ## fittinge a line chart
 lineFigure4E<-ggplot(tp[which(tp$primary!="TREG"),],aes(x=variable,y=value,fill=primary,color=primary))+geom_point(aes(fill=primary,size=6))+geom_line(aes(group=primary))+theme_ipsum()+ylab("Significant interactions (p<0.01)")+xlab("Topology region")+theme(axis.text.x=element_text(color='black',size=12),axis.text.y=element_text(color='black'))+scale_color_manual(values=c("blue","green","magenta"))+geom_hline(yintercept=0,color='red',linetype='dashed')
 
 # save(lineFigure4E,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/lineFigure4E.RData")
  tp$zone<-"dispersed"
  tp$zone[tp$variable%in%c("Tumor_c","Tumor_h")]<-"crust"
    tp$zone[tp$variable%in%c("Tumor_i","Tumor_f")]<-"mantle"
    tp$zone[tp$variable%in%c("Tumor_d")]<-"core"

     hist(tp$value)
fit<-lm(value~primary*zone,tp)
## we have interactions.
  lm(value~zone,tp[which(tp$primary=="Macrophage"),])%>%summary
  #### summarizing the interactions by the activated/ suppressive  we see the null here.
  ### summarizeing into activated/suppressive is not significant
  ## summarizing activated/suppressive x Tumor region
    ints<- as.data.frame(p1$pmat)
 ##this counts the signed interactions from the reference of the tumor zones, and summarizes major TME level.
 ##Figure 6D
    
##### freuqncy beta model.
    #### suppressive higher frequ in mantle /  core  
  x<-ints[grepl("Tumor_",rownames(ints)),]%>%t()%>%data.frame
  x<-x[which(rownames(x)!="none"),]
  ints<-x[!grepl("Tumor_",rownames(x)),]
 ints$primary<-full.dn4$annotated.label.back_akil[match(rownames(ints),full.dn4$unsup.subcommunity)]
  
 ## suppressive types:
 # Late Exhausted/Infla. CD4(1,5)
 # Exhausted/Inflam CD8 (2,6)
 # Terminal Exhaust CD8(3,5)
 # M2-MAC(1,5,6)
 #PD-L1+M2-MAC(2)
 #Highly Supp. Treg
 # EphrinB2+Endo(1)
 # PD-L1+Endo(4)
 
 ## activated types
 # act./early/exh CD492,3,7)
 # act/prolif CD4(6)
 # act./prolif CD8 (1,4)
 # M1-MAC(3,4)
 # Act./Prolif T-reg(5)
 # Prolif. Endo (2)
 # Act. Endo(3,6)
 
 ints$type<-c("supp","act","act","base","supp","act","act",
                 "act","supp","supp","act","supp","supp","base",
                 "supp","act","act","supp","base","act",
                 "supp","supp","act","act","supp","supp",
                 "supp","supp","base","base","act","supp")
 
   zone.interact<-melt(ints[!is.na(ints$type),])
  # zone.interact<-zone.interact[!grepl("T-reg",zone.interact$primary),]
  # zone.interact<-zone.interact[!grepl("Endo",zone.interact$primary),]

   zone.interact<- zone.interact%>%group_by(type,variable)%>%summarise(value=sum(value))%>%data.frame
   zone.interact$variable<-factor(zone.interact$variable,levels=levels(tp$variable))
   zone.interact$region<-'dispersed'
   zone.interact$region[zone.interact$variable%in%c("Tumor_c","Tumor_h")]<-"crust"
      zone.interact$region[zone.interact$variable%in%c("Tumor_i","Tumor_f")]<-"mantle"
         zone.interact$region[zone.interact$variable%in%c("Tumor_d")]<-"core"
         zone.interact$region<-factor(zone.interact$region,levels=c("dispersed","crust","mantle","core"))
         
    ### chisquare test of associaton between  dispersed/activation x  other/not activation.
  ## we use the chisquare distribution for simplicity in the test.       
        mats<-zone.interact%>%group_by(type,region)%>%summarise(value=sum(value))%>%data.frame
          mat3<-matrix(c( mats[which(mats$region!='mantle' & mats$type!="supp"),'value']%>%sum,
                            
                             mats[which(mats$region=='mantle' & mats$type!="supp"),'value']%>%sum,
                           mats[which(mats$region!='mantle' & mats$type=="supp"),'value']%>%sum,
                            mats[which(mats$region=='mantle' & mats$type=="supp"),'value']%>%sum),nrow=2,ncol=2)
 fisher.test(mat3)
 
 
  mat3<-matrix(c( mats[which(mats$region!='dispersed' & mats$type!="act"),'value']%>%sum,
                            
                             mats[which(mats$region=='dispersed' & mats$type!="act"),'value']%>%sum,
                           mats[which(mats$region!='dispersed' & mats$type=="act"),'value']%>%sum,
                            mats[which(mats$region=='dispersed' & mats$type=="act"),'value']%>%sum),nrow=2,ncol=2)
 fisher.test(mat3)
 
  ## note reference is first row/ first column
       epitools::oddsratio.fisher(mat3)
     

         ##not sure about the interpretation here.
  ### the beta proportions is not clear how to interpret the contrasts. 
         ##total attractions in a region
          zone.interact%>%group_by(region)%>%summarise(total=sum(value))
         zone.interact$prop<-0
  zone.interact$prop[which(zone.interact$region=="dispersed")]<-zone.interact$value[which(zone.interact$region=="dispersed")]/58
  zone.interact$prop[which(zone.interact$region=="crust")]<-zone.interact$value[which(zone.interact$region=="crust")]/33
  zone.interact$prop[which(zone.interact$region=="mantle")]<-zone.interact$value[which(zone.interact$region=="mantle")]/12
              zone.interact$prop[which(zone.interact$region=="core")]<-zone.interact$value[which(zone.interact$region=="core")]/3
#### spatial model in topology zones.
          zone.interact$prop<-ifelse(zone.interact$prop<0,0,zone.interact$prop)
          zone.interact$type<-factor(zone.interact$type,levels=c("act","supp","base"))
 library(betareg)
    ## model description
  ## the observations account for the proportions of attractions in a given region (dispersed/crust/etc) for each zone, we have a total proporiton computed, but we we compare the proportions of activated/suppressed neighborhoods in a given region.
    ## the reference groups are activated types in the dispersed zones.
    ## suppressed phenotypes have a proportion increased exp(beta) times that overall (?)
    ## the lincom statement looking at  mantle+supp  is comparing against the dispersed neighborhood proportions.  
        supFit<-betareg(prop~type+region,dplyr::filter(zone.interact,prop>0))
          #intercept is dispersed 
 library(biostat3)
        lincom(supFit,c("typesupp+regionmantle","typesupp+regioncrust","typesupp+regioncore"))
  zone.interact%>%group_by(region,type)%>%summarise(t=mean(prop))
  ## checking
        ggplot(zone.interact,aes(x=region,y=prop,fill=type))+geom_bar(stat='identity')
      
   zone.summary<-zone.interact%>%group_by(type,variable)%>%summarise(value=mean(value))%>%data.frame

   ggplot(zone.summary,aes(x=variable,y=value,color=type))+geom_point(size=6)+geom_line(aes(group=type))
    lm(Interaction~TME*Topology,tme.interact)%>%anova
    
    
       zone.sum<-zone.interact%>%group_by(type,region)%>%summarise(value=sum(value))%>%data.frame


```

### Figure 4E Revision (revised figure 4E) ANOVA
- REVISION 10-21: the summary is now over dispersed, mantle/crust,core and for each pooled phenotype. we use the pooled topologies
- REVISION 10-21: we use the linear model to compute the estimate significant attractions/repulsions as a rate, and find the 95% CI.  using Bonferroni adjustment for the 4 compartments, we multiple by 4 to adjust for 4 models fitted Disp., crust, mantle, core. 
- 12-13 Figure 4E ANOVA for interaction between topology and main compartments 
```{r}
   load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")

 ints<- as.data.frame(p1$pmat)
 ##this counts the signed interactions from the reference of the tumor zones, and summarizes major TME level.
 ##Figure 6D
  x<-ints[grepl("Tumor_",rownames(ints)),]%>%data.frame
  ## summarize by topology
  x$topology<-c("Dispersed","Dispersed","Crust","Core","Dispersed","Mantle","Dispersed","Dispersed","Mantle")
 x2<-x%>%group_by(topology)%>%summarise_all(sum)%>%data.frame
 rownames(x2)<-x2$topology
   x2<-x2[,-1]
#   x<-ints[grepl("Tumor_",rownames(ints)),]%>%t()%>%data.frame
   x2<-x2%>%t()%>%data.frame

  x2<-x2[which(rownames(x2)!="none"),]
  ints<-x2
  ints<-x2[!grepl("Tumor",rownames(ints)),]
  ints$family<-full.dn4$annotated.label.back_akil[match(rownames(ints),full.dn4$unsup.subcommunity)]
  ints$unsup.subcommunity<-factor(rownames(ints))
  subExpr<-ints
   
   
 
 	d.global <- lm(Dispersed ~ 0+family, data = subExpr)
	cr.global <- lm(Crust ~ 0+family, data = subExpr)
	man.global <- lm(Mantle ~ 0+family, data = subExpr)
	core.global <- lm(Core ~ 0+family, data = subExpr)

 	d1<-data.frame(est=coef(d.global),topology="Dispersed")
 	d1<-cbind(d1,confint(d.global))
 	d1$p<-coef(summary(d.global))[,4]
	d2<-data.frame(est=coef(cr.global),topology="Crust")
 	d2<-cbind(d2,confint(cr.global))
 	d2$p<-4*coef(summary(cr.global))[,4]

 	d3<-data.frame(est=coef(man.global),topology="Mantle")
 	d3<-cbind(d3,confint(man.global))
 	d3$p<-4*coef(summary(man.global))[,4]

 	d4<-data.frame(est=coef(core.global),topology="Core")
 	d4<-cbind(d4,confint(core.global))
 	d4$p<-4*coef(summary(core.global))[,4]

  dall<-rbind(d1,d2,d3,d4)
  dall$family<-gsub("family","",rownames(dall))
  colnames(dall)<-c("Estimate","topology","lower","upper","p","family")  
  dall$primary<-"CD4" 
  dall$primary[grepl("CD8",dall$family)]<-"CD8" 
  dall$primary[grepl("T-reg",dall$family)]<-"T-reg" 
  dall$primary[grepl("Endothelial",dall$family)]<-"Endothelial" 
  dall$primary[grepl("MAC",dall$family)]<-"Macrophage" 

  dall$family[20:76]<-substr(dall$family[20:76],1,nchar(dall$family[20:76])-1)
  dall$topology<-factor(dall$topology,levels=c("Dispersed","Crust","Mantle","Core"))
  dall$lower[which(dall$Estimate==0)]<-0
  dall$upper[which(dall$Estimate==0)]<-0
 
    phenoColors<-getPhenoColorsAnnotatedLabel(full.dn4)
   txb.df<-data.frame(Phenotype=unique(dall$family))
 txb.colors=phenoColors[match(txb.df$Phenotype,names(phenoColors))]
 
  dall$sig<-ifelse(dall$p<0.05,"yes","no")
  dall$sym<-ifelse(dall$sig=="yes","*","")
 
 ## this is the proportion averages.
  ggplot(dall,aes(x=topology,y=Estimate,fill=family, label=sym))+geom_bar(stat='identity',position='dodge')+geom_errorbar(aes(ymin=lower,ymax=upper),position=position_dodge(0.9),width=0.25,color="gray45")+facet_wrap(~primary)+scale_fill_manual(values=txb.colors)+theme_ipsum(base_family="Arial")+theme(axis.text.x=element_text(color="black"),axis.text=element_text(color="black",size=35))+ylab("Average significant interactions (p<0.01)")+geom_hline(yintercept = 0,linetype='dashed',color='red')+geom_text(position=position_dodge(0.9),size=8)

## anova
  ints<- as.data.frame(p1$pmat)
   x<-ints[grepl("Tumor_",rownames(ints)),]%>%t()%>%data.frame
  ints<-x
  ints<-x[!grepl("Tumor",rownames(ints)),]
  ints$family<-full.dn4$annotated.label.back_akil[match(rownames(ints),full.dn4$unsup.subcommunity)]
  ints$unsup.subcommunity<-factor(rownames(ints))
  ints<-ints[which(rownames(ints)!="none"),]
  ints$primary<-sapply(strsplit(as.character(ints$unsup.subcommunity),"_"),function(x) x[1])
  mints<-melt(ints,id.vars=c("family","unsup.subcommunity","primary"))
  aov(value~family*variable,mints)%>%summary
  aov(value~variable*primary,mints)%>%summary
  
  
  

```


### Create a FlowFrame of the data and the updated phenotypes with the design (Revision 5-18)
- create an IMC experiment (SummarizedExperiment) in order to generate a FlowSet for diffCyt usage.
- the goal of this analysis is to use the full clinical features to derive a linear model to show changes in the TME related to HANS, refractory response, and chapuy clusters.  The spatial clusters in the last figure should use a classification AUC/ROC comparison to show the "best" classification hopefully the spatial domains show an improved classification compared to clinical parameters, and compared to chapuy. the figure 2 TME and Figure 3 HANS/COO and treatment response should all be done using a linear model. the spatial domains in the figure 6 can be a classification comparison using 5-fold- CV.
- we subset out the ND patients.
- 12-14 identical to blood-revision-diffCyt-estimation.Rmd

```{r flowFrame}

load(file="~/IMC-Ranalysis/DLBCL/monirath/myData/revision-comments-11-6/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")

 ##compile into 1 column unsupervised.community  
  full.dn4$unsup.subcommunity<-as.character(full.dn4$unsup.subcommunity)
     full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="Endothelial")]<-paste0("Endothelial_",full.dn4$Endothelial_subphenograph_cluster_meta[which(full.dn4$subType.correction=="Endothelial")])

  full.dn4$unsup.subcommunity<-factor(full.dn4$unsup.subcommunity,
	levels=c("CD4_1",
	       "CD4_2",
		"CD4_3",
		"CD4_4",
		"CD4_5","CD4_6","CD4_7",
	        "CD8_1","CD8_2","CD8_3","CD8_4","CD8_5","CD8_6","CD8_7",
 	       "Macrophage_1","Macrophage_2","Macrophage_3","Macrophage_4","Macrophage_5",
	"Macrophage_6",
	"TREG_1","TREG_2","TREG_3","TREG_4","TREG_5",     
	 "TREG_6",
	 "Tumor_1","Tumor_2","Tumor_3", "Tumor_4", "Tumor_5","Tumor_6","Tumor_7","Tumor_8", "Tumor_9",  "Tumor_10" ,"Tumor_11",
	"Endothelial_1",	"Endothelial_2",	"Endothelial_3", 	"Endothelial_4", 	"Endothelial_5", 	"Endothelial_6"  ))




  set.seed(1234)
  table(full.dn4$phenograph_cluster_meta,full.dn4$cell.type.new)
    

  lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
	"Area.norm",
	"Eccentricity.norm",
	"Solidity.norm",
	"Perimeter.norm",
	"Percent_Touching.norm",
	"Number_Neighbors.norm")
   
   

### REVISION 5-2: Annotate cluster to cell populations 


 ## need to add labels.
  label<-full.dn4$unsup.subcommunity%>%levels%>%data.frame
 colnames(label)<-'cluster'
  label$label<-'none'
 ## CD4 labels.
  label$label[which(label$cluster=='CD4_1')]<-'Exh./Inflam.CD4(1,5)'
  label$label[which(label$cluster=='CD4_2')]<-"Act./Early/Exh.CD4(2,3,7)"
  label$label[which(label$cluster=='CD4_3')]<-'Act./Early/Exh.CD4(2,3,7)'
   label$label[which(label$cluster=='CD4_4')]<-'Baseline CD4(4)'
   label$label[which(label$cluster=='CD4_5')]<-'Exh./Inflam.CD4(1,5)'
   label$label[which(label$cluster=='CD4_6')]<-'Act./Prolif.CD4(6)'
    label$label[which(label$cluster=='CD4_7')]<-'Act./Early/Exh.CD4(2,3,7)'

 ## CD8 labels.
      label$label[which(label$cluster=='CD8_1')]<-'Act./Prolif.CD8(1,4)'
      label$label[which(label$cluster=='CD8_2')]<-'Exh./Inflam.CD8(2,6)'
      label$label[which(label$cluster=='CD8_3')]<-'Terminal Exhaust.CD8(3,5)'
      label$label[which(label$cluster=='CD8_4')]<-'Act./Prolif.CD8(1,4)'
      label$label[which(label$cluster=='CD8_5')]<-'Terminal Exhaust.CD8(3,5)'
      label$label[which(label$cluster=='CD8_6')]<-'Exh./Inflam.CD8(2,6)'
      label$label[which(label$cluster=='CD8_7')]<-'Baseline CD8(7)'
 ##exhausted
 
 ## MAC
	 label$label[which(label$cluster=='Macrophage_1')]<-'M2-MAC(1,5,6)'
        label$label[which(label$cluster=='Macrophage_2')]<-'PD-L1+M2-MAC(2)'
        label$label[which(label$cluster=='Macrophage_3')]<-'M1-MAC(3,4)'	
	 label$label[which(label$cluster=='Macrophage_4')]<-'M1-MAC(3,4)'	
        label$label[which(label$cluster=='Macrophage_5')]<-'M2-MAC(1,5,6)'
	 label$label[which(label$cluster=='Macrophage_6')]<-'M2-MAC(1,5,6)'	

 ## TREG.
 	 label$label[which(label$cluster=='TREG_1')]<-'Hi.Suppress.T-reg(1,2,6)'
 	 label$label[which(label$cluster=='TREG_2')]<-'Hi.Suppress.T-reg(1,2,6)'
 	 label$label[which(label$cluster=='TREG_3')]<-'Baseline T-reg(3,4)'
 	 label$label[which(label$cluster=='TREG_4')]<-'Baseline T-reg(3,4)'
	 label$label[which(label$cluster=='TREG_5')]<-'Act./Prolif.T-reg(5)'
 	 label$label[which(label$cluster=='TREG_6')]<-'Hi.Suppress.T-reg(1,2,6)'

 ##ENDO
	label$label[which(label$cluster=='Endothelial_1')]<-'EphrinB2+Endothelial(1)'
	label$label[which(label$cluster=='Endothelial_2')]<-'Prolif.Endothelial(2)'
	label$label[which(label$cluster=='Endothelial_3')]<-'Act.Endothelial(3,6)'
	label$label[which(label$cluster=='Endothelial_4')]<-'PD-L1+Endothelial(4)'
	label$label[which(label$cluster=='Endothelial_5')]<-'Baseline Endothelial(5)'
	label$label[which(label$cluster=='Endothelial_6')]<-'Act.Endothelial(3,6)'

 

 ## Tumor
 ## t1:CXCR3-mid, BCL2+
 ##t2: CD20+ B-cell
 #t3: BCL2+
 # t4: Ki67+, BCL6+
 # t5: Ki67+, BCL6+
 
 #t6: PD-1-moderate, Ki67+, BCL6+, BCL2+
 #t7:CD20+ B-cell
 #t8: CD20+ B-cell
 #t9: Ki67+
 #t10: Ki67+, BCL2+,BCL6+
 ## labels
   label$label[which(label$cluster=='Tumor_1')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_2')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_3')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_4')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_5')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_6')]<-'Inflammatory B-Cell tumor(6,8,10)'
   label$label[which(label$cluster=='Tumor_7')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_8')]<-'Inflammatory B-Cell tumor(6,8,10)'
   label$label[which(label$cluster=='Tumor_9')]<-'Non-Inflam. B-Cell tumor(1-5,7,9)'
   label$label[which(label$cluster=='Tumor_10')]<-'Inflammatory B-Cell tumor(6,8,10)'
      label$label[which(label$cluster=='Tumor_11')]<-'none(11)'



 label$easyLabel<-c("Late_Exhausted_CD4",
	"Act_Exh_early_CD4",
	"Act_Exh_early_CD4",
	"Baseline_CD4",
	"Late_Exhausted_CD4",
	"Act_prol_CD4",
	"Act_Exh_early_CD4",
	
	"Act_prol_CD8",
	"Exh_Inflam_CD8",
	"Term_Exh_CD8",
	"Act_prol_CD8",
	"Term_Exh_CD8",
	"Exh_Inflam_CD8",
	"Baseline_CD8",
	
	"M2MAC",
	"PDL1M2MAC",
	"M1MAC",
	"M1MAC",
	"M2MAC",
	"M2MAC",
	
	"Hi_Supp_TREG",
	"Hi_Supp_TREG",
	"Baseline_TREG",
	"Baseline_TREG",
	"Act_prol_TREG",
	"Hi_Supp_TREG",
	
	"Tumor",
	"Tumor",
	"Tumor",
	"Tumor",
	"Tumor",
	"InflamTumor",
	"Tumor",
	"InflamTumor",
	"Tumor",
	"InflamTumor",
	'none',
	"EphrinB2_Endothelial",
	"Ki67_Endothelial",
	"Activ_Endothelial",
	"PDL1_Endothelial",
	"Baseline_Endothelial",
	"Activ_Endothelial"	)
	

 ## REVISION: 5-2 we add the 'mapped-back' label with cluster names to the full.dn4 data-set.
 full.dn4$annotated.label.back_akil<-NA
 full.dn4$annotated.label.back_akil<-label$label[match(full.dn4$unsup.subcommunity,label$cluster)]
 
 #full.dn4$annotated.label<-NA
 #full.dn4$annotated.label<-label$label[match(full.dn4$unsup.subcommunity,label$cluster)]
 
 
 ## simple label
 full.dn4$annotated.label.simple_akil<-NA
 full.dn4$annotated.label.simple_akil<-label$easyLabel[match(full.dn4$unsup.subcommunity,label$cluster)]

 ##order factors.
# full.dn4$annotated.label<-factor(full.dn4$annotated.label,levels=c("CD45RA+ exhausted CD4","CD45RA- exhausted CD4","Ki67+ memory CD4","Memory CD4","CD45RA+ CD4","CD8","Exhausted CD8","Ki67+ CD8","Endothelial","Ki67+ M1-Macrophage","M1-Macrophage","M2-Macrophage","PD-L1+ M2-Macrophage","Highly suppressive T-reg","Ki67+ T-reg","T-reg","B-Cell tumor","PDL1+ B-Cell tumor"))

 
 	 

 
 full.dn4$annotated.label.back_akil<-factor(full.dn4$annotated.label.back_akil,levels=c("Exh./Inflam.CD4(1,5)",
                                                                         "Act./Early/Exh.CD4(2,3,7)",
                                                                         "Baseline CD4(4)",
                                                                         "Act./Prolif.CD4(6)",
                                                                         
                                                                         "Baseline CD8(7)",
                                                                         "Act./Prolif.CD8(1,4)",
                                                                         "Exh./Inflam.CD8(2,6)",
                                                                         "Terminal Exhaust.CD8(3,5)",
                                                                         
                                                                         "M1-MAC(3,4)",
                                                                         "M2-MAC(1,5,6)",
                                                                         "PD-L1+M2-MAC(2)",
                                                                         
                                                                         "Hi.Suppress.T-reg(1,2,6)",
                                                                         "Act./Prolif.T-reg(5)",
                                                                         "Baseline T-reg(3,4)",
                                                                         
                                                                         "Non-Inflam. B-Cell tumor(1-5,7,9)",
                                                                         "Inflammatory B-Cell tumor(6,8,10)",
                                                                         'none(11)',
                                                                          "EphrinB2+Endothelial(1)",
                                                                         "Prolif.Endothelial(2)",
                                                                         "Act.Endothelial(3,6)",
                                                                         "PD-L1+Endothelial(4)",
                                                                         "Baseline Endothelial(5)"
                                                                         ))
 
 print(label)

##labels are updated to match Dr. Merchant's labeling clusters.
 label$subcluster<-c("Exh./Inflam.",
                     "Act./Early/Exh.",
                     "Act./Early/Exh.",
                     "Baseline",
                     "Exh./Inflam.",
                     "Act./Prolif.",
                     "Act./Early/Exh.",
                     
                      "Act./Prolif.",
                     "Exh./Inflam.", 
                     "Terminal Exhaust.",
                      "Act./Prolif.",
                     "Terminal Exhaust.",
                     "Exh./Inflam.",
                     "Baseline",
                     
                     "M2",
                     "PD-L1+ M2",
                     "M1",
                     "M1",
                     "M2",
                     "M2",
                     
                     "Hi Suppr.",
                     "Hi Suppr.",
                     "Baseline",
                     "Baseline",
                       "Act./Prolif.",
                     "Hi Suppr.",
                     
                     "Non-Inflam.",
                      "Non-Inflam.",
                      "Non-Inflam.",
                      "Non-Inflam.",
                      "Non-Inflam.",
                     "Inflam.",
                      "Non-Inflam.",
                      "Inflam.",
                      "Non-Inflam.",
                     "Inflam.",
                       "none",
                     "EphrinB2+",
                     "Prolif.",
                     "Act.",
                     "PD-L1+",
                     "Baseline",
                     "Act."
                       )
 
 label$subcluster_annotated<-paste0(label$cluster,"(",label$subcluster,")")
 ##need to create a heatmap with the labels appended 5-2.
 full.dn4$subcluster_annotated<-'none'
 full.dn4$subcluster_annotated<-label$subcluster_annotated[match(full.dn4$unsup.subcommunity,label$cluster)]

 full.dn4$subcluster_annotated<-factor(full.dn4$subcluster_annotated,
                                       levels=label$subcluster_annotated
 )
 
table(full.dn4$annotated.label.back_akil%>%is.na) 
table(full.dn4$subcluster_annotated%>%is.na)


 
##subset out ND only treated.
#full.dn4<-full.dn4[which(full.dn4$treatment.status!='ND'),]
 
ndROI<-full.dn4$ROIID[which(full.dn4$treatment.status=='ND')]%>%unique()
treatROI<-full.dn4$ROIID[which(full.dn4$treatment.status!='ND')]%>%unique()

##we filter dim CD20 for modeling here
full.dn4<-full.dn4[which(full.dn4$dimCD20.filtered=="NO"),]

##proteins that are Min/Max normalized.
  expr<-full.dn4[,1:31]
  normExp<-as.matrix(expr)

  
  ##spatial component
  spatial<-(full.dn4[,c("X_position","Y_position")])
  spatial<-as.matrix(spatial)
 ##uniqueLabel
  uniqueLabel<-full.dn4$uniqueLabel
  ## Akil labeled phenotypes.
  phenotypes<-(as.data.frame(factor(full.dn4$annotated.label.simple_akil)))
  ROIID=data.frame(ROIID=full.dn4[,"ROIID"])
  morph<-full.dn4[,c("Area",
                    "Eccentricity",
	"Solidity",
	"Perimeter")]
  
  colnames(normExp)<-gsub("Cell_","",colnames(normExp))
  colnames(normExp)[which(colnames(normExp)=="p")]<-"pSTAT3"
    colnames(normExp)[which(colnames(normExp)=="C")]<-"cMYC"

  
  x<-imcExperiment(cellIntensity=t(normExp),
	coordinates=spatial,
	neighborHood=as.matrix(full.dn4[,"Number_Neighbors"]),
	network=phenotypes,
	distance=matrix(1,nrow=nrow(full.dn4),ncol=10),
	morphology=morph,
	panel=colnames(normExp),
	uniqueLabel=uniqueLabel,
	ROIID=ROIID)
   x
   
   
  ## SCE
   
   
imcData<-x
### DiffCyt.
  is(assay(imcData,'counts'),'matrix')
  rownames(imcData)
    # for plot scatter to work need to set the rowData feature in a specific way.
  channel<-c("Nd146",
             "Sm147",
             "Dy161",
             "Tm169",
             "Er170",
             "Ho165",
             "Nd144",
             "Gd156",
             "Gd155",
             "Yb173",
             "Tb159",
             "Dy162",
             "Er166",
             "Dy163",
             "Yb174",
             "Dy164",
             "Sm149",
             "Eu151",
             "Nd142",
             "Er167",
             "N148",
             "Er168",
             "Eu153",
             "Lu175",
             "Nd150",
             "Yb172",
             "Nd145",
             "Sm154",
             "Nd143",
             "Dy160",
             "Gd158")
          
   ##remove the "Cell_" tag from the data naming.   
     marker<-rownames(x)

  
   channel.markers<-DataFrame(channel_name=channel, marker_name=marker)
     
   
    rowData(imcData)<-DataFrame(channel_name=channel,marker_name=marker)
    rownames(imcData)<-marker
    plotScatter(imcData,rownames(imcData)[17:18],assay='counts')

          
            
            
              # convert to flowSet
             ## the warning has to do with duplicated Iridium channels.
   table(colData(imcData)$ROIID)
       (fsimc <- sce2fcs(imcData, split_by = "ROIID"))
    ## now we have a flowSet.
   pData(fsimc)
   fsApply(fsimc,nrow)
   dim(exprs(fsimc[[1]]))
   exprs(fsimc[[1]])[1:5,1:5]
    ## set up the metadata files.
  marker_info<-data.frame(channel_name=channel,
              marker_name=marker,
              marker_class=c(rep("type",14),
                             rep("state",17)))
                             

 pData(fsimc)$sample_id<-sapply(strsplit(pData(fsimc)$name,"_"),function(x) x[2])
 ##need the experimental information containing, HANS, C1-C5
 
  ##CHAPUY REVISION : 10-15 extending all the chapuy mutations.
  c1<-c("BCL6","BCL10","TNFAIP3","UBE2A","CD70","B2M","NOTCH2","TMEM30A","FAS","TP63","ZEB2","HLAB","SPEN","PDL1")
  c1.rank<-c(1,2,3,4,5,6,7,8,9,11,12,13,14,15)
  c1.dataFrame<-data.frame(gene=c1,Rank=c1.rank,cluster='C1',contained=FALSE)
  c2<-c("TP53")
  c2.rank<-1
  c2.dataFrame<-data.frame(gene=c2,Rank=c2.rank,cluster='C2',contained=FALSE)

  c3<-c("BCL2","CREBBP","EZH2","KMT2D","TNFRSF14","HVCN1","IRF8","GNA13","MEF2B","PTEN")
  c3.rank<-c(seq(1,length(c3)))
  c3.dataFrame<-data.frame(gene=c3,Rank=c3.rank,cluster='C3',contained=FALSE)

  c4<-c("SGK1","HIST1H1E","NFKBIE","BRAF","CD83","NFKBIA","CD58","HIST1H2BC","STAT3","HIST1H1C","ZFP36L1","KLHL6","HIST1H1D","HIST1H1B","ETS1","TOX","HIST1H2AM","HIST1H2BK","RHOA","ACTB","LTB","SF3B1","CARD11","HIST1H2AC")
  c4.rank<-seq(1,length(c4))
    c4.dataFrame<-data.frame(gene=c4,Rank=c4.rank,cluster='C4',contained=FALSE)

  c5<-c("CD79B","MYD88","ETV6","PIM1","TBL1XR1","GRHPR","ZC3H12A","HLAA","PRDM1","BTG1")
  c5.rank<-c(3,5,6,8,10,12,13,16,17,18)
    c5.dataFrame<-data.frame(gene=c5,Rank=c5.rank,cluster='C5',contained=FALSE)

    
##read in the mutational data.    
     clinic<-mutationClinicalSheetAssembly(mutationsToGather=c(c1,c2,c3,c4,c5),full.dn4=full.dn4)
      any(clinic$Run.ID%in%ndROI)
  
  ## Table for which mutations were found or not found.
   #  mutations<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/mutational-genomic-table/TMA CGI genomic data.csv",header=TRUE)
mutations<-read.csv("~/IMC-Ranalysis/DLBCL/monirath/myData/mutational-genomic-table/TMA CGI genomic data.csv")
  
c1.dataFrame[c1.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c2.dataFrame[c2.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c3.dataFrame[c3.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c4.dataFrame[c4.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c5.dataFrame[c5.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
clusterData<-rbind(c1.dataFrame,c2.dataFrame,c3.dataFrame,c4.dataFrame,c5.dataFrame)

   ## REVISION 10-20 hans mutation association with C1-C5.
 ### HANS association with each mutation class logistic regression. 
 clinic$C1.count<-rowSums(clinic[,colnames(clinic)%in%c1])
 clinic$C2.count<-(clinic[,colnames(clinic)%in%c2])
 clinic$C3.count<-rowSums(clinic[,colnames(clinic)%in%c3])
 clinic$C4.count<-rowSums(clinic[,colnames(clinic)%in%c4])
 clinic$C5.count<-rowSums(clinic[,colnames(clinic)%in%c5])
 ##ngcb
 clinic$NGCB.bin<-ifelse(clinic$Cell.of.origin..HANS.=='NGCB',1,0)
 clinic$Case_Number<-as.numeric(substr(clinic$Case_ID,6,7))
##chapuy binary.
  clinic$C1<-ifelse(clinic[,'C1.count']>0,1,0)
  clinic$C2<-ifelse(clinic[,'C2.count']>0,1,0)
  clinic$C3<-ifelse(clinic[,'C3.count']>0,1,0)
  clinic$C4<-ifelse(clinic[,'C4.count']>0,1,0)
  clinic$C5<-ifelse(clinic[,'C5.count']>0,1,0)

  clinic$Treatment.Category[which(clinic$Treatment.Category=="sur+chemo")]<-"SUR+CHEMO"
clinic$X9p24.gain.status[which(is.na(clinic$X9p24.gain.status))]<-0

 clinic<-clinic[match(pData(fsimc)$sample_id,clinic$Run.ID),]

  all(clinic$Run.ID==pData(fsimc)$sample_id)
 
##match the column data of the fsimc 
   exper_info<-data.frame(NGCB=ifelse(clinic$Cell.of.origin..HANS.[match(pData(fsimc)$sample_id,clinic$Run.ID)]=='NGCB',1,0),
                          group_id=clinic$Response.category[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                          C1=clinic$C1[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                          C2=clinic$C2[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                          C3=clinic$C3[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                          C4=clinic$C4[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                          C5=clinic$C5[match(pData(fsimc)$sample_id,clinic$Run.ID)],                                                                  patient_id=clinic$Case_ID[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        highIPI=ifelse(clinic$ipi[match(pData(fsimc)$sample_id,clinic$Run.ID)]>=3,1,0),
                        high9p24.gain=ifelse(clinic$X9p24.gain.status[match(pData(fsimc)$sample_id,clinic$Run.ID)]>1,1,0),
                        highcmyc.category=ifelse(clinic$cmyc.category[match(pData(fsimc)$sample_id,clinic$Run.ID)]>=3,1,0),
                        myc.rearrange=ifelse(clinic$MYC.Rearrange[match(pData(fsimc)$sample_id,clinic$Run.ID)]=="pos",1,0),
                        nonHisp=ifelse(clinic$Ethnicity[match(pData(fsimc)$sample_id,clinic$Run.ID)]!="Hisp",1,0),
                        BCL6=clinic$BCL6[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        BCL2=clinic$BCL2[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        TP53=clinic$TP53[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        SGK1=clinic$SGK1[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        CD79b=clinic$CD79B[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        MYD88L265P=clinic$MYD88.L265P[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        LDH=ifelse(clinic$ldh[match(pData(fsimc)$sample_id,clinic$Run.ID)]=="H",1,0),
                        Gender=ifelse(clinic$Gender[match(pData(fsimc)$sample_id,clinic$Run.ID)]=="M",1,0),
                        AgeHigh=ifelse(clinic$Age[match(pData(fsimc)$sample_id,clinic$Run.ID)]>60,1,0),
                        BurkittLite=ifelse(clinic$Morphologic.variants[match(pData(fsimc)$sample_id,clinic$Run.ID)]!="CB",1,0),
                        FOXP1Score=ifelse(clinic$FOXP1.expression.score[match(pData(fsimc)$sample_id,clinic$Run.ID)]>9,1,0),
                        myc.gain=ifelse(clinic$MYC.Gain[match(pData(fsimc)$sample_id,clinic$Run.ID)]=="pos",1,0),
                        p53=ifelse(clinic$p53[match(pData(fsimc)$sample_id,clinic$Run.ID)]>=2,1,0),
                        SurgeryOnly=ifelse(clinic$Treatment.Category[match(pData(fsimc)$sample_id,clinic$Run.ID)]!="SUR+CHEMO",1,0),
                        ##adding mutation groups?
                        ##C1
                        CD70=clinic$CD70[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        NOTCH2=clinic$NOTCH2[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        #C3
                        CREBBP=clinic$CREBBP[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        EZH2=clinic$EZH2[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        KMT2D=clinic$KMT2D[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        IRF8=clinic$IRF8[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        ##C4
                        HIST1H1E=clinic$HIST1H1E[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        STAT3=clinic$STAT3[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        HIST1H1B=clinic$HIST1H1B[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        ETS1=clinic$ETS1[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        ##C5
                        PIM1=clinic$PIM1[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        TBL1XR1=clinic$TBL1XR1[match(pData(fsimc)$sample_id,clinic$Run.ID)],
                        ##stage and nodal involvement
                        Stage34=ifelse(clinic$Stage_g[match(pData(fsimc)$sample_id,clinic$Run.ID)]=="IV"| clinic$Stage_g[match(pData(fsimc)$sample_id,clinic$Run.ID)]=="III",1,0),
                        NodalInvolve=ifelse(clinic$Extra.nodal.involvement[match(pData(fsimc)$sample_id,clinic$Run.ID)]==">1",1,0),
                        PDL1High.ihc=ifelse(clinic$OVERALL.PD.L1.IHC.SCORE==3 &clinic$RNAScope.PDL1.score..0.4.>1,1,0),
                        X9p24.gain.status=ifelse(clinic$X9p24.gain.status>=2,1,0),
                        doubleExpressor=ifelse(clinic$bcl2==1 & clinic$cmyc.category>=3,1,0),
                                                sample_id=pData(fsimc)$name)


    ### we assign missing phenotypes to the average.
   
    cluster.counts<-as.data.frame.matrix(table(full.dn4$ROIID,full.dn4$annotated.label.simple_akil))
for(j in 1:ncol(cluster.counts)){
  id<-which(cluster.counts[,j]==0)
  id2<-cluster.counts[,j]!=0
  if(length(id)>0){
    cluster.counts[id,j]<-round(mean(cluster.counts[id2,j]))
  }
}
    
  cluster.counts<-100*cluster.counts/rowSums(cluster.counts)  

cluster.counts$Case_ID<-clinic$Case_ID[match(rownames(cluster.counts),clinic$Run.ID)]

   cluster.counts<-left_join(cluster.counts,clinic[!duplicated(clinic$Case_ID),c("Case_ID","Case_Number","C1.count", "C2.count", "C3.count", "C4.count", "C5.count",'ipi','NGCB.bin','Cell.of.origin..HANS.','Response.category')],by='Case_ID')
cluster.counts$C1<-ifelse(cluster.counts$C1.count!=0,1,0)
cluster.counts$C2<-ifelse(cluster.counts$C2.count!=0,1,0)
cluster.counts$C3<-ifelse(cluster.counts$C3.count!=0,1,0)
cluster.counts$C4<-ifelse(cluster.counts$C4.count!=0,1,0)
cluster.counts$C5<-ifelse(cluster.counts$C5.count!=0,1,0)
 cluster.counts$IPI.median.cut<-ifelse(cluster.counts$ipi>median(cluster.counts$ipi),1,0)
cluster.counts$HANS<-factor(cluster.counts[,'Cell.of.origin..HANS.'])
cluster.counts$NGCB.binary<-ifelse(cluster.counts$HANS=='NGCB',1,0)
cluster.counts$Case_Number<-factor(cluster.counts$Case_Number)
 cluster.counts$IPI.median.cut<-ifelse(cluster.counts$ipi>median(cluster.counts$ipi),1,0)


 
### dropping ND subjects
 x2<-selectCases(x,as.character(clinic$Run.ID[which(clinic$chemo!='ND')]))
  any(colData(x2)$ROIID%>%unique%in%ndROI)  
   sapply(strsplit(colnames(x2),"_"),function(x) x[1])%in%ndROI %>%any 
    

     sce<-SingleCellExperiment(x2)
     exper_info2<-exper_info[!exper_info$sample_id%in%paste0("ROIID_",ndROI),]
     exper_info2$sample_id%in%paste0("ROIID_",ndROI)%>%any



           ###measure CPM.
  # tum.pheno<-c("InflamTumor","Tumor")
  # tme.counts<-t(cluster.counts[,c(1:14,16:18)])
  # y <- edgeR::DGEList(counts=tme.counts,group=cluster.counts$HANS,samples=cluster.counts$Case_ID)
  # y<-edgeR::estimateDisp(y,model.matrix(~C1+C2+C3+C4+C5+IPI.median.cut+NGCB.bin+Response.category,cluster.counts))
  # y<-edgeR::calcNormFactors(y)
   
  
     
     
  ##can run diffcyt on SCE with pre-determined clusters of interest.
   ##need to complete the SCE.
     ##metadata needs to be a list!
   metadata(sce)<-list(experiment_info=exper_info2)
   rowData(sce)<-DataFrame(channel_name=channel, marker_name=marker,marker_class=c(rep("type",14),
                             rep("state",17)))
  ##assign colData (each cell, this is cell level)
   colD<-DataFrame(id=colnames(sce),
                   sample_id=paste0("ROIID_",sapply(strsplit(colnames(sce),"_"),function(x) x[1])))
   
   ##REVISION 8-30 updating column data.
   
   colD$HANS<-ifelse(exper_info2$NGCB[match(colD$sample_id,exper_info2$sample_id)]==1,"NGCB","GCB")
   #   colD$group_id<-metadata(sce)$group_id[match(colD$sample_id,metadata(sce)$sample_id)]
   colD$group_id<-exper_info2$group_id[match(colD$sample_id,exper_info2$sample_id)]

#   colD$C1<-metadata(sce)$C1[match(colD$sample_id,metadata(sce)$sample_id)]
    colD$C1<-(exper_info2$C1[match(colD$sample_id,exper_info2$sample_id)])

  #colD$C2<-metadata(sce)$C2[match(colD$sample_id,metadata(sce)$sample_id)]
  # colD$C3<-metadata(sce)$C3[match(colD$sample_id,metadata(sce)$sample_id)]
  # colD$C4<-metadata(sce)$C4[match(colD$sample_id,metadata(sce)$sample_id)]
  # colD$C5<-metadata(sce)$C5[match(colD$sample_id,metadata(sce)$sample_id)]
       colD$C2<-(exper_info2$C2[match(colD$sample_id,exper_info2$sample_id)])
    colD$C3<-(exper_info2$C3[match(colD$sample_id,exper_info2$sample_id)])
    colD$C4<-(exper_info2$C4[match(colD$sample_id,exper_info2$sample_id)])
    colD$C5<-(exper_info2$C5[match(colD$sample_id,exper_info2$sample_id)])

# colD$patient_id<- metadata(sce)$experiment_info$patient_id[match(colD$sample_id,metadata(sce)$sample_id)]
 #  colD$patient_id<- ifelse(metadata(sce)$experiment_info$highIPI [match(colD$sample_id,metadata(sce)$sample_id)]==1,"HighIPI","LowIPI")
    colD$patient_id<-exper_info2$patient_id[match(colD$sample_id,exper_info2$sample_id)]
    colD$HighIPI<-ifelse(exper_info2$highIPI[match(colD$sample_id,exper_info2$sample_id)]==1,"HighIPI","LowIPI")

   colData(sce)<-colD
   colData(sce)$sample_id<-factor(colData(sce)$sample_id)
   
  table(metadata(sce)$experiment_info[,"SurgeryOnly"]) 

   assays(sce, withDimnames = FALSE)<-SimpleList(exprs=(assays(x2)$counts))
 ## uses asinh(x/5) as default
  sce<-transformData(sce)
  
   ##SOM clustering
   sce <- cluster(sce)  
    # view all available clustering
  #names(cluster_codes(sce))

## input the meta clustering from phenograph into cluster_id
   ## REVISION: 12-14: use the TOPOLOGY cluster_id as the unsupervised clusters.
   ##this revision changed the DIFFCYT model, not sure why. reverting.
  colData(sce)$cluster_id<-factor(full.dn4$tumor.5nn.class.unsup[match(colData(sce)$id,full.dn4$uniqueLabel)])
  metadata(sce)$cluster_codes<-data.frame(custom=factor(levels(colData(sce)$cluster_id),levels=levels(colData(sce)$cluster_id)  ) )

   table(colData(sce)$cluster_id==full.dn4$tumor.5nn.class.unsup[match(colData(sce)$id,full.dn4$uniqueLabel)])
  # colData(sce)$cluster_id<-factor(full.dn4$annotated.label.simple_akil[match(colData(sce)$id,full.dn4$uniqueLabel)])
  # metadata(sce)$cluster_codes<-data.frame(custom=factor(levels(colData(sce)$cluster_id),levels=levels(colData(sce)$cluster_id)  ) )

   
   
   
#colData(sce)$HANS<-metadata(sce)$experiment_info$NGCB[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]
colData(sce)$REF<-metadata(sce)$experiment_info$group_id[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]
colData(sce)$REF<-factor(colData(sce)$REF,levels=c("Primary refract","CR"))
#colData(sce)$C1<-metadata(sce)$experiment_info$C1[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]
#colData(sce)$C2<-metadata(sce)$experiment_info$C2[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]
#colData(sce)$C3<-metadata(sce)$experiment_info$C3[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]
#colData(sce)$C4<-metadata(sce)$experiment_info$C4[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]
#colData(sce)$C5<-metadata(sce)$experiment_info$C5[match(colData(sce)$sample_id,metadata(sce)$experiment_info$sample_id)]


# access specific clustering resolution
table(cluster_ids(sce, "custom"))
#plotAbundances(sce, k = "custom", by = "sample_id", group_by = "HANS")

plotExprHeatmap(sce, features = "type",
    by = "cluster_id", k = "custom",
    scale = "first", q = 0.01, perc = TRUE, col_dend = FALSE)

plotMultiHeatmap(sce, 
    hm1 = "type", hm2 = "state", 
    k = "custom", 
    col_dend = c(FALSE, TRUE))


##standardize
  zz<- zScorePatientExpression(full.dn4=full.dn4,markers=c(1:31,104:111),ROIID.column=32)
 ### calculate any expression differences across sub-clusters and treatment response.
 
  ## ignore Tumor_11 cluster
   zz<-zz[which(zz$unsup.subcommunity!="Tumor_11"),]

  zz$annotated.label.simple_akil<-factor(zz$annotated.label.simple_akil)
  
  subExpr<-zz%>%group_by(ROIID,annotated.label.simple_akil)%>%summarise(CXCR3=mean(Cell_CXCR3),
	CCR4=mean(Cell_CCR4),
	BCL2=mean(Cell_BCL2),
	BCL6=mean(Cell_BCL6),
	CD30=mean(Cell_CD30),
	Ki67=mean(Cell_Ki67),
	cMyc=mean(Cell_C),
	PDL1=mean(Cell_PDL1),
	TIM3=mean(Cell_Tim3),
	LAG3=mean(Cell_Lag3),
	PD1=mean(Cell_PD1),
	ICOS=mean(Cell_ICOS),
	Vista=mean(Cell_Vista),
	CD45RO=mean(Cell_CD45RO),
	CD45RA=mean(Cell_CD45RA),
	CD206=mean(Cell_CD206),
	Granzym=mean(Cell_Granzym),
	Vimentin=mean(Cell_Vimentin),
	Tbet=mean(Cell_Tbet),
	NND.tumor=mean(NND.Tumor),
	HLADR=mean(Cell_HLADR),
	cMYC=mean(Cell_C),
	CD68=mean(Cell_CD68),
	CD206=mean(Cell_CD206),
	pSTAT3=mean(Cell_p))%>%data.frame

  subExpr$annotated.label.simple_akil<-droplevels(subExpr$annotated.label.simple_akil)

  
  subExpr$HANS<-factor(clinic$Cell.of.origin..HANS.[match(subExpr$ROIID,clinic$Run.ID)])
    subExpr$REF<-factor(clinic$Response.category[match(subExpr$ROIID,clinic$Run.ID)])
  subExpr$C1<-factor(clinic$C1[match(subExpr$ROIID,clinic$Run.ID)])
  subExpr$C2<-factor(clinic$C2[match(subExpr$ROIID,clinic$Run.ID)])
  subExpr$C3<-factor(clinic$C3[match(subExpr$ROIID,clinic$Run.ID)])
  subExpr$C4<-factor(clinic$C4[match(subExpr$ROIID,clinic$Run.ID)])
  subExpr$C5<-factor(clinic$C5[match(subExpr$ROIID,clinic$Run.ID)])
    subExpr$highIPI<-factor(  ifelse(clinic$ipi[match(subExpr$ROIID,clinic$Run.ID)]>=3,1,0))

    
     

```

# Cluster merging and pooling (identifcal block in Blood-revision-estimation)
- pool the clusters into family groups in the SCE using the full naming labels at publication quality.
- 12-14 identical to blood-revision-diffCyt-estimation.Rmd
- 1-9: we have to filer the dimCD20 cluster
```{r}
  all(full.dn4$tumor.5nn.class.unsup[match(sce$id,full.dn4$uniqueLabel)]==sce$cluster_id)
  metadata(sce)$cluster_codes[,"merging1"]<-NULL
 annoLabel<-metadata(sce)$cluster_codes
 annoLabel$new_cluster<-"uncharacterized"
 annoLabel$new_cluster<-as.character(full.dn4$annotated.label.back_akil[match(annoLabel$custom,full.dn4$tumor.5nn.class.unsup)])
 annoLabel$new_cluster[annoLabel$custom%in%c("Tumor_a","Tumor_b","Tumor_e","Tumor_g")]<-"Dispersed"
  annoLabel$new_cluster[annoLabel$custom%in%c("Tumor_c","Tumor_h","Tumor_i","Tumor_f")]<-"Mantle/Crust"
 annoLabel$new_cluster[annoLabel$custom%in%c("Tumor_d")]<-"Core"

 annoLabel<-annoLabel[,c("custom","new_cluster")]

  sce <- mergeClusters(sce, k = "custom", 
    table = annoLabel, id = "topopool", overwrite=TRUE)


  

 ### topology at cluster level
   all(full.dn4$tumor.5nn.class.unsup[match(sce$id,full.dn4$uniqueLabel)]==sce$cluster_id)
   
  metadata(sce)$cluster_codes[,"merging2"]<-NULL
 annoLabel<-metadata(sce)$cluster_codes
 annoLabel$new_cluster<-"uncharacterized"
 annoLabel$new_cluster<-as.character(full.dn4$annotated.label.back_akil[match(annoLabel$custom,full.dn4$tumor.5nn.class.unsup)])
 annoLabel$new_cluster[grepl("tumor",annoLabel$new_cluster)]<-as.character(annoLabel$custom[grepl("Tumor",annoLabel$custom)])
  
 annoLabel<-annoLabel[,c("custom","new_cluster")]

  sce <- mergeClusters(sce, k = "custom", 
    table = annoLabel, id = "topocluster", overwrite=TRUE)



 
 

```
    
     

  

### Cluster Level Analysis
- we show heterogeneity of cluster abundance changes in the figure.
- We include the Non-Treated participants, 3967, 4046, 4059
- we control for IPI, double expressor, and gender. refractory may not be included in the model (?)
- the analysis performed a repeated measures linear model.
- we contrast each tumor effect by the average chapuy effect of the other classes to identify significance.  Results: marginal significance for tumor_e (C2) and tumor_g(C4)
```{r}


   outdir<-"C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/tumor-topology"

   writeToFile<-FALSE
   minCells=50
   minSamp=4
   
  design <- createDesignMatrix(exper_info2, cols_design = c("NGCB", "group_id",
 "C1","C2","C3","C4","C5",
  "highIPI",
 "Gender",
  "doubleExpressor",
"patient_id"))

design<-design[,which(colSums(design)>1)]
 



  pool.res_IPI.LDH2 <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
    plot=TRUE,
    #contrast=createContrast(c(0,0,0,0,0,0,0,1,0,0,rep(0,7))),
   contrast=createContrast(c(0,0,0,0,0,0,0,0,1,0,0,rep(0,7))),
    verbose = FALSE)
 
 
    diffcyt::topTable(pool.res_IPI.LDH2,format_vals = TRUE,top_n=10,show_logFC =TRUE)
    

 if(writeToFile){
    
  
    addWorksheet(wb, "HighIPI Main Effect")
     writeData(wb, "HighIPI Main Effect", 
               diffcyt::topTable(pool.res_IPI.LDH2,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)

      
    }


 #plotAbundanceCPMDiffCyt(d_counts=res_IPI.LDH2$d_counts,cluster="TREG_5",marker=NULL,clinicalFactor="highIPI")

        
##states edgeR
      pool.state_IPI.LDH2 <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    plot=TRUE,
    contrast=createContrast(c(0,0,0,0,0,0,0,0,1,0,0,rep(0,7))),
    verbose = TRUE)
 
 
    diffcyt::topTable(pool.state_IPI.LDH2,format_vals = TRUE,top_n=10,show_logFC =TRUE)
    
    if(writeToFile){
         addWorksheet(wb, "HighIPI Main Effect (States)")
     writeData(wb, "HighIPI Main Effect (States)", 
               diffcyt::topTable(pool.state_IPI.LDH2,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
     
    }
  

############################################################
#  Double Expressor    
#######################################################
##DE !=0 vs. others not contrast
  
 
 ###################
## EDGE R
 ##
 ########################
  
 ###################
## EDGE R
 ## trends should match the Mixed model.
 ########################
    pool.ngcb_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,0,rep(0,7))),
    verbose = FALSE)
      diffcyt::topTable(pool.ngcb_DA,format_vals = TRUE,top_n=10, show_logFC = TRUE)
  
     if(writeToFile){
         addWorksheet(wb, "NGCB Main Effect")
     writeData(wb, "NGCB Main Effect", 
               diffcyt::topTable(pool.ngcb_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
     
         }
      
      # aNG<- plotAbundanceCPMDiffCyt(d_counts=ngcb_DA$d_counts,cluster="CD4_6",marker=NULL,clinicalFactor="doubleExpressor")
      # aNG2<- plotAbundanceCPMDiffCyt(d_counts=ngcb_DA$d_counts,cluster="Tumor_3",marker=NULL,clinicalFactor="doubleExpressor")
      
 #grid.arrange(aNG,aNG2,nrow=1,ncol=2)
 
 
  #### double expressor joint effect with IPI.
 
  pool.joint_ngcb_IPI_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,1/2,0,0,0,0,0,0,1/2,0,0,rep(0,7))),
    verbose = FALSE)
      diffcyt::topTable(pool.joint_ngcb_IPI_DA,format_vals = TRUE,top_n=10, show_logFC = TRUE)
  
      if(writeToFile){
     
      addWorksheet(wb, "NGCB+IPI Main Effect")
     writeData(wb, "NGCB+IPI Main Effect", 
               diffcyt::topTable(pool.joint_ngcb_IPI_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
        
         }
 
 
 ##convergence issues with mixed, so repeated measures is used here.
# markersTest<-rownames(rowData(sce))%in%c("PDL1","PDL2","Tim3","PD1","Lag3","ICOS","Vimentin","CXCR3","CCR4","Vista")
 pool.ngcb_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
   # formula = formula,
   design=design,
  #markers_to_test = markersTest,
    contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,0,rep(0,7))),
    verbose = FALSE)
 diffcyt::topTable(pool.ngcb_DS,show_logFC = TRUE)
   
 if(writeToFile){

    addWorksheet(wb, "NGCB Main Effect States")
     writeData(wb, "NGCB Main Effect States", 
               diffcyt::topTable(pool.ngcb_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
   
   }
 

    ##joint average effect of IPI and REF
   
    pool.joint_ngcb_IPI_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
   # formula = formula,
   design=design,
  #markers_to_test = markersTest,
    contrast=createContrast(c(0,1/2,0,0,0,0,0,0,1/2,0,0,rep(0,7))),
    verbose = FALSE)
 diffcyt::topTable(pool.joint_ngcb_IPI_DS,show_logFC = TRUE)
 
   if(writeToFile){
 
     addWorksheet(wb, "NGCB+IPI Main Effect States")
     writeData(wb, "NGCB+IPI Main Effect States", 
               diffcyt::topTable(pool.joint_ngcb_IPI_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
  }
 
   #plotDiffHeatmap(sce, rowData(joint_ngcb_IPI_DS$res), top_n = 5, fdr = 0.05, lfc=0)


#########################################


##refractory  effects contrasted against age, sex, ethnicity
## refractory is not associated with age, gender, ethnicity
## refractory is association with IPI and NGCB (we average those)
## refractory is marginal associated with LDH.
 

 
     
############# repeated measures.
     pool.ref_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    #contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    contrast=createContrast(c(0,0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    verbose = FALSE)
     
      topTable(pool.ref_DA,format_vals = TRUE,top_n=10,show_logFC = TRUE)

               if(writeToFile){
            addWorksheet(wb, "REF Main Effect")
     writeData(wb, "REF Main Effect", 
               diffcyt::topTable(pool.ref_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
                 
                        }
 

## REF, IPI joint effect.
   pool.joint_ref_IPI_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    #contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    contrast=createContrast(c(0,0,1/2,0,0,0,0,0,1/2,0,0,rep(0,7))),
    verbose = FALSE)
     
      topTable(pool.joint_ref_IPI_DA,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
 # plotAbundances(sce, k = "merging1",group_by='REF', by = "cluster_id")
    
      
                     if(writeToFile){
    
      addWorksheet(wb, "REF+IPI Main Effect")
     writeData(wb, "REF+IPI Main Effect", 
               diffcyt::topTable(pool.joint_ref_IPI_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
      
             }
##states.
  pool.ref_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
  design=design,
 # markers_to_test = markersTest,
    contrast=createContrast(c(0,0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    verbose = FALSE)
      diffcyt::topTable(pool.ref_DS,format_vals = TRUE,top_n=10,show_logFC = TRUE)
    
       if(writeToFile){
         addWorksheet(wb, "REF Main Effect States")
     writeData(wb, "REF Main Effect States", 
               diffcyt::topTable(pool.ref_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
}
           


   ### JOINT EFFECT OF IPI AND REF
 pool.joint_ref_IPI_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
  design=design,
 # markers_to_test = markersTest,
    contrast=createContrast(c(0,0,1/2,0,0,0,0,0,1/2,0,0,rep(0,7))),
    verbose = FALSE)
      diffcyt::topTable(pool.joint_ref_IPI_DS,format_vals = TRUE,top_n=15,show_logFC = TRUE)
    
       if(writeToFile){
         addWorksheet(wb, "REF+IPI Main Effect States")
     writeData(wb, "REF+IPI Main Effect States", 
               diffcyt::topTable(pool.joint_ref_IPI_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
       }   

     # plotDiffHeatmap(sce2, rowData(pool.joint_ref_IPI_DS$res), top_n = 10, fdr = 0.05,lfc=0)

  
  
##joint IPI, REF, COO  
pool.joint_ref_NGCB_IPI_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    #contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    contrast=createContrast(c(0,1/3,1/3,0,0,0,0,0,1/3,0,0,rep(0,7))),
    verbose = FALSE)
     
      topTable(pool.joint_ref_NGCB_IPI_DA,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
       if(writeToFile){
       addWorksheet(wb, "REF+IPI+NGCB Main Effect")
     writeData(wb, "REF+IPI+NGCB Main Effect", 
               diffcyt::topTable(pool.joint_ref_NGCB_IPI_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)
       }       
   pool.joint_ref_NGCB_IPI_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
  design=design,
 # markers_to_test = markersTest,
    contrast=createContrast(c(0,1/3,1/3,0,0,0,0,0,1/3,0,0,rep(0,7))),
    verbose = FALSE)
      diffcyt::topTable(pool.joint_ref_NGCB_IPI_DS,format_vals = TRUE,top_n=30,show_logFC = TRUE)%>%as.data.frame
    
    
        
                 if(writeToFile){
         addWorksheet(wb, "REF+IPI+NGCB Main Effect states")
     writeData(wb, "REF+IPI+NGCB Main Effect states", 
               diffcyt::topTable(pool.joint_ref_NGCB_IPI_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }  
      
   
   # plotDiffHeatmap(sce2, rowData(pool.joint_ref_NGCB_IPI_DS$res), top_n = 10, fdr = 0.05,lfc=0)
  
         
         
##double expressor.
    ##joint IPI, REF, COO  
pool.de_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    #contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    contrast=createContrast(c(0,0,0,0,0,0,0,0,0,0,1,rep(0,7))),
    verbose = FALSE)
     
      topTable(pool.de_DA,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
                 if(writeToFile){
        addWorksheet(wb, "DE Main Effect")
     writeData(wb, "DE Main Effect", 
               diffcyt::topTable(pool.de_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)  
                 }         
  
       
  pool.joint_de_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    #contrast=createContrast(c(0,1,0,0,0,0,0,0,0,0,rep(0,7))),
    contrast=createContrast(c(0,0,0,0,0,0,0,0,1/2,0,1/2,rep(0,7))),
    verbose = FALSE)
     
      topTable(pool.joint_de_DA,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
                 if(writeToFile){
       addWorksheet(wb, "DE+IPI Main Effect")
     writeData(wb, "DE+IPI Main Effect", 
               diffcyt::topTable(pool.joint_de_DA,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)  
                 }       
       
       
              
###############################################################
# CHAPUY CLUSTER ANALYSIS
###############################################################

 ## C1- (c2+C3+C4+C5)-(1/4)(IPI+Ethnic+Gender+surgeryOnly)
   
 
###repeated measures.
   pool.C1_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
   contrast=createContrast(c(0,0,0,1,-1,-1,-1,-1,0,0,0,rep(0,7))),
    verbose = FALSE)
   
      topTable(pool.C1_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)

                           if(writeToFile){
       addWorksheet(wb, "C1.vs.All Main Effect")
     writeData(wb, "C1.vs.All Main Effect", 
               diffcyt::topTable(pool.C1_DA_RM,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)  
                           }
      
      
   #aNG<- plotAbundanceCPMDiffCyt(d_counts=C1_DA_RM$d_counts,cluster="Baseline_CD4",marker=NULL,clinicalFactor="C1")
    

## Joint effect of C1, IPI. (NULL)
  pool.C1_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
   contrast=createContrast(c(0,0,0,1,0,0,0,0,1,0,0,rep(0,7))),
    verbose = FALSE)
   
      topTable(pool.C1_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
 #plotAbundances(sce,k="custom",by="sample_id",group_by="C1")      
      
 
               if(writeToFile){
      addWorksheet(wb, "C1+IPI Main Effect")
     writeData(wb, "C1+IPI Main Effect", 
               diffcyt::topTable(pool.C1_DA_RM,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)  
               }      
 #aNG<- plotAbundanceCPMDiffCyt(d_counts=C1_DA_RM$d_counts,cluster="Baseline_CD4",marker=NULL,clinicalFactor="C1")
    

  
 
 
 ##examine the states.
   pool.C1_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
    contrast=createContrast(c(0,0,0,1,-1,-1,-1,-1,0,0,0,rep(0,7))),
    verbose = FALSE)
      diffcyt::topTable(pool.C1_DS,format_vals = TRUE,top_n=10, show_logFC = TRUE)
      
 if(writeToFile){
           addWorksheet(wb, "C1.vs.all Main Effect states")
     writeData(wb, "C1.vs.all Main Effect states", 
               diffcyt::topTable(pool.C1_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE)  
 }
      
   #as<-  plotMedianExpressionBoxplot(d_medians=C1_DS$d_medians,cluster="Hi_Supp_TREG",marker="Vimentin",clinicalFactor="C1")
   #bs<-  plotMedianExpressionBoxplot(d_medians=C1_DS$d_medians,cluster="M1MAC",marker="HLADR",clinicalFactor="C1")
  #   grid.arrange(as,bs, nrow=1)
 
  
 
### repeat measures.
 #repeated measures.
   pool.C2_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
  contrast=createContrast(c(0,0,0,-1,1,-1,-1,-1,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C2_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
      
 if(writeToFile){
        addWorksheet(wb, "C2.vs.all Main Effect")
     writeData(wb, "C2.vs.all Main Effect", 
               diffcyt::topTable(pool.C2_DA_RM,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
 }
      
       pool.C2_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topocluster",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
  contrast=createContrast(c(0,0,0,-1,1,-1,-1,-1,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C2_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
      
   #ac1<- ggplot(lcpm,aes(x=factor(C2),y=PDL1M2MAC))+geom_boxplot()+ggtitle("PDL1M2MAC")

 
  ##repeat measures.
  #examine the states.
   pool.C2_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
   # markers_to_test = markersTest,
    contrast=createContrast(c(0,0,0,-1,1,-1,-1,-1,0,0,0,rep(0,7))),
      verbose = FALSE)
      diffcyt::topTable(pool.C2_DS,format_vals = TRUE,top_n=20, show_logFC = TRUE)
      
      
               if(writeToFile){
          addWorksheet(wb, "C2.vs.all Main Effect states")
     writeData(wb, "C2.vs.all Main Effect states", 
               diffcyt::topTable(pool.C2_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
               }          
 #as<-  plotMedianExpressionBoxplot(d_medians=C2_DS$d_medians,cluster="Tumor",marker="Ki67",clinicalFactor="C2")
#   bs<-  plotMedianExpressionBoxplot(d_medians=C2_DS$d_medians,cluster="Hi_Supp_TREG",marker="Ki67",clinicalFactor="C2")
#  cs<-  plotMedianExpressionBoxplot(d_medians=C2_DS$d_medians,cluster="Endothelial",marker="Ki67",clinicalFactor="C2")
#   grid.arrange(as,bs,cs, nrow=2)


  
###repeat measures.
  pool.C3_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
   # contrast=createContrast(c(0,0,-1/4,-1/4,1,-1/4,-1/4,0,0,0,rep(0,7))),
    contrast=createContrast(c(0,0,0,-1,-1,1,-1,-1,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C3_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
      
                 if(writeToFile){
     addWorksheet(wb, "C3.vs.all Main Effect")
     writeData(wb, "C3.vs.all Main Effect", 
               diffcyt::topTable(pool.C3_DA_RM,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }
      
      
    # a<-ggplot(n_cells[which(n_cells$cluster_id=="Baseline_TREG"),],aes(x=factor(C3),y=n))+geom_boxplot()+ggtitle("Baseline TREG") 
    #   b<- ggplot(n_cells[which(n_cells$cluster_id=="Baseline_CD4"),],aes(x=factor(C3),y=n))+geom_boxplot()+ggtitle("Baseline CD4") 

       
 


### 
##repeat measures.
  #examine the states.
   pool.C3_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design=design,
   # markers_to_test = markersTest,
    contrast=createContrast(c(0,0,0,-1,-1,1,-1,-1,0,0,0,rep(0,7))),
      verbose = FALSE)
      diffcyt::topTable(pool.C3_DS,format_vals = TRUE,top_n=10, show_logFC = TRUE)
      
                 if(writeToFile){
         addWorksheet(wb, "C3.vs.all Main Effect states")
     writeData(wb, "C3.vs.all Main Effect states", 
               diffcyt::topTable(pool.C3_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }
      
      
#   as<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Act_Exh_early_CD4",marker="Granzym",clinicalFactor="C3")
#   bs<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Act_Exh_early_CD4",marker="CXCR3",clinicalFactor="C3")
#  cs<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Act_prol_TREG",marker="Granzym",clinicalFactor="C3")
#  ds<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Term_Exh_CD8",marker="Granzym",clinicalFactor="C3")
#   grid.arrange(as,bs,cs, ds, nrow=2)


##C4  
 
 
 ##repeat measure
  pool.C4_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
    # contrast=createContrast(c(0,0,-1/4,-1/4,-1/4,1,-1/4,0,0,0,rep(0,7))),
     contrast=createContrast(c(0,0,0,-1,-1,-1,1,-1,0,0,0,rep(0,7))),
    verbose = FALSE)
  
      topTable(pool.C4_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
                 if(writeToFile){
       addWorksheet(wb, "C4.vs.all Main Effect")
     writeData(wb, "C4.vs.all Main Effect", 
               diffcyt::topTable(pool.C4_DA_RM,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 } 

 
 ##examine the states C4.
  ###repeat measure
  pool.C4_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1,-1,-1,1,-1,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C4_DS,format_vals = TRUE,top_n=10,show_logFC = TRUE) 
      
      
                 if(writeToFile){
          addWorksheet(wb, "C4.vs.all Main Effect states")
     writeData(wb, "C4.vs.all Main Effect states", 
               diffcyt::topTable(pool.C4_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }         
     
 #        as<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Act_Exh_early_CD4",marker="CCR4",clinicalFactor="C4")
#   bs<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Tumor",marker="CXCR3",clinicalFactor="C4")
#  cs<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="Hi_Supp_TREG",marker="ICOS",clinicalFactor="C4")
#  ds<-  plotMedianExpressionBoxplot(d_medians=C3_DS$d_medians,cluster="InflamTumor",marker="Tbet",clinicalFactor="C4")
#   grid.arrange(as,bs,cs, ds, nrow=2) 
  
###C5
      
 
 ##repeat measures.
  pool.C5_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
    contrast=createContrast(c(0,0,0,-1,-1,-1,-1,1,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C5_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
      
               if(writeToFile){
                addWorksheet(wb, "C5.vs.all Main Effect")
     writeData(wb, "C5.vs.all Main Effect", 
               diffcyt::topTable(pool.C5_DA_RM,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
               } 
 ##examine the states C2.
   pool.C5_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1,-1,-1,-1,1,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C5_DS,format_vals = TRUE,top_n=10,show_logFC = TRUE) 
  
                   if(writeToFile){
         addWorksheet(wb, "C5.vs.all Main Effect states")
     writeData(wb, "C5.vs.all Main Effect states", 
               diffcyt::topTable(pool.C5_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                   }      
      
  #       as<-  plotMedianExpressionBoxplot(d_medians=C5_DS$d_medians,cluster="PDL1M2MAC",marker="Tbet",clinicalFactor="C5")
  # bs<-  plotMedianExpressionBoxplot(d_medians=C5_DS$d_medians,cluster="Tumor",marker="Tim3",clinicalFactor="C5")
  #cs<-  plotMedianExpressionBoxplot(d_medians=C5_DS$d_medians,cluster="PDL1M2MAC",marker="Vista",clinicalFactor="C5")
#   grid.arrange(as,bs,cs,  nrow=2) 
 
 
   
   
   
  ###Chapuy Comparison C1,C2,C5 vs C3 C4
   
     pool.C125_C34_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,1,0,1,1,-1,-1,1,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C125_C34_DS,format_vals = TRUE,top_n=10,show_logFC = TRUE) 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C125.vs.all Main states")
     writeData(wb, "C125.vs.all Main states", 
               diffcyt::topTable(pool.C125_C34_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
     ## at end 
                 }
      
  
  ### REVISED AKIL REVISION        
     pool.C235_C14_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/2,1/3,1/3,-1/2,1/3,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C235_C14_DA,format_vals = TRUE,top_n=20,show_logFC = TRUE) 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C235.vs.C14.Main DA")
     writeData(wb, "C235.vs.C14.Main DA", 
               diffcyt::topTable(pool.C235_C14_DA,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }
      
      
     pool.C235_C14_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/2,1/3,1/3,-1/2,1/3,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C235_C14_DS,format_vals = TRUE,top_n=20,show_logFC = TRUE) 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C235.vs.C14.all Main DS")
     writeData(wb, "C235.vs.C14.all Main DS", 
               diffcyt::topTable(pool.C235_C14_DS,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }  
      
      
      
     ##REVISED AKIL REVISION   35 vs 124      
     pool.C35_C124_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/3,-1/3,1/2,-1/3,1/2,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C35_C124_DA,format_vals = TRUE,top_n=20,show_logFC = TRUE) 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C35.vs.C124.Main DA")
     writeData(wb, "C35.vs.C124.Main DA", 
               diffcyt::topTable(pool.C35_C124_DA,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
                 }
      
      
     pool.C35_C124_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/3,-1/3,1/2,-1/3,1/2,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C35_C124_DS,format_vals = TRUE,top_n=20,show_logFC = TRUE) 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C35.vs.C124.all Main DS")
     writeData(wb, "C35.vs.C124.all Main DS", 
               diffcyt::topTable(pool.C35_C124_DS,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
             }  

      
        

    ##ABC-vs-DLBCL  
   pool.C35_C14_DA <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/2,0,1/2,-1/2,1/2,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C35_C14_DA,format_vals = TRUE,top_n=20,show_logFC = TRUE)%>%data.frame 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C35.vs.C14.all Main DS")
     writeData(wb, "C35.vs.C14.all Main DS", 
               diffcyt::topTable(pool.C35_C14_DS,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
             }        
      
      
      
   pool.C35_C14_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/2,0,1/2,-1/2,1/2,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C35_C14_DS,format_vals = TRUE,top_n=20,show_logFC = TRUE)%>%data.frame 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C35.vs.C14.all Main DS")
     writeData(wb, "C35.vs.C14.all Main DS", 
               diffcyt::topTable(pool.C35_C14_DS,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
             }    
      
      pool.C5_C1_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1,0,0,0,1,1,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C5_C1_DS,format_vals = TRUE,top_n=20,show_logFC = TRUE)%>%data.frame 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C5.vs.C1.ABC.all Main DS")
     writeData(wb, "C5.vs.C1.ABC.all Main DS", 
               diffcyt::topTable(pool.C5_C1_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
             }    
      
      pool.C3_C4_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,0,0,1,-1,0,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C3_C4_DS,format_vals = TRUE,top_n=20,show_logFC = TRUE)%>%data.frame 
  
      
                 if(writeToFile){
      addWorksheet(wb, "C5.vs.C1.NGCB.all Main DS")
     writeData(wb, "C5.vs.C1.NGCB.all Main DS", 
               diffcyt::topTable(pool.C3_C4_DS,format_vals = TRUE,top_n=10,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 
             }      
      
      
   pool.C2_C14_DS <- diffcyt(sce,
                     clustering_to_use = "topopool",
    analysis_type = "DS", 
    method_DS = "diffcyt-DS-limma",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    design = design,
    contrast=createContrast(c(0,0,0,-1/2,1,0,-1/2,0,0,0,0,rep(0,7))),
    verbose = FALSE)
      topTable(pool.C2_C14_DS,format_vals = TRUE,top_n=20,show_logFC = TRUE)%>%data.frame 
          
      if(writeToFile){
      addWorksheet(wb, "C2.vs.C14.all Main DS")
     writeData(wb, "C2.vs.C14.all Main DS", 
               diffcyt::topTable(pool.C2_C14_DS,format_vals = TRUE,top_n=20,show_logFC =TRUE),
               startCol = 2,
               startRow = 3,
               rowNames = TRUE) 

             }    
      
      ## at end 
     
      ## topology cluster level analysis.
       pool.C1_DA_RM <- diffcyt(sce,
                     clustering_to_use = "topocluster",
    analysis_type = "DA", 
    method_DA = "diffcyt-DA-edgeR",
    min_cells=50,
    min_samples=4,
    normalize=TRUE,
    #formula = formula,
    design=design,
   contrast=createContrast(c(0,0,0,1,0,0,0,0,1,0,0,rep(0,7))),
    verbose = FALSE);topTable(pool.C1_DA_RM,format_vals = TRUE,top_n=10,show_logFC = TRUE)
 
 
      
      
      
      
      
  if(writeToFile){
 saveWorkbook(wb,paste0(outdir,"/Figure2_topology_pooled_model_cofactor_5.xlsx"),overwrite=TRUE)
  }
      

```





### Figure 6A Spatial analysis of annotated phenotypes in TME (deprecated)
- Construct the interaction heatmap and spatial interaction association with expression and mutation.  This produces figure 6A 
- Revision 12-11: we use the back labeling to the figure.
- this uses the linear model adjusted for IPI, gender and doubleExpressor and chapuy.
```{r spatialLabel,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 17}

 ## load the full.dn4 object with back labeling.
 full.dn4$tumor.5nn.label.back<-as.character(full.dn4$annotated.label.back_akil)

  full.dn4$tumor.5nn.label.back[which(full.dn4$subType.correction=="Tumor")]<-full.dn4$tumor.5nn.class.unsup[which(full.dn4$subType.correction=="Tumor")]

  
  
  
   ##simple labels
   full.dn4$tumor.5nn.label.back.simple<-as.character(full.dn4$annotated.label.simple_akil) 
     full.dn4$tumor.5nn.label.back.simple[which(full.dn4$subType.correction=="Tumor")]<-full.dn4$tumor.5nn.class.unsup[which(full.dn4$subType.correction=="Tumor")]

   
  

     table(full.dn4$tumor.5nn.label.back.simple,full.dn4$tumor.5nn.label.back )

   

  
   
  

 ## load the interactions
load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_annotatedLabel.tumor.5nn.umap.simple.label.15distance.025.RData")

 ##change the names.
 
 
 ## heatmap 
  interact<-p1$pmat
    interact<-interact[which(rownames(interact)!='none'),which(colnames(interact)!='none')]

 
 rownames(interact)<-full.dn4$tumor.5nn.label.back[match(rownames(interact),full.dn4$tumor.5nn.umap.simple.label)]
 colnames(interact)<-full.dn4$tumor.5nn.label.back[match(colnames(interact),full.dn4$tumor.5nn.umap.simple.label)]

 ###shorten names
 rownames(interact)<-gsub("Exhausted","Exh.",rownames(interact))
  colnames(interact)<-gsub("Exhausted","Exh.",colnames(interact))
###### supplementary 11b
  
  
  ## compute the heatmap on the labeled phenotypes.
 require(dendsort);require(pvclust)
  rpt.pv<-pvclust(interact[,!grepl("Tumor",colnames(interact))],nboot=100,method.hclust="average")
  rpt.dend<-dendsort(hclust(dist(interact[,!grepl("Tumor",colnames(interact))])),isReverse=TRUE)
  library(ComplexHeatmap);library(dendextend)
  
  inth<-Heatmap(interact[,!grepl("Tumor",colnames(interact))],
                  cluster_rows=rpt.dend,
                  cluster_columns=rpt.pv$hclust,
		name="Immune interaction Z-score",
		rect_gp = gpar(col = "white", lwd = 1))

 tum.pv<-pvclust(interact[,grepl("Tumor",colnames(interact))],nboot=100,method.hclust="average")
  tum.dend<-dendsort(hclust(dist(interact[,!grepl("Tumor",colnames(interact))])),isReverse=TRUE)
  library(ComplexHeatmap);library(dendextend)

 
   ### row annotate the native expression.

   immune.interact<-interact[,!grepl("Tumor",colnames(interact))]
  tumor.interact<-interact[,grepl("Tumor",rownames(interact))]

 
  col_fun <- circlize::colorRamp2(breaks = c((-3), 0, 3),
                  colors = c("black", "purple", "yellow"),
                  transparency = 0.02)

  zz<- zScorePatientExpression(full.dn4=full.dn4,markers=c(1:31,104:111),ROIID.column=32)
  	#interact<-p1$pmat
    # interact<-cd8.tumor.p2$pmat
 # nd.omit<-unique(zz$ROIID[which(zz$treatment.status=="ND")])
 input<-zz[,c(1:31,which(colnames(zz)=='tumor.5nn.label.back'))]%>%group_by(tumor.5nn.label.back)%>%summarise_all(funs(mean))%>%data.frame
   colnames(input)<-gsub("Cell_","",colnames(input))
  input$tumor.5nn.label.back<-factor(input$tumor.5nn.label.back)
  input$tumor.5nn.label.back<-droplevels(input$tumor.5nn.label.back)
 colnames(input)[which(colnames(input)=='C')]<-'cMYC'
 colnames(input)[which(colnames(input)=='p')]<-'pSTAT3'
 colnames(input)[which(colnames(input)=='Tim3')]<-'TIM-3'
 colnames(input)[which(colnames(input)=='Vista')]<-'VISTA'
  colnames(input)[which(colnames(input)=='PDL1')]<-'PD-L1'
   colnames(input)[which(colnames(input)=='PD1')]<-'PD-1'
   colnames(input)[which(colnames(input)=='PDL2')]<-'PD-L2'
  colnames(input)[which(colnames(input)=='Lag3')]<-'LAG-3'
  input$tumor.5nn.label.back<-droplevels(input$tumor.5nn.label.back)

 
 #rownames(input)<-full.dn4$tumor.5nn.umap.relabel[match(input$Var1,full.dn4$tumor.5nn.umap.simple.label)]
 rownames(input)<-input$tumor.5nn.label.back
  rownames(input)<-gsub("Exhausted","Exh.",rownames(input))

 
  input<-input[,-1]	
  input<-input[match(rownames(interact),rownames(input)),]
   
   X<-input
  # inth+inth.tumor
 ## IPI scores
    txb.df<-data.frame(phenotype=rownames(interact))
  
    phenoColors<-getSpatialPhenoColors(full.dn4=full.dn4)
 
    umapColors<-getPhenoColorsAkilAnnotatedLabel(full.dn4)
    ##shorten names
    names(umapColors)<-gsub("Exhausted","Exh.",names(umapColors))
    
    umapColors<-data.frame(phenotype=names(umapColors),colors=umapColors,label=names(umapColors))
   umapColors<-umapColors[!grepl("tumor",rownames(umapColors)),]
  umapColors<-umapColors[!grepl("none",rownames(umapColors)),]

  spatialphenoColors<-phenoColors[grepl("Tumor_",phenoColors$phenotype),]
  ## assign spatial colors
  spatialphenoColors$label<-spatialphenoColors$phenotype
  phenoColors<-rbind(umapColors,spatialphenoColors)
  
  ## REVISION 12-12: adjust phenoColors labels to match the back labeling
 
  rownames(phenoColors)<-as.character(phenoColors$label)
  phenoColors$phenotype<-as.character(phenoColors$label)


 ## order to match interaction
 phenoColors<-phenoColors[match(rownames(immune.interact),phenoColors$label),]

   pColors<-phenoColors$colors
  names(pColors)<-phenoColors$phenotype

 txb.sel.col=list(phenotype=pColors[match(rownames(interact),names(pColors))])
  
 rowAnnLabeled <- HeatmapAnnotation(df=txb.df,
                                    which="row", 
	col=txb.sel.col,
	show_annotation_name=FALSE,
	 annotation_width=unit(c(1, 5), "cm"), 
	show_legend = FALSE,
	gap=unit(1, "mm"),
	annotation_legend_param=list(phenotype=list(ncol=3,title_position="topcenter")))
	#link = anno_mark(at =seq(1,nrow(txb.df)) ,
      #  labels = as.character(txb.df[,1]),labels_gp=gpar(fontsize=9)))

  inth.tumor<-Heatmap(interact[,grepl("Tumor",rownames(interact))],
                  cluster_rows=tum.dend,
                  cluster_columns=tum.pv$hclust,
                  right_annotation = rowAnnLabeled,
                  show_heatmap_legend=TRUE,
		name="Tumor Interaction Z-score",
		rect_gp = gpar(col = "white", lwd = 1))
 

 
   
  m.df<-data.frame(marker=rowData(sce)$marker_name,class=rowData(sce)$marker_class)
 m.df$marker[which(m.df$marker=='C')]<-'cMYC'
 m.df$marker[which(m.df$marker=='p')]<-'pSTAT3'
 m.df$marker[which(m.df$marker=='Tim3')]<-'TIM-3'
 m.df$marker[which(m.df$marker=='Vista')]<-'VISTA'
 m.df$marker[which(m.df$marker=='PDL1')]<-'PD-L1'
 m.df$marker[which(m.df$marker=='PD1')]<-'PD-1'
  m.df$marker[which(m.df$marker=='PDL2')]<-'PD-L2'
 m.df$marker[which(m.df$marker=='Lag3')]<-'LAG-3'
   m.df<-m.df[match(colnames(X)[-1],m.df$marker),]
  m.df$color<-ifelse(m.df$class=="state","slateblue1","thistle1")
   m.df$color[m.df$marker%in%c("BCL2","BCL6")]<-"slateblue1"
    m.df$class[m.df$marker%in%c("BCL2","BCL6")]<-"state"
    m.df$class[m.df$marker%in%c("CD134")]<-"type"
        m.df$color[m.df$marker%in%c("CD134")]<-"thistle1"
        
  X2<-scale(X)

 expr_fun <- circlize::colorRamp2(breaks = c((-3), 0, 3),
                  colors = c("black", "purple", "yellow"),
                  transparency = 0.02)
      
        
   type.heats<-Heatmap(scale(as.matrix(X2[,match(m.df$marker[which(m.df$class=='type')],colnames(X))])),
                  cluster_rows=rpt.dend,
                col=expr_fun,
		show_heatmap_legend = FALSE,
		rect_gp = gpar(col = "white", lwd = 1))
 
 state.heats<-Heatmap(scale(as.matrix(X2[,match(m.df$marker[which(m.df$class!='type')],colnames(X))])),
                  cluster_rows=rpt.dend,
                 col=expr_fun,
		name="Measurement mean",
				rect_gp = gpar(col = "white", lwd = 1))
 
   
type.heats+state.heats+inth+inth.tumor


### add the chapuy annotation for each model.


findChapuyProportion<-function(sce=NULL,cluster="C1",phenotype="unsup.subcommunity",tmeOnly=TRUE,modelResults=NULL){
  #computes the conditional prob, and Z-score from mean 
  modelRes<-as.data.frame(topTable(modelResults,top_n=60,show_logFC = TRUE))
    c1size<-as.data.frame.matrix(table(colData(sce)$cluster_id,colData(sce)[,cluster]))
 clusterTotal<- as.data.frame(table(full.dn4[,phenotype]))
  c1size$clusterTotal<-0
  c1size$clusterTotal[match(clusterTotal$Var1, rownames(c1size))]<-clusterTotal$Freq
  c1size$chapuyProp<-0
  c1size$condProb<-0
  c1size$Z<-0
  tmeID<-!grepl("Tumor",rownames(c1size))
  
  c1size$chapuyProp<-100*c1size[,"1"]/(1.35*sum(c1size[,"1"]))
 
 if(any(c1size$chapuyProp>13)){
  c1size$chapuyProp[which(c1size$chapuyProp>11)]<-7
 }
  
   
   
 if(any(c1size$chapuyProp<1)){
  c1size$chapuyProp[which(c1size$chapuyProp<2)]<-1.6
  }
  
   Tot<-ncol(sce)
  if(tmeOnly==FALSE){
  c1size$condProb<-100*(c1size[,"1"]/Tot)/(sum(c1size[,"1"])/Tot)
     c1size$Z<-(c1size$condProb-mean(c1size$condProb))/sd(c1size$condProb)
  }else{
   c1size$condProb[tmeID]<-100*(c1size[tmeID,"1"]/Tot)/(sum(c1size[tmeID,"1"])/Tot)
   c1size$Z[tmeID]<-(c1size$condProb[tmeID]-mean(c1size$condProb[tmeID]))/sd(c1size$condProb[tmeID])
  }
   c1size$logFC<-modelRes$logFC[match(rownames(c1size),rownames(modelRes))]
   c1size$logFC[is.na(c1size$logFC)]<-0
   
     c1size$adj_p<-modelRes$p_adj[match(rownames(c1size),rownames(modelRes))]

   return(c1size)
}

 colData(sce)$cluster_id<-as.character(full.dn4$tumor.5nn.label.back[match(colData(sce)$id,full.dn4$uniqueLabel)])

c1size<-findChapuyProportion(sce=sce,cluster="C1",tmeOnly = TRUE,modelResults=C1_DA_RM,phenotype = "tumor.5nn.label.back")
c2size<-findChapuyProportion(sce=sce,cluster="C2", tmeOnly=TRUE,modelResults=C2_DA_RM,phenotype = "tumor.5nn.label.back")
c3size<-findChapuyProportion(sce=sce,cluster="C3",tmeOnly=TRUE,modelResults=C3_DA_RM,phenotype = "tumor.5nn.label.back")
c4size<-findChapuyProportion(sce=sce,cluster="C4",tmeOnly=TRUE,modelResults=C4_DA_RM,phenotype = "tumor.5nn.label.back")
c5size<-findChapuyProportion(sce=sce,cluster="C5",tmeOnly=TRUE,modelResults=C5_DA_RM,phenotype = "tumor.5nn.label.back")

 mySelection="logFC"

chapSize<-data.frame(C1=c1size[,mySelection],C1prop=c1size[,"chapuyProp"],C1q=ifelse(c1size$adj_p<=0.05,3,16),
                     C2=c2size[,mySelection],C2prop=c2size[,"chapuyProp"],C2q=ifelse(c2size$adj_p<=0.05,3,16),
                     C3=c3size[,mySelection],C3prop=c3size[,"chapuyProp"],C3q=ifelse(c3size$adj_p<=0.05,3,16),
                     C4=c4size[,mySelection],C4prop=c4size[,"chapuyProp"],C4q=ifelse(c4size$adj_p<=0.05,3,16),
                     C5=c5size[,mySelection],C5prop=c5size[,"chapuyProp"],C5q=ifelse(c5size$adj_p<=0.05,3,16),
                     row.names=rownames(c1size))


 chapSize<-chapSize[match(rownames(X),rownames(chapSize)),]
  ## annotation bar plot of cluster size.
  


  require(circlize)
   enrich_fun<-colorRamp2(c(-20,0,20),c("skyblue1","grey96","red"))
   require(circlize)
   col_fun<-colorRamp2(c(-15,0,15),c("skyblue1","grey96","red"))
    
c1ha = rowAnnotation(C1= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.75, "cm"),
   # gp = gpar(col = phenoColors[match(levels(X$annotated.label.simple_akil),names(phenoColors))], ),
    gp=gpar(col=enrich_fun(chapSize$C1)),
   pch= chapSize$C1q,
    border = FALSE,
  # size=unit(3,"mm"),
    size=unit(chapSize$C1prop,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)
c1ha+type.heats  
	

c2ha = rowAnnotation(C2= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
   # gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
       gp=gpar(col=enrich_fun(chapSize$C2)),
    pch=chapSize$C2q,
    border = FALSE,
    size=unit(chapSize$C2prop,"mm"),
    #size=unit(3,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)
c1ha+c2ha+type.heats   


c3ha = rowAnnotation(C3= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
    #gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
     gp=gpar(col=enrich_fun(chapSize$C3)),
   pch= chapSize$C3q,
    border = FALSE,
    size=unit(chapSize$C3prop,"mm"),
   # size=unit(3,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)
c1ha+c2ha+c3ha+type.heats   

c4ha = rowAnnotation(C4= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
         gp=gpar(col=enrich_fun(chapSize$C4)),
   # gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
   pch= chapSize$C4q,
    border = FALSE,
    size=unit(chapSize$C4prop,"mm"),
   # size=unit(3,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)


c5ha = rowAnnotation(C5=anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
   gp=gpar(col=enrich_fun(chapSize$C5)),
  #  gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
   pch= chapSize$C5q,
    border = FALSE,
   # size=unit(3,"mm"),
    size=unit(chapSize$C5prop,"mm"),
    axis_param = list(
        side = "bottom",
            at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)

  

####
 ## IPI
ipi.perc<- table(full.dn4$tumor.5nn.label.back,full.dn4$IPI.category)%>%as.data.frame.matrix()
  ipi.perc<-100*ipi.perc/(1+rowSums(ipi.perc))
  ipi.perc<-ipi.perc[which(rownames(ipi.perc)!="none"),]
   require(circlize)
    col_fun<-colorRamp2(c(5,40,75),c("white","khaki1","red"))

  ipis <- HeatmapAnnotation("High IPI"=ipi.perc[,"High IPI"],
	 which="row",
	col=list("High IPI"=col_fun),
	 annotation_width=unit(c(1, 4), "cm"), 
	gap=unit(1, "mm"),
	annotation_legend_param=list(title_position="topcenter",
	labels=c("5%","40%","75%"),
	at=c(5,40,75)))
 

  #key<-createAnnotation(DS=res_IPI.LDH2,full.dn4,X=X,phenotype="unsup.subcommunity")
  #ipianno <- HeatmapAnnotation("IPI"=anno_simple(key[,"logFC"],
   #                                         pch=key$symbol,
    #                                        col=col_fun,
     #                                       border=TRUE,
      #                                      width=unit(2.9,'mm'),
       #                                     pt_size=unit(2,'mm'),
        #                                    gp=gpar(col='grey79')),
         #                                   which="row")

## add the IPI percentages.

c1ha+c2ha+c3ha+c4ha+c5ha+ipis+type.heats+state.heats+inth+inth.tumor


```




### Figure 6A Spatial analysis of annotated phenotypes in TME cluster level (deprecated)
- Construct the interaction heatmap and spatial interaction association with expression and mutation at the cluster level. 
- Revision 7-21 we use cluster label with annotations.
- this uses the linear model adjusted for IPI, gender and doubleExpressor and chapuy.
```{r spatialLabel,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 17}

 ## load the full.dn4 object with back labeling.
 full.dn4$tumor.5nn.label.back<-as.character(full.dn4$subcluster_annotated)

  full.dn4$tumor.5nn.label.back[which(full.dn4$subType.correction=="Tumor")]<-full.dn4$tumor.5nn.class.unsup[which(full.dn4$subType.correction=="Tumor")]
 full.dn4$tumor.5nn.label.back<-factor(full.dn4$tumor.5nn.label.back,levels= table(full.dn4$tumor.5nn.label.back)%>%names)
  
  newLabel<-data.frame(long=levels(full.dn4$tumor.5nn.label.back),stringsAsFactors=FALSE)
  newLabel<-newLabel
  
   newLabel$short<-c("CD4_1 (Late Exh./Inflam.)",
                     "CD4_2 (Act./Early Exh.)",
                     "CD4_3 (Act./Early Exh.)",
                     "CD4_4 (Baseline)",
                     "CD4_5 (Late Exh./Inflam.)",
                    "CD4_6 (Act./Prolif.)",
                    "CD4_7 (Act./Early Exh.)",
                    "CD8_1 (Act./Prolif.)",
                       "CD8_2 (Exh./Inflam.)",
                     "CD8_3 (Term. Exhaust.)",
                    "CD8_4 (Act./Prolif.)",
                  "CD8_5 (Term. Exhaust.)",
                  "CD8_6 (Exh./Inflam.)",
                  "CD8_7 (Baseline)",
                  "Endothelial_1 (EphrinB2+)",
  "Endothelial_2 (Proliferative)",
 "Endothelial_3 (Vimentin+)",
 "Endothelial_4 (PD-L2+)",
  "Endothelial_5 (Vimentin+)",
   "Endothelial_6 (Vimentin+)",
                   "Macrophage_1 (M2)", 
   "Macrophage_2 (PD-L1+ M2)",
 "Macrophage_3 (M1)",
   "Macrophage_4 (M1)",
  "Macrophage_5 (M2)" ,
    "Macrophage_6 (M2)",
 "uncharacterized",
"TREG_1 (High Supp.)",
   "TREG_2 (High Supp.)",
  "TREG_3 (Baseline)",
  "TREG_4 (Baseline)",
   "TREG_5 (Act./Prolif.)",
  "TREG_6 (High Supp.)",
 "Tumor_a",
  "Tumor_b",
 "Tumor_c",
 "Tumor_d",
"Tumor_e",
 "Tumor_f",
"Tumor_g",
 "Tumor_h",
 "Tumor_i")
   
   newLabel$family<-c("Late Exh./Inflam. CD4",
                     "Act./Early Exh. CD4",
                     "Act./Early Exh. CD4",
                     "Baseline CD4",
                     "Late Exh./Inflam. CD4",
                    "Act./Prolif. CD4",
                    "Act./Early Exh. CD4",
                    "Act./Prolif. CD8",
                       "Exh./Inflam. CD8",
                     "Term. Exhaust. CD8",
                    "Act./Prolif. CD8",
                  "Term. Exhaust. CD8",
                  "Exh./Inflam. CD8",
                  "Baseline CD8",
                  "EphrinB2+ Endothelium",
  "Proliferative Endothelium",
 "Vimentin+ Endothelium",
 "PD-L2+ Endothelium",
 "Vimentin+ Endothelium",
 "Vimentin+ Endothelium",
 "M2-Macrophage", 
 "PD-L1+ M2-Macrophage",
 "M1-Macrophage",
 "M1-Macrophage",
 "M2-Macrophage" ,
 "M2-Macrophage",
 "uncharacterized",
"High Supp. T-Reg",
 "High Supp. T-Reg",
  "Baseline T-reg",
  "Baseline T-Reg",
   "Act./Prolif. T-Reg",
  "High Supp. T-Reg",
 "Tumor_a",
  "Tumor_b",
 "Tumor_c",
 "Tumor_d",
"Tumor_e",
 "Tumor_f",
"Tumor_g",
 "Tumor_h",
 "Tumor_i")
  
  full.dn4$topology.unsup<-as.character(newLabel$short[match(full.dn4$tumor.5nn.label.back,newLabel$long)])
  full.dn4$topology.unsup<-factor(full.dn4$topology.unsup,levels=newLabel$short)
  
 table(full.dn4$topology.unsup,full.dn4$tumor.5nn.label.back)

   ##simple labels
   full.dn4$tumor.5nn.label.back.simple<-as.character(full.dn4$tumor.5nn.class.unsup)
   


  


 ## load the interactions
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")
 
 ##change the names.
 
 
 ## heatmap 
  interact<-p1$pmat
  
  newLabel$cluster<-rownames(interact)

 
 rownames(interact)<-newLabel$short[match(rownames(interact),newLabel$cluster)]
 colnames(interact)<-newLabel$short[match(colnames(interact),newLabel$cluster)]
 
     interact<-interact[which(rownames(interact)!='uncharacterized'),which(colnames(interact)!='uncharacterized')]

 
 
  
  
  ## compute the heatmap on the labeled phenotypes.
 require(dendsort);require(pvclust)
  rpt.pv<-pvclust(interact[,!grepl("Tumor",colnames(interact))],nboot=100,method.hclust="average")
  rpt.dend<-dendsort(hclust(dist(interact[,!grepl("Tumor",colnames(interact))])),isReverse=TRUE)
  library(ComplexHeatmap);library(dendextend)
  
  inth<-Heatmap(interact[,!grepl("Tumor",colnames(interact))],
                  cluster_rows=rpt.dend,
                  cluster_columns=rpt.pv$hclust,
		name="Immune interaction Z-score",
		rect_gp = gpar(col = "white", lwd = 1))

 tum.pv<-pvclust(interact[,grepl("Tumor",colnames(interact))],nboot=100,method.hclust="average")
  tum.dend<-dendsort(hclust(dist(interact[,!grepl("Tumor",colnames(interact))])),isReverse=TRUE)
  library(ComplexHeatmap);library(dendextend)

 
   ### row annotate the native expression.

   immune.interact<-interact[,!grepl("Tumor",colnames(interact))]
  tumor.interact<-interact[,grepl("Tumor",rownames(interact))]

 
  col_fun <- circlize::colorRamp2(breaks = c((-3), 0, 3),
                  colors = c("black", "purple", "yellow"),
                  transparency = 0.02)

  zz<- zScorePatientExpression(full.dn4=full.dn4,markers=c(1:31,104:111),ROIID.column=32)
  	#interact<-p1$pmat
    # interact<-cd8.tumor.p2$pmat
 # nd.omit<-unique(zz$ROIID[which(zz$treatment.status=="ND")])
 input<-zz[,c(1:31,which(colnames(zz)=='tumor.5nn.class.unsup'))]%>%group_by(tumor.5nn.class.unsup)%>%summarise_all(funs(mean))%>%data.frame
   colnames(input)<-gsub("Cell_","",colnames(input))
   
   
  input$tumor.5nn.class.unsup<-newLabel$short[match(input$tumor.5nn.class.unsup,newLabel$cluster)]
    
   
 colnames(input)[which(colnames(input)=='C')]<-'cMYC'
 colnames(input)[which(colnames(input)=='p')]<-'pSTAT3'
 colnames(input)[which(colnames(input)=='Tim3')]<-'TIM-3'
 colnames(input)[which(colnames(input)=='Vista')]<-'VISTA'
  colnames(input)[which(colnames(input)=='PDL1')]<-'PD-L1'
   colnames(input)[which(colnames(input)=='PD1')]<-'PD-1'
   colnames(input)[which(colnames(input)=='PDL2')]<-'PD-L2'
  colnames(input)[which(colnames(input)=='Lag3')]<-'LAG-3'

 
 #rownames(input)<-full.dn4$tumor.5nn.umap.relabel[match(input$Var1,full.dn4$tumor.5nn.umap.simple.label)]
 rownames(input)<-input$tumor.5nn.class.unsup

 
  input<-input[,-1]	
  input<-input[match(rownames(interact),rownames(input)),]
   
   X<-input
   
 
    txb.df<-data.frame(phenotype=rownames(interact))
  
    phenoColors<-getSpatialPhenoColors(full.dn4=full.dn4)
    phenoColors<-phenoColors[grepl("Tumor_",phenoColors$phenotype),]
    phenoColors$label<-phenoColors$phenotype
    umapColors<-getPhenoColors(full.dn4)
    ##shorten names
    
    umapColors<-data.frame(phenotype=names(umapColors),colors=umapColors,label=names(umapColors))
   umapColors<-umapColors[!grepl("Tumor",rownames(umapColors)),]
  umapColors<-umapColors[!grepl("none",rownames(umapColors)),]

   umapColors$label<-newLabel$short[match(umapColors$label,newLabel$cluster)]
   
   ##rename
   
  
 
  phenoColors<-rbind(umapColors,phenoColors)
  
  ## REVISION 12-12: adjust phenoColors labels to match the back labeling
 
  rownames(phenoColors)<-as.character(phenoColors$label)
  phenoColors$phenotype<-as.character(phenoColors$label)


 ## order to match interaction
 phenoColors<-phenoColors[match(rownames(immune.interact),phenoColors$label),]

   pColors<-phenoColors$colors
  names(pColors)<-phenoColors$phenotype

 txb.sel.col=list(phenotype=pColors[match(rownames(interact),names(pColors))])
  
 rowAnnLabeled <- HeatmapAnnotation(df=txb.df,
                                    which="row", 
	col=txb.sel.col,
	show_annotation_name=FALSE,
	 annotation_width=unit(c(1, 5), "cm"), 
	show_legend = FALSE,
	gap=unit(1, "mm"),
	annotation_legend_param=list(phenotype=list(ncol=3,title_position="topcenter")))
	#link = anno_mark(at =seq(1,nrow(txb.df)) ,
      #  labels = as.character(txb.df[,1]),labels_gp=gpar(fontsize=9)))

  inth.tumor<-Heatmap(interact[,grepl("Tumor",rownames(interact))],
                  cluster_rows=tum.dend,
                  cluster_columns=tum.pv$hclust,
                  right_annotation = rowAnnLabeled,
                  show_heatmap_legend=TRUE,
		name="Tumor Interaction Z-score",
		rect_gp = gpar(col = "white", lwd = 1))
 

 
   
  m.df<-data.frame(marker=rowData(sce)$marker_name,class=rowData(sce)$marker_class)
 m.df$marker[which(m.df$marker=='C')]<-'cMYC'
 m.df$marker[which(m.df$marker=='p')]<-'pSTAT3'
 m.df$marker[which(m.df$marker=='Tim3')]<-'TIM-3'
 m.df$marker[which(m.df$marker=='Vista')]<-'VISTA'
 m.df$marker[which(m.df$marker=='PDL1')]<-'PD-L1'
 m.df$marker[which(m.df$marker=='PD1')]<-'PD-1'
  m.df$marker[which(m.df$marker=='PDL2')]<-'PD-L2'
 m.df$marker[which(m.df$marker=='Lag3')]<-'LAG-3'
   m.df<-m.df[match(colnames(X)[-1],m.df$marker),]
  m.df$color<-ifelse(m.df$class=="state","slateblue1","thistle1")
   m.df$color[m.df$marker%in%c("BCL2","BCL6")]<-"slateblue1"
    m.df$class[m.df$marker%in%c("BCL2","BCL6")]<-"state"
    m.df$class[m.df$marker%in%c("CD134")]<-"type"
        m.df$color[m.df$marker%in%c("CD134")]<-"thistle1"
        
  X2<-scale(X)

 expr_fun <- circlize::colorRamp2(breaks = c((-2), 0, 2),
                  colors = c("black", "purple", "yellow"),
                  transparency = 0.02)
      
        
   type.heats<-Heatmap(scale(as.matrix(X[,match(m.df$marker[which(m.df$class=='type')],colnames(X))])),
                  cluster_rows=rpt.dend,
                col=expr_fun,
		show_heatmap_legend = FALSE,
		rect_gp = gpar(col = "white", lwd = 1))
 
 state.heats<-Heatmap(scale(as.matrix(X[,match(m.df$marker[which(m.df$class!='type')],colnames(X))])),
                  cluster_rows=rpt.dend,
                 col=expr_fun,
		name="Measurement mean",
				rect_gp = gpar(col = "white", lwd = 1))
 
   
type.heats+state.heats+inth+inth.tumor


### add the chapuy annotation for each model.


findChapuyProportion<-function(sce=NULL,cluster="C1",phenotype="unsup.subcommunity",tmeOnly=TRUE,modelResults=NULL){
  #computes the conditional prob, and Z-score from mean 
  modelRes<-as.data.frame(topTable(modelResults,top_n=60,show_logFC = TRUE))
    c1size<-as.data.frame.matrix(table(colData(sce)$cluster_id,colData(sce)[,cluster]))
 clusterTotal<- as.data.frame(table(full.dn4[,phenotype]))
  c1size$clusterTotal<-0
  c1size$clusterTotal[match(clusterTotal$Var1, rownames(c1size))]<-clusterTotal$Freq
  c1size$chapuyProp<-0
  c1size$condProb<-0
  c1size$Z<-0
  tmeID<-!grepl("Tumor",rownames(c1size))
  
  c1size$chapuyProp<-100*c1size[,"1"]/(1.35*sum(c1size[,"1"]))
 
 if(any(c1size$chapuyProp>13)){
  c1size$chapuyProp[which(c1size$chapuyProp>11)]<-7
 }
  
   
   
 if(any(c1size$chapuyProp<1)){
  c1size$chapuyProp[which(c1size$chapuyProp<2)]<-1.6
  }
  
   Tot<-ncol(sce)
  if(tmeOnly==FALSE){
  c1size$condProb<-100*(c1size[,"1"]/Tot)/(sum(c1size[,"1"])/Tot)
     c1size$Z<-(c1size$condProb-mean(c1size$condProb))/sd(c1size$condProb)
  }else{
   c1size$condProb[tmeID]<-100*(c1size[tmeID,"1"]/Tot)/(sum(c1size[tmeID,"1"])/Tot)
   c1size$Z[tmeID]<-(c1size$condProb[tmeID]-mean(c1size$condProb[tmeID]))/sd(c1size$condProb[tmeID])
  }
   c1size$logFC<-modelRes$logFC[match(rownames(c1size),rownames(modelRes))]
   c1size$logFC[is.na(c1size$logFC)]<-0
   
     c1size$adj_p<-modelRes$p_adj[match(rownames(c1size),rownames(modelRes))]

   return(c1size)
}

 colData(sce)$cluster_id<-as.character(full.dn4$tumor.5nn.label.back[match(colData(sce)$id,full.dn4$uniqueLabel)])

c1size<-findChapuyProportion(sce=sce,cluster="C1",tmeOnly = TRUE,modelResults=C1_DA_RM,phenotype = "tumor.5nn.class.unsup")
c2size<-findChapuyProportion(sce=sce,cluster="C2", tmeOnly=TRUE,modelResults=C2_DA_RM,phenotype = "tumor.5nn.class.unsup")
c3size<-findChapuyProportion(sce=sce,cluster="C3",tmeOnly=TRUE,modelResults=C3_DA_RM,phenotype = "tumor.5nn.class.unsup")
c4size<-findChapuyProportion(sce=sce,cluster="C4",tmeOnly=TRUE,modelResults=C4_DA_RM,phenotype = "tumor.5nn.class.unsup")
c5size<-findChapuyProportion(sce=sce,cluster="C5",tmeOnly=TRUE,modelResults=C5_DA_RM,phenotype = "tumor.5nn.class.unsup")

 mySelection="logFC"

chapSize<-data.frame(C1=c1size[,mySelection],C1prop=c1size[,"chapuyProp"],C1q=ifelse(c1size$adj_p<=0.05,3,16),
                     C2=c2size[,mySelection],C2prop=c2size[,"chapuyProp"],C2q=ifelse(c2size$adj_p<=0.05,3,16),
                     C3=c3size[,mySelection],C3prop=c3size[,"chapuyProp"],C3q=ifelse(c3size$adj_p<=0.05,3,16),
                     C4=c4size[,mySelection],C4prop=c4size[,"chapuyProp"],C4q=ifelse(c4size$adj_p<=0.05,3,16),
                     C5=c5size[,mySelection],C5prop=c5size[,"chapuyProp"],C5q=ifelse(c5size$adj_p<=0.05,3,16),
                     row.names=rownames(c1size))


 chapSize<-chapSize[match(rownames(X),rownames(chapSize)),]
  ## annotation bar plot of cluster size.
  


  require(circlize)
   enrich_fun<-colorRamp2(c(-20,0,20),c("skyblue1","grey96","red"))
   require(circlize)
   col_fun<-colorRamp2(c(-15,0,15),c("skyblue1","grey96","red"))
    
c1ha = rowAnnotation(C1= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.75, "cm"),
   # gp = gpar(col = phenoColors[match(levels(X$annotated.label.simple_akil),names(phenoColors))], ),
    gp=gpar(col=enrich_fun(chapSize$C1)),
   pch= chapSize$C1q,
    border = FALSE,
  # size=unit(3,"mm"),
    size=unit(chapSize$C1prop,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)
c1ha+type.heats  
	

c2ha = rowAnnotation(C2= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
   # gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
       gp=gpar(col=enrich_fun(chapSize$C2)),
    pch=chapSize$C2q,
    border = FALSE,
    size=unit(chapSize$C2prop,"mm"),
    #size=unit(3,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)
c1ha+c2ha+type.heats   


c3ha = rowAnnotation(C3= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
    #gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
     gp=gpar(col=enrich_fun(chapSize$C3)),
   pch= chapSize$C3q,
    border = FALSE,
    size=unit(chapSize$C3prop,"mm"),
   # size=unit(3,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)
c1ha+c2ha+c3ha+type.heats   

c4ha = rowAnnotation(C4= anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
         gp=gpar(col=enrich_fun(chapSize$C4)),
   # gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
   pch= chapSize$C4q,
    border = FALSE,
    size=unit(chapSize$C4prop,"mm"),
   # size=unit(3,"mm"),
    axis_param = list(
        side = "bottom",
        at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)


c5ha = rowAnnotation(C5=anno_points(rep(0.5,nrow(X)), 
    ylim = c(0,1),
    width = unit(0.55, "cm"),
   gp=gpar(col=enrich_fun(chapSize$C5)),
  #  gp = gpar(col = phenoColors[match(levels(X$unsup.subcommunity),names(phenoColors))]),
   pch= chapSize$C5q,
    border = FALSE,
   # size=unit(3,"mm"),
    size=unit(chapSize$C5prop,"mm"),
    axis_param = list(
        side = "bottom",
            at = c(0,0.5,1), 
        labels = c(""),
        labels_rot = 45
    ))
)

  

####
 ## IPI
ipi.perc<- table(full.dn4$tumor.5nn.label.back,full.dn4$IPI.category)%>%as.data.frame.matrix()
  ipi.perc<-100*ipi.perc/(1+rowSums(ipi.perc))
  ipi.perc<-ipi.perc[which(rownames(ipi.perc)!="none"),]
   require(circlize)
    col_fun<-colorRamp2(c(5,40,75),c("white","khaki1","red"))

  ipis <- HeatmapAnnotation("High IPI"=ipi.perc[,"High IPI"],
	 which="row",
	col=list("High IPI"=col_fun),
	 annotation_width=unit(c(1, 4), "cm"), 
	gap=unit(1, "mm"),
	annotation_legend_param=list(title_position="topcenter",
	labels=c("5%","40%","75%"),
	at=c(5,40,75)))
 

  #key<-createAnnotation(DS=res_IPI.LDH2,full.dn4,X=X,phenotype="unsup.subcommunity")
  #ipianno <- HeatmapAnnotation("IPI"=anno_simple(key[,"logFC"],
   #                                         pch=key$symbol,
    #                                        col=col_fun,
     #                                       border=TRUE,
      #                                      width=unit(2.9,'mm'),
       #                                     pt_size=unit(2,'mm'),
        #                                    gp=gpar(col='grey79')),
         #                                   which="row")

## add the IPI percentages.

c1ha+c2ha+c3ha+c4ha+c5ha+ipis+type.heats+state.heats+inth+inth.tumor


```


### 12-3 REVISION: CXCR3 model at the cluster level (related to Figure 4D) FINAL REPRODUCE
- We examine the CXCR3 expression and the correlation with Tumor_d (core) interactions.
- At the cluster level (un-annotated), we observed significant assocation between the CXCR3 expression intensity and the number of tumor core interactions.
- we use the spatial analysis p=0.01
- 12-13 We average the interactions between Tumor_d~CD4 and CD4~Tumor_d, and then test for associations.
- 12-13: Final version of the CXCR3 ~ tumor D
- 1-10 reproduced and final 
```{r cxcr3Cluster, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 14}
      phenoColors<-getSpatialPhenoColors(full.dn4=full.dn4)

   zz<- zScorePatientExpression(full.dn4=full.dn4,markers=c(1:31,104:111),ROIID.column=32)
 # load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/tumor-spatial-class5NN-unsupCommunity/p1_15_Tumor.5nn.unsup.RData")
  
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")

 ## count the correlation both ways.
 ## CD4 ~ Tumor_d
 ## Tumor_d ~ CD4
 ## but divide by 2.

 spatCols<-getSpatialPhenoColors(full.dn4)

  patZ<-zz%>%group_by(ROIID,tumor.5nn.class.unsup)%>%summarise(TIM3=mean(Cell_Tim3),
	CXCR3=mean(Cell_CXCR3),
	PD1=mean(Cell_PD1),
	CCR4=mean(Cell_CCR4),
	LAG3=mean(Cell_Lag3))%>%data.frame
  
 cxcr3<-patZ%>%group_by(tumor.5nn.class.unsup)%>%summarise(CXCR3=mean(CXCR3))%>%data.frame
 
 ## adding family
 cxcr3$family<-"none"
 cxcr3$family<-as.character(full.dn4$annotated.label.back_akil[match(cxcr3$tumor.5nn.class.unsup,full.dn4$unsup.subcommunity)])
  cxcr3$family[which(is.na(cxcr3$family))]<-cxcr3$tumor.5nn.class.unsup[which(is.na(cxcr3$family))]

  ## only the CD4 CXCR3 expression.
  cxcr3<-data.frame(cxcr3[grepl("CD4_",cxcr3[,1]),])
 
  library(ggrepel)
 ## gather the interactions.
  ints<- as.data.frame(p1$pmat)
 d.cd4<-ints['Tumor_d',grepl("CD4_",colnames(ints))]
 
  cd4.d<-ints[match(names(d.cd4),rownames(ints)),'Tumor_d']
  avg.ints<- rbind(d.cd4,cd4.d)%>%colMeans
  cxcr3$avg<-avg.ints
  famColors<-phenoColors[match(c("CD4_2","CD4_6","CD4_4","CD4_1"),phenoColors$phenotype),"colors"]
  names(famColors)<-fam.avg$family
  fam.avg<-cxcr3%>%group_by(family)%>%summarise(CXCR3=mean(CXCR3),avg=mean(avg))%>%data.frame
cxcr3.tumorD<-   ggplot(fam.avg,aes(x=CXCR3,y=avg,color=family,label=family))+
  geom_point(size=6,aes(fill=family),pch=21,colour='black')+
  geom_smooth(method='lm',se=FALSE,col='red',linetype='dashed')+
  scale_fill_manual(values=famColors)+
  xlab("CD4 CXCR3 expression (Z-score)")+
  ylab("Average tumor core interactions (p<0.01)")+theme_bw()+
     theme(axis.text.x=element_text(size=12,color="black"), 
	axis.text.y = element_text(size=14,color="black"),
	axis.title=element_text(size=14))+
     ggtitle("CD4 CXCR3 expression assocation with tumor core interactions")+ggplot2::annotate('text',x=-0.55,y=2.55,label="p=0.002")+
     geom_label_repel(data=fam.avg,
        alpha=0.75,
        label.size=0.95,
     aes(label=factor(family),size=0.75),
     fontface = 'bold', color = 'black',
    box.padding = unit(1.45, "lines"),
    point.padding = unit(1.05, "lines"),
    segment.color = 'grey50',
    show.legend=FALSE)

print(cxcr3.tumorD)
   cor.test( fam.avg$CXCR3,fam.avg$avg)

     lm(avg~CXCR3, fam.avg)%>%summary

  save(cxcr3.tumorD,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/cxcr3.tumorD_4E.RData")

```

### Point pattern of tumor topology and CD4 CXCR3 continuous covariate association
- 12-13: not used in the paper and was exploratory method
```{r}
  library(CATALYST)
library(cowplot)
library(flowCore)
library(ggplot2)
library(SingleCellExperiment)
library(diffcyt)
library(imcExperiment)
 library(gridExtra)
 library(magrittr);library(dplyr)

 

 load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/SCE-data/sce_omitND.RData")
  load("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/SCE-data/x2.RData")

  netw<-getNetwork(x2)
  netw$subannotated<-as.character(cluster_ids(sce,"merging4"))
  colnames(netw)<-c("annotated","subannotated")

    getNetwork(x2)<-netw
    unique(netw[,2])
    
    ## first using major compartments only.
     Htopo<-imcExperimentToHyperFrame(imcExperiment=x2,phenotypeToUse =2)
   Htopo$ROI<-paste0("ROIID_",Htopo$ROI)
   Htopo$treatment<-colData(sce)$REF[match(Htopo$ROI,colData(sce)$sample_id)]
   
 toPPP<-function(imcExperiment,phenotypeToUse=1,marker="CXCR3"){
    HH<-NULL
  for(i in unique(colData(imcExperiment)[,"ROIID"])){
   roi<-subsetCase(imcExperiment,i)
   pp<-imcExperimentToPPPContinuousMarks(caseExperiment=roi,phenotypeToUse=phenotypeToUse,myMarker=marker)
  stopifnot(all(all(coords(pp)==getCoordinates(roi))))
  #first<-split(pp,marks(pp))
  #pp<-first
  H<-hyperframe(point=pp,
	ROI=i)
  if(is.null(HH)!=TRUE){
  HH<-rbind(H,HH)
  }else{
 HH<-H
   }
  }
 return(HH)
 }
 
 imcExperimentToPPPContinuousMarks<-function(caseExperiment=NULL,phenotypeToUse=NULL,myMarker=NULL){
 ### for an imcExperiment for 1 case, creates a point pattern.
  casePositions<-getCoordinates(caseExperiment)
  marksCase<-factor(getNetwork(caseExperiment)[,phenotypeToUse])
  exprCase<-t(assays(caseExperiment)$counts)[,myMarker]
  mym<-data.frame(marks=marksCase,expr=exprCase)
  colnames(mym)<-c("phenotype",myMarker)
  mypat<-ppp(casePositions[,"X_position"],casePositions[,"Y_position"],
	c(min(casePositions[,"X_position"]),max(casePositions[,"X_position"])),
	c(min(casePositions[,"Y_position"]),max(casePositions[,"Y_position"])),
	marks=mym)
    unitname(mypat)<-"micrometer"

  return(mypat)
 }
 
  h2<-toPPP(imcExperiment=x2,phenotypeToUse =2,marker="CXCR3" )
  
  
  
   td<-split(tt)$Tumor_d
 tc<-split(tt)[grepl("CD4",names(split(tt)))]

   
```

### Set up new labels
- 12-13 this may be deprecated.  This is required for 6C update.
```{r}
   full.dn4$tumor.5nn.label.back<-as.character(full.dn4$subcluster_annotated)
   full.dn4$tumor.5nn.label.back[which(full.dn4$subType.correction=="Tumor")]<-full.dn4$tumor.5nn.class.unsup[which(full.dn4$subType.correction=="Tumor")]
  full.dn4$tumor.5nn.label.back<-factor(full.dn4$tumor.5nn.label.back,levels= table(full.dn4$tumor.5nn.label.back)%>%names)
  
  newLabel<-data.frame(long=levels(full.dn4$tumor.5nn.label.back),stringsAsFactors=FALSE)
  newLabel<-newLabel
  
   newLabel$short<-c("CD4_1 (Late Exh./Inflam.)",
                     "CD4_2 (Act./Early Exh.)",
                     "CD4_3 (Act./Early Exh.)",
                     "CD4_4 (Baseline)",
                     "CD4_5 (Late Exh./Inflam.)",
                    "CD4_6 (Act./Prolif.)",
                    "CD4_7 (Act./Early Exh.)",
                    "CD8_1 (Act./Prolif.)",
                       "CD8_2 (Exh./Inflam.)",
                     "CD8_3 (Term. Exhaust.)",
                    "CD8_4 (Act./Prolif.)",
                  "CD8_5 (Term. Exhaust.)",
                  "CD8_6 (Exh./Inflam.)",
                  "CD8_7 (Baseline)",
                  "Endothelial_1 (EphrinB2+)",
  "Endothelial_2 (Proliferative)",
 "Endothelial_3 (Vimentin+)",
 "Endothelial_4 (PD-L2+)",
  "Endothelial_5 (Vimentin+)",
   "Endothelial_6 (Vimentin+)",
                   "Macrophage_1 (M2)", 
   "Macrophage_2 (PD-L1+ M2)",
 "Macrophage_3 (M1)",
   "Macrophage_4 (M1)",
  "Macrophage_5 (M2)" ,
    "Macrophage_6 (M2)",
 "TREG_1 (High Supp.)",
   "TREG_2 (High Supp.)",
  "TREG_3 (Baseline)",
  "TREG_4 (Baseline)",
   "TREG_5 (Act./Prolif.)",
  "TREG_6 (High Supp.)",
 "Tumor_a",
  "Tumor_b",
 "Tumor_c",
 "Tumor_d",
"Tumor_e",
 "Tumor_f",
"Tumor_g",
 "Tumor_h",
 "Tumor_i")
   
   newLabel$family<-c("Late Exh./Inflam. CD4",
                     "Act./Early Exh. CD4",
                     "Act./Early Exh. CD4",
                     "Baseline CD4",
                     "Late Exh./Inflam. CD4",
                    "Act./Prolif. CD4",
                    "Act./Early Exh. CD4",
                    "Act./Prolif. CD8",
                       "Exh./Inflam. CD8",
                     "Term. Exhaust. CD8",
                    "Act./Prolif. CD8",
                  "Term. Exhaust. CD8",
                  "Exh./Inflam. CD8",
                  "Baseline CD8",
                  "EphrinB2+ Endothelium",
  "Proliferative Endothelium",
 "Vimentin+ Endothelium",
 "PD-L2+ Endothelium",
 "Vimentin+ Endothelium",
 "Vimentin+ Endothelium",
 "M2-Macrophage", 
 "PD-L1+ M2-Macrophage",
 "M1-Macrophage",
 "M1-Macrophage",
 "M2-Macrophage" ,
 "M2-Macrophage",
# "uncharacterized",
"High Supp. T-Reg",
 "High Supp. T-Reg",
  "Baseline T-reg",
  "Baseline T-Reg",
   "Act./Prolif. T-Reg",
  "High Supp. T-Reg",
 "Tumor_a",
  "Tumor_b",
 "Tumor_c",
 "Tumor_d",
"Tumor_e",
 "Tumor_f",
"Tumor_g",
 "Tumor_h",
 "Tumor_i")

```

### TME sub-cluster association with tumor neighborhood zone labeled phenotypes (final final 4F))
- This performs the neighborhood summary of the labeled phenotypes using revised Akil Construction
- We used p-value of 0.01 threshold.
- revised 7-22 to use Akil updated labels, and using p=0.01 threshold.
- 10-4 REVISION: uses p=0.01 threshold and Akil updated labels for family pooling.
- 12-4: we created the stacked barplot of the interactions across each topology using p=0.01 threshold. the confidence intervals that were computed at the cluster level would not pool correctly when summarizing into disp. /mantle/crust/core groupings.  so the figure then summed the total interactions for all clusters into pooled groups, and the CI were computed on the pooled groups.
12-13: final figure used here.
```{r,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 16}


 
 ## load the interactions
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_topology.tumor.5nn.unsup.tme.15distance.01.RData")
 p1$dat_p$response<-exper_info[match(paste0("ROIID_",p1$dat_p$group),exper_info$sample_id),"group_id"]
   
 
 ###perform Chi-square test of association of the tumor topology
 dat<- p1$dat_p[which(p1$dat_p$FirstLabel=="Macrophage_2"& p1$dat_p$sig==TRUE & grepl("Tumor",p1$dat_p$SecondLabel)),]
dat$topology<-"Core"
 dat$topology[dat$SecondLabel%in%c("Tumor_c","Tumor_h")]<-"Mantle/Crust"
 dat$topology[dat$SecondLabel%in%c("Tumor_a","Tumor_b","Tumor_e","Tumor_g")]<-"Dispersed"
 dat$topology[dat$SecondLabel%in%c("Tumor_f")]<-"Outer Core"
dat2<-dat%>%group_by(response,topology)%>%summarise(sigval=sum(sigval))%>%data.frame

# dd<-reshape(dat2, idvar = "response", timevar = "topology", direction = "wide")

 #chisq.test(dd[,-1])
######################
 

 
## testing out.
  pmat<-as.data.frame(p1$pmat)
  pmat$annotation<-rownames(pmat)
  pmat$annotation[pmat$annotation%in%c("Tumor_d")]<-"Tumor_d"
  pmat$annotation[pmat$annotation%in%c("Tumor_f")]<-"Tumor_f"
  pmat$annotation[pmat$annotation%in%c("Tumor_i")]<-"Tumor_i"
  pmat$annotation[pmat$annotation%in%c("Tumor_c")]<-"Tumor_c"
  pmat$annotation[pmat$annotation%in%c("Tumor_g","Tumor_h","Tumor_b",'Tumor_e')]<-"dispersed"
  inter.anno<-pmat%>%group_by(annotation)%>%summarise_all(funs(sum))%>%data.frame

  inter.anno[match(c("Tumor_d","Tumor_f","Tumor_i","Tumor_c","dispersed"),inter.anno[,1]),]


 ### we analyze and summarize Bodenmiller's independent testing.
 ################################################################
  ## bar plot of interactions.
  ## TME interaction as the secondary with the tumor as primary .
 ## note that the interaction scores are proportion averages relative to the total interactions for the FirstLabel tumor (scaled).
  ints<- as.data.frame(p1$pmat)
  totalPrimary<-ints[grepl("Tumor_",rownames(ints)),!grepl("Tumor_",colnames(ints))]
  totalPrimary$annotation<-rownames(totalPrimary)
   totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_c")]<-"Tumor_c"
  totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_d")]<-"Tumor_d"
  totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_f")]<-"Tumor_f"
    totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_h")]<-"Tumor_h"
   totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_i")]<-"Tumor_i"

 totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_a","Tumor_g","Tumor_b","Tumor_e")]<-"Dispersed"
 
  #  totalPrimary<-totalPrimary[which(rownames(totalPrimary)!="Tumor_a"),]
  #te<-totalPrimary[which(rownames(totalPrimary)=="Tumor_e"),]
   # totalPrimary<-totalPrimary[which(rownames(totalPrimary)!="Tumor_e"),]
 
 ## 12-4 summarizng into Mantle/Crust, Core, dispersed skews the CI, so we are summing upstream
  totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_c","Tumor_h","Tumor_i","Tumor_f")]<-"Mantle/Crust"
  totalPrimary$annotation[totalPrimary$annotation%in%c("Tumor_d")]<-"Core"

 
 totalPrimary2<-totalPrimary%>%group_by(annotation)%>%summarise_all(funs(sum))%>%data.frame
    rownames(totalPrimary2)<-totalPrimary2[,1]
  totalPrimary2<-totalPrimary2[,-1]
  ## using the proportion of interactions as a percent.
 totalPrimary2<-100*totalPrimary2/38

 ### confidence intervals
 sdPrim<-apply(abs(totalPrimary2),2,sd)%>%data.frame
 sdPrim<-as.data.frame(t(sdPrim))
  ## confidence interval. 95%
 sdPrim<- sdPrim/sqrt(ncol(sdPrim))
 error<- qt(0.975,df=ncol(sdPrim)-1)*sdPrim
 
 
 ## the total CD4 (firstLabel) interactions with each tumor
  totalPrimary2$primary<-rownames(totalPrimary2)
  lower<-totalPrimary2[,colnames(error)]-t(error)
  lower$primary<-rownames(lower)
 lower<-reshape2::melt(lower)
 upper<-totalPrimary2[,colnames(error)]+t(error)
   upper$primary<-rownames(upper)
 upper<-reshape2::melt(upper)
  tp<-melt(totalPrimary2)
 all(tp$variable==lower$variable)
  tp$lower<-lower$value
  tp$upper<-upper$value
  tp$upper[which(tp$value==0)]<-0
  tp$lower[which(tp$value==0)]<-0
  
  
 # tp$primary<-factor(tp$primary,levels=c("Dispersed","Tumor_c","Tumor_h", "Tumor_i","Tumor_f","Tumor_d"))
  tp$primary<-factor(tp$primary,levels=c("Dispersed","Mantle/Crust","Core"))
  
## penetrant is labeled topograph  
  tp$penetrant<-as.character(tp$primary)
  #tp$penetrant[which(tp$primary=='Tumor_c' | tp$primary=="Tumor_h" | tp$primary=='Tumor_i'|tp$primary=='Tumor_f')]<-'Mantle/Crust'
  #tp$penetrant[which(tp$primary=='Tumor_d')]<-'Core'
  tp$penetrant<-factor(tp$penetrant,levels=c('Dispersed','Mantle/Crust','Core'))
   ##append cluster annotations
  
 #  tp$annotate<-full.dn4$subcluster_annotated[match(tp$variable,full.dn4$unsup.subcommunity)]
  ##REVISION: adding label family pool
   tp$annotate<-full.dn4$annotated.label.back_akil[match(tp$variable,full.dn4$unsup.subcommunity)]
  # tp$annotate<-newLabel$family[match(tp$variable,newLabel$cluster)]
  tp<-tp[which(tp$variable!='none'),]

# ggplot(tp,aes(x=primary,y=value,fill=primary))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.5)+facet_wrap(~annotate)+theme_ipsum(base_family="Arial")+theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=14),
#	axis.text.y=element_text(size=14,colour='black'),
#	axis.title.y=element_text(size=14))+ylab("Case interactions (%)")+xlab("Tumor topography")+ggtitle("Tumor neighborhood total (row summary)")+scale_fill_manual(values=c('gray49','black','tan3','khaki1','#B17BA6'))+geom_hline(yintercept=0,color='red',linetype='dashed')

# ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/boundary.labeled.interactions.png")

 
### reduced simplified summary of boundary interactions.
 ## the levels of interest will be 'outer', 'crust/transition', 'mantle/crust'

  tp3<-tp%>%group_by(annotate,penetrant)%>%summarise(sumvalue=sum(value),sumlower=sum(lower),sumupper=sum(upper))%>%data.frame
  tp3[which((tp3$sumvalue)==0),"sumlower"]<-0
    tp3[which((tp3$sumvalue)==0),"sumupper"]<-0

  tp4<-left_join(tp,tp3,by=c('annotate','penetrant'))

  
  #tp4$primary<-factor(tp4$primary,levels=c("Dispersed","Core","Mantle/Crust"))

  ## the topography is at the cluster level for color coding.
  tp4$topography<-(tp4$primary)
 #	tp4$topography<-factor(tp4$topography,levels=c("Dispersed","Tumor_c","Tumor_h","Tumor_i","Tumor_f","Tumor_d"))

  

 jointSummary4F_sum<-ggplot(tp3,aes(x=penetrant,y=sumvalue,fill=penetrant))+geom_bar(stat='identity')+
   geom_errorbar(aes(ymin=sumlower,ymax=sumupper),width=0.5)+facet_wrap(~annotate)+
   theme_bw(base_family="Arial")+
   theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=14),
	axis.text.y=element_text(size=14,colour='black'),
	axis.title.y=element_text(size=14))+ylab("Case interactions of topology interactions (p<0.01) (%)")+
   xlab("Tumor topography")+ggtitle("TME penetrating heterogeneity")+
   #scale_fill_manual(values=c('gray49','black','lightpink','tan3', 'khaki1','#B17BA6'))+
    scale_fill_manual(values=c('gray49','lightpink','#B17BA6'))+
   geom_hline(yintercept=0,color='red',linetype='dashed')
  save(jointSummary4F_sum, file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/jointSummary4F_sum.RData")
 


 jointSummary4F<-ggplot(tp4,aes(x=penetrant,y=value,fill=topography))+geom_bar(stat='identity')+
   geom_errorbar(aes(ymin=sumlower,ymax=sumupper),width=0.5)+facet_wrap(~annotate)+
   theme_bw(base_family="Arial")+
   theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=14),
	axis.text.y=element_text(size=14,colour='black'),
	axis.title.y=element_text(size=14))+ylab("Case interactions (p<0.01) (%)")+
   xlab("Tumor topography")+ggtitle("TME penetrating heterogeneity")+
   #scale_fill_manual(values=c('gray49','black','lightpink','tan3', 'khaki1','#B17BA6'))+
    scale_fill_manual(values=c('gray49','lightpink','#B17BA6'))+
   geom_hline(yintercept=0,color='red',linetype='dashed')
 
 
 save(jointSummary4F, file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/jointSummary4F.RData")
# ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/labeled-revision-p025/boundary.labeled.interactions.tumor.neighborhood.rowSummary.png")

  
 
  #levels(tp4$annotate)<- tp4$annotate%>%unique


 ## CD8
# ggplot(tp4[tp4$annotate%in%c("Ki67+ CD8","CD8","Exhausted CD8")
 cd84F<-   ggplot(tp4[tp4$annotate%in%c("Baseline CD8(7)","Activated/Prolif. CD8(1,4)","Exhausted/Inflam. CD8(2,6)","Terminally Exhaust. CD8(3,5)"),]
  ,aes(x=penetrant,y=value.x,fill=topography))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower.y,ymax=upper.y),width=0.5)+facet_wrap(~annotate)+
     theme_ipsum(base_family="Arial")+
     theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=16),
	axis.text.y=element_text(size=16,colour='black'),
	axis.title=element_text(size=16))+
     ylab("Case interactions (p<0.01) (%)")+
     xlab("Tumor topography")+
     ggtitle("TME penetrating heterogeneity")+
     scale_fill_manual(values=c('gray49','lightpink','#B17BA6'))+
    # scale_fill_manual(values=c('gray49','black','lightpink','tan3', 'khaki1','#B17BA6'))+
     geom_hline(yintercept=0,color='red',linetype='dashed')
 save(cd84F, file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/cd84F.RData")

 #ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/labeled-revision-p025/boundary.labeled.interactions.tumor.neighborhood.CD8.rowSummary.png")


 ## MAC
 #ggplot(tp4[tp4$annotate%in%c("Ki67+ M1-Macrophage","M2-Macrophage","PD-L1+ M2-Macrophage")
 mac4F<-ggplot(tp4[grepl("Mac",tp4$annotate),],
 aes(x=penetrant,y=value.x,fill=topography))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower.y,ymax=upper.y),width=0.5)+
   facet_wrap(~annotate)+theme_ipsum(base_family="Arial")+theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=16),
	axis.text.y=element_text(size=16,colour='black'),
	axis.title.y=element_text(size=16))+ylab("Case interactions (p<0.01) (%)")+
   xlab("Tumor topography")+ggtitle("TME penetrating heterogeneity")+
  scale_fill_manual(values=c('gray49','lightpink','#B17BA6'))+
  # scale_fill_manual(values=c('gray49','tan3','black','khaki1','#B17BA6'))+
   geom_hline(yintercept=0,color='red',linetype='dashed')
 save(mac4F, file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/mac4F.RData")

# ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/labeled-revision-p025/boundary.labeled.interactions.tumor.neighborhood.MAC.rowSummary.png")

 ## TREG
#ggplot(tp4[tp4$annotate%in%c("Ki67+ T-reg","T-reg","Highly suppressive T-reg")
 #tp4[which(tp4$primary=="Dispersed"& tp4$annotate=="Highly Suppress. T-reg(1,2,6)"),"value.y"]<-8.1

 treg4F<-ggplot(tp4[grepl("T-reg",tp4$annotate),],
 aes(x=penetrant,y=value.x,fill=topography))+
   geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower.y,ymax=upper.y),width=0.5)+
   facet_wrap(~annotate)+theme_ipsum(base_family="Arial")+
   theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=16),
	axis.text.y=element_text(size=16,colour='black'),
	axis.title.y=element_text(size=16))+
   ylab("Case interactions (p<0.01) (%)")+
   xlab("Tumor topography")+ggtitle("TME penetrating heterogeneity")+
    scale_fill_manual(values=c('gray49','lightpink','#B17BA6'))+
   #scale_fill_manual(values=c('gray49','tan3','black','khaki1','#B17BA6'))+
   geom_hline(yintercept=0,color='red',linetype='dashed')
 save(treg4F, file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/treg4F.RData")

#ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Figure6-final/labeled-revision-p025/boundary.labeled.interactions.tumor.neighborhood.TREG.rowSummary.png")



 ## CD4
#ggplot(tp4[tp4$annotate%in%c("Ki67+ memory CD4","CD45RA+ CD4","Memory CD4","CD45RA+ exhausted CD4",
#	"CD45RA- exhausted CD4","CD45RA+TIM3+ CD4")
cd4_4F<- ggplot(tp4[grepl("CD4",tp4$annotate),],
        aes(x=penetrant,y=value.x,fill=topography))+
   geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower.y,ymax=upper.y),width=0.5)+
   facet_wrap(~annotate)+theme_ipsum(base_family="Arial")+theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=16),
	axis.text.y=element_text(size=16,colour='black'),
	axis.title.y=element_text(size=16))+
   ylab("Case interactions (p<0.01) (%)")+
   xlab("Tumor topography")+
   ggtitle("TME penetrating heterogeneity")+
   scale_fill_manual(values=c('gray49','lightpink','#B17BA6'))+
 # scale_fill_manual(values=c('gray49','tan3','black','khaki1','#B17BA6'))+
   geom_hline(yintercept=0,color='red',linetype='dashed')
  save(cd4_4F, file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure4/cd4_4F.RData")

 
 
 fit<-lm(value.x~0+penetrant,tp4)
  cf<-fit%>%coef
  ci<-fit%>%confint%>%as.data.frame
  cd<-cbind(ci,cf)
  pp<-summary(fit)%>%coef%>%data.frame
  
  colnames(cd)<-c("lower","upper","Estimate")
   cd$topology<-c("Dispersed","Mantle/Crust","Core")
   cd$topology<-factor(cd$topology,levels=c("Dispersed","Mantle/Crust","Core"))
    ## add stars to bar
   cd$label<-ifelse(pp[,4]<0.05,"*","")
   
   
   
 ggplot(cd,aes(x=topology,y=Estimate,fill=topology))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.5)+scale_fill_manual(values=c("gray49","khaki1","#B17BA6"))+theme_ipsum()+theme(axis.text.x = element_text(angle = 90, hjust = 1,colour='black',size=14),
	axis.text.y=element_text(size=14,colour='black'),
	axis.title.y=element_text(size=14))+geom_text(data=cd,label=cd$label,nudge_y=c(1.5,0.9),size=10)+ylab("Cohort significant interaction (%)")
 
 

```



### Chapuy Cluster construction
- chapuy model association (perhaps deprecate)
- 
```{r}
 c1<-c("BCL6","BCL10","TNFAIP3","UBE2A","CD70","B2M","NOTCH2","TMEM30A","FAS","TP63","ZEB2","HLAB","SPEN","PDL1")
  c1.rank<-c(1,2,3,4,5,6,7,8,9,11,12,13,14,15)
  c1.dataFrame<-data.frame(gene=c1,Rank=c1.rank,cluster='C1',contained=FALSE)
  c2<-c("TP53")
  c2.rank<-1
  c2.dataFrame<-data.frame(gene=c2,Rank=c2.rank,cluster='C2',contained=FALSE)

  c3<-c("BCL2","CREBBP","EZH2","KMT2D","TNFRSF14","HVCN1","IRF8","GNA13","MEF2B","PTEN")
  c3.rank<-c(seq(1,length(c3)))
  c3.dataFrame<-data.frame(gene=c3,Rank=c3.rank,cluster='C3',contained=FALSE)

  c4<-c("SGK1","HIST1H1E","NFKBIE","BRAF","CD83","NFKBIA","CD58","HIST1H2BC","STAT3","HIST1H1C","ZFP36L1","KLHL6","HIST1H1D","HIST1H1B","ETS1","TOX","HIST1H2AM","HIST1H2BK","RHOA","ACTB","LTB","SF3B1","CARD11","HIST1H2AC")
  c4.rank<-seq(1,length(c4))
    c4.dataFrame<-data.frame(gene=c4,Rank=c4.rank,cluster='C4',contained=FALSE)

  c5<-c("CD79B","MYD88","ETV6","PIM1","TBL1XR1","GRHPR","ZC3H12A","HLAA","PRDM1","BTG1")
  c5.rank<-c(3,5,6,8,10,12,13,16,17,18)
    c5.dataFrame<-data.frame(gene=c5,Rank=c5.rank,cluster='C5',contained=FALSE)



   
   #clinic<-mutationClinicalSheetAssembly(mutationsToGather=c("TP53","BCL2","BCL6","CD79B","SGK1","MYD88","MYC","NOTCH1","NOTCH2","EZH2"),full.dn4=full.dn4)
  clinic<-mutationClinicalSheetAssembly(mutationsToGather=c(c1,c2,c3,c4,c5),full.dn4=full.dn4)

  ## Table for which mutations were found or not found.
     mutations<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/mutational-genomic-table/TMA CGI genomic data.csv",header=TRUE)

c1.dataFrame[c1.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c2.dataFrame[c2.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c3.dataFrame[c3.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c4.dataFrame[c4.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
c5.dataFrame[c5.dataFrame$gene%in%unique(mutations[,8]),'contained']<-TRUE
clusterData<-rbind(c1.dataFrame,c2.dataFrame,c3.dataFrame,c4.dataFrame,c5.dataFrame)



  table(c1%in%unique(mutations[,8]))
  
  tableSum<-data.frame(Mutation=c(c1,c2,c3,c4,c5),class=c(rep("C1",length(c1)),rep("C2",length(c2)),rep("C3",length(c3)),rep("C4",length(c4)),rep("C5",length(c5))  ) )
 tableSum$found<-"NO"
 
 tableSum$found[tableSum$Mutation%in%c1[c1%in%unique(mutations[,8])]]<-'YES'
 tableSum$found[tableSum$Mutation%in%c2[c2%in%unique(mutations[,8])]]<-'YES'
 tableSum$found[tableSum$Mutation%in%c3[c3%in%unique(mutations[,8])]]<-'YES'
 tableSum$found[tableSum$Mutation%in%c4[c4%in%unique(mutations[,8])]]<-'YES'
 tableSum$found[tableSum$Mutation%in%c5[c5%in%unique(mutations[,8])]]<-'YES'

 table(tableSum$found,tableSum$class)
 
 ## REVISION 10-20 hans mutation association with C1-C5.
 ### HANS association with each mutation class logistic regression. 
 clinic$C1.count<-rowSums(clinic[,colnames(clinic)%in%c1])
 clinic$C2.count<-(clinic[,colnames(clinic)%in%c2])
 clinic$C3.count<-rowSums(clinic[,colnames(clinic)%in%c3])
 clinic$C4.count<-rowSums(clinic[,colnames(clinic)%in%c4])
 clinic$C5.count<-rowSums(clinic[,colnames(clinic)%in%c5])
 
 clinic$NGCB.bin<-ifelse(clinic$Cell.of.origin..HANS.=='NGCB',1,0)
 clinic$Case_Number<-as.numeric(substr(clinic$Case_ID,6,7))

 ## Fitting multiple models and selecting a hierarchical model for phenotypes.
cluster.counts<-as.data.frame.matrix(table(full.dn4$ROIID,full.dn4$tumor.5nn.umap.simple.label))
cluster.counts<-100*cluster.counts/rowSums(cluster.counts)
cluster.counts$Run.ID<-as.numeric(rownames(cluster.counts))
cluster.counts<-left_join(cluster.counts,clinic[,c("Run.ID","Case_Number","C1.count", "C2.count", "C3.count", "C4.count", "C5.count",'ipi','p53')],by='Run.ID')
cluster.counts$C1<-ifelse(cluster.counts$C1.count!=0,1,0)
cluster.counts$C2<-ifelse(cluster.counts$C2.count!=0,1,0)
cluster.counts$C3<-ifelse(cluster.counts$C3.count!=0,1,0)
cluster.counts$C4<-ifelse(cluster.counts$C4.count!=0,1,0)
cluster.counts$C5<-ifelse(cluster.counts$C5.count!=0,1,0)


cluster.counts$HANS<-factor(clinic[match(cluster.counts$Run.ID,clinic$Run.ID),'Cell.of.origin..HANS.'])
cluster.counts$NGCB.binary<-ifelse(cluster.counts$HANS=='NGCB',1,0)


 
 ## need to model C2-C5 and append to the figure.
arModelExtract<-function(cluster.counts=NULL,myForm="C1 ~ CD4_1+CD4_2+CD4_3+CD4_4+CD4_5+CD4_6+CD4_7+ipi"){
  myForm<-as.formula(myForm)
  fit.gee.ind.c <- geeglm(myForm, id = Case_Number, data = cluster.counts,
family = binomial(link = "logit"), corstr = "independence", scale.fix = TRUE)
  fit.gee.ar1.c <- update(fit.gee.ind.c, corstr = "ar1")
  summary(fit.gee.ar1.c)
 results<- coef(summary(fit.gee.ar1.c))
 print(results)
 return(results[which(results[,4]<0.059),])
}
 library(geepack)
 library(MuMIn)
 ##CD4_3,  CD4_2 (week)
# arModelExtract(cluster.counts=cluster.counts,myForm="C1 ~ CD45RA_CD42+CD45RANeg_exhaustedCD4+CD45RA_exhaustedCD41+Ki67_memoryCD46+MemoryCD4347+Ki67_CD814+ipi")
 
 
 
   ## we focus the model on TME and not tumor associations with mutations.    
   arModelExtract(cluster.counts=cluster.counts,myForm="C1 ~  Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary")
   
    c1.final.noIPI <- geeglm(C1 ~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary, 
                         id = Case_Number,
                         data = cluster.counts,
                         family = binomial(link = "logit"),
                         corstr = "exchangeable", scale.fix = TRUE);summary(c1.final.noIPI)
 c1.final.noIPI.result<-coef(summary(c1.final.noIPI))
 c1.final.noIPI.result$Bonf<-c1.final.noIPI.result[,4]

   
    arModelExtract(cluster.counts=cluster.counts,myForm="C2 ~  Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary")
      c2.final.noIPI <- geeglm(C2 ~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary, 
                         id = Case_Number,
                         data = cluster.counts,
                         family = binomial(link = "logit"),
                         corstr = "exchangeable", scale.fix = TRUE);summary(c2.final.noIPI)
 c2.final.noIPI.result<-coef(summary(c2.final.noIPI))
 c2.final.noIPI.result$Bonf<-c2.final.noIPI.result[,4]

   
    
 arModelExtract(cluster.counts=cluster.counts,myForm="C3 ~  Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary")
    c3.final.noIPI <- geeglm(C3 ~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary, 
                         id = Case_Number,
                         data = cluster.counts,
                         family = binomial(link = "logit"),
                         corstr = "exchangeable", scale.fix = TRUE);summary(c3.final.noIPI)
 c3.final.noIPI.result<-coef(summary(c3.final.noIPI))
 c3.final.noIPI.result$Bonf<-c3.final.noIPI.result[,4]

 
     arModelExtract(cluster.counts=cluster.counts,myForm="C4 ~  Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary")
       c4.final.noIPI <- geeglm(C4 ~  Tumor_a+Tumor_b+Tumor_d+Tumor_e+Tumor_g+NGCB.binary, 
                         id = Case_Number,
                         data = cluster.counts,
                         family = binomial(link = "logit"),
                         corstr = "exchangeable", scale.fix = TRUE);summary(c4.final.noIPI)
 c4.final.noIPI.result<-coef(summary(c4.final.noIPI))
 c4.final.noIPI.result$Bonf<-c4.final.noIPI.result[,4]
     
   arModelExtract(cluster.counts=cluster.counts,myForm="C5 ~  Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary")
     c5.final.noIPI <- geeglm(C5 ~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+ipi+NGCB.binary, 
                         id = Case_Number,
                         data = cluster.counts,
                         family = binomial(link = "logit"),
                         corstr = "exchangeable", scale.fix = TRUE);summary(c5.final.noIPI)
 c5.final.noIPI.result<-coef(summary(c5.final.noIPI))
 c5.final.noIPI.result$Bonf<-c5.final.noIPI.result[,4]
 
 

```







## Tumor D ~ CXCR3 model (this finalized version is in "blood-chl.vs.dlbcl-revision.Rmd)
- the finalized version is posted earlier, this is an improved model, but the final version here in this script averages D~CD4  + CD4~D and has the best model.
```{r}
   netw<-getNetwork(x2)
  netw$subannotated<-as.character(cluster_ids(sce,"merging4"))
  colnames(netw)<-c("annotated","subannotated")

    getNetwork(x2)<-netw
    unique(netw[,2])
   
   
    Hpool<-imcExperimentToHyperFrameContinuousMarks(imcExperiment=x2,phenotypeToUse =2)
   Hpool$ROI<-paste0("ROIID_",Hpool$ROI)
   Hpool$treatment<-colData(sce)$REF[match(Hpool$ROI,colData(sce)$sample_id)]

   H<-Hpool[which(H$ROI!="ROIID_3969"),]
  
##create a list of each ROI
## creates zone masks
 pstatList<-list()
 
  for(ROI in H$ROI){
    print(ROI)
 roi_4193<-H$point[[which(H$ROI==ROI)]]
 #allppp<-split(roi_4193)
 allppp<-split(roi_4193,factor(marks(roi_4193)$marks))
 phenos<-names(allppp)
  ##CD4/CD8/TREG/MAC/Endo are possible TME.
#cd4.mask<-intensityMask(H,ROI=ROI,tme="CD4")
cd4.mask<-intensityMaskContinuousMarks(H,ROI=ROI,tme="CD4")
 cd4.im<-cd4.mask$mask.image
cd4.tess<-cd4.mask$tess
cd4.tess$image<-cd4.im
# cd8.mask<-intensityMask(H,ROI=ROI,tme="CD8")
 cd8.mask<-intensityMaskContinuousMarks(H,ROI=ROI,tme="CD8")
cd8.im<-cd8.mask$mask.image
cd8.tess<-cd8.mask$tess
  cd8.tess$image<-cd8.im

# treg.mask<-intensityMask(H,ROI=ROI,tme="TREG")
 treg.mask<-intensityMaskContinuousMarks(H,ROI=ROI,tme="TREG")

   treg.tess<-treg.mask$tess
  treg.im<-treg.mask$mask.image
  treg.tess$image<-treg.im
##rename levels of TREG mask.
  
  #  mac.mask<-intensityMask(H,ROI=ROI,tme="MAC")
   mac.mask<-intensityMaskContinuousMarks(H,ROI=ROI,tme="Macrophage")
   mac.tess<-mac.mask$tess
   mac.im<-mac.mask$mask.image
   mac.tess$image<-mac.mask$mask.image
   
   
 ## protein pSTAT3 masks.
   test<-roi_4193
   marks(test)<-marks(test)[,c("CXCR3")]
  ##create a mask based on expression.
  #elv<-expr.im$pSTAT3
  elv<-Smooth(test)
  b<-quantile(elv,probs=(0:4)/4,type=2)
  Zcut<-cut(elv,breaks=b,labels=c(0,0,0,"CXCR3"))
  pstat3mask<-tess(image=Zcut)
  
   test<-roi_4193
   marks(test)<-marks(test)[,c("Lag3")]
  ##create a mask based on expression.
  #elv<-expr.im$pSTAT3
  elv<-Smooth(test)
  b<-quantile(elv,probs=(0:4)/4,type=2)
  Zcut<-cut(elv,breaks=b,labels=c(0,0,0,"Lag3"))
  lag3mask<-tess(image=Zcut)
   
  
  ##treg zone pSTAT3 overlap
  cd4.cxcr3<-intersect.tess(cd4.im,pstat3mask)
  
  

roi<-list(    cd4.zone=cd4.mask$zone,
              cd8.zone=cd8.mask$zone,
              treg.zone=treg.mask$zone,
              mac.zone=mac.mask$zone,
              ppp=roi_4193,
              cd4=cd4.im,
              cd8=cd8.im,
              mac=mac.im,
              treg=treg.im,
              cxcr3Global=pstat3mask,
              Lag3Global=lag3mask,
              cd4cxcr3=cd4.cxcr3
              )

    pstatList[[(length(pstatList)+1)]]<-roi
   }##ROI list 
 
    names(pstatList)<-H$ROI
   
 
    

  
 topologyModelFit<-function(ROI=NULL,referencePoint="Tumor_d",roiList=NULL){
    print(ROI)
  ppp<-roiList[[which(names(roiList)==ROI)]]$ppp
  spl<-split(ppp)
  
  ##model the activated CD8 TME point dendency
    pdl1m2<-spl[[referencePoint]]
    marks(pdl1m2)<-marks(pdl1m2)$marks
  ## relate CD4 ~ TH1 zones + exhausted CD8 + (grouped TREG) + Activated CD8
    th1zone<-roiList[[ROI]]$cd4cxcr3
  maczone<-roiList[[ROI]]$mac
   tregzone<-roiList[[ROI]]$treg
    cd8zone<-roiList[[ROI]]$cd8

  levels(th1zone$image)<-ifelse(grepl("CXCR3",levels(th1zone$image)),"CXCR3CD4","CD4")
  model1<- ppm(pdl1m2~th1zone+maczone+tregzone+cd8zone)
  return(model1)
  }
  topologyModelFit(ROI=targetROI[1],roiList=pstatList)
  ## 
 targetROI<- H$ROI[which(with(H,length(which(marks(point)$marks=="Tumor_d")))!=0)]
 # targetROI<-targetROI[which(targetROI!="ROIID_4184"& targetROI!="ROIID_3969")] 
   tr<- lapply(targetROI,function(x) topologyModelFit(ROI=x,roiList=pstatList))
  names(tr)<-targetROI
  highRisk<-exper_info2[match(targetROI,exper_info2$sample_id),"highIPI"] 

  dd<-data.frame(roi=targetROI,IPI=highRisk)
  dd$C1<-exper_info2$C1[match(dd$roi,exper_info2$sample_id)]
    dd$C2<-exper_info2$C2[match(dd$roi,exper_info2$sample_id)]
  dd$C3<-exper_info2$C3[match(dd$roi,exper_info2$sample_id)]
  dd$C4<-exper_info2$C4[match(dd$roi,exper_info2$sample_id)]
  dd$C5<-exper_info2$C5[match(dd$roi,exper_info2$sample_id)]

  ## TH1
  th1.error<-sapply(tr,function(x) sqrt(diag(vcov(x)))["th1zoneCXCR3CD4"])
  th1.coef<-sapply(tr,function(x) coef(x)["th1zoneCXCR3CD4"])
  dd$CD4<-th1.coef/th1.error

   ## exh hot spot    
  exh8.error<-sapply(tr,function(x) sqrt(diag(vcov(x)))["maczoneMacrophage"])
  exh8.coef<-sapply(tr,function(x) coef(x)["maczoneMacrophage"]) 
  dd$MAC<-exh8.coef/exh8.error

  ## Terminal  
 ex.error2<-sapply(tr,function(x) sqrt(diag(vcov(x)))["tregzoneTREG"])
  ex.coef2<-sapply(tr,function(x) coef(x)["tregzoneTREG"]) 
  dd$TREG<-ex.coef2/ex.error2
  
  
 rownames(dd)<-dd$roi
 aov(value~variable+IPI+C1+C2+C3+C4+C5,tt)%>%summary
  
  
  lm(CD4~C1+C2+C3+C4+C5+IPI,dd)%>%summary
    lm(MAC~C1+C2+C3+C4+C5+IPI,dd)%>%summary
    lm(TREG~C1+C2+C3+C4+C5+IPI,dd)%>%summary

  
      
    
    
     
```




### CXCR3 model at the cluster level (deprecated original version.)
 We examine the CXCR3 expression and the correlation with Tumor_d (core) interactions.
 This is supplementary Figure 6B.  This is the CXCR3 interactions with tumor core at the TME cluster (un-annotated) level.  At the cluster level (un-annotated), we observed significant assocation between the CXCR3 expression intensity and the number of tumor core interactions.
- 12-13 this section used the cluster association and is deprecated, included here for log only.
```{r cxcr3Cluster, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 14}
 load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/tumor-spatial-class5NN-unsupCommunity/p1_15_Tumor.5nn.unsup.RData")
  load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/full.dn4.5nn2.CD8.MC.RData")
   zz<- zScorePatientExpression(full.dn4=full.dn4.5nn2,markers=c(1:31,104:111),ROIID.column=32)

 ## count the correlation both ways.
 ## CD4 ~ Tumor_d
 ## Tumor_d ~ CD4
 ## but divide by 2.

 spatCols<-getSpatialPhenoColors(full.dn4.5nn2)

  patZ<-zz%>%group_by(ROIID,tumor.5nn.class.unsup)%>%summarise(TIM3=mean(Cell_Tim3),
	CXCR3=mean(Cell_CXCR3),
	PD1=mean(Cell_PD1),
	CCR4=mean(Cell_CCR4),
	LAG3=mean(Cell_Lag3))%>%data.frame
 cxcr3<-patZ%>%group_by(tumor.5nn.class.unsup)%>%summarise(CXCR3=mean(CXCR3))%>%data.frame
  ## only the CD4 CXCR3 expression.
 cxcr3<-data.frame(cxcr3[grepl("CD4_",cxcr3[,1]),])
 library(ggrepel)
 ## gather the interactions.
  ints<- as.data.frame(p1$pmat)
 d.cd4<-ints['Tumor_d',grepl("CD4_",colnames(ints))]
 cd4.d<-ints[match(names(d.cd4),rownames(ints)),'Tumor_d']
  avg.ints<- rbind(d.cd4,cd4.d)%>%colMeans

 #cd4ints<-t(totalPrimary2['core',grepl("CD4_",colnames(totalPrimary2))])%>%data.frame(tumor.5nn.class.unsup=rownames(t(totalPrimary2['core',grepl("CD4_",colnames(totalPrimary2))])))%>%data.frame
 cd4ints<-data.frame(avg.ints)
 cd4ints$tumor.5nn.class.unsup<-rownames(cd4ints)

  cxcr3<-left_join(cxcr3,cd4ints,by='tumor.5nn.class.unsup')
  cor.test( cxcr3$CXCR3,cxcr3$avg.ints)
  cxc<-ggplot(cxcr3,aes(x=CXCR3,y=avg.ints))+geom_point(size=6,aes(fill=tumor.5nn.class.unsup),pch=21,colour='black')+geom_smooth(method='lm',se=FALSE,col='red')+scale_fill_manual(values=spatCols[match(cxcr3$tumor.5nn.class.unsup,spatCols$phenotype),'colors'])+xlab("CD4 CXCR3 expression (Z-score)")+ylab("Signif. tumor core interactions (%)")+theme_ipsum(base_family='Arial')+theme(axis.text.x=element_text(size=12), 
	axis.text.y = element_text(size=14))+ggtitle("CD4 CXCR3 expression assocation with tumor core interactions")+ggplot2::annotate('text',x=-0.51,y=4.2,label="p-value=0.024, R^2=0.82")+geom_text_repel(label=cxcr3$tumor.5nn.class.unsup,point.padding=0.5)+geom_vline(xintercept=0,linetype='dashed',colour='black')

  print(cxc)
   cor.test( cxcr3$CXCR3,cxcr3$avg.ints)

  #ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Tumor-analysis-5NN/CXCR3.Core.interactions.Correlation.CD4.png")

  
### deprecated shit
   
   deprecated<-function(){   
 alltumor<-ints[grepl("Tumor",rownames(ints)),grepl("CD4",colnames(ints))]
 alltumor$topography<-"Dispersed"
 alltumor$topography[rownames(alltumor)%in%c("Tumor_c","Tumor_h","Tumor_f","Tumor_i")]<-"Mantle"
 alltumor$topography[which(rownames(alltumor)=="Tumor_d")]<-"Core"
  
 top.ints<-alltumor%>%group_by(topography)%>%summarise_all(funs(mean))%>%as.data.frame
 rownames(top.ints)<-top.ints[,1]
 top.ints<-top.ints[,-1]
 top.ints<-as.data.frame(t(top.ints))
 top.ints$tumor.5nn.class.unsup<-rownames(top.ints)
 
  top.ints$family<-as.character(full.dn4$annotated.label.back_akil[match(top.ints$tumor.5nn.class.unsup,full.dn4$unsup.subcommunity)])
  

 
 cxcr3<-left_join(cxcr3,top.ints,by="tumor.5nn.class.unsup") 
  rownames(cxcr3)<-cxcr3$tumor.5nn.class.unsup
  #cxcr3<-cxcr3[,-1]
  cxcr3<-cxcr3[,which(colnames(cxcr3)!="family.x")]
  colnames(cxcr3)[which(colnames(cxcr3)=="family.y")]<-"family"
lm(CXCR3~Core+Mantle+Dispersed,cxcr3)%>%summary
 sink("~/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-model-worksheets/Figure2-9-20/tumor-topology/topology.MLR.txt")
lm(CXCR3~Core+Mantle+Dispersed,cxcr3)%>%summary
sink()
 cxcr3$CD4<-rownames(cxcr3)
c2<-melt(cxcr3,id.vars=c("CXCR3","CD4"))
 ggplot(c2,aes(x=value,y=CXCR3,colour=variable))+geom_point()+geom_smooth(method='lm',se=FALSE)+facet_wrap(~variable)+theme_ipsum(base_family="Arial")+xlab("CD4 Spatial interaction with tumor topography (q<0.01)")+ylab("CD4 CXCR3 mean intensity (Z-score)")
 

  spatCols$phenotype<-full.dn4$annotated.label.back_akil[match(spatCols$phenotype,full.dn4$unsup.subcommunity)]
   cxc<-ggplot(cxcr3,aes(x=CXCR3,y=Mantle))+geom_point(size=6,aes(fill=family),pch=21,colour='black')+
     geom_smooth(method='lm',se=FALSE,col='red')+
     scale_fill_manual(values=spatCols[match(cxcr3$family,spatCols$phenotype),'colors'])+
     xlab("CD4 CXCR3 expression (Z-score)")+ylab("Tumor Mantle interactions (p<0.01) (%)")+
     theme_ipsum(base_family='Arial')+theme(axis.text.x=element_text(size=12), 
	axis.text.y = element_text(size=14),axis.text=element_text(color="black"))+
     ggtitle("CD4 CXCR3 expression assocation with tumor mantle interactions")+
     #ggplot2::annotate('text',x=-0.51,y=0.9,size=4,label="p-value=0.0589, Adj. R^2=0.78")+
      geom_label_repel(
        alpha=0.75,
        label.size=0.85,
     aes(label=factor(family),size=1.5),
     fontface = 'bold', color = 'black',
    box.padding = unit(0.45, "lines"),
    point.padding = unit(0.15, "lines"),
    segment.color = 'grey50',
    show.legend=FALSE)
   
    
  
  
 
 #cd4ints<-t(totalPrimary2['core',grepl("CD4_",colnames(totalPrimary2))])%>%data.frame(tumor.5nn.class.unsup=rownames(t(totalPrimary2['core',grepl("CD4_",colnames(totalPrimary2))])))%>%data.frame
 cd4ints<-data.frame(avg.ints)
 cd4ints$tumor.5nn.class.unsup<-rownames(cd4ints)

  cxcr3<-left_join(cxcr3,cd4ints,by='tumor.5nn.class.unsup')
  cor.test( cxcr3$CXCR3,cxcr3$avg.ints)
  cxc<-ggplot(cxcr3,aes(x=CXCR3,y=avg.ints))+geom_point(size=6,aes(fill=tumor.5nn.class.unsup),pch=21,colour='black')+geom_smooth(method='lm',se=FALSE,col='red')+scale_fill_manual(values=spatCols[match(cxcr3$tumor.5nn.class.unsup,spatCols$phenotype),'colors'])+xlab("CD4 CXCR3 expression (Z-score)")+ylab("Signif. tumor core interactions (%)")+theme_ipsum(base_family='Arial')+theme(axis.text.x=element_text(size=12), 
	axis.text.y = element_text(size=14))+ggtitle("CD4 CXCR3 expression assocation with tumor core interactions")+ggplot2::annotate('text',x=-0.51,y=4.2,label="p-value=0.024, R^2=0.82")+geom_text_repel(label=cxcr3$tumor.5nn.class.unsup,point.padding=0.5)+geom_vline(xintercept=0,linetype='dashed',colour='black')

  print(cxc)
   cor.test( cxcr3$CXCR3,cxcr3$avg.ints)

  #ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Tumor-analysis-5NN/CXCR3.Core.interactions.Correlation.CD4.png")

  

   
   
    #### family level.
   load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/spatial-analysis/p1_15_annotatedLabel.tumor.5nn.umap.simple.label.15distance.025.RData")
   
    patZ<-zz%>%group_by(ROIID, tumor.5nn.label.back)%>%summarise(TIM3=mean(Cell_Tim3),
	CXCR3=mean(Cell_CXCR3),
	PD1=mean(Cell_PD1),
	CCR4=mean(Cell_CCR4),
	LAG3=mean(Cell_Lag3))%>%data.frame
 cxcr3<-patZ%>%group_by(tumor.5nn.label.back)%>%summarise(CXCR3=mean(CXCR3))%>%data.frame
  ## only the CD4 CXCR3 expression.
 cxcr3<-data.frame(cxcr3[grepl("CD4",cxcr3[,1]),])
 library(ggrepel)
 ## gather the interactions.
  ints<- as.data.frame(p1$pmat)
 d.cd4<-ints['Tumor_d',grepl("CD4",colnames(ints))]
 cd4.d<-ints[match(names(d.cd4),rownames(ints)),'Tumor_d']
  avg.ints<- rbind(d.cd4,cd4.d)%>%colMeans

 #cd4ints<-t(totalPrimary2['core',grepl("CD4_",colnames(totalPrimary2))])%>%data.frame(tumor.5nn.class.unsup=rownames(t(totalPrimary2['core',grepl("CD4_",colnames(totalPrimary2))])))%>%data.frame
 cd4ints<-data.frame(avg.ints)
 cd4ints$tumor.5nn.label.back<-full.dn4$tumor.5nn.label.back[match(rownames(cd4ints),full.dn4$annotated.label.simple_akil)]

  cxcr3<-left_join(cxcr3,cd4ints,by='tumor.5nn.label.back')
  cor.test( cxcr3$CXCR3,cxcr3$avg.ints)
  cxc2<-ggplot(cxcr3,aes(x=CXCR3,y=avg.ints))+geom_point(size=6,aes(fill=tumor.5nn.label.back),pch=21,colour='black')+geom_smooth(method='lm',se=FALSE,col='red')+scale_fill_manual(values=phenoColors[match(cxcr3$tumor.5nn.label.back,phenoColors$phenotype),'colors'])+xlab("CD4 CXCR3 expression (Z-score)")+ylab("Signif. tumor core interactions (%)")+theme_ipsum(base_family='Arial')+theme(axis.text.x=element_text(size=12), 
	axis.text.y = element_text(size=14))+ggtitle("CD4 CXCR3 expression assocation with tumor core interactions")+ggplot2::annotate('text',x=-0.51,y=4.2,label="p-value=0.066, R^2=0.87")+geom_text_repel(label=cxcr3$tumor.5nn.label.back,point.padding=0.5)+geom_vline(xintercept=0,linetype='dashed',colour='black')

  print(cxc)
   cor.test( cxcr3$CXCR3,cxcr3$avg.ints)
   }
   
   

```

### Write CXCR3 cluster to table
```{r, eval=FALSE}
  writeData(fig6,"CD4_5.CXCR3.LMER",
	data.frame(estimate=  cor.test( cxcr3$CXCR3,cxcr3$avg.ints)$estimate,p= cor.test( cxcr3$CXCR3,cxcr3$avg.ints)$p.value,row.names="correlation.CD45.Td"),
	rowNames=TRUE,withFilter=FALSE,colNames=TRUE,startCol=1,startRow=17)
 
  openxlsx::saveWorkbook(fig6,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/Tumor-analysis-5NN/Figure6Table.xlsx",overwrite=TRUE)


```



### Tumor topology Clark-evans with Chapuy and IPI revision (MLR)
- 7-30: revision perform MLR with Chapuy, COO and IPI.
- performs the clark-evans test in MLR with Chapuy and IPI association, comparing within DLBCL significance is achieved by using topology d,c,f,g,i.
- also compares across lymph node examining COO and RLN.
- creates Figure 5D, and also performs multviariate
- requires the flowFrame section to be ran.
- 12-12 produces the cartoons and boxplot for revised Figure 3.
- 1-8: uses exper_info object from the clinical sheet, removed dependency on fsimc object from outside script.
- 1-8: includes CoxPh model estimates for the Clark-Evans (we observe the null and do not report).
```{r}
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/revision-comments-11-6/steidl-lymphode-classification/steidl-RLN-classification-analysis-Case4-1-23/rln.dn4.nbhd.5nn.1-23.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/clinic.RData")


   table(rln.nbhd.5nn$primary.phenotype)
    ## 
   rln.nbhd.5nn$case%>%table
  
   rln<-NULL
  rois<-c("Case1_6" , "Case2_7",  "Case3_8" , "Case4_9",  "Case5_10")
  for(i in rois){
  print(i)
  data<-subset(rln.nbhd.5nn,rln.nbhd.5nn$ROIID==i)
  data<-data[which(data$primary.phenotype=='BCell'),]
     pp<-phenographToPPP(case=data,marksCase=factor(data$primary.phenotype),shift=FALSE)
   H<-hyperframe(point=pp,
	#response=unique(data$response),
	#COO=unique(data$COO),
	ROI=i)
   if(is.null(rln)!=TRUE){
  rln<-rbind(H,rln)
  }else{
 rln<-H
  } 
 } ###

  rln$z<-with(rln,clarkevans.test(point,alternative='clustered',correction='Donnelly')$statistic)
  
HH3<-NULL
   rois<-unique(full.dn4$ROIID)
  for(i in rois){
  print(i)
  data<-subset(full.dn4,full.dn4$ROIID==i)
     pp<-phenographToPPP(case=data,marksCase=factor(data$tumor.5nn.class.unsup),shift=FALSE)
   H<-hyperframe(point=pp,
	response=unique(data$response),
	ROI=i)
   if(is.null(HH3)!=TRUE){
  HH3<-rbind(H,HH3)
  }else{
 HH3<-H
  } 
 } ###

 HH3$COO<-full.dn4$COO[match(HH3$ROI,full.dn4$ROIID)]
  mantle<-HH3
  
 man4<-with(HH3, point[which(marks(point)%in%unique(c(
	'Tumor_c',
	'Tumor_d',
	'Tumor_f',
	'Tumor_g',
	'Tumor_i')))])
   names(man4)<-HH3$ROI
  id<-which(sapply(man4,npoints)!=0)
  man4<-hyperframe(point=man4[id],ROI=names(man4)[id])
  man4$COO<-man4$COO[match(man4$ROI,HH3$ROI)]
 ## border corrections on Clark-Evans statistic.
  man4$z<-with(man4,clarkevans.test(point,alternative='clustered',correction='Donnelly')$statistic)
   man4$COO<-HH3$COO[match(man4$ROI,HH3$ROI)]

    ## drop tumor_d core to identify best model.
  man3<-with(HH3, point[which(marks(point)%in%unique(c(
	'Tumor_c',
	'Tumor_i',
	'Tumor_f',
	'Tumor_g')))])
   names(man3)<-mantle$ROI
   id<-which(sapply(man3,npoints)!=0)
  man3<-hyperframe(point=man3[id],ROI=names(man3)[id])
  man3$COO<-mantle$COO[match(man3$ROI,mantle$ROI)]
 ## border corrections on Clark-Evans statistic.
  man3$z<-with(man3,clarkevans.test(point,alternative='clustered',correction='Donnelly')$statistic) 
  
  
  ####load and add the clinical data.
   exper_info<-clinic
   exper_info$sample_id<-paste0("ROIID_",exper_info$Run.ID)
   exper_info$highIPI<-ifelse(exper_info$ipi>=3,1,0)
 
 man3$C1<-exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),c("C1")]
 man3$C2<-exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),c("C2")]
 man3$C3<-exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),c("C3")]
 man3$C4<-exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),c("C4")]
 man3$C5<-exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),c("C5")]
 man3$sample_id<-factor(exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),c("sample_id")])
 man3$highIPI<-exper_info[match(paste0("ROIID_",man3$ROI),exper_info$sample_id),"highIPI"]

 all(man4$ROI==man3$ROI)

 man4$C1<-man3$C1
 man4$C2<-man3$C2
 man4$C3<-man3$C3
 man4$C4<-man3$C4
 man4$C5<-man3$C5
  man4$highIPI<-man3$highIPI


  ###using man4 object all tumor topology (man4 contains c,d,f,g,i)
  cedata<-data.frame(COO=man4$COO,
                     z=man4$z,
                     row.names=man4$ROI,
                      C1=factor(ifelse(man4$C1==1,"C1","non-C1")),
                      C2=factor(ifelse(man4$C2==1,"C2","non-C2")),
                      C3=factor(ifelse(man4$C3==1,"C3","non-C3")),
                      C4=factor(ifelse(man4$C4==1,"C4","non-C4")),
                      C5=factor(ifelse(man4$C5==1,"C5","non-C5")),
                      highIPI=factor(ifelse(man4$highIPI==1,"highIPI","lowIPI")))
  
     ce.fit<-lm(z~C1+C2+C3+C4+C5+highIPI,cedata)

    clarkEvans.fit<-aov(z~COO+C1+C2+C3+C4+C5+highIPI,cedata)
  clarkEvans.fit%>%TukeyHSD
  
   clarkEvans.fit<-aov(z~COO+C1+C2+C4+C5+highIPI,cedata)
  clarkEvans.fit%>%TukeyHSD
  
	library(rstatix)
   stat.test<-  clarkEvans.fit%>%tukey_hsd()
   stat.test
   
  ### creating model compared against RLN, and DLBCL. 
  ##uses man3 object, best model by dropping tumor_d  
     cedata<-data.frame(COO=man3$COO,z=man3$z,row.names=man3$ROI)
  cedata$case<-full.dn4$case[match(rownames(cedata),full.dn4$ROIID)]
  cedata$case<-factor(cedata$case)
  
 ## RLN clark evans.
   rlnZ<-data.frame(COO="RLN",z=rln$z,case=rln$ROI)
  ce.d<-rbind(cedata,rlnZ)
  ce.d$COO<-relevel(ce.d$COO,ref='RLN')


 ce.d$COO <- as.factor(ce.d$COO)
  clarkEvans.fit<-aov(z~COO,ce.d)
  clarkEvans.fit%>%TukeyHSD
	library(rstatix)
   stat.test_rln<-  clarkEvans.fit%>%tukey_hsd()
   
 library(ggpubr)
  ggboxplot(ce.d, x = "COO", y = "z",fill='COO') +
  stat_pvalue_manual(
    stat.test_rln, label = "p.adj", 
    y.position = c(1.8, 1.6, 1.4)
 )+ylab("Clark-evans test statistic")+scale_fill_manual(values=c('khaki2','red','royalblue2'))+xlab("HANS classification")+ggtitle("Spatial clustering of tumor topography")

  
  ## save output for Tucker.
  figure3Box<- ggboxplot(ce.d, x = "COO", y = "z",fill='COO') +
  stat_pvalue_manual(
    stat.test_rln, label = "p.adj", 
    y.position = c(1.8, 1.6, 1.4)
 )+ylab("Clark-evans test statistic")+scale_fill_manual(values=c('khaki2','red','royalblue2'))+xlab("HANS classification")+ggtitle("Spatial clustering of tumor topography")

  save(figure3Box,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/figure3Box.RData")
  
  ### Cox model.
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/SCE-data/sce_omitND.RData")
   missingSurv<-c("3978",  "3979" , "39792", "4070","4071")
      noTreat<-full.dn4$ROIID[which(full.dn4$treatment.status=="ND")]%>%unique
      missingSurv<-paste0("ROIID_",missingSurv)
      noTreat<-paste0("ROIID_",noTreat)
      
   clinic2<- metadata(sce)$experiment_info
   clinic2$os_d<-full.dn4$os_d[match(clinic2$sample_id,paste0("ROIID_",full.dn4$ROIID))]
   clinic2$os<-full.dn4$os[match(clinic2$sample_id,paste0("ROIID_",full.dn4$ROIID))]
  clinic2$os_ref<-full.dn4$response[match(clinic2$sample_id,paste0("ROIID_",full.dn4$ROIID))]

### death status  
  clinic2$status<-ifelse(clinic2$os=="event",1,0)
    
## refractory status
      clinic2$refstatus<-ifelse(clinic2$os_ref=="Primary refract",1,0)

    
clinic2$C1f<-factor(clinic2$C1)
clinic2$C2f<-factor(clinic2$C2)
clinic2$C3f<-factor(clinic2$C3)
clinic2$C4f<-factor(clinic2$C4)
clinic2$C5f<-factor(clinic2$C5)
 ## individual hazard plots
man4$sample_id<-paste0("ROIID_",man4$ROI)
 clinic3<-left_join(clinic2,as.data.frame(man3[,c("sample_id","z")]),by="sample_id")
 clinic3<-clinic3[!is.na(clinic3$z),]

 coxph(Surv(os_d,status)~z+highIPI,clinic3)

chpcx<-coxph(Surv(os_d,status)~C1+C2+C3+C4+C5+highIPI,clinic3)
 

  
  
  
  
  ## plotting cartoons
    plotShape<-function(path="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-shape-objects-revised/final-shapes/",ROI="3964",cols=NULL,full.dn4,compartment=NULL,phenotype=NULL,phenotype.column.toShape=NULL){
  # sh<-dir("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-shape-objects-revised/final-shapes/")
    sh<-dir(path)
  n<-sh[grepl(ROI,sh)]
  load(paste0(path,n))
  spdf$cellType2<-sapply(strsplit(as.character(spdf$cellType),"_"),function(x) x[1])
  mycase<-full.dn4[which(full.dn4$ROIID==ROI),]
  tumor.id<-mycase$uniqueLabel[which(mycase[,compartment]==phenotype) ]
  spdf$uniqueLabel<-spdf$id
  spdf<-left_join(spdf,mycase[,c("uniqueLabel",phenotype.column.toShape)],by="uniqueLabel")
   spdf$cellType3<-factor(spdf[,phenotype.column.toShape],levels=unique(spdf[,phenotype.column.toShape]))
 test<- ggplot(spdf,aes(x=long,y=lat,group=id,fill=cellType3)) +
      geom_polygon( size=0.1,colour="black") +scale_fill_manual(values=cols[match(levels(spdf$cellType3),cols$phenotype),"colors"])+ggtitle(ROI)+theme_bw()
 
  # print(test)
  return(test)
    }
    
     color_clusters <- c(#DC050C","#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
                    "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
                     "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
                     "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
                     "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
                     "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")
  

  phenoColors<-data.frame(colors=color_clusters[1:19],phenotype=as.character(unique(roi$unsup.subcommunity)))
  phenoColors$colors[grepl("Tumor_",phenoColors$phenotype)]<-'grey'
  phenoColors$colors[grepl("CD4_1",phenoColors$phenotype)]<-'red'
    phenoColors$colors[grepl("CD8",phenoColors$phenotype)]<-'green'
        phenoColors$colors[grepl("TREG_",phenoColors$phenotype)]<-'purple'
            phenoColors$colors[grepl("Macrophage_",phenoColors$phenotype)]<-'lightpink'

 tmeColors<-data.frame(colors=c("magenta","purple","blue","green","red","grey"),
                        phenotype=c("CD68 Macrophage","FOXP3 T REG","T Helper","T cytotoxic","Endothelial","Tumor"))

  
	c30<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/cartoons/cartoon-shapeFiles/final-shapes/",
          ROI='4183',
	cols=tmeColors,
	full.dn4,
	compartment="rename",
	phenotype="CD4_1",
	phenotype.column.toShape="rename")
  print(c30)

  c17<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/cartoons/cartoon-shapeFiles/final-shapes/",
          ROI='3974',
	cols=tmeColors,
	full.dn4,
	compartment="rename",
	phenotype="CD4_1",
	phenotype.column.toShape="rename")
  print(c17)

 save(c30,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/c30_cartoon.RData")  
 save(c17,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/c17_cartoon.RData")  

  ### plot with the topology colors.
  ## need the RLN colors.
  rlnColors<-data.frame(colors=c("magenta","purple","blue","green","red","grey","white","yellow","azure","azure","cyan"),
                        phenotype=c("MAC","TREG","CD4","CD8","Endothelial","BCell","empty","DC","HRS","Myeloid.Granulocytes","NK"))
  
  ##case 4_9 shows BCells 
	case4_9<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/shape-objects-2-4/",
          ROI='Case4_9_',
	cols=tmeColors,
	full.dn4,
	compartment="rename",
	phenotype="CD4_1",
	phenotype.column.toShape="rename")
  print(case4_9)
 
  save(case4_9,file=
          "C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/case4_9.RLN.TMEColors.RData")
  
  
### now show the topology colored images.
   full.dn4$topology.TME<-as.character(full.dn4$rename)
   full.dn4$topology.TME[full.dn4$rename=="Tumor"]<-full.dn4$tumor.5nn.class.unsup[full.dn4$rename=="Tumor"]
    spatcols<-getSpatialPhenoColors(full.dn4=full.dn4)
   spatcols2<-data.frame(phenotype=unique(full.dn4$topology.TME))
     spatcols2$colors<-spatcols$colors[match(spatcols2$phenotype,spatcols$phenotype)]   
     spatcols2$colors[spatcols2$phenotype=="T cytotoxic"]<-"green"
      spatcols2$colors[spatcols2$phenotype=="Endothelial"]<-"red"
      spatcols2$colors[spatcols2$phenotype=="T Helper"]<-"blue"
      spatcols2$colors[spatcols2$phenotype=="CD68 Macrophage"]<-"magenta"
            spatcols2$colors[spatcols2$phenotype=="FOXP3 T REG"]<-"purple"
      spatcols2$colors[spatcols2$phenotype=="none"]<-"azure"
      spatcols2$colors[spatcols2$phenotype=="Tumor_d"]<-"plum4"


  
	c30.top<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/cartoons/cartoon-shapeFiles/final-shapes/",
          ROI='4183',
	cols=spatcols2,
	full.dn4,
	compartment="topology.TME",
	phenotype="CD4_1",
	phenotype.column.toShape="topology.TME")
  print(c30.top)

  c17.top<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/cartoons/cartoon-shapeFiles/final-shapes/",
          ROI='3974',
	cols=spatcols2,
	full.dn4,
	compartment="topology.TME",
	phenotype="CD4_1",
	phenotype.column.toShape="topology.TME")
  print(c17.top)

 save(c30.top,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/c30_4183_cartoon_topologyColors.RData")  
 save(c17.top,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/c17_3974_cartoon_topologyCOlors.RData")  

  ### plot with the topology colors.
  ## need the RLN colors.
  topoColors<-data.frame(colors=c("magenta","purple","blue","green","red","white","yellow","azure","azure","cyan","gray89","plum4","khaki1","black","lightslategrey","orange"),
                        phenotype=c("MAC","TREG","CD4","CD8","Endothelial","empty","DC","HRS","Myeloid.Granulocytes","NK","2","3","4","5","6","7"))
  
  ##case 4_9 shows BCells 
	case4_9<-plotShape(path="C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/shape-objects-2-4/",
          ROI='Case4_9_',
	cols=topoColors,
	full.dn4,
	compartment="rename",
	phenotype="CD4_1",
	phenotype.column.toShape="rename")
  print(case4_9)
 
  save(case4_9,file=
          "C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Final-figures-Tucker/Figure3/case4_9.RLN.TMEColors.RData")
  
  
```


## Revision comments 3-20-2022
Reviewer asked what the topological cluster association related to the phenotype clusters are.  We perform beta regression on the proportions
```{r}
  tumor<-full.dn4[which(full.dn4$compartment=="Tumor"),]
 library(corrplot)
library(RColorBrewer)

 corDat<-as.data.frame.matrix(table(tumor$tumor.5nn.class,tumor$annotated.label.back_akil))
 corDat<-corDat[,which(colSums(corDat)>0)]
 corDat<-corDat/sum(corDat)

 ## cluster level
  absol.prop<-as.data.frame.matrix(table(full.dn4.5nn2$case,full.dn4.5nn2$tumor.5nn.class.unsup))
  absol.prop<-absol.prop/rowSums(absol.prop)
    absol.prop<-absol.prop[,grepl("Tumor",colnames(absol.prop))]
  absol.prop$dispersed<-absol.prop[,c("Tumor_a","Tumor_b","Tumor_e","Tumor_g")]%>%rowSums
  absol.prop$mantleCrust<-absol.prop[,c("Tumor_c","Tumor_h","Tumor_i","Tumor_f")]%>%rowSums

  
  absol.prop$HANS<-full.dn4.5nn2$COO[match(rownames(absol.prop),full.dn4.5nn2$case)]
 absol.prop$NGCB<-ifelse(absol.prop$HANS=="NGCB",1,0)
  absol.prop$IPI<-full.dn4.5nn2$IPI[match(rownames(absol.prop),full.dn4.5nn2$case)]

    absol.prop2<-as.data.frame.matrix(table(full.dn4.5nn2$case,full.dn4.5nn2$unsup.subcommunity))
  absol.prop2<-absol.prop2/rowSums(absol.prop2)
  absol.prop2<-absol.prop2[,grepl("Tumor",colnames(absol.prop2))]
  absol.prop2$inflamTumor<-absol.prop2[,c("Tumor_6","Tumor_8","Tumor_10")]%>%rowSums
  absol.prop2$Tumor<-absol.prop2[,c("Tumor_1","Tumor_2","Tumor_3","Tumor_4","Tumor_5","Tumor_7","Tumor_9")]%>%rowSums
  ## join the topology and tumor cluster
  
  joinProp<-cbind(absol.prop,absol.prop2)
  joinProp$IPI<-ifelse(joinProp$IPI>3,1,0)
  
  ## beta regression
  library(betareg)
#### modeling   
   dispFit<- betareg(dispersed~inflamTumor+Tumor+IPI+NGCB,dplyr::filter(joinProp,dispersed>0))
  confint(dispFit)
  
   mantFit<-betareg(mantleCrust~inflamTumor+Tumor+IPI+NGCB,dplyr::filter(joinProp,mantleCrust>0))
   confint(mantFit)
   
    coreFit<-betareg(Tumor_d~inflamTumor+Tumor+IPI+NGCB,dplyr::filter(joinProp,Tumor_d>0))
    summary(coreFit)
    confint(coreFit)
    


  spatCol<-getSpatialPhenoColors(full.dn4.5nn2)

  tum.lr<-glm(NGCB~Tumor_a+Tumor_b+Tumor_c+Tumor_d+Tumor_e+Tumor_f+Tumor_g+Tumor_h+Tumor_i+IPI,family='binomial',absol.prop)
  mlr.lr<-tum.lr%>%confint%>%data.frame
  mlr.lr$estimate<-coef(tum.lr)
   mlr.lr<-mlr.lr[order(mlr.lr$estimate),]
  mlr.lr<-mlr.lr[grepl("Tumor_",rownames(mlr.lr)),]
  mlr.lr$phenotype<-rownames(mlr.lr)
  mlr.lr$phenotype<-factor(mlr.lr$phenotype,levels=rownames(mlr.lr))

  ggplot(mlr.lr,aes(x=phenotype,y=estimate))+geom_point(size=4,aes(color=phenotype))+geom_errorbar(aes(ymin=X2.5..,ymax=X97.5..),width=0.2,color='grey39')+scale_color_manual(values=spatCol$colors[match(levels(mlr.lr$phenotype),spatCol$phenotype)])+theme_ipsum(base_family='Arial')+theme(axis.text.x = element_text(angle = 90, hjust = 1))+xlab("Spatial phenotype")+ylab("Non-GC log-odds (IPI adjusted)")+ggtitle("Spatial Phenotype Non-GC association")+geom_hline(yintercept=0,linetype='dashed',color='red')
 
  
  
 
 
 
 ## pooled level
 
 
```
