---
title: "Imaging mass cytometry reveals tumor and immune spatial and phenotypic clusters associated with clinical outcomes in diffuse large B cell lymphoma (DLBCL)"
author: " Anthony Colombo+,Monirath Hav+, Erik Gerdtsson, Mohan Singh, Denaly Chen, James Hicks, Peter Kuhn, Imran Siddiqi, Akil Merchant*"
date: "`r Sys.Date()`"
output: html_document
---


# {.tabset}  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup {.tabset}

### Helper Functions

```{r, eval=FALSE}


zScorePatientExpression<-function(full.dn4=NULL,markers=NULL,ROIID.column=NULL){
  X3<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(sd))%>%data.frame
  X2<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(mean))%>%data.frame
   ###
   z<-full.dn4
  for(cl in unique(full.dn4$ROIID)){
   for(mark in colnames(X2)[-1]){
  z[which(z$ROIID==cl),which(colnames(z)==mark)]<- ( z[which(z$ROIID==cl),which(colnames(z)==mark)]-X2[which(X2$ROIID==cl),which(colnames(X2)==mark)])/(X3[which(X3$ROIID==cl),which(colnames(X3)==mark)])
   }
 }
 return(z)
 }



metaClusterPhenotype<-function(dn4=NULL,phenotype=NULL,myPheno=NULL,markers=NULL,plotPCA=FALSE,k2=15){
 #myPheno is the column for the phenotype.
    
   erik<-subset(dn4,dn4[,myPheno]==phenotype)
  ## subMeta holds the casePheno 
  dn4[,paste0(phenotype,"_subMeta")]<-NA
    erik[,paste0(phenotype,"_subMeta")]<-NA
 require(Rphenograph)
   roiData<-as.data.frame.matrix(table(erik$ROIID,erik[,myPheno]))

   for(roi in rownames(roiData)[which(roiData[,phenotype]>45)]){
     myroi<-subset(erik,erik$ROIID==roi)
     graph<-Rphenograph(myroi[,markers],k=45)
     myroi[,paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
    dn4[match(myroi$uniqueLabel,dn4$uniqueLabel),paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
     erik[match(myroi$uniqueLabel,erik$uniqueLabel),paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
   }

  ErikMeta=erik[,
	c(markers,
	"ROIID",
	paste0(phenotype,"_subMeta"))] %>% group_by(ROIID,.dots=paste0(phenotype,"_subMeta")) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
     
          ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c(markers)], k = k2)
     subphenograph_cluster_meta<- factor(membership(erikPheno[[2]]))
     table(subphenograph_cluster_meta)
     ErikMeta=cbind(ErikMeta, subphenograph_cluster_meta)
     ### plot PCA, label each case.  COO,  status.  ## 
   if( any(colnames(erik)== paste0(phenotype,"_subphenograph_cluster_meta"))){
     erik<-erik[,which(colnames(erik)!= paste0(phenotype,"_subphenograph_cluster_meta"))]
 } 

   myLocal=left_join(erik, ErikMeta[,c("ROIID",paste0(phenotype,"_subMeta"),"subphenograph_cluster_meta")], by = c("ROIID"))
  colnames(myLocal)[which(colnames(myLocal)=="subphenograph_cluster_meta")]<-paste0(phenotype,"_subphenograph_cluster_meta") 
  dn4[,paste0(phenotype,"_subphenograph_cluster_meta")]<-NA
  dn4[match(myLocal$uniqueLabel,dn4$uniqueLabel),paste0(phenotype,"_subphenograph_cluster_meta")]<-myLocal[,which(colnames(myLocal)==paste0(phenotype,"_subphenograph_cluster_meta"))]

  plot_clustering_heatmap_wrapper(expr=ErikMeta[,c(markers)], 
  cell_clustering=ErikMeta[,"subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
 return(dn4)
}



plot_clustering_heatmap_wrapper<-function(expr=NULL, 
  cell_clustering=NULL, cluster_merging = NULL,useMedians=TRUE,useQuantiles=FALSE,
color_clusters=NULL){
  ## cluster_merging should be character merging vector of 29 or so clusters. this is input from the ANNOTATION object.  not a matched object level.
   expr<-as.matrix(expr)
  if(useQuantiles==TRUE){
  ##max min normalization
  library(matrixStats)
  rng <- colQuantiles(expr, probs = c(0.001, 0.999))
  expr01 <- t((t(expr) - rng[, 1]) / (rng[, 2] - rng[, 1]))
  expr01[expr01 < 0] <- 0
  expr01[expr01 > 1] <- 1
  }else{
   expr01<-expr 

  }

 if(is.null(color_clusters)==TRUE){
  color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")
 }

  # Calculate the median expression 
  library(dplyr)
  expr_median <- data.frame(expr, cell_clustering = cell_clustering) %>%
   dplyr::group_by(cell_clustering) %>% 
    dplyr::summarize_all(funs(median))
 if(useMedians==TRUE){
  expr01_median <- data.frame(expr01, cell_clustering = cell_clustering) %>%
    group_by(cell_clustering) %>% 
    dplyr::summarize_all(funs(median))
   }else{
   expr01_median <- data.frame(expr01, cell_clustering = cell_clustering) %>%
    dplyr::group_by(cell_clustering) %>% 
   dplyr::summarize_all(funs(mean))
 }
  # Calculate cluster frequencies
  clustering_table <- as.numeric(table(cell_clustering))
   print(dim(expr_median))
  # This clustering is based on the markers that were used for the main clustering
  d <- dist(expr_median[, colnames(expr)], method = "euclidean")
  cluster_rows <- hclust(d, method = "average")
   expr_heat <- as.matrix(expr01_median[, colnames(expr01)])
  rownames(expr_heat) <- expr01_median$cell_clustering
 
  labels_row <- paste0(rownames(expr_heat), " (", 
    round(clustering_table / sum(clustering_table) * 100, 2), "%)")
  labels_col <- colnames(expr_heat)
  
  # Row annotation for the heatmap
  annotation_row <- data.frame(cluster = factor(expr01_median$cell_clustering))
  rownames(annotation_row) <- rownames(expr_heat)
  
  color_clusters <- color_clusters[1:nlevels(annotation_row$cluster)]
  names(color_clusters) <- levels(annotation_row$cluster)
  annotation_colors <- list(cluster = color_clusters)
  annotation_legend <- FALSE
  
  if(!is.null(cluster_merging)){
    
    annotation_row$cluster_merging <- factor(cluster_merging)
    color_clusters <- color_clusters[1:nlevels(factor(cluster_merging))]
    names(color_clusters) <- levels(factor(cluster_merging))
    annotation_colors$cluster_merging <- color_clusters
    annotation_legend <- TRUE
  }
  
  # Colors for the heatmap
  library(RColorBrewer);library(pheatmap)
  color <- colorRampPalette(rev(brewer.pal(n = 9, name = "RdYlBu")))(100)
  
  pheatmap::pheatmap(expr_heat, color = color, 
    cluster_cols = FALSE, cluster_rows = cluster_rows, 
    labels_col = labels_col, labels_row = labels_row, 
    display_numbers = TRUE, number_color = "black", 
    fontsize = 12, fontsize_number = 12,
    annotation_row = annotation_row, annotation_colors = annotation_colors,
    annotation_legend = annotation_legend)
  
}



mynormalize<-function(data=NULL,percentile=NULL){
   if(percentile>1){
     percentile<-percentile/100
   }
    if(is.null(percentile)==TRUE){
    minValues<-apply(data,2,min)
    maxValues<-apply(data,2,max) 
   }else{
    minValues<-apply(data,2,function(x) quantile(x,1-percentile)  )
    maxValues<-apply(data,2,function(x) quantile(x,percentile)  )
   }
    hv2<-maxValues-minValues
    dataXR<-data
    stopifnot(all( colnames(data)==names(minValues)))
    stopifnot(all(colnames(data)==names(maxValues)))
    stopifnot(all(colnames(data)==names(hv2)))
    for(j in 1:ncol(data)){
    dataXR[,j]<-(data[,j]-minValues[j])/hv2[j]
    }
    dataXR[dataXR<0]<-0
    dataXR[dataXR>1]<-1

  stopifnot(all(apply(dataXR,2,max)==1))
  stopifnot(all(apply(dataXR,2,min)==0))
   return(dataXR)
}



mutationClinicalSheetAssembly<-function(mutationsToGather=c("TP53","BCL2","BCL6","CD79B","SGK1","MYD88","MYC","NOTCH1","NOTCH2","EZH2","CREBBP"),full.dn4=NULL){
 ## the mutation of MYD88 is specifically controlled for L265P region.
 ## the column header for the non-syn SNP annotation for MYD88 is AA.Mutation.Cosmic_v70

   mutations<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/mutational-genomic-table/TMA CGI genomic data.csv")
   colnames(mutations)<-gsub("USC076","USC76",colnames(mutations))
  #  clinic<-read.csv("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/patient.clinical.table.csv",header=T)
     library(XLConnect) 
   wb<-XLConnect::loadWorkbook("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/clinicalDataRevisedFinal_final.xlsx")
   clinic <- readWorksheet(wb, 1);  
  clinic$MB<-gsub(" ","",clinic$End..MB)
 mut.tag<-colnames(mutations)[9:67]
  mutt<-data.frame(mut.tag,id=sapply(strsplit(gsub("\\.","_",mut.tag),"_"),function(x) x[1])
) 

  mutt$id[which(mutt[,"id"]=="USC076")]<-"USC76"
  tag<- as.character(clinic[,"MB"])
  tag<-gsub(" ","",tag)
  tag<-data.frame(tag=tag,stringsAsFactors=FALSE)
  inters<-intersect(tag[,1],mutt$id)
   key<-data.frame(MB=tag[match(inters,tag[,1]),],stringsAsFactors=FALSE)
  key$id<-mutt[match(inters,mutt[,2]),1]
 ##fix a column header
 ## USC76 has a funcy header.
 ##
  casemuts<-mutations[,match(key$id,colnames(mutations))]
  casemuts$gene.name<-mutations[,"gene_name..Homo_sapiens_ensembl_v74_mRNA."]
  casemuts$chromosome<-as.character(mutations$Chromosome)
 casemuts$aminoAcid<-as.character(mutations$AA.Mutation.Cosmic_v70)
 casemuts$aminoAcid<-gsub( "\"","",mutations$AA.Mutation.Cosmic_v70)
  casemuts$gene.name<-as.character(casemuts$gene.name)
 ##for MYD88, the gene name will have the aminoAcid appended.
  casemuts$gene.name[which(casemuts$gene.name=="MYD88" & casemuts$aminoAcid=="L265P, L265P")]<-"MYD88.L265P"
  casemuts<-casemuts[which(casemuts$aminoAcid!="S219C, S219C, S219C, S219C" & casemuts$aminoAcid!="M232T, M232T" & casemuts$aminoAcid!="0.999, 0.999" & casemuts$aminoAcid!="S251N, S243N, S251N, S243N"),]
  casemuts[which(casemuts$gene.name=="MYD88" & casemuts$chromosome==""),"gene.name"]<-"MYD88.L265P"
 casemuts[which(casemuts$gene.name=="MYD88.L265P" & casemuts$chromosome==""),1:20]<-ifelse(casemuts[which(casemuts$gene.name=="MYD88.L265P" & casemuts$chromosome!=""),1:20]!=0,1,0)
  casem<- casemuts[which(casemuts$chromosome==""),]
 mutationsToGather[which(mutationsToGather=="MYD88")]<-"MYD88.L265P"
  dat<-casem[casem$gene.name%in%mutationsToGather,]
   ### 
   for(i in mutationsToGather){
  key[,i]<-0
 key[match(colnames(dat)[which(dat[dat$gene.name==i,]==1)],key$id),i]<-1
   }
  write.csv(key,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/Mutational.CNV.categorical.table.csv")
  ## append to the clinical sheet
  clinic2<-clinic[,!colnames(clinic)%in%mutationsToGather]
 clinic2<-left_join(clinic2,key[,c("MB",mutationsToGather)],by="MB")
  for(i in mutationsToGather){
  clinic2[which(is.na(clinic2[,i])), i]<-0
  }
 ##add the factor to control for replicates
  replicate<-model.matrix(~0+clinic2$MB)
  replicate<-replicate[,which(colSums(replicate)>1)]  
 colnames(replicate)<-substring(colnames(replicate),11)
  clinic2<-data.frame(clinic2,replicate)
 clinic2$WES.avail<-FALSE
  clinic2$WES.avail[match(key$MB,clinic2$MB)]<-TRUE
 clinic2$Case_ID<-paste0("Case_",clinic2$Col1)
  return(clinic2)

}


 ## test for enriched populations.
 abundanceTest<-function(p=NULL,x=NULL,y=NULL){
  myForm<-formula(paste0(p,"~0+",y))

 fit<-lm(myForm,x)
 myInt<-confint(fit)%>%data.frame
  myInt<-cbind(myInt,coef(fit))
  myInt$phenotype<-p
  diffit<-lm(as.formula(paste0(p,"~",y)),x)%>%summary%>%coef
  myInt$REF.vs.CR.p.value<-diffit[2,"Pr(>|t|)"]
  return(myInt)
  } 
```

### Phenotype calling in DLBCL


```{r calls}
  library(Rphenograph)

##create full random network all ROI.
 source("~/Documents/imageAnalysis/Ansell/cluster-phenograph-analysis/clusterFunctions.R")
  load("~/Documents/imageAnalysis/DLBCL/monirath/myData/final-scripts/Figure-MasterKey/supporting-data-back/ERIK.dlbcl.RData")
 
  #clinic<-read.csv("~/DLBCL/clinicalDataFinal.csv",header=TRUE,sep="\t") 
 # clinic<-read.csv("~/DLBCL/clinicalDataRevisedFinal.csv",header=TRUE,sep=",") 
   clinic<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/USC_CGI_DLBCL_Data_Final.csv",header=T)
   lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p")
		#"Cell_DNA",
		#"Cell_DNA2")
   

  dlbcl$uniqueLabel<-paste0(dlbcl$ROIID,"_",dlbcl$CellId)
 ## these ROIs were not in the data sheet.
  missingSurv<-c("3978","3972", "3979" , "39792", "4070","4071","4187")
  dlbcl<-dlbcl[!dlbcl$ROIID%in%missingSurv,]
  dlbcl<-dlbcl[which(dlbcl$ROIID!="3972"),]
   dlbcl$ROIID[which(dlbcl$ROIID=="39722")]<-"3972"
  dlbcl$uniqueLabel<-paste0(dlbcl$ROIID,"_",dlbcl$CellId)
  dlbcl$case<-paste0("Case_",clinic$X[match(dlbcl$ROIID,clinic$Run.ID)])

 
   dn<-dlbcl[,c(lineage_markers,induc)]
   adn<-dn/10000
   adn<-asinh(adn)
 
  #dn4<-adn/apply(adn,2,max)
   full.dn4<-apply(adn,2,function(x) (x-min(x))/max(x))
   full.dn4<-data.frame(full.dn4)
   full.dn4$ROIID<-dlbcl$ROIID
   full.dn4$CellId<-dlbcl$CellId
   full.dn4$uniqueLabel<-paste0(full.dn4$ROIID,"_",full.dn4$CellId)
   all(full.dn4$uniqueLabel==dlbcl$uniqueLabel)
   full.dn4$X_position<-dlbcl$X_position
   full.dn4$Y_position<-dlbcl$Y_position
   full.dn4$cell.type.new<-dlbcl$cell.type.new
   full.dn4$case<-dlbcl$case
    full.dn4$response<-clinic$Response.category[match(full.dn4$ROIID,clinic$Run.ID)]
  full.dn4$replicate.number<-"first"
   full.dn4$replicate.number[full.dn4$ROIID%in%c("3971","3974","3969","3977","4193","3969","4185","4186","4184")]<-"second"
   targs<-c(lineage_markers,induc)
   full.dn4$Cell_CCR4.Tim3.PDL1<-full.dn4$Cell_CCR4*full.dn4$Cell_Tim3*full.dn4$Cell_PDL1
   full.dn4$Cell_CXCR3.Tbet<-full.dn4$Cell_CXCR3*full.dn4$Cell_Tbet
   full.dn4$Cell_Lag3.PD1<-full.dn4$Cell_Lag3*full.dn4$Cell_PD1
    full.dn4$Cell_Tim3.PD1<-full.dn4$Cell_Tim3*full.dn4$Cell_PD1

  ### add treatment info
  full.dn4$treatment.status<-clinic$chemo[match(full.dn4$ROIID,clinic$Run.ID)]
  full.dn4$treatment.status<-clinic$chemo[match(full.dn4$ROIID,clinic$Run.ID)]
  full.dn4$MYC.Gain<-clinic$MYC.Gain[match(full.dn4$ROIID,clinic$Run.ID)]
  full.dn4$X9p24.Status<-clinic$X9p24.Status[match(full.dn4$ROIID,clinic$Run.ID)]
  
   ##BCL2 is associated with C3 (Chapuy)
   full.dn4$bcl2<-clinic$bcl2[match(full.dn4$ROIID,clinic$Run.ID)]
   full.dn4$bcl6<-clinic$bcl6[match(full.dn4$ROIID,clinic$Run.ID)]
  ##survival
    full.dn4$os<-clinic$os[match(full.dn4$ROIID,clinic$Run.ID)]
    full.dn4$os_d<-clinic$os_d[match(full.dn4$ROIID,clinic$Run.ID)]
    full.dn4$phenograph_cluster_meta<-dlbcl$phenograph_cluster_meta[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]
 ##add COO
   full.dn4$COO<-clinic$Cell.of.origin..HANS.[match(full.dn4$ROIID,clinic$Run.ID)]
  
   ###add in the mutational cluster info
  mut<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/mutational-genomic-table/clinical.table.morphological.cluster.information.csv")

    full.dn4$Mutation.C1<-mut$Morphological.Cluster1[match(full.dn4$ROIID,mut$Run.ID)]
    full.dn4$Mutation.C2<-mut$Morphological.Cluster2[match(full.dn4$ROIID,mut$Run.ID)]
      full.dn4$Mutation.C3<-mut$Morphological.Cluster3[match(full.dn4$ROIID,mut$Run.ID)]
    full.dn4$Mutation.C4<-mut$Morphological.Cluster4[match(full.dn4$ROIID,mut$Run.ID)]
    full.dn4$Mutation.C5<-mut$Morphological.Cluster5[match(full.dn4$ROIID,mut$Run.ID)] 


      plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers[c(3,5:12,14)],"Cell_BCL2","Cell_Ki67")], 
  cell_clustering=full.dn4[,"phenograph_cluster_meta"], cluster_merging = NULL,useMedians=TRUE)




   savePlot(file="original.phenotype.MC.png")
   anno<-data.frame("1"="Endothelial",
		"2"="CD8",
		"3"="CD4",
		"4"="Tumor",
		"5"="CD8",
		"6"="CD4",
		"7"="Treg",
		"8"="Macrophages",
		"9"="Tumor",
		"10"="Tumor",
		"11"="CD4",
		"12"="Tumor",
		"13"="Tumor",
		"14"="Tumor")
  anno<-as.data.frame(t(anno))
  anno$cluster<-rownames(anno)



```



### Identifying sub-populations

```{r DLBCLsubType}
   full.dn4$subType<-as.character(full.dn4$cell.type.new)
     full.dn4$subType.correction<-as.character(full.dn4$cell.type.new)
 full.dn4$subType.reassignedLabel<-as.character(full.dn4$subType)


   for (i in unique(as.character(full.dn4$cell.type.new))[-1]){
    ROI=data.frame(subset(full.dn4,full.dn4$subType==i))
      set.seed(42)
    graph<- Rphenograph(ROI[,c(lineage_markers,induc)], k = 50)
    phenograph_cluster<- factor(membership(graph[[2]]))
   full.dn4$subType[match(ROI$uniqueLabel,full.dn4$uniqueLabel)]=paste0(i,"_",phenograph_cluster)
  
 }
 
  ##plot the sub-phenographs co-expression.
  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tumor"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Tumor.subTypes.png")
  
  ## CD8 reassign
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_28")]<-"Tcyt"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_28")]<-"Tumor_28.Reassigned.CD8"
   ##MAC re-assign
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_1")]<-"Macrophages"
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_27")]<-"Macrophages"
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_34")]<-"Macrophages"
 
  full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_1")]<-"Tumor_1.Reassigned.MAC"

 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_27")]<-"Tumor_27.Reassigned.MAC"

 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_34")]<-"Tumor_34.Reassigned.MAC"



  ##CD4 re-assign
    full.dn4$subType.correction[which(full.dn4$subType=="Tumor_10")]<-"Th"
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_41")]<-"Th"
  full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_10")]<-"Tumor_10.Reassigned.CD4"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_41")]<-"Tumor_41.Reassigned.CD4"

    #Endo re-assign
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_25")]<-"Endothelial"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_25")]<-"Tumor_25.Reassigned.Endo"

 ##Th re-assign
  full.dn4$subType.correction[which(full.dn4$subType=="Tumor_20")]<-"Th"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_20")]<-"Tumor_20.Reassigned.CD4"

  full.dn4$Cell_DNA<-0.5
  full.dn4$Cell_DNA2<-0.5
  full.dn4$Cell_DNA[match(dna$uniqueLabel,full.dn4$uniqueLabel)]<-dna$Cell_DNA
  full.dn4$Cell_DNA2[match(dna$uniqueLabel,full.dn4$uniqueLabel)]<-dna$Cell_DNA2

  a<-ggplot(full.dn4,aes(x=subType.correction,y=Cell_DNA))+geom_boxplot()+theme_bw(base_size=14)
  b<-ggplot(full.dn4,aes(x=subType.correction,y=Cell_DNA2))+geom_boxplot()+theme_bw(base_size=14)
  library(gridExtra) 
 grid.arrange(a,b)
  ggsave("DNA.channel.distribution.png")
 
 
 



 ##plot the tumor correction
    plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tumor"),"subType.correction"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Tumor.subType.correction.png")

 ### tumor with DNA channels.
   plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers,induc,"Cell_DNA","Cell_DNA2")], 
  cell_clustering=full.dn4[,"subType.correction"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.all.subType.correction.DNA.Channel.png")



  
   plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tumor"),"subType.reassignedLabel"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Tumor.subType.reassignedLabel.png")
   
########### META-CLUSTER EACH POPULATION ############################
 ##meta-cluster the tumor population
   full.dn42<-metaClusterPhenotype(dn4=full.dn4,phenotype="Tumor",myPheno="subType.correction",markers=c(lineage_markers,induc))
   save(full.dn42,file="full.dn42.tumor.subMeta.RData")
 
  cd8.subMeta<-metaClusterPhenotype(dn4=full.dn4,phenotype="Tcyt",myPheno="subType.correction",markers=c(lineage_markers,induc)) 
     save( cd8.subMeta,file=" cd8.subMeta.cd8.subMeta.RData")

  th.subMeta<-metaClusterPhenotype(dn4=full.dn4,phenotype="Th",myPheno="subType.correction",markers=c(lineage_markers,induc))
    save(th.subMeta,file="th.subMeta.th.subMeta.RData") 
 
  treg.subMeta<-metaClusterPhenotype(dn4=full.dn4,phenotype="Treg",myPheno="subType.correction",markers=c(lineage_markers,induc))
    save(treg.subMeta,file="treg.subMeta.treg.subMeta.RData") 


 mac.subMeta<-metaClusterPhenotype(dn4=full.dn4,phenotype="Macrophages",myPheno="subType.correction",markers=c(lineage_markers,induc))
    save(mac.subMeta,file="mac.subMeta.mac.subMeta.RData") 

 endo.subMeta<-metaClusterPhenotype(dn4=full.dn4,phenotype="Endothelial",myPheno="subType.correction",markers=c(lineage_markers,induc))
    save(endo.subMeta,file="endo.subMeta.endo.subMeta.RData") 

 #### assemble the submeta.
 full.dn4$Tumor_subMeta[match(full.dn42$uniqueLabel,full.dn4$uniqueLabel)]<-full.dn42$Tumor_subMeta
 full.dn4$Tumor_subphenograph_cluster_meta[match(full.dn42$uniqueLabel,full.dn4$uniqueLabel)]<-full.dn42$Tumor_subphenograph_cluster_meta
  
  x<-full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)]
  id<-full.dn4[which(full.dn4$cell.type.new=="Tumor" & !is.na(full.dn4$Tumor_subphenograph_cluster_meta)),"uniqueLabel"]

  plot_clustering_heatmap_wrapper(expr=full.dn4[match(id,full.dn4$uniqueLabel),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[match(id,full.dn4$uniqueLabel),"Tumor_subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Tumor.subType.subMetaCluster.png")
   
 ####  CD8
 load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/subcluster-data/cd8.subMeta.cd8.subMeta.RData")
 full.dn4$Tcyt_subMeta[match(cd8.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-cd8.subMeta$Tcyt_subMeta
 full.dn4$Tcyt_subphenograph_cluster_meta[match(cd8.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-cd8.subMeta$Tcyt_subphenograph_cluster_meta
  
  id<-full.dn4[which(full.dn4$cell.type.new=="Tcyt" & !is.na(full.dn4$Tcyt_subphenograph_cluster_meta)),"uniqueLabel"]

  plot_clustering_heatmap_wrapper(expr=full.dn4[match(id,full.dn4$uniqueLabel),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[match(id,full.dn4$uniqueLabel),"Tcyt_subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.CD8.subType.subMetaCluster.png")
   

 ## CD4 
  ####  CD8
 load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/subcluster-data/th.subMeta.th.subMeta.RData")
 full.dn4$Th_subMeta[match(th.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-th.subMeta$Th_subMeta
  full.dn4$Th_subphenograph_cluster_meta[match(th.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-th.subMeta$Th_subphenograph_cluster_meta
  
  id<-full.dn4[which(full.dn4$cell.type.new=="Th" & !is.na(full.dn4$Th_subphenograph_cluster_meta)),"uniqueLabel"]

  plot_clustering_heatmap_wrapper(expr=full.dn4[match(id,full.dn4$uniqueLabel),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[match(id,full.dn4$uniqueLabel),"Th_subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.CD4.subType.subMetaCluster.png")
   

  ### Treg
  load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/subcluster-data/treg.subMeta.treg.subMeta.RData")
 full.dn4$Treg_subMeta[match(treg.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-treg.subMeta$Treg_subMeta
  full.dn4$Treg_subphenograph_cluster_meta[match(treg.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-treg.subMeta$Treg_subphenograph_cluster_meta
  
  id<-full.dn4[which(full.dn4$cell.type.new=="Treg" & !is.na(full.dn4$Treg_subphenograph_cluster_meta)),"uniqueLabel"]

  plot_clustering_heatmap_wrapper(expr=full.dn4[match(id,full.dn4$uniqueLabel),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[match(id,full.dn4$uniqueLabel),"Treg_subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.treg.subType.subMetaCluster.png")
   
 ## MAC
  load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/subcluster-data/mac.subMeta.mac.subMeta.RData")
 full.dn4$Macrophages_subMeta[match(mac.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-mac.subMeta$Macrophages_subMeta
  full.dn4$Macrophages_subphenograph_cluster_meta[match(mac.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-mac.subMeta$Macrophages_subphenograph_cluster_meta
  
  id<-full.dn4[which(full.dn4$cell.type.new=="Macrophages" & !is.na(full.dn4$Macrophages_subphenograph_cluster_meta)),"uniqueLabel"]

  plot_clustering_heatmap_wrapper(expr=full.dn4[match(id,full.dn4$uniqueLabel),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[match(id,full.dn4$uniqueLabel),"Macrophages_subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.mac.subType.subMetaCluster.png")
   

  ## endo
  load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/subcluster-data/endo.subMeta.endo.subMeta.RData")
 full.dn4$Endothelial_subMeta[match(endo.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-endo.subMeta$Endothelial_subMeta
  full.dn4$Endothelial_subphenograph_cluster_meta[match(endo.subMeta$uniqueLabel,full.dn4$uniqueLabel)]<-endo.subMeta$Endothelial_subphenograph_cluster_meta
  
  id<-full.dn4[which(full.dn4$cell.type.new=="Endothelial" & !is.na(full.dn4$Endothelial_subphenograph_cluster_meta)),"uniqueLabel"]

  plot_clustering_heatmap_wrapper(expr=full.dn4[match(id,full.dn4$uniqueLabel),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[match(id,full.dn4$uniqueLabel),"Endothelial_subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.endo.subType.subMetaCluster.png")
   
 full.dn4$Cell_DNA=0.4
 full.dn4$Cell_DNA2=0.4 
 full.dn4$Cell_DNA[match(dna$uniqueLabel,full.dn4$uniqueLabel)]<-dna$Cell_DNA
  full.dn4$Cell_DNA2[match(dna$uniqueLabel,full.dn4$uniqueLabel)]<-dna$Cell_DNA2

 ### add morphology
 full.dn4$Area<-dlbcl$Area[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]
  full.dn4$Eccentricity<-dlbcl$Eccentricity[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]
  full.dn4$Solidity<-dlbcl$Solidity[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]
    full.dn4$Perimeter<-dlbcl$Perimeter[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]
     full.dn4$Percent_Touching<-dlbcl$Percent_Touching[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]
    full.dn4$Number_Neighbors<-dlbcl$Number_Neighbors[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel)]

   full.dn4$DNA.Area<-full.dn4$Cell_DNA/full.dn4$Area


  save(full.dn4,file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.RData")

 ##### end of the meta-clustering of the sub-types.


 #### global sub-types.

 ## Tcyt
 plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tcyt"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tcyt"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.CD8.subTypes.png")
  

 ##Th
 plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Th"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Th"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.CD4.subTypes.png")

 ##Th_12 has FOXP3+ 
     full.dn4$subType.correction[which(full.dn4$subType=="Th_12")]<-"Treg"
  full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Th_12")]<-"Th_12.Reassigned.Treg"

      full.dn4$subType.correction[which(full.dn4$subType=="Th_21")]<-"Treg"
full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Th_21")]<-"Th_21.Reassigned.Treg"

     full.dn4$subType.correction[which(full.dn4$subType=="Th_9")]<-"Tcyt"
full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Th_9")]<-"Th_9.Reassigned.CD8"

 ##plot the corrected
   plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Th"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Th"),"subType.reassignedLabel"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Th.subType.reassignedLabel.png")
  




 #MAC
  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Macrophages"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Macrophages"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.MAC.subTypes.png")


 ##Endo
   plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Endothelial"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Endothelial"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Endo.subTypes.png")
##Treg
  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Treg"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Treg"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.treg.subTypes.png")

 ##plot the heatmap of corrected
  plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers,induc)], 
  cell_clustering=full.dn4[,"subType.correction"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.corrected.subTypes.png")

 melts<-melt(full.dn4[,c("Cell_CD20","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3","Cell_CD4","Cell_CD3","DNA.Area","subType.correction")])
 
 ggplot(melts, aes(x=subType.correction,y=value))+geom_boxplot()+facet_wrap(~variable)+ theme(axis.text.x = element_text(angle = 90, hjust = 1))

 ggsave("boxplot.corrected.phenotype.png")


  color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")

   
 ###plot the meta-cluster heatmap.
  metaData<-full.dn4[,c(lineage_markers,induc)]
  colnames(metaData)<-gsub("Cell_","",colnames(metaData))
  metaData<-metaData[,c("CD20","CD4","CD3","CD8","CD68","CD31","FOXP3")]

 

 ##patient distributions
  ##meta cluster key
  annotation_row<-data.frame(cluster=seq(1,14))
  annotation_row$cluster<-factor(annotation_row$cluster)
  annotation_row$colors <- color_clusters[1:nlevels(annotation_row$cluster)]



      plot_clustering_heatmap_wrapper(expr=metaData, 
  cell_clustering=full.dn4[,"phenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=annotation_row$colors)
  savePlot(file="full.allROI.metaCluster.png")

  ## FIX ME: show the adjusted meta-cluster populations.
 table(full.dn4$phenograph_cluster_meta,full.dn4$subType.correction)
 full.dn4$phenograph_cluster_meta_correction<-full.dn4$phenograph_cluster_meta

 ## we see that tumor cluster 9 has low CD20 expression and re-assign this meta-cluster to other components.
 ## within MC9 assign to endo
 id<-which(full.dn4$phenograph_cluster_meta==9 & full.dn4$subType.correction=="Endothelial")
 full.dn4$phenograph_cluster_meta_correction[id]<-1
 ## within MC9 assign to mac
  id<-which(full.dn4$phenograph_cluster_meta==9 & full.dn4$subType.correction=="Macrophages")
 full.dn4$phenograph_cluster_meta_correction[id]<-8
  id<-which(full.dn4$phenograph_cluster_meta==9 & full.dn4$subType.correction=="Tcyt")
 full.dn4$phenograph_cluster_meta_correction[id]<-2
  id<-which(full.dn4$phenograph_cluster_meta==9 & full.dn4$subType.correction=="Th")
 full.dn4$phenograph_cluster_meta_correction[id]<-3

 ##filter out low CD20 expressions
 ## there are 35,133 CD20 cells that had low CD20 expression and not corrected into immune component.
  id<-which(full.dn4$phenograph_cluster_meta_correction==9 & full.dn4$Cell_CD20<0.25)
 full.dn4$dimCD20.filtered<-'NO'
 full.dn4$dimCD20.filtered[id]<-'YES'
 ## meta.cluster 9 has dim CD20
 table(full.dn4[id,"phenograph_cluster_meta"])


    plot_clustering_heatmap_wrapper(expr=metaData[-id,], 
  cell_clustering=full.dn4[-id,"phenograph_cluster_meta_correction"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=annotation_row$colors)
  savePlot(file="full.allROI.metaCluster.CD20.lowFiltered.threshold.0.25.png")

 full.dn4<-full.dn4[-id,]
 write.csv(100*length(id)/nrow(dlbcl),file="Cluster9.0.25.filtered.csv")
  
    mat<-full.dn4
    mat<-as.data.frame.matrix(table(mat$ROIID,mat$phenograph_cluster_meta))
  mat$ROIID<-rownames(mat)
  mat$ROIID[which(mat$ROIID=="39722")]<-"3972"
  mat<-mat[mat$ROIID%in%clinic$Run.ID,]
  mat$case<-factor(paste0("Case",clinic$X[match(mat$ROIID,clinic$Run.ID)]))
  caseLevs<-levels(mat$case)
  mat$case<-factor(mat$case,levels=caseLevs[order(nchar(caseLevs),caseLevs)])
 
 mat<-mat[,c(1:14,16)]%>%group_by(case)%>%summarise_all(funs(sum))%>%data.frame
 mat[,-1]<-mat[,-1]/rowSums(mat[,-1])
 colnames(mat)[-1]<-gsub("X","", colnames(mat)[-1])

  mat<-reshape2::melt(mat)
  levs<-levels(factor(mat$variable))
  mat$variable<-factor(mat$variable,levels=levs[order(nchar(levs),levs)])
 
 roiMeta<-ggplot(mat,aes(x=case,y=value,fill=variable))+geom_bar(stat='identity')+scale_fill_manual(values=annotation_row$colors)+theme_bw(base_size=14)+theme(axis.text.x=element_text(angle=90,hjust=1))+ylab("Meta-cluster cluster frequency")+ggtitle("Meta-clusters per case including all ROI")
  roiMeta
   ggsave("patient.metacluster.distribution.percentage.normalized.png")
  
 ###frequency
   mat<-full.dn4
    mat<-as.data.frame.matrix(table(mat$ROIID,mat$phenograph_cluster_meta))

   mat$ROIID<-rownames(mat)
   mat<-mat[mat$ROIID%in%clinic$Run.ID,]
    mat$case<-factor(paste0("Case",clinic$X[match(mat$ROIID,clinic$Run.ID)]))  

  mat<-reshape2::melt(mat)
  levs<-levels(factor(mat$variable))
  mat$variable<-factor(mat$variable,levels=levs[order(nchar(levs),levs)])
  mat$case<-factor(paste0("Case",clinic$X[match(mat$ROIID,clinic$Run.ID)]))
  caseLevs<-levels(mat$case)
  mat$case<-factor(mat$case,levels=caseLevs[order(nchar(caseLevs),caseLevs)])
 
 roiMeta<-ggplot(mat,aes(x=case,y=value,fill=variable))+geom_bar(stat='identity')+scale_fill_manual(values=annotation_row$colors)+theme_bw(base_size=14)+theme(axis.text.x=element_text(angle=90,hjust=1))+ylab("Meta-cluster cluster frequency")+ggtitle("Meta-clusters per case including all ROI")
  roiMeta
   ggsave("patient.metacluster.distribution.all.png")


 write.csv(table(full.dn4$ROIID,full.dn4$subType.correction),file="final.rephenotype.csv")


 ### save the filtered data 
  save(full.dn4,file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")


```


### the full DLBCL data without any filtering complete re-assignments.
 We have the full complete DLBCL data, but drop ROIs with missing clinical information. 
```{r,eval=FALSE}

load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.RData")

 ##matches the dimensions. no filtering. just re-assignments.
  dim(dlbcl[match(full.dn4$uniqueLabel,dlbcl$uniqueLabel),])
  missingRoi<-dlbcl$uniqueLabel[!dlbcl$uniqueLabel%in%full.dn4$uniqueLabel]
 droppedRoi<-data.frame(label=missingRoi,ROI=dlbcl$ROIID[match(missingRoi,dlbcl$uniqueLabel)])
 ## removed ROI: 3978, 4187, 4071, 4070, 397992, 3979, and 3978 becuase they did not have clinical information.
 ## ROI: 39722 (N=10,020) was renamed to 3972, and the 1st pass of 3972 (N=2697) was omitted.
 dlbcl[!dlbcl$ROIID%in%missingSurv,]%>%dim
 ## total dropping only ROI is 692,463
 
```


### Analysis for Figure 1
 Tumor cluster 9 is dim for CD20, but negative for other markers and retained.
```{r, eval=FALSE}

  set.seed(1234)
  table(full.dn4$phenograph_cluster_meta,full.dn4$cell.type.new)
    

  lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
	"Area.norm",
	"Eccentricity.norm",
	"Solidity.norm",
	"Perimeter.norm",
	"Percent_Touching.norm",
	"Number_Neighbors.norm")


  plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers)], 
  cell_clustering=full.dn4[,"phenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)

  anno<-data.frame("1"="Endothelial",
		"2"="CD8",
		"3"="CD4",
		"4"="Tumor",
		"5"="CD8",
		"6"="CD4",
		"7"="Treg",
		"8"="Macrophages",
		"9"="Tumor",
		"10"="Tumor",
		"11"="CD4",
		"12"="Tumor",
		"13"="Tumor",
		"14"="Tumor")
  anno<-as.data.frame(t(anno))
  anno$cluster<-rownames(anno)


 ## table of re-assigned labels.
 table(full.dn4$subType.reassignedLabel)

```


### Meta-cluster heatmap of primary phenotypes
To study phenotypes we have the full cohort, but could alternatively remove the dim CD20 (ambiguous) tumor cells for phenotype studies only.
```{r, eval=FALSE}

 ErikMeta=full.dn4[,
	c("Cell_CD20","Cell_CD3","Cell_CD4","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3",
	"ROIID",
	"phenograph_cluster_meta")] %>% group_by(ROIID,phenograph_cluster_meta) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame

   plot_clustering_heatmap_wrapper(expr=full.dn4[,c("Cell_CD20","Cell_CD3","Cell_CD4","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3")], 
  cell_clustering=full.dn4[,"subType.correction"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 #savePlot(file="allCells.corrected.png")

  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$immune.status=="immune"),c("Cell_CD20","Cell_CD3","Cell_CD4","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3")], 
  cell_clustering=full.dn4[which(full.dn4$immune.status=="immune"),"subType.correction"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 #savePlot(file="TME.corrected.png")


```

 ### Unsupervised community detection on each compartment
 For phenotype analysis, we separated the dimCD20 population.  We analyzed the CD20 bright clusters and clustered them without the dimCD20 population.  This ensures phenotype accuracy.
 Tumor sub-cluster 11 is the dim CD20 cluster, when studying phenotypes we did not emphasize this cluster.
```{r,eval=FALSE}
# meta-cluster each TME population.

######## unsupervised community detection.
 ## subPhenograph the CD4.
  full.dn4$CD4.subPhenograph<-NA
   morphs<-mynormalize(data=full.dn4[,c("Area",
	"Eccentricity",
	"Solidity",
	"Perimeter",
	"Percent_Touching",
	"Number_Neighbors")],percentile=0.99)
  colnames(morphs)<-paste0(colnames(morphs),".norm")
  full.dn4<-cbind(full.dn4,morphs)

  lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
	"Area.norm",
	"Eccentricity.norm",
	"Solidity.norm",
	"Perimeter.norm",
	"Percent_Touching.norm",
	"Number_Neighbors.norm")
 ##meta cluster the CD4.
  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="CD4",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
  savePlot(file="CD4_subMeta.cluster.png")
  ## CD8
  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="CD8",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
   savePlot(file="CD8_subMeta.cluster.png")
  table(full.dn4$subType.correction,full.dn4$CD8_subphenograph_cluster_meta)
 ## TREG
  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="TREG",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
 savePlot(file="TREG_subMeta.cluster.png")

 table(full.dn4$subType.correction,full.dn4$TREG_subphenograph_cluster_meta)
 ##MAC
 full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="Macrophage",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
 savePlot(file="MAC_subMeta.cluster.png")
  table(full.dn4$subType.correction,full.dn4$Macrophage_subphenograph_cluster_meta)

 ## performed originally by ensuring CD20 was positively expressed.
 ## dim(full.dn4[which(full.dn4.nof$dimCD20.filtered=='YES'),])
  full.dn4_2<-metaClusterPhenotype(dn4=full.dn4[which(full.dn4.nof$dimCD20.filtered=='YES'),],
	phenotype="Tumor",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))


 savePlot(file="Tumor_subMeta.cluster.png")

 


 ##compile into 1 column unsupervised.community  
  full.dn4$unsup.subcommunity<-as.character(full.dn4$subType.correction)
 full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="CD4")]<-paste0("CD4_",full.dn4$CD4_subphenograph_cluster_meta[which(full.dn4$subType.correction=="CD4")])
 ## CD8
  full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="CD8")]<-paste0("CD8_",full.dn4$CD8_subphenograph_cluster_meta[which(full.dn4$subType.correction=="CD8")])
 ## TREG
 full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="TREG")]<-paste0("TREG_",full.dn4$TREG_subphenograph_cluster_meta[which(full.dn4$subType.correction=="TREG")])
 ## MAC
 full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="Macrophage")]<-paste0("Macrophage_",full.dn4$Macrophage_subphenograph_cluster_meta[which(full.dn4$subType.correction=="Macrophage")])
 ## Tumor
  full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="Tumor")]<-paste0("Tumor_",full.dn4$Tumor_subphenograph_cluster_meta[which(full.dn4$subType.correction=="Tumor")])

  full.dn4$unsup.subcommunity<-factor(full.dn4$unsup.subcommunity,
	levels=c("CD4_1",
	       "CD4_2",
		"CD4_3",
		"CD4_4",
		"CD4_5","CD4_6","CD4_7",
	        "CD8_1","CD8_2","CD8_3","CD8_4","CD8_5","CD8_6","CD8_7",
 	       "Macrophage_1","Macrophage_2","Macrophage_3","Macrophage_4","Macrophage_5",
	"Macrophage_6",
	"TREG_1","TREG_2","TREG_3","TREG_4","TREG_5",     
	 "TREG_6",
	 "Tumor_1","Tumor_2","Tumor_3", "Tumor_4", "Tumor_5","Tumor_6","Tumor_7",      	"Tumor_8", "Tumor_9",  "Tumor_10" ,"Tumor_11", "Endothelial" ))


 ## add in the exonic screening from CGI
  ## need column for WES.available = YES/NO
 full.dn4$WES.available<-"FALSE"
 full.dn4$WES.available<-clinic$WES.avail[match(full.dn4$case,clinic$Case_ID)]

  ## need the column key for the USCIDs which match the clinical mutational panel ID.
 full.dn4$MB<-clinic$MB[match(full.dn4$case,clinic$Case_ID)]

 ## add all the mutation information from the clinical sheet.
 ### TP53
   full.dn4$TP53<-clinic$TP53[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$BCL2<-clinic$BCL2[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$BCL6<-clinic$BCL6[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$CD79B<-clinic$CD79B[match(full.dn4$case,clinic$Case_ID)]
     full.dn4$SGK1<-clinic$SGK1[match(full.dn4$case,clinic$Case_ID)]
    full.dn4$MYD88.L265P<-clinic$MYD88.L265P[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$MYC<-clinic$MYC[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$NOTCH1<-clinic$NOTCH1[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$NOTCH2<-clinic$NOTCH2[match(full.dn4$case,clinic$Case_ID)]
   full.dn4$EZH2<-clinic$EZH2[match(full.dn4$case,clinic$Case_ID)]

  ## add in the IPI class
    full.dn4$IPI<-clinic$ipi[match(full.dn4$case,clinic$Case_ID)]

 ## add in 9p24 status
  clinic$X9p24.indicator<-ifelse(clinic$X9p24.gain.status!=0,1,0)
   clinic$X9p24.indicator[which(is.na(clinic$X9p24.indicator))]<-0

     full.dn4$X9p24.indicator<-clinic$X9p24.indicator[match(full.dn4$case,clinic$Case_ID)]

 # save(full.dn4,file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")



   plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers,induc)], 
  cell_clustering=full.dn4[,"unsup.subcommunity"], useMedians=TRUE,useQuantiles=FALSE,
color_clusters=NULL)
 #  savePlot(file="subcommunities.png")
 ##### end of the meta-clustering of the sub-types.



```

### Comparing macrophage proportions GCB vs NGCB
Monette, et.al (Ken Young) identified ABC had increased densities of MAC in ABC vs. GCB.
```{r,eval=FALSE}


 load(file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")


 x<-table(full.dn4$ROIID,full.dn4$subType.correction)%>%as.data.frame.matrix
 x<-x[,which(colnames(x)!='none')]
 x<-100*x/rowSums(x)
 x$COO<-full.dn4$COO[match(rownames(x),full.dn4$ROIID)]

   abundResults<-c()
 for(p in colnames(x)[which(colnames(x)!='COO')]){
  d<-abundanceTest(p,x)
  abundResults<-rbind(abundResults,d)
 }

  colnames(abundResults)[1:3]<-c("lower","upper",'estimate')
 abundResults$label<-label$label[match(abundResults$phenotype,label$easyLabel)]
  abundResults$response<-rep(c("CR","Primary refract"),nrow(abundResults)/2)

 abundResults$label<-factor(abundResults$label,levels=abundResults[order(abundResults$estimate,decreasing=TRUE),'label']%>%unique)
  responderColors<-getResponderColors()[1:2]
 names(responderColors)<-c("CR","Primary refract")




```


