---
title: "Single-cell spatial analysis of tumor immune architecture in diffuse large B cell lymphoma"
author: " Anthony Colombo+,Monirath Hav+, Erik Gerdtsson, Mohan Singh, Alexander Xu, Denaly Chen, Jane Houldsworth, Rita Shaknovich, Tomohiro Aoki, Lauren Chong, Katsuyoshi Takata, Elizabeth Chavez, Christian Steidl, James Hicks, Peter Kuhn, Imran Siddiqi, Akil Merchant*"
date: "`r Sys.Date()`"
output: html_document
fig_width: 15
fig_height: 15 
---


# {.tabset}  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup {.tabset}


### Helper Functions
```{r, eval=TRUE}

 logisticRegressionClinicalOdds<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn=NULL,mycFeature=NULL,groupA=1){
    message(pheno)
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
	ta<-as.data.frame(scale(ta))
    ta$feature<-full.dn4[match(rownames(ta),full.dn4$case),mycFeature]
    ta$feature<-ifelse(ta$feature==groupA,1,0)
  colnames(ta)[which(colnames(ta)=="feature")]<-mycFeature
    myForm<-formula(paste0(mycFeature,"~",pheno))
	fit<-glm(myForm,ta,family='binomial')
	ci<-exp(confint(fit))
  dd<-exp(coef(glm(myForm,ta,family='binomial')))
   dd<-cbind(ci,dd)%>%data.frame
  colnames(dd)<-c("lower.2.5%","upper.97.5%",paste0("OddsRatio.",groupA))
  pv<-summary(fit)%>%coef%>%data.frame
  dd$p<-pv[,4]
	return(dd[pheno,])
 }


multinomialRegressionClinicalOdds.RLNRatio<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn=NULL,mycFeature=NULL,groupA=1){
    message(pheno)
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
	ta<-as.data.frame(scale(ta))
    ta$feature<-full.dn4[match(rownames(ta),full.dn4$case),mycFeature]

      colnames(ta)[which(colnames(ta)=="feature")]<-mycFeature
    myForm<-formula(paste0(mycFeature,"~",pheno))
#	fit<-glm(myForm,ta,family='binomial')
   fit<- nnet::multinom(myForm, data = ta)
   z <- summary(fit)$coefficients/summary(fit)$standard.errors
   p <- (1 - pnorm(abs(z), 0, 1)) * 2
   z<-data.frame(z,p=p)
  return(z)
  } 





 getIntegratedColors<-function(immune=NULL){
   require(circlize)
    
  primary.phenotype<-as.character(immune$subannotation2)
   primary.phenotype<-factor(primary.phenotype)
   cd8s<-colorRampPalette(c("darkseagreen","green"),10)
  cd4s<-colorRampPalette(c("lightsteelblue1","blue"),10)
  macs<-colorRampPalette(c("magenta","maroon4"),10)
   tregs<-colorRampPalette(c("plum2","purple"),10)
   tumrs<-colorRampPalette(c("grey24","grey64"),10)
   phenoColors<-c(cd4s(8),
	cd8s(6),macs(9),tregs(8))

 names(phenoColors)<-levels(factor(primary.phenotype))
     return(phenoColors)

}


zScorePatientExpression<-function(full.dn4=NULL,markers=NULL,ROIID.column=NULL){
  X3<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(sd))%>%data.frame
  X2<-full.dn4[,c(markers,ROIID.column)]%>%group_by(ROIID)%>%summarise_all(funs(mean))%>%data.frame
   ###
   z<-full.dn4
  for(cl in unique(full.dn4$ROIID)){
   for(mark in colnames(X2)[-1]){
  z[which(z$ROIID==cl),which(colnames(z)==mark)]<- ( z[which(z$ROIID==cl),which(colnames(z)==mark)]-X2[which(X2$ROIID==cl),which(colnames(X2)==mark)])/(X3[which(X3$ROIID==cl),which(colnames(X3)==mark)])
   }
 }
 return(z)
 }

### make JoyPlot
   makeJoyPlot<-function(z,marker=NULL){
   ltlist<-list()
     for(roi in unique(z$ROIID)){
      lt = data.frame(density(z[which(z$ROIID==roi),marker])[c("x","y")])
    ltlist[[(length(ltlist)+1)]]<-lt
     }
    marker1= rowAnnotation(Z= anno_joyplot(ltlist, width = unit(4, "cm"), 
    gp = gpar(fill = color_clusters),
	 transparency = 0.35)) 
  marker1@anno_list[[1]]@name <-gsub("Cell_","",marker)
  names(marker1@anno_list)<-gsub("Cell_","",marker)
    return(marker1)
   }


 
metaClusterExperiment<-function(dn4=NULL,markers=NULL,plotPCA=FALSE){
   
   dn4$casePheno<-NA
   require(Rphenograph)
   for(roi in unique(dn4$ROIID)){
	message(roi)
     myroi<-subset(dn4,dn4$ROIID==roi)
     graph<-Rphenograph(myroi[,markers],k=45)
     myroi[,"casePheno"]<-membership(graph[[2]])
     dn4$casePheno[match(myroi$uniqueLabel,dn4$uniqueLabel)]<-myroi$casePheno
     }
  ## grab all the meta clusters.
  ErikMeta=dn4[,
	c(markers,
	"ROIID",
	"casePheno")] %>% group_by(ROIID,casePheno) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
     
          ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c(markers)], k = 15)
     phenograph_cluster_meta<- factor(membership(erikPheno[[2]]))
     table(phenograph_cluster_meta)
     ErikMeta=cbind(ErikMeta,phenograph_cluster_meta)
     ### plot PCA, label each case.  COO,  status.  ## 

   myLocal=left_join(dn4, ErikMeta[,c("ROIID",
	"casePheno","phenograph_cluster_meta")], by = c("ROIID", "casePheno"))

  dn4[,"phenograph_cluster_meta"]<-NA
  dn4[match(myLocal$uniqueLabel,dn4$uniqueLabel),"phenograph_cluster_meta"]<-myLocal[,"phenograph_cluster_meta"]

  plot_clustering_heatmap_wrapper(expr=ErikMeta[,c(markers)], 
  cell_clustering=ErikMeta[,"phenograph_cluster_meta"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 return(dn4)
}


 
plot_clustering_heatmap_wrapper<-function(expr=NULL, 
  cell_clustering=NULL, cluster_merging = NULL,useMedians=TRUE,useQuantiles=FALSE,
color_clusters=NULL){
  ## cluster_merging should be character merging vector of 29 or so clusters. this is input from the ANNOTATION object.  not a matched object level.
   expr<-as.matrix(expr)
  if(useQuantiles==TRUE){
  ##max min normalization
  library(matrixStats)
  rng <- colQuantiles(expr, probs = c(0.001, 0.999))
  expr01 <- t((t(expr) - rng[, 1]) / (rng[, 2] - rng[, 1]))
  expr01[expr01 < 0] <- 0
  expr01[expr01 > 1] <- 1
  }else{
   expr01<-expr 

  }

 if(is.null(color_clusters)==TRUE){
  color_clusters<- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")
 }

  # Calculate the median expression 
  library(dplyr)
  expr_median <- data.frame(expr, cell_clustering = cell_clustering) %>%
    group_by(cell_clustering) %>% 
    summarize_all(funs(median))
 if(useMedians==TRUE){
  expr01_median <- data.frame(expr01, cell_clustering = cell_clustering) %>%
    group_by(cell_clustering) %>% 
    summarize_all(funs(median))
   }else{
   expr01_median <- data.frame(expr01, cell_clustering = cell_clustering) %>%
    group_by(cell_clustering) %>% 
    summarize_all(funs(mean))
 }
  # Calculate cluster frequencies
  clustering_table <- as.numeric(table(cell_clustering))
  
  # This clustering is based on the markers that were used for the main clustering
  d <- dist(expr_median[, colnames(expr)], method = "euclidean")
  cluster_rows <- hclust(d, method = "average")
   expr_heat <- as.matrix(expr01_median[, colnames(expr01)])
  rownames(expr_heat) <- expr01_median$cell_clustering
 
  labels_row <- paste0(rownames(expr_heat), " (", 
    round(clustering_table / sum(clustering_table) * 100, 2), "%)")
  labels_col <- colnames(expr_heat)
  
  # Row annotation for the heatmap
  annotation_row <- data.frame(cluster = factor(expr01_median$cell_clustering))
  rownames(annotation_row) <- rownames(expr_heat)
  
  color_clusters <- color_clusters[1:nlevels(annotation_row$cluster)]
  names(color_clusters) <- levels(annotation_row$cluster)
  annotation_colors <- list(cluster = color_clusters)
  annotation_legend <- FALSE
  
  if(!is.null(cluster_merging)){
    
    annotation_row$cluster_merging <- factor(cluster_merging)
    color_clusters <- color_clusters[1:nlevels(factor(cluster_merging))]
    names(color_clusters) <- levels(factor(cluster_merging))
    annotation_colors$cluster_merging <- color_clusters
    annotation_legend <- TRUE
  }
  
  # Colors for the heatmap
  library(RColorBrewer);library(pheatmap)
  color <- colorRampPalette(rev(brewer.pal(n = 9, name = "RdYlBu")))(100)
  
  pheatmap(expr_heat, color = color, 
    cluster_cols = FALSE, cluster_rows = cluster_rows, 
    labels_col = labels_col, labels_row = labels_row, 
    display_numbers = TRUE, number_color = "black", 
    fontsize = 10, fontsize_number = 7,
    annotation_row = annotation_row, annotation_colors = annotation_colors,
    annotation_legend = annotation_legend)
  
}


 
 randomSampleUniqueLabel<-function(which.ROIID=NULL,N=NULL,full.dn4=NULL){
   dat<-subset(full.dn4,ROIID==which.ROIID)
   ids<-sample(dat$uniqueLabel,N,replace=FALSE) 
  return(ids)
 }


## phenograph to spatial point patter (PPP) object
 phenographToPPP<-function(case=NULL,marksCase=NULL,shift=FALSE){
 suppressPackageStartupMessages( require(spatstat))
  ##case: the case is the phenographed CSV 
  ##returns the PPP object, with marks as the original data frame.
  ##note that the window minimum is set to 0. This **SHOULD** not hopefully interfere with the spatial coordinates.
 ##this change of min 0 was necessary for clusterset function which conflicted with the window range.
  if(shift==TRUE){
  minX<-min(case[,"X_position"])
  minY<-min(case[,"Y_position"])
 case[,"X_position"]<-case[,"X_position"]-minX
 case[,"Y_position"]<-case[,"Y_position"]-minY
    mypat<-ppp(case[,"X_position"],case[,"Y_position"],c(min(case[,"X_position"]),summary(case[,"X_position"])[6]),c(min(case[,"Y_position"]),summary(case[,"Y_position"])[6]),marks=marksCase)
  ##shifting via spatstat causes non-stationary errors. FIXME: use manual shifting of coordinates prior   
 #mypat<-spatstat::shift(mypat,c(-minX,-minY))
  }else{
    #mypat<-ppp(case[,"X_position"],case[,"Y_position"],c(0,summary(case[,"X_position"])[6]),c(0,summary(case[,"Y_position"])[6]),marks=marksCase)
 mypat<-ppp(case[,"X_position"],case[,"Y_position"],c(min(case[,"X_position"]),max(case[,"X_position"])),c(min(case[,"Y_position"]),max(case[,"Y_position"])),marks=marksCase) 
 } 

 ## scale in  micrometers
   unitname(mypat)<-"micrometer"

  return(mypat)
 }



## 12-4 revision:  update the file paths for windows.
integrateHD.DLBCL<-function(hd=NULL,dlbcl=NULL){
  #load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/revision-comments-11-6/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
  
 ## for each marker use a tertile break point to identify phenotypes as a preliminary pass.
   #load("~/Documents/imageAnalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
  dlbcl.feats<-c("Cell_CCR4",
	"Cell_CD206","Cell_CD20","Cell_CD30","Cell_CD3",
	"Cell_CD45RA","Cell_CD45RO","Cell_CD4","Cell_CD68","Cell_CD8",
	"Cell_CXCR3","Cell_DNA","Cell_DNA2","Cell_EphrinB2","Cell_FOXP3","Cell_HLADR",
	"Cell_ICOS","Cell_Ki67","Cell_Lag3","Cell_PD1","Cell_PDL1","Cell_Tim3",
	"Cell_Tbet","Area","ROIID","uniqueLabel")
  dlbcl.sub<-full.dn4[,dlbcl.feats]
  dlbcl.sub$experiment<-"DLBCL"
  colnames(dlbcl.sub)<-gsub("Cell_","",colnames(dlbcl.sub))
 colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="Lag3")]<-"LAG3"
  colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="Tim3")]<-"TIM3"
   colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="DNA")]<-"DNA1"
   colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="CD8")]<-"CD8a"
  hd.feats<-c("CCR4",
	"CD206","CD20","CD30","CD3","CD45RA","CD45RO","CD4","CD68",
	"CD8a","CXCR3","DNA1","DNA2","EphrinB2","FOXP3","HLADR",
	"ICOS","Ki67","LAG3","PD1","PDL1","TIM3","Tbet","Area","ROIID","uniqueLabel")
  hd<-dn4[,hd.feats]
  hd$experiment<-"HD"
  all(colnames(dlbcl.sub)==colnames(hd))
   integ<-rbind(dlbcl.sub,hd)
  
  ## patient means, patient SD,  experiment means , experiment SD
   markers<-c("CCR4",
	"CD206","CD20","CD30","CD3","CD45RA","CD45RO","CD4","CD68",
	"CD8a","CXCR3","DNA1","DNA2","EphrinB2","FOXP3","HLADR",
	"ICOS","Ki67","LAG3","PD1","PDL1","TIM3","Tbet","Area")
  X3<-integ[,c(markers,"ROIID")]%>%group_by(ROIID)%>%summarise_all(funs(sd))%>%data.frame
  X2<-integ[,c(markers,"ROIID")]%>%group_by(ROIID)%>%summarise_all(funs(mean))%>%data.frame
  X3_1<-integ[,c(markers,"experiment")]%>%group_by(experiment)%>%summarise_all(funs(sd))%>%data.frame
  X2_1<-integ[,c(markers,"experiment")]%>%group_by(experiment)%>%summarise_all(funs(mean))%>%data.frame
   ###
   z<-integ
  for(cl in unique(integ$ROIID)){
  message(cl)
   for(mark in colnames(X2)[-1]){
  means<-(z[which(z$ROIID==cl),which(colnames(z)==mark)]-X2_1[match(unique(z$experiment[which(z$ROIID==cl)]), X2_1$experiment),which(colnames(X2_1)==mark)])
  pool.sd<-X3_1[match(unique(z$experiment[which(z$ROIID==cl)]), X3_1$experiment),which(colnames(X3_1)==mark)]
 z[which(z$ROIID==cl),which(colnames(z)==mark)]<-means/pool.sd
   }
 }
 z2<-zScorePatientExpression(full.dn4=z,
	markers=which(colnames(z)%in%markers),
	ROIID.column=which(colnames(z)=='ROIID'))
  z2[,markers]<-scale(z2[,markers]) 
  return(z2)
}



plotIntegrationPlots<-function(integ=NULL){

 
   z<-integ
   z<-z[!is.nan(z$DNA1),]
     z<-z[!is.nan(z$DNA2),]
   d<-data.frame(ROIID=unique(z$ROIID))
   # names(roiCol)<-d[,1]
  ### add ROI annotation.
 d$experiment[match(z$ROIID,d$ROIID)]<-z$experiment 
 rois<-HeatmapAnnotation(experiment=(d$experiment),
	which="row")

     cd20<-makeJoyPlot(z,marker="CD20")
   cd206<-makeJoyPlot(z,marker="CD206")
    CD3<-makeJoyPlot(z,marker="CD3")
   CD4<-makeJoyPlot(z,marker="CD4")
   CD45RA<-makeJoyPlot(z,marker="CD45RA")
   cd45ro<-makeJoyPlot(z,marker="CD45RO")
   cd68<-makeJoyPlot(z,marker="CD68")
    cd8<-makeJoyPlot(z,marker="CD8a")
   foxp3<-makeJoyPlot(z,marker="FOXP3")
   hladr<-makeJoyPlot(z,marker="HLADR")
   ccr4<-makeJoyPlot(z,marker="CCR4")
    cxcr3<-makeJoyPlot(z,marker="CXCR3")
    icos<-makeJoyPlot(z,marker="ICOS")
   ki<-makeJoyPlot(z,marker="Ki67")
   lag<-makeJoyPlot(z,marker="LAG3")
   pd1<-makeJoyPlot(z,marker="PD1")
   pdl1<-makeJoyPlot(z,marker="PDL1")
    tim3<-makeJoyPlot(z,marker="TIM3")
   tbet<-makeJoyPlot(z,marker="Tbet")
    eph<-makeJoyPlot(z,marker="EphrinB2") 
     cd30<-makeJoyPlot(z,marker="CD30") 
      dna1<-makeJoyPlot(z,marker="DNA1") 
      dna2<-makeJoyPlot(z,marker="DNA2") 

  draw(cd20+cd206+CD3+CD4+CD45RA+cd45ro+rois)
  draw(cd68+cd8+rois)
 # savePlot(file="integration.zscores.expression.png")
 draw(foxp3+hladr+ccr4+cxcr3+icos+ki+rois)
  draw(lag+pd1+pdl1+tim3+rois)
  #savePlot(file="integration.zscores.expression2.png")
  draw(tbet+eph+cd30+dna1+dna2+rois)
  # savePlot(file="integration.zscores.expression3.png")

}


  # 12-4 revision  update windows path
seuratDownSampleIntegrationEvaluation<-function(downSample=TRUE){
  
 #load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
   ## for each marker use a tertile break point to identify phenotypes as a preliminary pass.
  # load("~/Documents/imageAnalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
#  load("C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/revision-comments-11-6/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
  
 ## for each marker use a tertile break point to identify phenotypes as a preliminary pass.
   #load("~/Documents/imageAnalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
  
    z<-integrateHD.DLBCL()
	z<-z[!grepl("Tonsil_",z$ROIID),]
	z<-z[!grepl("Placenta_",z$ROIID),]
	z<-z[!is.na(z$DNA1),]
	z<-z[!is.na(z$DNA2),]
	
  ### evaluate the integration.  use seurat.
  if(downSample==TRUE){
  hd_ids<-sapply(unique(z[which(z$experiment!="DLBCL"),"ROIID"]),function(x) randomSampleUniqueLabel(which.ROIID=x,N=1900,full.dn4=z[which(z$experiment!="DLBCL"),]))
   hd_ids<-as.vector(hd_ids)
  hd_ids2<-sapply(unique(z[which(z$experiment=="DLBCL"),"ROIID"]),function(x) randomSampleUniqueLabel(which.ROIID=x,N=1900,full.dn4=z[which(z$experiment=="DLBCL"),]))
  hd_ids2<-as.vector(hd_ids2)
   downSamp<-z[match(c(hd_ids2,hd_ids),z$uniqueLabel),]
  }else{
 	downSamp<-z
  }
  	 rownames(downSamp)<-downSamp$uniqueLabel
	raw<-t(downSamp[,1:23])
    ## create summarized experiment
 	pbmc <- CreateSeuratObject(counts =raw )	
	groups<-downSamp$experiment
	pbmc <- AddMetaData(object = pbmc, metadata = downSamp$experiment, col.name = "tech")
	pbmc <- ScaleData(pbmc, verbose = FALSE)
	## the VST will error, choose mvp or disp
	pbmc <- FindVariableFeatures(pbmc, selection.method = "disp", nfeatures =15)
	pbmc <- RunPCA(pbmc, npcs = 10, verbose = FALSE)
	#pbmc <- RunUMAP(pbmc, reduction = "pca", dims = 1:20)
	#pbmc <- FindNeighbors(pbmc, dims = 1:10)
	#pbmc <- FindClusters(pbmc, resolution = 0.5)
	#p1 <- DimPlot(pbmc, reduction = "pca", group.by = "tech")
	#normalDat<-GetAssayData(object = pbmc, slot = "scale.data")
	#normalDat<-t(normalDat)

  return(pbmc)
  }




metaClusterPhenotype<-function(dn4=NULL,phenotype=NULL,myPheno=NULL,markers=NULL,plotPCA=FALSE,k2=15){
 #myPheno is the column for the phenotype.
    set.seed(1234)
   erik<-subset(dn4,dn4[,myPheno]==phenotype)
  ## subMeta holds the casePheno 
  dn4[,paste0(phenotype,"_subMeta")]<-NA
    erik[,paste0(phenotype,"_subMeta")]<-NA
 require(Rphenograph)
   roiData<-as.data.frame.matrix(table(erik$ROIID,erik[,myPheno]))

   for(roi in rownames(roiData)[which(roiData[,phenotype]>45)]){
     myroi<-subset(erik,erik$ROIID==roi)
     graph<-Rphenograph(myroi[,markers],k=45)
     myroi[,paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
    dn4[match(myroi$uniqueLabel,dn4$uniqueLabel),paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
     erik[match(myroi$uniqueLabel,erik$uniqueLabel),paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
   }

  ErikMeta=erik[,
	c(markers,
	"ROIID",
	paste0(phenotype,"_subMeta"))] %>% group_by(ROIID,.dots=paste0(phenotype,"_subMeta")) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
     
          ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c(markers)], k = k2)
     subphenograph_cluster_meta<- factor(membership(erikPheno[[2]]))
     table(subphenograph_cluster_meta)
     ErikMeta=cbind(ErikMeta, subphenograph_cluster_meta)
     ### plot PCA, label each case.  COO,  status.  ## 
   if( any(colnames(erik)== paste0(phenotype,"_subphenograph_cluster_meta"))){
     erik<-erik[,which(colnames(erik)!= paste0(phenotype,"_subphenograph_cluster_meta"))]
 } 

   myLocal=left_join(erik, ErikMeta[,c("ROIID",paste0(phenotype,"_subMeta"),"subphenograph_cluster_meta")], by = c("ROIID"))

  colnames(myLocal)[which(colnames(myLocal)=="subphenograph_cluster_meta")]<-paste0(phenotype,"_subphenograph_cluster_meta") 
  dn4[,paste0(phenotype,"_subphenograph_cluster_meta")]<-NA
  dn4[match(myLocal$uniqueLabel,dn4$uniqueLabel),paste0(phenotype,"_subphenograph_cluster_meta")]<-myLocal[,which(colnames(myLocal)==paste0(phenotype,"_subphenograph_cluster_meta"))]

  plot_clustering_heatmap_wrapper(expr=ErikMeta[,c(markers)], 
  cell_clustering=ErikMeta[,"subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
 return(dn4)
}


metaClusterExperiment<-function(dn4=NULL,markers=NULL,plotPCA=FALSE){
   ##metaclusters across ROI.
   dn4$casePheno<-NA
   require(Rphenograph)
   for(roi in unique(dn4$ROIID)){
	message(roi)
     myroi<-subset(dn4,dn4$ROIID==roi)
     graph<-Rphenograph(myroi[,markers],k=45)
     myroi[,"casePheno"]<-membership(graph[[2]])
     dn4$casePheno[match(myroi$uniqueLabel,dn4$uniqueLabel)]<-myroi$casePheno
     }
  ## grab all the meta clusters.
  ErikMeta=dn4[,
	c(markers,
	"ROIID",
	"casePheno")] %>% group_by(ROIID,casePheno) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
     
          ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c(markers)], k = 15)
     phenograph_cluster_meta<- factor(membership(erikPheno[[2]]))
     table(phenograph_cluster_meta)
     ErikMeta=cbind(ErikMeta,phenograph_cluster_meta)
     ### plot PCA, label each case.  COO,  status.  ## 

   myLocal=left_join(dn4, ErikMeta[,c("ROIID",
	"casePheno","phenograph_cluster_meta")], by = c("ROIID", "casePheno"))

  dn4[,"phenograph_cluster_meta"]<-NA
  dn4[match(myLocal$uniqueLabel,dn4$uniqueLabel),"phenograph_cluster_meta"]<-myLocal[,"phenograph_cluster_meta"]

  plot_clustering_heatmap_wrapper(expr=dn4[,c(markers)], 
  cell_clustering=dn4[,"phenograph_cluster_meta"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 return(dn4)
}

 plotMeta<-function(dn4=NULL){

 markers<-c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3")
   set.seed(3242020)
   ErikMeta=dn4[,
	c(markers,
	"ROIID",
	"casePheno")] %>% group_by(ROIID,casePheno) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
     
           ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c(markers)], k = 45)
     phenograph_cluster_meta<- factor(membership(erikPheno[[2]]))
     table(phenograph_cluster_meta)
     ErikMeta=cbind(ErikMeta,phenograph_cluster_meta)

   myLocal=left_join(dn4, ErikMeta[,c("ROIID",
	"casePheno","phenograph_cluster_meta")], by = c("ROIID", "casePheno"))

  dn4[,"phenograph_cluster_meta"]<-NA
  dn4[match(myLocal$uniqueLabel,dn4$uniqueLabel),"phenograph_cluster_meta"]<-myLocal[,"phenograph_cluster_meta"]

   plot_clustering_heatmap_wrapper(expr=ErikMeta[,c(markers)], 
  cell_clustering=ErikMeta[,"phenograph_cluster_meta"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 return(dn4)
  }

getExperimentColors<-function(){
  cd4s<-colorRampPalette(c("khaki1","plum2"),10)
  cd4s<-cd4s(2)
 names(cd4s)<-c("DLBCL","HD")
  return(cd4s)
}

##REVISION 11-1: we included the RLN in the bar plot TME comparison, and use a LM instead of t-test.
integratedPrimaryCommunityBarPlot<-function(cleandat=NULL,experiment='experiment'){
   tums<-cleandat[cleandat$annotation%in%c("CD4","CD8","TREG","Macrophage"),]
   tums$response<-factor(as.character(tums$response))
  ## filter prevalance
  x<-as.data.frame.matrix(table(tums[,experiment],tums$annotation))
  x<-x/rowSums(x)
  toFilter<-which(apply(apply(x,2,function(x) x==0),2,function(y) any(y==TRUE)))
  tums<-tums[!tums$annotation%in%names(toFilter),]

 ## need to plot all the tumor proportions CR vs. REF
  #response vars=c("darkslateblue","firebrick2"))
 
 computeImmuneProportions<-function(full.dn4=NULL,primary=NULL,clinicalFactor=NULL){
  tumor.abund<-as.data.frame.matrix(table(tums$ROIID,tums$annotation))
  ## immune proportions
  tumor.abund<-100*(tumor.abund/(rowSums(tumor.abund)))

  tumor.abund$response<-full.dn4[,clinicalFactor][match(rownames(tumor.abund),full.dn4$ROIID)]
  tumor.means<-tumor.abund%>%group_by(response)%>%summarise_all(funs(mean))%>%data.frame
  rownames(tumor.means)<-tumor.means[,1]
  tumor.means<-tumor.means[,-1]
  tumor.means<-t(tumor.means)
  tumor.means<-as.data.frame(tumor.means)
  tumor.means$pheno<-rownames(tumor.means)
  #tumor.means<-melt(tumor.means)
  tumors<-reshape(tumor.means,varying=1:3,v.names='response',timevar='pheno',direction='long')
  tumors$pheno<-colnames(tumor.means)[tumors$pheno]
  tumors$id<-tumor.means$pheno[tumors$id]
  tumor.means<-tumors
  #tumor.means<-data.frame(means=c(tumor.means[,1],tumor.means[,2]),response=c(rep(colnames(tumor.means)[1],nrow(tumor.means)),rep(colnames(tumor.means)[2],nrow(tumor.means))),names=rep(rownames(tumor.means,2)))
  tumor.sd<-tumor.abund%>%group_by(response)%>%summarise_all(funs(sd))%>%data.frame
  tumor.se<-tumor.sd[,-1]/sqrt(table(tumor.abund$response))
   tumor.se<-t(tumor.se)
   colnames(tumor.se)<-tumor.sd$response
   tumor.se<-as.data.frame(tumor.se)
     tumor.se$pheno<-rownames(tumor.se)
   tumorSE<-reshape(tumor.se,varying=1:3,v.names='response',timevar='pheno',direction='long')
 tumorSE$pheno<-colnames(tumor.se)[tumorSE$pheno]
  tumorSE$id<-tumor.se$pheno[tumorSE$id]
  tumor.se<-tumorSE
     # tumor.means$se<-c(tumor.se[,1],tumor.se[,2])
  ## need to plot all the tumor proportions CR vs. REF
  tumor.means<- left_join(tumor.means,tumor.se,by=c("pheno","id"))
  colnames(tumor.means)[which(colnames(tumor.means)=='response.x')]<-'means'
    colnames(tumor.means)[which(colnames(tumor.means)=='response.y')]<-'se'
  return(tumor.means)
 }



 ## add the immune populations.
  ## MAC, TREG, CD4, CD8.
   cd4.prop<-computeImmuneProportions(full.dn4=tums,primary="CD4",clinicalFactor=experiment)
  cd4.prop<-cd4.prop[which(cd4.prop$means>0),]
    cd4s<-getExperimentColors()
    cd4s<-c(cd4s,'lightcoral')
    names(cd4s)[3]<-'RLN'
    cd4.exp<-ggplot(cd4.prop,aes(x=id,y=means,fill=pheno))+geom_bar(stat='identity',position=position_dodge(),colour='black')+geom_errorbar(aes(ymin=abs(means-2*se), ymax=means+2*se), width=.2,
                 position=position_dodge(.9))+theme_bw(base_size=14)+ theme(axis.text.x = element_text(angle = 45, hjust = 1,color='black',size=17),
	axis.text.y=element_text(size=15,color='black'))+ylab("Relative proportion (%)")+xlab("TME communities")+scale_fill_manual("Cohort",values=cd4s)
  cd4.exp
 ### immune proportions
  #savePlot(file="TME.primary.immune.proportions.png")

 ## make a sheet of t-test results.
 computeImmuneProportionsTTest<-function(full.dn4=NULL,primary=NULL,clinicalFactor=NULL){
  tumor.abund<-as.data.frame.matrix(table(tums$ROIID,tums$annotation))
  ## immune proportions
  tumor.abund<-100*(tumor.abund/(rowSums(tumor.abund)))
  tumor.abund$response<-full.dn4[,clinicalFactor][match(rownames(tumor.abund),full.dn4$ROIID)]
  tres<-t.test(tumor.abund[which(tumor.abund$response=="DLBCL"),primary],
 	tumor.abund[which(tumor.abund$response=="HD"),primary])
	   ## need to plot all the tumor proportions CR vs. REF
  return(tres)
 }
 
  computeImmuneProportionsLinearModel<-function(full.dn4=NULL,primary=NULL,clinicalFactor=NULL){
  tumor.abund<-as.data.frame.matrix(table(tums$ROIID,tums$annotation))
  ## immune proportions
  tumor.abund<-100*(tumor.abund/(rowSums(tumor.abund)))
  tumor.abund$response<-full.dn4[,clinicalFactor][match(rownames(tumor.abund),full.dn4$ROIID)]
  form<-as.formula(paste0(primary,"~response"))
  tres<-coef(summary(lm(form,tumor.abund)))
  tres<-as.data.frame(tres)
  tres$phenotype<-primary
	   ## need to plot all the tumor proportions CR vs. REF
  return(tres)
 }
 
    cd4.t<-computeImmuneProportionsTTest(full.dn4=tums,primary="CD4",clinicalFactor=experiment)
    cd4.lm<-as.data.frame(computeImmuneProportionsLinearModel(full.dn4=tums,primary="CD4",clinicalFactor=experiment))
    cd4.lm$bonf.p<-cd4.lm[,4]*4
 	require(broom)
	cd.4<-tidy(cd4.t)%>%data.frame
	cd.4$bonf.p<-cd.4$p.value*4
      cd8.t<-computeImmuneProportionsTTest(full.dn4=tums,primary="CD8",clinicalFactor="experiment")
      cd8.lm<-as.data.frame(computeImmuneProportionsLinearModel(full.dn4=tums,primary="CD8",clinicalFactor=experiment))
  cd8.lm$bonf.p<-cd8.lm[,4]*4
	cd.8<-tidy(cd8.t)%>%data.frame
	cd.8$bonf.p<-cd.8$p.value*4  
  mac.t<-computeImmuneProportionsTTest(full.dn4=tums,primary="Macrophage",clinicalFactor="experiment")
	mac<-tidy(mac.t)%>%data.frame
	mac$bonf.p<-mac$p.value*4    
	mac.lm<-as.data.frame(computeImmuneProportionsLinearModel(full.dn4=tums,primary="Macrophage",clinicalFactor=experiment))
 mac.lm$bonf.p<-mac.lm[,4]*4

treg.t<-computeImmuneProportionsTTest(full.dn4=tums,primary="TREG",clinicalFactor=experiment)
  treg<-tidy(treg.t)%>%data.frame
	treg$bonf.p<-treg$p.value*4
	treg.lm<-as.data.frame(computeImmuneProportionsLinearModel(full.dn4=tums,primary="TREG",clinicalFactor=experiment))
treg.lm$bonf.p<-treg.lm[,4]*4
 
 # library("xlsx")
#    write.xlsx(cd4.lm, file = "proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes.xlsx",
 #     sheetName = "CD4", append = FALSE)
  #  write.xlsx(cd8.lm, file = "proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes.xlsx",
 #      sheetName = "CD8", append = TRUE)
 # write.xlsx(mac.lm, file = "proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes.xlsx",
 #    sheetName = "Macrophage", append = TRUE) 
 # write.xlsx(treg.lm, file = "proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes.xlsx",
 #    sheetName = "TREG", append = TRUE)

    return(cd4.exp)
 }

chiHelperMeanScale.IntegrationComparisonEnrichment<-function(pheno=NULL,full.dn4=NULL,
	phenotypeColumn="unsup.subcommunity",
	conditionOnPrimary=FALSE,contrastFactor='COO',
	groupA='GCB'){
 ### the features are average to scale the data.
	##only works for HANS COO
	##phenotype column is the column of phenoytpe levels.
    message(pheno)
    primary<-sapply(strsplit(pheno,"_"),function(x) x[1])
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
    ta$COO<-full.dn4[match(rownames(ta),full.dn4$case),contrastFactor]
    ta2<-ta[,c(pheno,"COO")]
    x11<-round(mean(ta2[which(ta2$COO==groupA),1]))+1
    x12<-round(mean(ta2[which(ta2$COO!=groupA),1]))+1
	##conditioned on the same phenotype family.
  if(conditionOnPrimary==TRUE){
    ta3<-ta[,grepl(primary,colnames(ta))]
  }else{
      ta3<-ta[,which(colnames(ta)!=pheno)]
  }
   ##ta3 object is negative for the phenotype
    ta3$COO<-full.dn4[match(rownames(ta3),full.dn4$case),contrastFactor]
   ta3<-ta3%>%group_by(COO)%>%summarise_all(funs(mean))%>%data.frame()
    x21<-round(mean(as.numeric(ta3[which(ta3$COO==groupA),which(colnames(ta3)!="COO")])))+1
    x22<-round(mean(as.numeric(ta3[which(ta3$COO!=groupA),which(colnames(ta3)!="COO")])))+1
    mat<-matrix(c(x11,x21,x12,x22),nrow=2,ncol=2)
	rownames(mat)<-c(paste0(pheno,"+"),paste0(pheno,"-"))
	colnames(mat)<-c(groupA,paste0("not.",groupA))
 ### need OR with confidence intervals
     require(epitools)
     dd<-oddsratio(mat,rev="both")
	return(dd)
  }
  

plotIntegrationPlots.IMMUNE<-function(integ=NULL){

   exCol<-getExperimentColors()
   z<-integ
   z<-z[!is.nan(z$DNA1),]
     z<-z[!is.nan(z$DNA2),]
   d<-data.frame(ROIID=unique(z$ROIID))
   # names(roiCol)<-d[,1]
  ### add ROI annotation.
 d$experiment[match(z$ROIID,d$ROIID)]<-z$experiment 
 rois<-HeatmapAnnotation(experiment=(d$experiment),
	which="row",
	col=list(experiment=exCol))

      cd3<-makeJoyPlot(z,marker="CD3")
   cd206<-makeJoyPlot(z,marker="CD206")
   CD4<-makeJoyPlot(z,marker="CD4")
   CD45RA<-makeJoyPlot(z,marker="CD45RA")
   cd45ro<-makeJoyPlot(z,marker="CD45RO")
   cd68<-makeJoyPlot(z,marker="CD68")
    cd8<-makeJoyPlot(z,marker="CD8a")
   foxp3<-makeJoyPlot(z,marker="FOXP3")
   ccr4<-makeJoyPlot(z,marker="CCR4")
    cxcr3<-makeJoyPlot(z,marker="CXCR3")
    pd1<-makeJoyPlot(z,marker="PD1")
   pdl1<-makeJoyPlot(z,marker="PDL1")
    tim3<-makeJoyPlot(z,marker="TIM3")
    dna1<-makeJoyPlot(z,marker="DNA1") 
    dna2<-makeJoyPlot(z,marker="DNA2") 
    hladr<-makeJoyPlot(z,marker="HLADR") 
    area<-makeJoyPlot(z,marker="Area") 
  draw(tim3+cd45ro+CD45RA+hladr+rois)
 # savePlot(file="immune.integration.zscores.expression.png")
 draw(ccr4+cxcr3+pd1+pdl1+rois)
 draw(dna1+dna2+area+rois)

# savePlot(file="immune.integration.zscores.expression2.png")
 

}


integratedBoxPlotsBatchEvaluation<-function(cleandat=NULL){
   require(limma) 
   immune<-cleandat[which(cleandat$annotation!="BCell" & cleandat$annotation!="HRS"),]
### BOXPLOTS OF BATCH EVALUATION.
  ## full plots.
  exCol<-getExperimentColors()
 pdl1<-cleandat%>%ggplot(aes(x=ROIID,y=PDL1,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 pd1<- cleandat%>%ggplot(aes(x=ROIID,y=PD1,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  tim3<-cleandat%>%ggplot(aes(x=ROIID,y=TIM3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 ccr4<-cleandat%>%ggplot(aes(x=ROIID,y=CCR4,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 cxcr3<-cleandat%>%ggplot(aes(x=ROIID,y=CXCR3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 lag3<-cleandat%>%ggplot(aes(x=ROIID,y=LAG3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  icos<-cleandat%>%ggplot(aes(x=ROIID,y=ICOS,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  ki67<-cleandat%>%ggplot(aes(x=ROIID,y=Ki67,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  hladr<-cleandat%>%ggplot(aes(x=ROIID,y=HLADR,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)

  area<-cleandat%>%ggplot(aes(x=ROIID,y=Area.norm,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)


  full.state<-grid.arrange(pdl1,pd1,tim3,ccr4,cxcr3,icos,hladr,ki67,lag3,nrow=3,ncol=3)
  ## we do not choose lag3 nor Ki67
  # savePlot(file="full.plots.states.png")


 cd68<-cleandat%>%ggplot(aes(x=ROIID,y=CD68,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  cd4<-cleandat%>%ggplot(aes(x=ROIID,y=CD4,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  cd8<-cleandat%>%ggplot(aes(x=ROIID,y=CD8a,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
  foxp3<-cleandat%>%ggplot(aes(x=ROIID,y=FOXP3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 cd3<-cleandat%>%ggplot(aes(x=ROIID,y=CD3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 dna<-cleandat%>%ggplot(aes(x=ROIID,y=DNA1,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 cd206<-cleandat%>%ggplot(aes(x=ROIID,y=CD206,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)
 
 cd45ra<-cleandat%>%ggplot(aes(x=ROIID,y=CD45RA,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)

 # full.lineage<-grid.arrange(cd4,cd8,cd68,foxp3,cd206,dna,nrow=2,ncol=3)
#  savePlot(file="full.plots.lineage.limma.png")


 ## TME only
  pdl1.tme<-immune%>%ggplot(aes(x=ROIID,y=PDL1,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
 pd1.tme<-immune%>%ggplot(aes(x=ROIID,y=PD1,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
    tim3.tme<-immune%>%ggplot(aes(x=ROIID,y=TIM3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
   ccr4.tme<-immune%>%ggplot(aes(x=ROIID,y=CCR4,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
   cxcr3.tme<-immune%>%ggplot(aes(x=ROIID,y=CXCR3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
   icos.tme<-immune%>%ggplot(aes(x=ROIID,y=ICOS,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
 lag3.tme<-immune%>%ggplot(aes(x=ROIID,y=LAG3,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
    ki67.tme<-immune%>%ggplot(aes(x=ROIID,y=Ki67,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI (TME)")+scale_fill_manual(values=exCol)
  

 area<-immune%>%ggplot(aes(x=ROIID,y=Area.norm,fill=experiment))+geom_boxplot()+theme_bw(base_size=14)+theme(axis.ticks.x=element_blank(),axis.text.x=element_blank())+xlab("ROI")+scale_fill_manual(values=exCol)


tme.state<-grid.arrange(pdl1.tme,
			pd1.tme,
			tim3.tme,
			ccr4.tme,
			cxcr3.tme,
			icos.tme,nrow=2,ncol=3)
  # savePlot(file="TME.plots.states.immune.stratified.png")


    ## plot the supplementary method
 
  tme.state2<-grid.arrange(lag3.tme,
			ki67.tme,
			area,
			nrow=2,ncol=2)
  # savePlot(file="TME.plots.states.immune.stratified.supplementary.extras.png")

 

}



 ## linear model to show that PDL1 is up in HD on MAC and PD1 is up on CD4/CD8 .
  empiricalFit<-function(cleandat=NULL,phenotype=NULL){
   patient<-cleandat%>%group_by(ROIID,annotation)%>%summarise(PD1=mean(PD1),PDL1=mean(PDL1),TIM3=mean(TIM3),CCR4=mean(CCR4),CXCR3=mean(CXCR3),ICOS=mean(ICOS),Ki67=mean(Ki67))%>%data.frame
   pat.pheno<-subset(patient,patient$annotation==phenotype)
   pat.pheno$experiment<-cleandat$experiment[match(pat.pheno$ROIID,cleandat$ROIID)]
   pat.pheno$normal<-cleandat$normalLymphnode[match(pat.pheno$ROIID,cleandat$ROIID)]
   rownames(pat.pheno)<-pat.pheno$ROIID
   pat<-pat.pheno[,c("PD1","PDL1","TIM3","ICOS","Ki67")]
    pat<-t(pat)
  design<-model.matrix(~experiment+normal,pat.pheno)
  require(limma)
  fit <- lmFit(pat,design)
 #  Moderated t-statistic
  fit <- eBayes(fit)
 print( topTable(fit,coef=2))
  return(fit)
 }


 grabDistanceRange<-function(synth=NULL,uniqueLabel=NULL,k=1,from=NULL,to=NULL){
  nnd<-as.data.frame(nndist(synth,k=k,by=marks(synth)))
  colnames(nnd)<-paste0("k_",k,"_",colnames(nnd))
  for(i in 1:ncol(nnd)){
  nnd[which(is.infinite(nnd[,i])),i] <-300
 }##
   nnd$marks<-marks(synth)
   nnd$K=1
   nnd$uniqueLabel<-uniqueLabel 
  return(nnd)
  }


distanceCalculationROI<-function(synth=NULL,zz=NULL,cluster.column="cd8.case.cluster"){
  require(Rphenograph)
  nnd1<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=1)
    nnd2<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=2)
  nnd3<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=3)
  nnd4<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=4)
  nnd5<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=5)
  #nnd6<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=6)
  #nnd7<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=7)
  #nnd8<-grabDistanceRange(synth,uniqueLabel=zz$uniqueLabel,k=8)
  
  ## assemble
  nnd<-cbind(nnd1[,grepl("k_",colnames(nnd1))],
	nnd2[,grepl("k_",colnames(nnd2))],
	nnd3[,grepl("k_",colnames(nnd3))],
	nnd4[,grepl("k_",colnames(nnd4))],
	nnd5[,grepl("k_",colnames(nnd5))])
	#nnd6[,grepl("k_",colnames(nnd6))],
	#nnd7[,grepl("k_",colnames(nnd7))],
	#nnd8[,grepl("k_",colnames(nnd8))])
    nnd$uniqueLabel<-zz$uniqueLabel
    nnd$marks<-marks(synth)
   stopifnot(all(nrow(nnd)==nrow(zz)))
     nnd$ROIID<-zz$ROIID
    nnd$cluster<-zz[,cluster.column]
  return(nnd)
}

 setUpMarkers<-function(){
 hd.lineage<-c( "CD11b",
                       "CD11c",
                       "CD14",
                       "CD15",
                       "CD16",
                       "CD206",
			"CD20",
			"CD30",
			"CD34",
			"CD3",
			"CD45RA",
			"CD45RO",
			"CD4",
			"CD68",
			"CD8a",
                       "FOXP3",
			"EphrinB2",
			"Eccentricity.norm",
			"Area.norm",
			"DNA1",
			"DNA2") 
 hd.induc<-c("X41BB",
		"CCR4",
		"CXCR3",
		"CXCR5",
		"Caspase3",
		"GATA3",
		"HLAABC",
		"HLADR",
		"ICOS",
 		"IDO",
		"Ki67",
		"LAG3",
		"NKG2D",
		"PD1",
		"PDL1",
		"TIM3",
		"Tbet")

  lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
		"Cell_DNA",
		"Cell_DNA2")
  
  myl<-list(hd.lineage,	
	hd.induc,
	lineage_markers,
	induc)
	names(myl)<-c("hd.lineage","hd.induc","lineage_markers","induc")
 
 return(myl)
   
 }


 color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")

 
  threeColors<-getExperimentColors()
    threeColors<-c(threeColors,'indianred3')
    names(threeColors)[3]<-'RLN'
    

    
   findSecondObjectNumber<-function(pp=NULL,minDistance=13.5){
  ##FIX ME : use apply nbhd
   x<-pairdist(pp)
    id<-apply(x,1,function(x) which(x<=minDistance))
  ##stack the list into long format
  names(id)<-as.character(seq(1,npoints(pp)))
  datRel<-stack(id)
 ##the second object near the first fixed object.
  colnames(datRel)<-c("Second Object Number","First Object Number")
   ##exclude self
   datRel<-datRel[which(datRel[,"First Object Number"]!=datRel[,"Second Object Number"]),]
   datRel[,1]<-as.numeric(datRel[,1])
  datRel[,2]<-as.numeric(datRel[,2])
   return(datRel)
  }

     
 createRelationTable<-function(dn5=NULL,myPheno=NULL,minDistance=12){
  ##for each ROI, create PPP
  dat_relation2<-c()
  for(roi in unique(dn5$ROIID)){
   message(roi)
   dats<-subset(dn5,dn5$ROIID==roi)
   synth3.cells<-phenographToPPP(case=dats,marks=factor(dats[,myPheno]))
   dat_relation<-findSecondObjectNumber(pp=synth3.cells,minDistance=minDistance)
  dat_relation$Relationship<-"Neighbors"
  dat_relation[,"First Object Name"]<-"cell"
  dat_relation[,"First Image Number"]<-roi
  dat_relation[,"Second Object Name"]<-"cell"
  dat_relation[,"Second Image Number"]<-roi
  dat_relation2<-rbind(dat_relation2,dat_relation)
  }
 return(data.table(dat_relation2))
}

    
 

makePermutationNBHD<-function(dat_cells=NULL,dat_relation=NULL,n_perm=100,pthreshold=0.05){
  d = prepare_tables(dat_cells, dat_relation)
  dat_baseline = apply_labels(d[[1]], d[[2]]) %>%
  aggregate_histo()


 set.seed(12312)
   dat_perm = rbindlist(mclapply(1:n_perm, function(x){
  dat_labels = shuffle_labels(d[[1]])
  apply_labels(dat_labels, d[[2]]) %>%
    aggregate_histo()
  },mc.cores = 1
  ), idcol = 'run') 
 
   dat_p <- calc_p_vals(dat_baseline, dat_perm, n_perm = n_perm, p_tresh = pthreshold) 
 ### plotting heatmap.
  
   pmat = dcast(dat_p, 'FirstLabel ~ SecondLabel', value.var = 'sigval', fun.aggregate = sum,
             fill=0, drop=F)
   rname = pmat$FirstLabel
   pmat<-as.data.table(pmat)
   pmat = pmat[,which(colnames(pmat)!="FirstLabel"),with=FALSE] %>%
    as.matrix()
row.names(pmat) <- rname
 ### he structures his probabilities greater than and less than.
  return(list(pmat=pmat,dat_p=dat_p))
 }
```

### Library
  
```{r,eval=TRUE}

  library(spatstat)
  library(data.table)
  library(dplyr)
  library(magrittr)
  library(dtplyr)  
  library(ggplot2)
  library(parallel)
  library(neighbouRhood)
  library(gplots)
  library(RColorBrewer)
  library(magrittr);library(dplyr)
  library(ComplexHeatmap)
 library(Seurat)
  library(hrbrthemes)
  library(lsmeans);library(afex)
 library(kBET)

 # load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
  ## we need to use the non-filtered full.dn4 and re-create the filtered object.
# load(file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/revision-comments-11-6/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")
 full.dn4<-full.dn4[which(full.dn4$dimCD20.filtered!='YES'),]

 ## for each marker use a tertile break point to identify phenotypes as a preliminary pass.
  # load("~/Documents/imageAnalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.RData")
load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/Figure4/original-paper-data/dn4.final.foldRevised.RData")


```
### Hodgkin Lymphoma phenotype clustering
- Using the Aoki et. al cohort, we phenotype the TME present using meta-clustering as previously performed in DLBCL. 
- 12-4 update the path for local windows use.
```{r HL,eval=FALSE}
 ## we used Image J to correct tissue folds/tears by masking out those regions on the ROI.
  # load("~/Documents/imageAnalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.M.RData")
 #load("~/Documents/imageAnalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.metaClustered.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/dn4.final.foldRevised.metaClustered.RData")

  ##ZZ-transform.
  hd.z<- zScorePatientExpression(full.dn4=dn4,markers=1:36,ROIID.column=37)
  hd.z2<- zScorePatientExpression(full.dn4=dn4,markers=41:42,ROIID.column=37)
  hd.z$Area.norm<-hd.z2$Area
  hd.z$Eccentricity.norm<-hd.z2$Eccentricity
  hd.z2<-NULL
 ## 
  hd.z$ROIID<-as.character(hd.z$ROIID)
  hd.z$ROIID[which(hd.z$ROIID=="Case22_ROI01_40_a0")]<-"Case22_ROI_01_40_a0"
  hd.z$ROIID<-factor(hd.z$ROIID,levels=unique(hd.z$ROIID)[order(nchar( sapply(strsplit(unique(hd.z$ROIID),"_"),function(x) x[1])))])

 ## color key for ROIs.
  color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")
   

 # load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/hd.z.foldRevised.allCases.Z.RData")

 ## load the DLBCL data.
 # load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure6-CD8-5NN-spatialModel-RICO/tumor.cd8.dlbcl/full.dn4.5nn2.CD8.MC.RData")

 # library(ComplexHeatmap)
 ## show the standardization.
  d<-data.frame(ROI=levels(hd.z$ROIID),col=c(color_clusters,color_clusters[1:14]) )
  
  roiCol<-d[,2] 
  names(roiCol)<-d[,1]
  ### add ROI annotation. 
 rois<-HeatmapAnnotation(ROI=levels(hd.z$ROIID),
	which="row",
	col=list(ROI=roiCol),
	annotation_legend_param=list(ncol=2))


   x41bb<-makeJoyPlot(hd.z,marker="X41BB")
   ccr4<-makeJoyPlot(hd.z,marker="CCR4")
    cd11b<-makeJoyPlot(hd.z,marker="CD11b")
   cd11c<-makeJoyPlot(hd.z,marker="CD11c")
    cd14<-makeJoyPlot(hd.z,marker="CD14")
   cd16<-makeJoyPlot(hd.z,marker="CD16")
   CD206<-makeJoyPlot(hd.z,marker="CD206")
   CD20<-makeJoyPlot(hd.z,marker="CD20")
   cd30<-makeJoyPlot(hd.z,marker="CD30")
   cd34<-makeJoyPlot(hd.z,marker="CD34")
    cd3<-makeJoyPlot(hd.z,marker="CD3")
    cd45ra<-makeJoyPlot(hd.z,marker="CD45RA")
    cd45ro<-makeJoyPlot(hd.z,marker="CD45RO")
    cd4<-makeJoyPlot(hd.z,marker="CD4")
    cd68<-makeJoyPlot(hd.z,marker="CD68")
    cd8<-makeJoyPlot(hd.z,marker="CD8a")
    cxcr3<-makeJoyPlot(hd.z,marker="CXCR3")
    cxcr5<-makeJoyPlot(hd.z,marker="CXCR5")
    caspase<-makeJoyPlot(hd.z,marker="Caspase3")
    dna1<-makeJoyPlot(hd.z,marker="DNA1")
    dna2<-makeJoyPlot(hd.z,marker="DNA2")
    ephrinb2<-makeJoyPlot(hd.z,marker="EphrinB2")
     foxp3<-makeJoyPlot(hd.z,marker="FOXP3")
     gata3<-makeJoyPlot(hd.z,marker="GATA3")
     hlaabc<-makeJoyPlot(hd.z,marker="HLAABC")
     hladr<-makeJoyPlot(hd.z,marker="HLADR")
     icos<-makeJoyPlot(hd.z,marker="ICOS")
     ido<-makeJoyPlot(hd.z,marker="IDO")
    ki67<-makeJoyPlot(hd.z,marker="Ki67")
   lag<-makeJoyPlot(hd.z,marker="LAG3")
      nkg2d<-makeJoyPlot(hd.z,marker="NKG2D")
   pd1<-makeJoyPlot(hd.z,marker="PD1")
   pdl1<-makeJoyPlot(hd.z,marker="PDL1")
    tim3<-makeJoyPlot(hd.z,marker="TIM3")
    tbet<-makeJoyPlot(hd.z,marker="Tbet")
 
  ## mean 0 and SD 1
   hd.z[,c(1:37)]%>%group_by(ROIID)%>%summarise_all(funs(mean))%>%data.frame%>%head
   hd.z[,c(1:37)]%>%group_by(ROIID)%>%summarise_all(funs(sd))%>%data.frame%>%head
   
  ######
  draw(x41bb+ccr4+cd11b+cd11c+cd14+cd16+rois)
   draw(CD206+CD20+cd30+cd34+cd3+cd45ra+rois)
  #savePlot(file="HD.zscores.expression_1.png")
 ###
  draw(cd45ro+cd4+cd68+cd8+cxcr3+cxcr5+rois)
  draw(caspase+dna1+dna2+ephrinb2+foxp3+gata3+rois)
 # savePlot(file="HD.zscores.expression2.png")
  ###
  draw(hlaabc+hladr+icos+ido+ki67+lag+rois)
  draw(nkg2d+pd1+pdl1+tim3+tbet+rois)
  # savePlot(file="HD.zscores.expression3.png")


 ### boxplots.
  library(ggplot2)
    hd.z[,c("CD20","CD206","CD3","CD30","CD34","CD4","CD8a","CD45RA","CD45RO","CD68","FOXP3","ROIID","subType.correction2")]%>%group_by(ROIID,subType.correction2)%>%summarise_all(funs(mean))%>%data.frame%>%reshape2::melt()%>%ggplot(aes(x=subType.correction2,y=value))+geom_boxplot()+facet_wrap(~variable)+theme_bw(base_size=14)+theme(axis.text.x = element_text(angle = 90, hjust = 1))+ylab("Measurement mean")
   # ggsave("Zscore.measurement.gatedPopulations.allHD.png")


 

 hd.lineage<-c( "CD11b",
                       "CD11c",
                       "CD14",
                       "CD15",
                       "CD16",
                       "CD206",
			"CD20",
			"CD30",
			"CD34",
			"CD3",
			"CD45RA",
			"CD45RO",
			"CD4",
			"CD68",
			"CD8a",
                       "FOXP3",
			"EphrinB2",
			"Eccentricity.norm",
			"Area.norm",
			"DNA1",
			"DNA2") 
 hd.induc<-c("X41BB",
		"CCR4",
		"CXCR3",
		"CXCR5",
		"Caspase3",
		"GATA3",
		"HLAABC",
		"HLADR",
		"ICOS",
		"IDO",
		"Ki67",
		"LAG3",
		"NKG2D",
		"PD1",
		"PDL1",
		"TIM3",
		"Tbet")
```

### Meta-cluster HL experiment
We meta-cluster using k1=45 and k2=15 to identify phenotypes in HL. this is slow.
```{r clusterHL,eval=FALSE}  
 
   hd.z<-metaClusterExperiment(dn4=hd.z,
	markers=hd.lineage,
	induc=hd.induc,
	plotPCA=FALSE) 




```

### Hodgkin Lymphoma TME expression
 After identifying 17 meta-clusters of all HL samples, we annotate and create labels for each cluster.
 - reproduces Figure S11
```{r cHLPheno,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 14}

 load("C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/noCoordinates-RData/dn4.global.meta.annotated.RData")
  table(hd.z$phenograph_cluster_meta)

 hd.lineage<-c( "CD11b",
                       "CD11c",
                       "CD14",
                       "CD15",
                       "CD16",
                       "CD206",
			"CD20",
			"CD30",
			"CD34",
			"CD3",
			"CD45RA",
			"CD45RO",
			"CD4",
			"CD68",
			"CD8a",
                       "FOXP3",
			"EphrinB2",
			"Eccentricity.norm",
			"Area.norm",
			"DNA1",
			"DNA2") 
 hd.induc<-c("X41BB",
		"CCR4",
		"CXCR3",
		"CXCR5",
		"Caspase3",
		"GATA3",
		"HLAABC",
		"HLADR",
		"ICOS",
		"IDO",
		"Ki67",
		"LAG3",
		"NKG2D",
		"PD1",
		"PDL1",
		"TIM3",
		"Tbet")
  


 ### meta-clustered data.
 #load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/hd.z.foldRevised.allCases.Z.RData")
    
  plot_clustering_heatmap_wrapper(expr=hd.z[,c(hd.lineage[1:17])], 
  cell_clustering=hd.z[,"phenograph_cluster_meta"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)


 
 anno<-data.frame("1"="DC",
		"2"="Macrophage",
		"3"="CD8",
		"4"="Granulocytic MDSC",
		"5"="CD4",
		"6"="Macrophage",
		"7"="CD4",
		"8"="BCell",
		"9"="CD4",
		"10"="Endothelial",
		"11"="CD4",
		"12"="HRS",
		"13"="TREG",
		"14"="CD4",
		"15"="HRS",
		"16"="CD4",
		"17"="Macrophage")
  anno<-as.data.frame(t(anno))
 anno$cluster<-gsub("X","",rownames(anno))
 hd.z$annotation<-anno[match(hd.z$phenograph_cluster_meta,anno$cluster),"V1"]
 

  ## Cluster 7 was an ambiguous cluster. 
  ## so we filter out the lowly expressed cells from the analysis.
  ### this will increase the phenotype accuracy, but add errors in the spatial analysis.
  ## the alternative is to keep phenotype ambiguity, with accurate spatial results.
  ## either way, there is a trade off.
   id<-which(hd.z[which(hd.z$phenograph_cluster_meta==7),"CD4"]<=(-0.7) |hd.z[which(hd.z$phenograph_cluster_meta==7),"CD3"]<=(-0.7)    )
  ## represents about 8% error in cohort.

  hd.z2<-hd.z
  hd.z2<-hd.z2[-id,]
   ErikMeta=hd.z2[,
	c(hd.lineage,
	"ROIID",
	"annotation")] %>% group_by(ROIID,annotation) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame

   plot_clustering_heatmap_wrapper(expr=hd.z[,c(hd.lineage[1:17])], 
  cell_clustering=hd.z[,"annotation"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 
 # savePlot(file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/steidl.allCases.metaCluster.noFiltering.png")





```

# Sub-phenotype population discovery on TME compartments
- 12-26 identify CD4, CD8, Mac sub-populations used for comparing spatial interactions DLBCL vs. HL
```{r}
 library(igraph)
 hd.z2<-metaClusterPhenotype(dn4=hd.z,phenotype="CD4",myPheno="annotation",markers=c(hd.lineage,hd.induc))
 hd.z2<-metaClusterPhenotype(dn4=hd.z2,phenotype="CD8",myPheno="annotation",markers=c(hd.lineage,hd.induc))
 hd.z2<-metaClusterPhenotype(dn4=hd.z2,phenotype="Macrophage",myPheno="annotation",markers=c(hd.lineage,hd.induc))
 hd.z2<-metaClusterPhenotype(dn4=hd.z2,phenotype="TREG",myPheno="annotation",markers=c(hd.lineage,hd.induc))

 hd.z2$subpopulation_meta<-as.character(hd.z2$annotation)
 
 hd.z2$subpopulation_meta[which(hd.z2$subpopulation_meta=="CD4")]<-paste0("CD4_",hd.z2$CD4_subphenograph_cluster_meta[which(hd.z2$subpopulation_meta=="CD4")])
 hd.z2$subpopulation_meta[which(hd.z2$subpopulation_meta=="CD8")]<-paste0("CD8_",hd.z2$CD8_subphenograph_cluster_meta[which(hd.z2$subpopulation_meta=="CD8")])
 hd.z2$subpopulation_meta[which(hd.z2$subpopulation_meta=="Macrophage")]<-paste0("Macrophage_",hd.z2$Macrophage_subphenograph_cluster_meta[which(hd.z2$subpopulation_meta=="Macrophage")])
 hd.z2$subpopulation_meta[which(hd.z2$subpopulation_meta=="TREG")]<-paste0("TREG_",hd.z2$TREG_subphenograph_cluster_meta[which(hd.z2$subpopulation_meta=="TREG")])
 
 save(hd.z2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/steidl.hl.CD4.CD8.MAC.TREG.subpop.RData")

```


### Manual gating of positive populations in cHL and DLBCL
* Need to move this section after the merging of the datasets
* include the RLN.
```{r,eval=FALSE}
   #load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")

 load("C:/Users/UOSC/Documents/IMC-Ranalysis/steidl/steidl-spatialAnalysis9-11/fold-revised-10-5/noCoordinates-RData/dn4.global.meta.annotated.RData")

 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/monirath/myData/revision-comments-11-6/full.dn4.subType.corrected.mutation.tsne.CD20dim.appended.un.filtered.RData")
 full.dn4<-full.dn4[which(full.dn4$dimCD20.filtered!='YES'),]
 
 library(Hmisc)
 hd.z$TIM3.cut<-cut2(hd.z$TIM3,g=3)
  hd.z$PD1.cut<-cut2(hd.z$PD1,g=3)
   hd.z$CCR4.cut<-cut2(hd.z$CCR4,g=3)
   hd.z$PDL1.cut<-cut2(hd.z$PDL1,g=3)
  hd.z$Ki67.cut<-cut2(hd.z$Ki67,g=3)
  hd.z$ICOS.cut<-cut2(hd.z$ICOS,g=3)

 hl.pdl1<-table(hd.z$annotation,hd.z$PDL1.cut)%>%as.data.frame.matrix
 hl.pdl1<-100* hl.pdl1/rowSums(hl.pdl1)
 hd.z$experiment<-'HL'
 hd.z$PDL1.thresh<-"PDL1-"
 hd.z$PDL1.thresh[which(hd.z$PDL1.cut=="[ 0.277, 6.219]")]<-"PDL1+"
 hd.z$PDL1.thresh[which(hd.z$PDL1.cut=="[-0.485, 0.277)")]<-"PDL1.mid"
### TIM3
  hd.z$TIM3.thresh<-"TIM3-"
  hd.z$TIM3.thresh[which(hd.z$TIM3.cut=="[ 0.214, 6.680]")]<-"TIM3+"
  hd.z$TIM3.thresh[which(hd.z$TIM3.cut=="[-0.505, 0.214)")]<-"TIM3.mid"
## PD1
  hd.z$PD1.thresh<-"PD1-"
  hd.z$PD1.thresh[which(hd.z$PD1.cut=="[ 0.180,11.437]")]<-"PD1+"
  hd.z$PD1.thresh[which(hd.z$PD1.cut=="[-0.512, 0.180)")]<-"PD1.mid"
  
  ##CCR4
   hd.z$CCR4.thresh<-"CCR4-"
  hd.z$CCR4.thresh[which(hd.z$CCR4.cut=="[ 0.245,13.652]")]<-"CCR4+"
  hd.z$CCR4.thresh[which(hd.z$CCR4.cut=="[-0.498, 0.245)")]<-"CCR4.mid"
  ##Ki67
   hd.z$Ki67.thresh<-"Ki67-"
  hd.z$Ki67.thresh[which(hd.z$Ki67.cut=="[-0.256,17.167]")]<-"Ki67+"
  hd.z$Ki67.thresh[which(hd.z$Ki67.cut=="[-0.417,-0.256)")]<-"Ki67.mid"
 ## ICOS
   hd.z$ICOS.thresh<-"ICOS-"
  hd.z$ICOS.thresh[which(hd.z$ICOS.cut=="[ 0.0546,14.7657]")]<-"ICOS+"
  hd.z$ICOS.thresh[which(hd.z$ICOS.cut=="[-0.5072, 0.0546)")]<-"ICOS.mid"


 ## add in the RLN status.
   hd.z$experiment[grepl("RLN",hd.z$diagnosis)]<-'RLN'


 ### dlbcl
    ## DLBCL Z-score.
     zz<- zScorePatientExpression(full.dn4=full.dn4,markers=c(1:31,104:111),ROIID.column=32)
 zz$TIM3.cut<-cut2(zz$Cell_Tim3,g=3)
  zz$PD1.cut<-cut2(zz$Cell_PD1,g=3)
   zz$CCR4.cut<-cut2(zz$Cell_CCR4,g=3)
   zz$PDL1.cut<-cut2(zz$Cell_PDL1,g=3)
  zz$Ki67.cut<-cut2(zz$Cell_Ki67,g=3)
  zz$ICOS.cut<-cut2(zz$Cell_ICOS,g=3)

   zz$experiment<-'DLBCL'

  zz$PDL1.thresh<-"PDL1-"
  zz$PDL1.thresh[which(zz$PDL1.cut=="[ 0.400, 9.438]")]<-"PDL1+"
 ## annotating the thresholds.
  zz$PDL1.thresh<-"PDL1-"
  zz$PDL1.thresh[which(zz$PDL1.cut=="[ 0.400, 9.438]")]<-"PDL1+"
  zz$PDL1.thresh[which(zz$PDL1.cut=="[-0.413, 0.400)")]<-"PDL1.mid"

### TIM3
  zz$TIM3.thresh<-"TIM3-"
  zz$TIM3.thresh[which(zz$TIM3.cut=="[ 0.359, 5.253]")]<-"TIM3+"
  zz$TIM3.thresh[which(zz$TIM3.cut=="[-0.482, 0.359)")]<-"TIM3.mid"
## PD1
  zz$PD1.thresh<-"PD1-"
  zz$PD1.thresh[which(zz$PD1.cut=="[ 0.333,15.748]")]<-"PD1+"
  zz$PD1.thresh[which(zz$PD1.cut=="[-0.484, 0.333)")]<-"PD1.mid"
  
  ##CCR4
  ## 
  zz$CCR4.thresh<-"CCR4-"
  zz$CCR4.thresh[which(zz$CCR4.cut=="[ 0.411, 7.598]")]<-"CCR4+"
  zz$CCR4.thresh[which(zz$CCR4.cut=="[-0.425, 0.411)")]<-"CCR4.mid"
  
  ##Ki67
   zz$Ki67.thresh<-"Ki67-"
  zz$Ki67.thresh[which(zz$Ki67.cut=="[ 0.211, 5.329]")]<-"Ki67+"
  zz$Ki67.thresh[which(zz$Ki67.cut=="[-0.615, 0.211)")]<-"Ki67.mid"
  
  ## icos
  zz$ICOS.thresh<-"ICOS-"
  zz$ICOS.thresh[which(zz$ICOS.cut=="[ 0.207,17.954]")]<-"ICOS+"
  zz$ICOS.thresh[which(zz$ICOS.cut=="[-0.634, 0.207)")]<-"ICOS.mid"

  library(hrbrthemes)


  joint<-data.frame(ROIID=c(as.character(hd.z$ROIID),as.character(zz$ROIID)),
	phenotype=c(as.character(hd.z$annotation),as.character(zz$subType.correction)),
	experiment=c(hd.z$experiment,zz$experiment),
	TIM3=c(as.character(hd.z$TIM3.thresh),as.character(zz$TIM3.thresh)),
	PD1=c(as.character(hd.z$PD1.thresh),as.character(zz$PD1.thresh)),
	CCR4=c(as.character(hd.z$CCR4.thresh),as.character(zz$CCR4.thresh)),
	Ki67=c(as.character(hd.z$Ki67.thresh),as.character(zz$Ki67.thresh)),
	PDL1=c(as.character(hd.z$PDL1.thresh),as.character(zz$PDL1.thresh)),
	ICOS=c(as.character(hd.z$ICOS.thresh),as.character(zz$ICOS.thresh)),
	X=c(hd.z$X_position,zz$X_position),
	Y=c(hd.z$Y_position,zz$Y_position))

all(joint$X==c(hd.z$X_position,zz$X_position))
all(joint$Y==c(hd.z$Y_position,zz$Y_position))

  pdl1.pos<-joint%>%group_by(phenotype,experiment)%>%summarise(PDL1.pos=mean(PDL1=="PDL1+"),SD=sd(PDL1=="PDL1+"))%>%data.frame
  pdl1.pos$SD[which(pdl1.pos$experiment=='DLBCL')]<- pdl1.pos$SD[which(pdl1.pos$experiment=='DLBCL')]/41
  pdl1.pos$SD[which(pdl1.pos$experiment=='HL')]<- pdl1.pos$SD[which(pdl1.pos$experiment=='HL')]/39
  ##add the RLN
    pdl1.pos$SD[which(pdl1.pos$experiment=='RLN')]<- pdl1.pos$SD[which(pdl1.pos$experiment=='RLN')]/10


  pdl1.barGraph<-ggplot(pdl1.pos[pdl1.pos$phenotype%in%c("CD8","CD4","Endothelial","Macrophage","TREG"),],aes(x=phenotype,y=PDL1.pos,fill=experiment))+geom_bar(stat='identity',position='dodge')+geom_errorbar(aes(ymin=PDL1.pos-1.96*SD,ymax=PDL1.pos+1.96*SD),position=position_dodge())+theme_ipsum(base_family="Arial")+scale_fill_manual(values=c('khaki1','plum2','red'))+ylab("PDL1+ by tertile gate (%)")+ggtitle("PDL1+ (%) by tertile gating within experiment")


 pd1.pos<-joint%>%group_by(phenotype,experiment)%>%summarise(PD1.pos=mean(PD1=="PD1+"),SD=sd(PD1=='PD1+'))%>%data.frame
 pd1.pos$SD[which(pd1.pos$experiment=='DLBCL')]<- pd1.pos$SD[which(pd1.pos$experiment=='DLBCL')]/41
  pd1.pos$SD[which(pd1.pos$experiment=='HL')]<- pd1.pos$SD[which(pd1.pos$experiment=='HL')]/39
    pd1.pos$SD[which(pd1.pos$experiment=='RLN')]<- pd1.pos$SD[which(pd1.pos$experiment=='RLN')]/10


  pd1.barGraph<-ggplot(pd1.pos[pd1.pos$phenotype%in%c("CD8","CD4","Endothelial","Macrophage","TREG"),],aes(x=phenotype,y=PD1.pos,fill=experiment))+geom_bar(stat='identity',position='dodge')+geom_errorbar(aes(ymin=PD1.pos-1.96*SD,ymax=PD1.pos+1.96*SD),position=position_dodge())+theme_ipsum(base_family="Arial")+scale_fill_manual(values=c('khaki1','plum2','red'))+ylab("PD1+ by tertile gate (%)")+ggtitle("PD1+ (%) by tertile gating within experiment")




 tim3.pos<-joint%>%group_by(phenotype,experiment)%>%summarise(TIM3.pos=mean(TIM3=="TIM3+"),SD=sd(TIM3=='TIM3+'))%>%data.frame
 tim3.pos$SD[which(tim3.pos$experiment=='DLBCL')]<- tim3.pos$SD[which(tim3.pos$experiment=='DLBCL')]/41
  tim3.pos$SD[which(tim3.pos$experiment=='HL')]<- tim3.pos$SD[which(tim3.pos$experiment=='HL')]/39
    tim3.pos$SD[which(tim3.pos$experiment=='RLN')]<- tim3.pos$SD[which(tim3.pos$experiment=='RLN')]/10


   tim3.barGraph<-ggplot(tim3.pos[tim3.pos$phenotype%in%c("CD8","CD4","Endothelial","Macrophage","TREG"),],aes(x=phenotype,y=TIM3.pos,fill=experiment))+geom_bar(stat='identity',position='dodge')+geom_errorbar(aes(ymin=TIM3.pos-1.96*SD,ymax=TIM3.pos+1.96*SD),position=position_dodge())+theme_ipsum(base_family="Arial")+scale_fill_manual(values=c('khaki1','plum2','red'))+ylab("TIM3+ by tertile gate (%)")+ggtitle("TIM3+ (%) by tertile gating within experiment")

library(gridExtra)
 grid.arrange(pdl1.barGraph,pd1.barGraph,tim3.barGraph,ncol=2,nrow=2)


hd.z$PD1.anno<-hd.z$annotation
 hd.z$PD1.anno[which(hd.z$PD1.anno=="TREG")]<-paste0(hd.z$PD1.thresh[which(hd.z$PD1.anno=="TREG")],hd.z$Ki67.thresh[which(hd.z$PD1.anno=="TREG")],hd.z$annotation[which(hd.z$PD1.anno=="TREG")])
zz$PD1.anno<-zz$subType.correction
  zz$PD1.anno[which(zz$PD1.anno=="TREG")]<-paste0(zz$PD1.thresh[which(zz$PD1.anno=="TREG")],zz$Ki67.thresh[which(zz$PD1.anno=="TREG")],zz$subType.correction[which(zz$PD1.anno=="TREG")])

```
# Distance analysis from PD1+ Treg to T-cells
- 1-3: nearest neighbor distances, and spatial interactions for CCR4+PD1+ observed the null but a secondary comparing  PD1+CCR4+ NND to PD1+.vs.PD1- CD4/CD8 showed that the overall average NND was shorter comparing PD1+CCR4+TREG to PD1+CD8/CD4 compared to PD1- overall.  The spatial differences comparing DLBCL to HL showed that there was a greater difference between PD1+CCR4+ NND to PD1+ vs. PD1-  in DLBCL compared to HL. this was significant in CD4, but trending in CD8. this suggests that there is a spatial relationship that has more spatial effects in DLBCL, trending, but not significant.
- 1-3: tally the total significant interactions between cHL and DLBCL (p<0.01) using cohort relative proportions.  DLBCL has significant effects of attraction between PD1+ Tcells and CCR4+PD1+ Tregs.
- 1-12: contains the section with attraction ANOVA comparison with CCR4+PD1+ vs. CCR4-PD1- in CD8 family.  also contains the codes for the comparison of CCR4+PD1+TREG with PD1+/mid/- CD8 attraction comparison.
```{r}

 zz$CCR4.PD1.anno<-zz$subType.correction
 zz$CCR4.PD1.anno[which(zz$CCR4.PD1.anno=="TREG")]<-paste0(zz$PD1.thresh[which(zz$CCR4.PD1.anno=="TREG")],zz$CCR4.thresh[which(zz$CCR4.PD1.anno=="TREG")],zz$subType.correction[which(zz$CCR4.PD1.anno=="TREG")])
 zz$CCR4.PD1.anno[which(zz$CCR4.PD1.anno=="CD4")]<-paste0(zz$PD1.thresh[which(zz$CCR4.PD1.anno=="CD4")],zz$subType.correction[which(zz$CCR4.PD1.anno=="CD4")])
 zz$CCR4.PD1.anno[which(zz$CCR4.PD1.anno=="CD8")]<-paste0(zz$PD1.thresh[which(zz$CCR4.PD1.anno=="CD8")],zz$subType.correction[which(zz$CCR4.PD1.anno=="CD8")])

 
   
   hd.z$CCR4.PD1.anno<-hd.z$annotation
 hd.z$CCR4.PD1.anno[which(hd.z$CCR4.PD1.anno=="TREG")]<-paste0(hd.z$PD1.thresh[which(hd.z$CCR4.PD1.anno=="TREG")],hd.z$CCR4.thresh[which(hd.z$CCR4.PD1.anno=="TREG")],hd.z$annotation[which(hd.z$CCR4.PD1.anno=="TREG")])
 
 hd.z$CCR4.PD1.anno[which(hd.z$CCR4.PD1.anno=="CD4")]<-paste0(hd.z$PD1.thresh[which(hd.z$CCR4.PD1.anno=="CD4")],hd.z$annotation[which(hd.z$CCR4.PD1.anno=="CD4")])
 hd.z$CCR4.PD1.anno[which(hd.z$CCR4.PD1.anno=="CD8")]<-paste0(hd.z$PD1.thresh[which(hd.z$CCR4.PD1.anno=="CD8")],hd.z$annotation[which(hd.z$CCR4.PD1.anno=="CD8")])

   
   
   
  
  findNND<-function(point=NULL,i=NULL,j=NULL){
    splitd<-split(point)
    nnD<-nncross(splitd[[i]],splitd[[j]])
    nnD$from<-i
    nnD$to<-j
    return(nnD)
  }   

 
############## SPATIAL PROPORTIONS ########################################3
 ############################################################################
 ## nearest distance does not work.  so we can do Vito's interactions using CCR4/PD1  ,  and ICOS/Ki67
  zz$CCR4.PD1.anno<-as.character(zz$unsup.subcommunity)
   zz$CCR4.PD1.anno[grepl("TREG",zz$CCR4.PD1.anno)]<-paste0(zz$PD1.thresh[grepl("TREG",zz$CCR4.PD1.anno)],zz$CCR4.thresh[grepl("TREG",zz$CCR4.PD1.anno)],zz$unsup.subcommunity[grepl("TREG",zz$CCR4.PD1.anno)])

   zz.tme<-zz[which(zz$subType.correction!="Tumor"),]

 dat_cells<-data.frame(ImageNumber=zz.tme$ROIID,
	ObjectNumber=zz.tme$CellId,
	label=as.character(zz.tme$CCR4.PD1.anno),
	group=zz.tme$ROIID)
  dat_cells<-data.table(dat_cells)
  
  dat_relation_15<-createRelationTable(dn5=zz.tme,
	myPheno="CCR4.PD1.anno",
	minDistance=15)
  
  save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1.RData")

  
 ## using the p=0.01 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

  save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_CCR4PD1.01.RData")

 dat_cells<-NULL
 dat_relation_15<-NULL
 zz.tme<-NULL
  
  ### HL
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/steidl.hl.CD4.CD8.MAC.TREG.subpop.RData")
  hd.z2$CCR4.PD1.anno<-hd.z2$subpopulation_meta
   hd.z2$CCR4.PD1.anno[grepl("TREG",hd.z2$CCR4.PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("TREG",hd.z2$CCR4.PD1.anno)],hd.z2$CCR4.thresh[grepl("TREG",hd.z2$CCR4.PD1.anno)],hd.z2$subpopulation_meta[grepl("TREG",hd.z2$CCR4.PD1.anno)])
  
  table(hd.z2$CCR4.PD1.anno)
  toFilt<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0",
           "Placenta_1_3_a0",
           "Placenta_2_4_a0",
           "Tonsil_1_1_a0",
           "Tonsil_2_2_a0")

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[which(!hd.z.tme$ROIID%in%toFilt),]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$CCR4.PD1.anno)
  
 dat_cells_hl<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$CCR4.PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_hl<-data.table(dat_cells_hl)
  
  dat_relation_15_hl<-createRelationTable(dn5=hd.z.tme,
	myPheno="CCR4.PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_hl.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_hl.RData")

  #filter tonsil and RLN.
 
 ## using the p=0.01 threshold.
  p1_hl<-makePermutationNBHD(dat_cells=dat_cells_hl,
	dat_relation=dat_relation_15_hl,
	n_perm=1000,pthreshold=0.01)

  save(p1_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_CCR4PD1.01.RData")

  
  ### RLN analysis.
  toRLN<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0"
          )

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[hd.z.tme$ROIID%in%toRLN,]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$CCR4.PD1.anno)
   hd.z.tme$CCR4.PD1.anno<-factor(hd.z.tme$CCR4.PD1.anno)
 hd.z.tme$ROIID<-factor(hd.z.tme$ROIID)
 
 dat_cells_rln<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$CCR4.PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_rln<-data.table(dat_cells_rln)
  
  dat_relation_15_rln<-createRelationTable(dn5=hd.z.tme,
	myPheno="CCR4.PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_rln.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_rln.RData")

  #filter tonsil and RLN.
 dat_cells_rln$label<-factor(dat_cells_rln$label)
 ## using the p=0.01 threshold.
  p1_rln<-makePermutationNBHD(dat_cells=dat_cells_rln,
	dat_relation=dat_relation_15_rln,
	n_perm=1000,pthreshold=0.01)

  save(p1_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_CCR4PD1.01.RData")

  
  dat_relation_15_hl<-NULL
 dat_relation_15<-NULL
 hd.z.tme<-NULL
 
 ################## gather the data into a linear model
 ## need to check the linear assumptions  
 ## residuals, interactions, and contrasting 
 
 ################################################
  ################################################
  #################FAMILY LEVEL ###############################
  ################################################
  ################################################
  ## examine the CCR4+/mid/- TREG association with CD8/CD4 at the family level.

  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_CCR4PD1.01.RData")
   load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_CCR4PD1.01.RData")
   load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_CCR4PD1.01.RData")

 
 p1$pmat[grepl("PD1\\+CCR4\\+",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]
 sum1<-100*p1$pmat[grepl("PD1\\+CCR4\\+",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)
 
 p1_hl$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]
  sum2<-100*p1_hl$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)
  
  p1_rln$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]
  sum3<-100*p1_rln$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)
  
  
 cd4.sum<-data.frame(CD4=c(sum1[grepl("CD4",names(sum1))],
                           sum2[grepl("CD4",names(sum2))],
                           sum3[grepl("CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("CD4",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="CD4"
                    )  
# cd8
 cd8.sum<-data.frame(CD8=c(sum1[grepl("CD8",names(sum1))],
                           sum2[grepl("CD8",names(sum2))],
                           sum3[grepl("CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("CD8",names(sum2))])),
                     rep("RLN",length(sum3[grepl("CD8",names(sum3))]))),
                      pheno="PD1+CCR4+",
                     interaction="CD8"
                     ) 
 lm(CD4~cohort,cd4.sum)%>%summary
 lm(CD8~cohort,cd8.sum)%>%summary
 
 
#####  Contrast 1
  p1$pmat[grepl("PD1\\-CCR4\\-",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]
 cont.sum1<-100*p1$pmat[grepl("PD1\\-CCR4\\-",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)
 
 p1_hl$pmat[grepl("PD1\\-CCR4\\-",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]
  cont.sum2<-100*p1_hl$pmat[grepl("PD1\\-CCR4\\-",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)
  
  p1_rln$pmat[grepl("PD1\\-CCR4\\-",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]
  cont.sum3<-100*p1_rln$pmat[grepl("PD1\\-CCR4\\-",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)

 cd4.sum2<-data.frame(CD4=c(cont.sum1[grepl("CD4",names(cont.sum1))],
                            cont.sum2[grepl("CD4",names(cont.sum2))],
                            cont.sum3[grepl("CD4",names(cont.sum3))]),
                      cohort=c(rep("DLBCL",length((cont.sum1[grepl("CD4",names(cont.sum1))]))),
                               rep("HL",length(cont.sum2[grepl("CD4",names(cont.sum2))])),
                               rep("RLN",length(cont.sum3[grepl("CD4",names(cont.sum3))]))),
                       pheno="PD1-CCR4-",
                     interaction="CD4"
                      )  
  
# cd8
 cd8.sum2<-data.frame(CD8=c(cont.sum1[grepl("CD8",names(cont.sum1))],
                            cont.sum2[grepl("CD8",names(cont.sum2))],
                            cont.sum3[grepl("CD8",names(cont.sum3))]
                            ),
                      cohort=c(rep("DLBCL",length((cont.sum1[grepl("CD8",names(cont.sum1))]))),
                               rep("HL",length(cont.sum2[grepl("CD8",names(cont.sum2))])),
                               rep("RLN",length(cont.sum3[grepl("CD8",names(cont.sum3))]))),
                       pheno="PD1-CCR4-",
                     interaction="CD8"
                     ) 
 
 lm(CD4~cohort,cd4.sum2)%>%summary
 lm(CD8~cohort,cd8.sum2)%>%summary
 
 # Contrast 2
  p1$pmat[grepl("PD1\\-CCR4\\+",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]
 cont.sum4<-100*p1$pmat[grepl("PD1\\-CCR4\\+",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)
 
 p1_hl$pmat[grepl("PD1\\-CCR4\\+",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]
  cont.sum5<-100*p1_hl$pmat[grepl("PD1\\-CCR4\\+",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)
  
  p1_rln$pmat[grepl("PD1\\-CCR4\\+",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]
  cont.sum6<-100*p1_rln$pmat[grepl("PD1\\-CCR4\\+",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)

 cd4.sum3<-data.frame(CD4=c(cont.sum4[grepl("CD4",names(cont.sum4))],
                            cont.sum5[grepl("CD4",names(cont.sum5))],
                            cont.sum6[grepl("CD4",names(cont.sum6))]),
                      cohort=c(rep("DLBCL",length((cont.sum4[grepl("CD4",names(cont.sum4))]))),
                               rep("HL",length(cont.sum5[grepl("CD4",names(cont.sum5))])),
                               rep("RLN",length(cont.sum6[grepl("CD4",names(cont.sum6))]))),
                       pheno="PD1-CCR4+",
                     interaction="CD4"
                      )  
  
# cd8
 cd8.sum3<-data.frame(CD8=c(cont.sum4[grepl("CD8",names(cont.sum4))],
                            cont.sum5[grepl("CD8",names(cont.sum5))],
                            cont.sum6[grepl("CD8",names(cont.sum6))]
                            ),
                      cohort=c(rep("DLBCL",length((cont.sum4[grepl("CD8",names(cont.sum4))]))),
                               rep("HL",length(cont.sum5[grepl("CD8",names(cont.sum5))])),
                               rep("RLN",length(cont.sum6[grepl("CD8",names(cont.sum6))]))),
                       pheno="PD1-CCR4+",
                     interaction="CD8"
                       ) 
 
 lm(CD4~cohort,cd4.sum3)%>%summary
 lm(CD8~cohort,cd8.sum3)%>%summary
 
 
 ## we test for interactions checking to see if stratified analysis is justified
############### MODELING ############################
##########################
 ########################
######################### 
 
 ## no interactions between phenotypes across cohorts.
  cd4.global<-rbind(cd4.sum,cd4.sum2,cd4.sum3)
  aov(CD4~cohort*pheno,cd4.global)%>%summary
  
  
   cd4.global$id<-rownames(cd4.global)
   cd4.global$id<-factor(cd4.global$id)
  
  
  ## marginal interaction p=0.09, but not significant
  cd8.global<-rbind(cd8.sum,cd8.sum2,cd8.sum3)
  ## fixing id
  cd8.global$id<-rownames(cd8.global)
  cd8.global$id<-factor(cd8.global$id)
    
  aov(CD8~cohort*pheno,cd8.global)%>%summary
  
  ## we will not stratify but form a single LM with contrasting.
  ## checking residuals
  colnames(cd8.global)[1]<-"Proportion"
  colnames(cd4.global)[1]<-"Proportion"
  tcells<-rbind(cd8.global,cd4.global)
  
  anova2<-lm(Proportion~cohort*pheno,cd8.global)
  qqnorm(anova2$residuals)
  qqline(anova2$residuals, datax=FALSE, distribution=qnorm)
  plot(anova2$residuals~anova2$fitted.values)
  plot(density(anova2$residuals))
 
   #library(multcomp);library(lme4)
#	spatial.global <-  lmer(Proportion~0+pheno+cohort+(1|interaction),tcells)
#	  alpha_noint <- rbind(
	#HL.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_1",36,-1)/37,
	#RLN.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_2",36,-1)/37,
  library(lsmeans);library(afex)
  #fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd8.global,return="aov")
  #ref1<-lsmeans(fit,c("pheno","cohort"))  
  
  ## dropping PD1-CCR4+
 fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd8.global[which(cd8.global$pheno!="PD1-CCR4+"),],return="aov")
  ref1<-lsmeans(fit,c("pheno","cohort")) 

  #c_list <- list(PD1CCR4.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
   #            PD1nCCR4n.DLBCL.vOther = c(1,0, 0, -1/2, 0, 0,-1/2,0,0),
    #            PD1nCCR4P.DLBCL.vOther = c(0,1, 0, 0, -1/2,0,0,-1/2,0),
     #          
      #          PD1CCR4.DLBCL=c(0,0, 1, 0, 0, 0,0,0,0),
       #        PD1CCR4.HL=c(0,0,0, 0,0,1,0,0,0),
        #       PD1CCR4.RLN=c(0,0,0,0,0,0,0,0,1),
         #      PD1nCCR4.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
          #     PD1nCCR4.HL=c(0,0,0, 0,1,0, 0,0,0),
        #       PD1nCCR4.RLN=c(0,0,0, 0,0,0, 0,1,0),
         #      PD1nCCR4n.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
          #     PD1nCCR4n.HL=c(0,0,0, 1,0,0, 0,0,0),
           #    PD1nCCR4n.RLN=c(0,0,0, 0,0,0, 1,0,0))
               
  ## dropping PD1-CCR4+
  c_list <- list(PD1CCR4.DLBCL.vOther = c(0,1, 0,-1/2, 0,-1/2),
               PD1nCCR4n.DLBCL.vOther = c(1,0, -1/2,0, -1/2,0),
            #    PD1nCCR4P.DLBCL.vOther = c(0,1, 0, 0, -1/2,0,0,-1/2,0),
               
                PD1CCR4.DLBCL=c(0,1, 0,0, 0,0),
               PD1CCR4.HL=c(0,0, 0,1, 0,0),
               PD1CCR4.RLN=c(0,0, 0,0, 0,1),
              
               PD1nCCR4n.DLBCL=c(1,0, 0,0, 0,0),
               PD1nCCR4n.HL=c(0,0, 1,0, 0,0),
               PD1nCCR4n.RLN=c(0,0, 0,0, 1,0))
  

  summary(contrast(ref1, c_list), adjust = "BH")
 cd8.rez.data<- confint(contrast(ref1, c_list), adjust = "BH")%>%data.frame
  cd8.rez.data$p.value<-summary(contrast(ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(ref1, c_list)), test = adjusted("BH"))

  
cd8.rez.data$cohort<-c(rep("contrast",2),rep(c("DLBCL","HL","RLN"),2))  
# cd8.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CCR4+",3),rep("PD1-CCR4+",3),rep("PD1-CCR4-",3))  
cd8.rez.data$phenotype<-c(rep("contrast",2),rep("PD1+CCR4+",3),rep("PD1-CCR4-",3))  

  
  ## CD4 anova
  

 # cd4fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd4.global,return="aov")
#  cd4ref1<-lsmeans(cd4fit,c("pheno","cohort"))  
  
  ## dropping PD1-CCR4+
  cd4fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd4.global[which(cd4.global$pheno!="PD1-CCR4+"),],return="aov")
  cd4ref1<-lsmeans(cd4fit,c("pheno","cohort")) 
  
  ## for CD4 we compare DLBCL to HL
   c_list2 <- list(PD1CCR4.DLBCL.vOther = c(0,1, 0,-1, 0,0),
               PD1nCCR4n.DLBCL.vOther = c(1,0, -1/2,0, -1/2,0),
            #    PD1nCCR4P.DLBCL.vOther = c(0,1, 0, 0, -1/2,0,0,-1/2,0),
               
                PD1CCR4.DLBCL=c(0,1, 0,0, 0,0),
               PD1CCR4.HL=c(0,0, 0,1, 0,0),
               PD1CCR4.RLN=c(0,0, 0,0, 0,1),
              
               PD1nCCR4n.DLBCL=c(1,0, 0,0, 0,0),
               PD1nCCR4n.HL=c(0,0, 1,0, 0,0),
               PD1nCCR4n.RLN=c(0,0, 0,0, 1,0))
   
   
  ### dropping a test PD1-CCR4+ to reduce DF
   c_list3 <-  list(PD1CCR4.DLBCL.vOther = c(0,1, 0,-1, 0,0),
               PD1nCCR4n.DLBCL.vOther = c(1,0, -1/2,0, -1/2,0))
               
              
   
  summary(contrast(cd4ref1, c_list2), adjust = "BH")
 cd4.rez.data<- confint(contrast(cd4ref1, c_list2), adjust = "BH")%>%data.frame
  cd4.rez.data$p.value<-summary(contrast(cd4ref1, c_list2), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd4ref1, c_list2)), test = adjusted("BH"))

  
cd4.rez.data$cohort<-c(rep("contrast",2),rep(c("DLBCL","HL","RLN"),2))  
  #cd4.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CCR4+",3),rep("PD1-CCR4+",3),rep("PD1-CCR4-",3)) 
cd4.rez.data$phenotype<-c(rep("contrast",2),rep("PD1+CCR4+",3),rep("PD1-CCR4-",3))  

################################
  
  
###
  ## we use this data just to check the model estimates in the anova.
 ## figure
 cd4.fit<- lm(CD4~0+cohort,cd4.sum)
 cd8.fit<-lm(CD8~0+cohort,cd8.sum)
 rez<-data.frame(Estimate=c(coef(cd8.fit)),
                            lower=confint(cd8.fit)[,1],
                            upper=confint(cd8.fit)[,2],
                 cohort=c("DLBCL","HL","RLN"),
                 phenotype="CCR4+PD1+TREG",
                 interaction="CD8")
 
 ## normal checking
  lm(CD8~cohort,cd8.sum)%>%residuals()%>%density%>%plot
  qqnorm(residuals(lm(CD8~cohort,cd8.sum)))
  qqline(residuals(lm(CD8~cohort,cd8.sum)))
 
 ## contrast 1
 cd4.fit<- lm(CD4~0+cohort,cd4.sum2)
 cd8.fit<-lm(CD8~0+cohort,cd8.sum2)
 rez2<-data.frame(Estimate=c(coef(cd8.fit)),
                            lower=confint(cd8.fit)[,1],
                            upper=confint(cd8.fit)[,2],
                 cohort=c("DLBCL","HL","RLN"),
                 phenotype="CCR4-PD1-TREG",
                 interaction="CD8"
                 )
 
 
 ## contrast 2
 cd4.fit<- lm(CD4~0+cohort,cd4.sum3)
 cd8.fit<-lm(CD8~0+cohort,cd8.sum3)
 rez3<-data.frame(Estimate=c(coef(cd8.fit)),
                            lower=confint(cd8.fit)[,1],
                            upper=confint(cd8.fit)[,2],
                 cohort=c("DLBCL","HL","RLN"),
                 phenotype="CCR4+PD1-TREG",
                 interaction="CD8")
 ## CD4
  cd4.fit<- lm(CD4~0+cohort,cd4.sum)
 cd8.fit<-lm(CD8~0+cohort,cd8.sum)
 rez4<-data.frame(Estimate=c(coef(cd4.fit)),
                            lower=confint(cd4.fit)[,1],
                            upper=confint(cd4.fit)[,2],
                 cohort=c("DLBCL","HL","RLN"),
                 phenotype="CCR4+PD1+TREG",
                 interaction="CD4")
 
 
 ## contrast 1
 cd4.fit<- lm(CD4~0+cohort,cd4.sum2)
 cd8.fit<-lm(CD8~0+cohort,cd8.sum2)
 rez5<-data.frame(Estimate=c(coef(cd4.fit)),
                            lower=confint(cd4.fit)[,1],
                            upper=confint(cd4.fit)[,2],
                 cohort=c("DLBCL","HL","RLN"),
                 phenotype="CCR4-PD1-TREG",
                 interaction="CD4"
                 )
 
 
 ## contrast 2
 cd4.fit<- lm(CD4~0+cohort,cd4.sum3)
 cd8.fit<-lm(CD8~0+cohort,cd8.sum3)
 rez6<-data.frame(Estimate=c(coef(cd4.fit)),
                            lower=confint(cd4.fit)[,1],
                            upper=confint(cd4.fit)[,2],
                 cohort=c("DLBCL","HL","RLN"),
                 phenotype="CCR4+PD1-TREG",
                 interaction="CD4")
 
 d<-rbind(rez,rez2,rez3,rez4,rez5,rez6)
  d$phenotype<-factor(d$phenotype,levels=c("CCR4+PD1+TREG","CCR4+PD1-TREG","CCR4-PD1-TREG"))
 
 #ggplot(d,aes(x=cohort,y=Estimate,color=interaction))+geom_point(position=position_dodge(width=0.5))+geom_errorbar(aes(ymin=lower,ymax=upper),position=position_dodge(width=0.5))+facet_wrap(~phenotype)
 #######################################
 
################ 
 # figure making of the ANOVA for CD8
   cd8.d<-cd8.rez.data[which(cd8.rez.data$cohort!="contrast"),]
   cd8.d$phenotype<-factor(cd8.d$phenotype,levels=c("PD1+CCR4+","PD1-CCR4+","PD1-CCR4-"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
  
  cohortColors<-threeColors
  names(cohortColors)[2]<-"HL"
  
  
  #label.df<-data.frame(phenotype=c("CCR4+PD1+TREG","CCR4+PD1-TREG","CCR4-PD1-TREG"),
   #                    Estimate=c(4.15,4.25,1.25),label=c("**","",""),cohort=NA)
  
  #label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
   #                    estimate = c(5.15,5.45,5.88, 4.25,4.55,4.85, 1.25,1.55,1.85),
    #                   cohort=NA,
     #                  label=c("***",rep("",8))
      #                )
  ## dropping PD1-CCR4+ comparison
   label.df <- data.frame(phenotype = c(1,1,1, 2,2,2),
                     estimate = c(4.9,5,5.1, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("***",rep("",5))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p2<- ggplot(cd8.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=4,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+4.1),lty=2)+ylab("Spatial interaction proportion with CD8 family (p<0.01)")+xlab("TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y = element_text(size=11))+ggtitle("CD8")+geom_hline(yintercept = 0,linetype='dashed',color='red')
  
  
## CD4
 
  cd4.d<-cd4.rez.data[which(cd4.rez.data$cohort!="contrast"),]
   cd4.d$phenotype<-factor(cd4.d$phenotype,levels=c("PD1+CCR4+","PD1-CCR4+","PD1-CCR4-"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
  
  cohortColors<-threeColors
  names(cohortColors)[2]<-"HL"
  
  

  #label.df<-data.frame(phenotype=c("CCR4+PD1+TREG","CCR4+PD1-TREG","CCR4-PD1-TREG"),
   #                    Estimate=c(4.15,4.25,1.25),label=c("**","",""),cohort=NA)
  
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(5.5,5.8,6.2, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p3<- ggplot(cd4.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=4,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      #geom_line(data=arc.df,aes(phenotype+1,Estimate+4.1),lty=2)+
   ylab("Spatial interaction proportion with CD4 family (p<0.01)")+xlab("TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y = element_text(size=11))+ggtitle("CD4")
  
 grid.arrange(p2,p3,nrow=1)
  save(p2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/spatial-comparison/p2_CD8.RData")
  save(p3,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/spatial-comparison/p3_CD4.RData")
 
######################################################
 ####################################################
######################################################
 ####################################################
######################################################
 ####################################################
######################################################
 ####################################################
######################################################
 ####################################################
 
 
### compare the contrast enriched in HL spatially.

  
  zz$CCR4.PD1.anno[grepl("CD4",zz$CCR4.PD1.anno)]<-paste0(zz$PD1.thresh[grepl("CD4",zz$CCR4.PD1.anno)],zz$unsup.subcommunity[grepl("CD4",zz$CCR4.PD1.anno)])
    zz$CCR4.PD1.anno[grepl("CD8",zz$CCR4.PD1.anno)]<-paste0(zz$PD1.thresh[grepl("CD8",zz$CCR4.PD1.anno)],zz$unsup.subcommunity[grepl("CD8",zz$CCR4.PD1.anno)])
 zz$CCR4.PD1.anno[grepl("Macrophage",zz$CCR4.PD1.anno)]<-paste0(zz$PDL1.thresh[grepl("Macrophage",zz$CCR4.PD1.anno)],zz$unsup.subcommunity[grepl("Macrophage",zz$CCR4.PD1.anno)])
 
 table(zz$CCR4.PD1.anno)
 
 zz.tme<-zz[which(zz$subType.correction!="Tumor"),]

 dat_cells<-data.frame(ImageNumber=zz.tme$ROIID,
	ObjectNumber=zz.tme$CellId,
	label=as.character(zz.tme$CCR4.PD1.anno),
	group=zz.tme$ROIID)
  dat_cells<-data.table(dat_cells)
  
  dat_relation_15<-createRelationTable(dn5=zz.tme,
	myPheno="CCR4.PD1.anno",
	minDistance=15)
  
  save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_PD1Tcells.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_PD1Tcells.RData")

  
 ## using the p=0.01 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

  save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_CCR4PD1_PD1Tcells.01.RData")

 dat_cells<-NULL
 dat_relation_15<-NULL
 zz.tme<-NULL
  
  ### HL
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/steidl.hl.CD4.CD8.MAC.TREG.subpop.RData")
 # hd.z2$ICOS.Ki67.anno<-hd.z2$subpopulation_meta
   #hd.z2$ICOS.Ki67.anno[grepl("TREG",hd.z2$ICOS.Ki67.anno)]<-paste0(hd.z2$ICOS.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],                                                     #       hd.z2$Ki67.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],
   #     hd.z2$PD1.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],
  #     hd.z2$subpopulation_meta[grepl("TREG",hd.z2$ICOS.Ki67.anno)])
  
   hd.z2$CCR4.PD1.anno<-hd.z2$subpopulation_meta
   hd.z2$CCR4.PD1.anno[grepl("TREG",hd.z2$CCR4.PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("TREG",hd.z2$CCR4.PD1.anno)],hd.z2$CCR4.thresh[grepl("TREG",hd.z2$CCR4.PD1.anno)],hd.z2$subpopulation_meta[grepl("TREG",hd.z2$CCR4.PD1.anno)])
 hd.z2$CCR4.PD1.anno[grepl("CD4",hd.z2$CCR4.PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("CD4",hd.z2$CCR4.PD1.anno)],hd.z2$subpopulation_meta[grepl("CD4",hd.z2$CCR4.PD1.anno)])
hd.z2$CCR4.PD1.anno[grepl("CD8",hd.z2$CCR4.PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("CD8",hd.z2$CCR4.PD1.anno)],hd.z2$subpopulation_meta[grepl("CD8",hd.z2$CCR4.PD1.anno)])
  hd.z2$CCR4.PD1.anno[grepl("Macrophage",hd.z2$CCR4.PD1.anno)]<-paste0(hd.z2$PDL1.thresh[grepl("Macrophage",hd.z2$CCR4.PD1.anno)],hd.z2$subpopulation_meta[grepl("Macrophage",hd.z2$CCR4.PD1.anno)])

   
  
  
  table(hd.z2$CCR4.PD1.anno)
  toFilt<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0",
           "Placenta_1_3_a0",
           "Placenta_2_4_a0",
           "Tonsil_1_1_a0",
           "Tonsil_2_2_a0")

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[which(!hd.z.tme$ROIID%in%toFilt),]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$CCR4.PD1.anno)
  
 dat_cells_hl<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$CCR4.PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_hl<-data.table(dat_cells_hl)
  
  dat_relation_15_hl<-createRelationTable(dn5=hd.z.tme,
	myPheno="CCR4.PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_PD1Tcells_hl.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_PD1Tcells_hl.RData")

  #filter tonsil and RLN.
 
 ## using the p=0.01 threshold.
  p1_hl<-makePermutationNBHD(dat_cells=dat_cells_hl,
	dat_relation=dat_relation_15_hl,
	n_perm=1000,pthreshold=0.01)

  save(p1_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_CCR4PD1_PD1Tcells.01.RData")

  
  ### RLN analysis.
  toRLN<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0"
          )

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[hd.z.tme$ROIID%in%toRLN,]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$CCR4.PD1.anno)
   hd.z.tme$CCR4.PD1.anno<-factor(hd.z.tme$CCR4.PD1.anno)
 hd.z.tme$ROIID<-factor(hd.z.tme$ROIID)
 
 dat_cells_rln<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$CCR4.PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_rln<-data.table(dat_cells_rln)
  
  dat_relation_15_rln<-createRelationTable(dn5=hd.z.tme,
	myPheno="CCR4.PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_PD1Tcells_rln.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_CCR4PD1_PD1Tcells_rln.RData")

  #filter tonsil and RLN.
 dat_cells_rln$label<-factor(dat_cells_rln$label)
 ## using the p=0.01 threshold.
  p1_rln<-makePermutationNBHD(dat_cells=dat_cells_rln,
	dat_relation=dat_relation_15_rln,
	n_perm=1000,pthreshold=0.01)

  save(p1_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_CCR4PD1_PD1Tcells.01.RData")

  
  dat_relation_15_hl<-NULL
 dat_relation_15<-NULL
 hd.z.tme<-NULL
 
 
# load the permutation results and model the hypothesis. 
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_CCR4PD1_PD1Tcells.01.RData")
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_CCR4PD1_PD1Tcells.01.RData")
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_CCR4PD1_PD1Tcells.01.RData")

 
 
########################################################################################
 ###########PD1+CCR4+TREG interaction with PD1+/PD1 mid/PD1-  Tcells#################################
########################################################################################
########################################################################################
########################################################################################

 
 p1$pmat[grepl("PD1\\+CCR4\\+",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]
 sum1<-100*p1$pmat[grepl("PD1\\+CCR4\\+",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)

 
 p1_hl$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]
  sum2<-100*p1_hl$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)


  p1_rln$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]
  sum3<-100*p1_rln$pmat[grepl("PD1\\+CCR4\\+",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)


  
 cd4.sum<-data.frame(CD4=c(sum1[grepl("PD1\\-CD4",names(sum1))],
                           sum2[grepl("PD1\\-CD4",names(sum2))],
                           sum3[grepl("PD1\\-CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD4",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PD1-CD4"
                    )  
 cd4.sum2<-data.frame(CD4=c(sum1[grepl("PD1\\+CD4",names(sum1))],
                           sum2[grepl("PD1\\+CD4",names(sum2))],
                           sum3[grepl("PD1\\+CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD4",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PD1+CD4"
                    )
 
 cd4.sum3<-data.frame(CD4=c(sum1[grepl("PD1.midCD4",names(sum1))],
                           sum2[grepl("PD1.midCD4",names(sum2))],
                           sum3[grepl("PD1.midCD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD4",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PD1.midCD4"
                    )
 cd4.sum.final<-rbind(cd4.sum,cd4.sum2,cd4.sum3)
# cd8
 cd8.sum<-data.frame(CD8=c(sum1[grepl("PD1\\-CD8",names(sum1))],
                           sum2[grepl("PD1\\-CD8",names(sum2))],
                           sum3[grepl("PD1\\-CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD8",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PD1-CD8"
                    )  
 cd8.sum2<-data.frame(CD8=c(sum1[grepl("PD1\\+CD8",names(sum1))],
                           sum2[grepl("PD1\\+CD8",names(sum2))],
                           sum3[grepl("PD1\\+CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD8",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PD1+CD8"
                    )
 
 cd8.sum3<-data.frame(CD8=c(sum1[grepl("PD1.midCD8",names(sum1))],
                           sum2[grepl("PD1.midCD8",names(sum2))],
                           sum3[grepl("PD1.midCD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD8",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PD1.midCD8"
                    )
 cd8.sum.final<-rbind(cd8.sum,cd8.sum2,cd8.sum3)
lm(CD4~cohort*interaction,cd4.sum.final)%>%summary
lm(CD8~cohort*interaction,cd8.sum.final)%>%summary
 
 ## PDL1+ MAC

 m.sum<-data.frame(MAC=c(sum1[grepl("PDL1\\-Macrophage",names(sum1))],
                           sum2[grepl("PDL1\\-Macrophage",names(sum2))],
                           sum3[grepl("PDL1\\-Macrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1\\-Macrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1\\-Macrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1\\-Macrophage",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PDL1-Macrophage"
                    )  
 m.sum2<-data.frame(MAC=c(sum1[grepl("PDL1\\+Macrophage",names(sum1))],
                           sum2[grepl("PDL1\\+Macrophage",names(sum2))],
                           sum3[grepl("PDL1\\+Macrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1\\+Macrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1\\+Macrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1\\+Macrophage",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PDL1+Macrophage"
                    )
 
 m.sum3<-data.frame(MAC=c(sum1[grepl("PDL1.midMacrophage",names(sum1))],
                           sum2[grepl("PDL1.midMacrophage",names(sum2))],
                           sum3[grepl("PDL1.midMacrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1.midMacrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1.midMacrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1.midMacrophage",names(sum3))]))),
                     pheno="PD1+CCR4+",
                     interaction="PDL1.midMacrophage"
                    )
 m.sum.final<-rbind(m.sum,m.sum2,m.sum3)


 
 
 ## we test for interactions checking to see if stratified analysis is justified
############### MODELING ############################
##########################
 ########################
######################### 
 
 ## no interactions between phenotypes across cohorts.
  cd4.global<-cd4.sum.final
  aov(CD4~cohort*interaction,cd4.global)%>%summary
  
  
   cd4.global$id<-rownames(cd4.global)
   cd4.global$id<-factor(cd4.global$id)
  
  
  ## marginal interaction p=0.09, but not significant
  cd8.global<-cd8.sum.final
  ## fixing id
  cd8.global$id<-rownames(cd8.global)
  cd8.global$id<-factor(cd8.global$id)
    
  aov(CD8~cohort*interaction,cd8.global)%>%summary
  
  
  ## macrphage
  m.global<-m.sum.final
  ## fixing id
  m.global$id<-rownames(m.global)
  m.global$id<-factor(m.global$id)
    
  aov(MAC~cohort*interaction,m.global)%>%summary
  
  
  
  ## we will not stratify but form a single LM with contrasting.
  ## checking residuals
  colnames(cd8.global)[1]<-"Proportion"
  colnames(cd4.global)[1]<-"Proportion"
    colnames(m.global)[1]<-"Proportion"

    ## trend test
    tester<-cd8.global[which(cd8.global$cohort=="DLBCL"),]
    tester$class<-0
    tester$class[which(tester$interaction=="PD1.midCD8")]<-1
    tester$class[which(tester$interaction=="PD1+CD8")]<-2
####    
    
    
 # tcells<-rbind(cd8.global,cd4.global)
  ### checking fit residuals.
  anova2<-lm(Proportion~cohort*pheno,cd8.global)
  qqnorm(anova2$residuals)
  qqline(anova2$residuals, datax=FALSE, distribution=qnorm)
  plot(anova2$residuals~anova2$fitted.values)
  plot(density(anova2$residuals))
 
   #library(multcomp);library(lme4)
#	spatial.global <-  lmer(Proportion~0+pheno+cohort+(1|interaction),tcells)
#	  alpha_noint <- rbind(
	#HL.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_1",36,-1)/37,
	#RLN.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_2",36,-1)/37,
  library(lsmeans);library(afex)
  fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd8.global,return="aov")
  ref1<-lsmeans(fit,c("interaction","cohort"))  
  
  ## dropping PD1-CCR4+
 # fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd8.global[which(cd8.global$pheno!="PD1-CCR4+"),],return="aov")
  #ref1<-lsmeans(fit,c("pheno","cohort")) 

  c_list <- list(PD1CD8.DLBCL.vOther = c(0,0,1, 0,0,-1/2,0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0,0, -1/2,0,0, -1/2,0,0),
                PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1/2,0, 0,-1/2,0),
               
                PD1CD8.DLBCL=c(0,0,1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0, 0,0,0, 0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
               
 
  

  summary(contrast(ref1, c_list), adjust = "BH")
 cd8.rez.data<- confint(contrast(ref1, c_list), adjust = "BH")%>%data.frame
  cd8.rez.data$p.value<-summary(contrast(ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(ref1, c_list)), test = adjusted("BH"))

  
cd8.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
cd8.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CD8+",3),rep("PD1midCD8+",3),rep("PD1-CD8+",3))  

  
  ## CD4 anova
  cd4fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd4.global,return="aov")
  cd4ref1<-lsmeans(cd4fit,c("interaction","cohort"))  
  

  c_list2 <- list(PD1CD8.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0, 0, -1,0,0, -1,0,0),
                PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1,0, 0,-1,0),
               
                PD1CD8.DLBCL=c(0,0, 1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0,0,0,0,0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
   
 
   
  summary(contrast(cd4ref1, c_list2), adjust = "BH")
  cd4.rez.data<- confint(contrast(cd4ref1, c_list2), adjust = "BH")%>%data.frame
  cd4.rez.data$p.value<-summary(contrast(cd4ref1, c_list2), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd4ref1, c_list2)), test = adjusted("BH"))

  
cd4.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
cd4.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CD4+",3),rep("PD1midCD4+",3),rep("PD1-CD4+",3))  

## macrophage
 mfit<-aov.car(Proportion~interaction*cohort+Error(id),data=m.global,return="aov")
 mref1<-lsmeans(mfit,c("interaction","cohort"))  
  

  c_list3 <- list(PDL1Mac.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
               PDL1nMac.DLBCL.vOther = c(1,0, 0, -1,0,0, -1,0,0),
                PDL1midMac.DLBCL.vOther = c(0,1,0, 0,-1,0, 0,-1,0),
               
                PDL1Mac.DLBCL=c(0,0, 1, 0,0,0, 0,0,0),
               PDL1Mac.HL=c(0,0,0, 0,0,1, 0,0,0),
               PDL1Mac.RLN=c(0,0,0,0,0,0,0,0,1),
               PDL1midMac.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PDL1midMac.HL=c(0,0,0, 0,1,0, 0,0,0),
               PDL1midMac.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PDL1nMac.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PDL1nMac.HL=c(0,0,0, 1,0,0, 0,0,0),
               PDL1nMac.RLN=c(0,0,0, 0,0,0, 1,0,0))
   
 
   
  summary(contrast(mref1, c_list3), adjust = "BH")
 m.rez.data<- confint(contrast(mref1, c_list3), adjust = "BH")%>%data.frame
  m.rez.data$p.value<-summary(contrast(mref1, c_list3), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(mref1, c_list3)), test = adjusted("BH"))

  
m.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
m.rez.data$phenotype<-c(rep("contrast",3),rep("PDL1+MAC+",3),rep("PDL1midMAC+",3),rep("PDL1-MAC+",3))  

################################
  
  
### PD1-CD4
 lm(CD4~0+cohort,cd4.sum)%>%summary
# PD1+CD4
 lm(CD4~0+cohort,cd4.sum2)%>%summary
 #PD1 mid CD4
  lm(CD4~0+cohort,cd4.sum3)%>%summary
  cd4.rez.data
  
  lm(CD8~0+cohort,cd8.sum)%>%summary
# PD1+CD4
 lm(CD8~0+cohort,cd8.sum2)%>%summary
 #PD1 mid CD4
  lm(CD8~0+cohort,cd8.sum3)%>%summary
  cd8.rez.data
 #######################################
 
################ 
 # figure making of the ANOVA for CD8
   cd8.d<-cd8.rez.data[which(cd8.rez.data$cohort!="contrast"),]
   cd8.d$phenotype<-factor(cd8.d$phenotype,levels=c("PD1+CD8+","PD1midCD8+","PD1-CD8+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
  
  cohortColors<-threeColors
  names(cohortColors)[2]<-"HL"
  

  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                      estimate = c(3.5,3.8,4, 4.25,4.55,4.85, 1.25,1.55,1.85),
                     cohort=NA,
                    label=c("*",rep("",8))
                  )
  
  

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 cd8p2<- ggplot(cd8.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2.7),lty=2)+ylab("Spatial interaction proportion (p<0.01)")+xlab("CCR4+PD1+TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y = element_text(size=11))+geom_hline(yintercept = 0,linetype="dashed",color='red')+ggtitle("CD8")
  
 ## write out p-values.
 write.csv(cd8.rez.data,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-model-worksheets/Figure5-1-19/Figure5.CD8.results.csv")
  write.csv(cd4.rez.data,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-model-worksheets/Figure5-1-19/Figure5.CD4.results.csv")

## CD4
 
  cd4.d<-cd4.rez.data[which(cd4.rez.data$cohort!="contrast"),]
   cd4.d$phenotype<-factor(cd4.d$phenotype,levels=c("PD1+CD4+","PD1midCD4+","PD1-CD4+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.3, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("*",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 cd4p3<- ggplot(cd4.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+ylab("Spatial interaction proportion (p<0.01)")+xlab("CCR4+PD1+TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y=element_text(size=11))+geom_hline(yintercept = 0,linetype="dashed",color='red')+ggtitle("CD4")
  
 
 ### macrophages
 
  m.d<-m.rez.data[which(m.rez.data$cohort!="contrast"),]
  m.d$phenotype<-factor(m.d$phenotype,levels=c("PDL1+MAC+","PDL1midMAC+","PDL1-MAC+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.2, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("*",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 macp4<- ggplot(m.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+ylab("Spatial interaction proportion (p<0.01)")+xlab("PD1+CCR4+TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y = element_text(size=11))+geom_hline(yintercept = 0,linetype="dashed",color='red')+ggtitle("MAC")
  
 
  grid.arrange(cd8p2,cd4p3,macp4,nrow=2)

    grid.arrange(cd8p2,cd4p3,nrow=2)

  ## save ggplot for Tucker
    save(cd8p2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/spatial-comparison/final/cd8p2.CD8.Figure5D.RData")
    save(cd4p3,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/spatial-comparison/final/cd4p3.CD4.Figure5D.RData")
     save(macp4,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/spatial-comparison/final/macp4.MAC.Figure5D.RData")
  ## as a positive control  examine the PDL1+MAC and the PD1+ Tcell interactions. checking Shipp model
####################################################  
  ####################################################  
  ####################################################  
  ###################SHIPP MODEL#################################  
  ####################################################  
  ####################################################  
  ####################################################  
  p1$pmat[grepl("PDL1\\+Macrophage",rownames(p1$pmat)),!grepl("Macrophage",colnames(p1$pmat))]
 sum1<-100*p1$pmat[grepl("PDL1\\+Macrophage",rownames(p1$pmat)),!grepl("Macrophage",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)

 
 p1_hl$pmat[grepl("PDL1\\+Macrophage",rownames(p1_hl$pmat)),!grepl("Macrophage",colnames(p1_hl$pmat))]
  sum2<-100*p1_hl$pmat[grepl("PDL1\\+Macrophage",rownames(p1_hl$pmat)),!grepl("Macrophage",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)


  p1_rln$pmat[grepl("PDL1\\+Macrophage",rownames(p1_rln$pmat)),!grepl("Macrophage",colnames(p1_rln$pmat))]
  sum3<-100*p1_rln$pmat[grepl("PDL1\\+Macrophage",rownames(p1_rln$pmat)),!grepl("Macrophage",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)


  
 cd4.sum<-data.frame(CD4=c(sum1[grepl("PD1\\-CD4",names(sum1))],
                           sum2[grepl("PD1\\-CD4",names(sum2))],
                           sum3[grepl("PD1\\-CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD4",names(sum3))]))),
                     pheno="PDL1+MAC+",
                     interaction="PD1-CD4"
                    )  
 cd4.sum2<-data.frame(CD4=c(sum1[grepl("PD1\\+CD4",names(sum1))],
                           sum2[grepl("PD1\\+CD4",names(sum2))],
                           sum3[grepl("PD1\\+CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD4",names(sum3))]))),
                     pheno="PDL1+MAC+",
                     interaction="PD1+CD4"
                    )
 
 cd4.sum3<-data.frame(CD4=c(sum1[grepl("PD1.midCD4",names(sum1))],
                           sum2[grepl("PD1.midCD4",names(sum2))],
                           sum3[grepl("PD1.midCD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD4",names(sum3))]))),
                     pheno="PDL1+MAC+",
                     interaction="PD1.midCD4"
                    )
 cd4.sum.final<-rbind(cd4.sum,cd4.sum2,cd4.sum3)
# cd8
 cd8.sum<-data.frame(CD8=c(sum1[grepl("PD1\\-CD8",names(sum1))],
                           sum2[grepl("PD1\\-CD8",names(sum2))],
                           sum3[grepl("PD1\\-CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD8",names(sum3))]))),
                     pheno="PDL1+MAC+",
                     interaction="PD1-CD8"
                    )  
 cd8.sum2<-data.frame(CD8=c(sum1[grepl("PD1\\+CD8",names(sum1))],
                           sum2[grepl("PD1\\+CD8",names(sum2))],
                           sum3[grepl("PD1\\+CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD8",names(sum3))]))),
                     pheno="PDL1+MAC+",
                     interaction="PD1+CD8"
                    )
 
 cd8.sum3<-data.frame(CD8=c(sum1[grepl("PD1.midCD8",names(sum1))],
                           sum2[grepl("PD1.midCD8",names(sum2))],
                           sum3[grepl("PD1.midCD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD8",names(sum3))]))),
                     pheno="PDL1+MAC+",
                     interaction="PD1.midCD8"
                    )
 cd8.sum.final<-rbind(cd8.sum,cd8.sum2,cd8.sum3)
lm(CD4~cohort*interaction,cd4.sum.final)%>%summary
lm(CD8~cohort*interaction,cd8.sum.final)%>%summary
 
 ### PDL1+MAC anova
  cd4.global<-cd4.sum.final
  aov(CD4~cohort*interaction,cd4.global)%>%summary
  
  
   cd4.global$id<-rownames(cd4.global)
   cd4.global$id<-factor(cd4.global$id)
  
  
  ## marginal interaction p=0.09, but not significant
  cd8.global<-cd8.sum.final
  ## fixing id
  cd8.global$id<-rownames(cd8.global)
  cd8.global$id<-factor(cd8.global$id)
    
  aov(CD8~cohort*interaction,cd8.global)%>%summary
  
  
  
  
  ## we will not stratify but form a single LM with contrasting.
  ## checking residuals
  colnames(cd8.global)[1]<-"Proportion"
  colnames(cd4.global)[1]<-"Proportion"

   #library(multcomp);library(lme4)
#	spatial.global <-  lmer(Proportion~0+pheno+cohort+(1|interaction),tcells)
#	  alpha_noint <- rbind(
	#HL.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_1",36,-1)/37,
	#RLN.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_2",36,-1)/37,
  library(lsmeans);library(afex)
  fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd8.global,return="aov")
  ref1<-lsmeans(fit,c("interaction","cohort"))  
  
  ## dropping PD1-CCR4+
 # fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd8.global[which(cd8.global$pheno!="PD1-CCR4+"),],return="aov")
  #ref1<-lsmeans(fit,c("pheno","cohort")) 

  c_list <- list(PD1CD8.DLBCL.vOther = c(0,0,1, 0,0,-1/2,0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0,0, -1/2,0,0, -1/2,0,0),
                PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1/2,0, 0,-1/2,0),
               
                PD1CD8.DLBCL=c(0,0,1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0,0,0,0,0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
               
 
  

  summary(contrast(ref1, c_list), adjust = "BH")
 cd8.rez.data<- confint(contrast(ref1, c_list), adjust = "BH")%>%data.frame
  cd8.rez.data$p.value<-summary(contrast(ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(ref1, c_list)), test = adjusted("BH"))

  
cd8.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
cd8.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CD8+",3),rep("PD1midCD8+",3),rep("PD1-CD8+",3))  

  
  ## CD4 anova
  cd4fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd4.global,return="aov")
  cd4ref1<-lsmeans(cd4fit,c("interaction","cohort"))  
  

  c_list2 <- list(PD1CD8.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0, 0, -1,0,0, -1,0,0),
                PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1,0, 0,-1,0),
               
                PD1CD8.DLBCL=c(0,0, 1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0,0,0,0,0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
   
 
   
  summary(contrast(cd4ref1, c_list2), adjust = "BH")
 cd4.rez.data<- confint(contrast(cd4ref1, c_list2), adjust = "BH")%>%data.frame
  cd4.rez.data$p.value<-summary(contrast(cd4ref1, c_list2), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd4ref1, c_list2)), test = adjusted("BH"))

  
cd4.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
cd4.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CD4+",3),rep("PD1midCD4+",3),rep("PD1-CD4+",3))  

 ###
  cd8.d<-cd8.rez.data[which(cd8.rez.data$cohort!="contrast"),]
   cd8.d$phenotype<-factor(cd8.d$phenotype,levels=c("PD1+CD8+","PD1midCD8+","PD1-CD8+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
  
  cohortColors<-threeColors
  names(cohortColors)[2]<-"HL"
  

  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                      estimate = c(3.5,3.8,4, 4.25,4.55,4.85, 1.25,1.55,1.85),
                     cohort=NA,
                    label=c("",rep("",8))
                  )
  
  

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p5<- ggplot(cd8.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2.7),lty=2)+ylab("Spatial interaction with CD8 proportion (p<0.01)")+xlab("PDL1+MAC phenotype")+theme(axis.text=element_text(color='black'))
  
## CD4
 
  cd4.d<-cd4.rez.data[which(cd4.rez.data$cohort!="contrast"),]
   cd4.d$phenotype<-factor(cd4.d$phenotype,levels=c("PD1+CD4+","PD1midCD4+","PD1-CD4+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.3, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p6<- ggplot(cd4.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+ylab("Spatial interaction with CD4 proportion (p<0.01)")+xlab("PDL1+MAC phenotype")+theme(axis.text=element_text(color='black'))
  
 
 zz$CCR4.PD1.anno.nnd<-sapply(strsplit(zz$CCR4.PD1.anno,"_"),function(x) x[1])
 hd.z2$CCR4.PD1.anno.nnd<-sapply(strsplit(hd.z2$CCR4.PD1.anno,"_"),function(x) x[1])

 ### nearest distance secondary model.
 ## conclusion there are trends showing that the difference in space is greater in DLBCL comparing NND from PD1+CD8/PD1-CD8 to PD1+CCR4+Treg.   Significant differences in CD4, and trends in CD8.
 HH3<-NULL
   rois<-zz$ROIID%>%unique()
  for(i in rois){
  print(i)
  data<-subset(zz,zz$ROIID==i)
     pp<-phenographToPPP(case=data,marksCase=factor(data$CCR4.PD1.anno.nnd),shift=FALSE)
   H<-hyperframe(point=pp,
	response=unique(data$experiment),
	ROI=i)
   if(is.null(HH3)!=TRUE){
  HH3<-rbind(H,HH3)
  }else{
 HH3<-H
  } 
 } ###
 
 
 
  HH4<-NULL
   rois<-hd.z2$ROIID%>%unique()
  for(i in rois){
  print(i)
  data<-subset(hd.z2,hd.z2$ROIID==i)
     pp<-phenographToPPP(case=data,marksCase=factor(data$CCR4.PD1.anno.nnd),shift=FALSE)
   H<-hyperframe(point=pp,
	response=unique(data$experiment),
	ROI=i)
   if(is.null(HH4)!=TRUE){
  HH4<-rbind(H,HH4)
  }else{
 HH4<-H
  } 
 } ###
   HH4<-HH4[which(HH4$ROI!="Tonsil_2_2_a0" & HH4$ROI!="Tonsil_1_1_a0" & HH4$ROI!="Placenta_2_4_a0" & HH4$ROI!="Placenta_1_3_a0"),]
 ### nearest distance from PD1+Ki67- Treg to any Tcell   compare  PD1- to any Tcell  as the axis.

  findNND<-function(point=NULL,i=NULL,j=NULL){
    splitd<-split(point)
    nnD<-nncross(splitd[[i]],splitd[[j]])
    nnD$from<-i
    nnD$to<-j
    return(nnD)
  }   

  
 ## DLBCL
  ## PD1/ CCR4+
  a<-with(HH3[which(with(HH3, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH3, any(grepl("PD1\\+CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1+CD8"))
    ##  PD1-CD8
 b<-with(HH3[which(with(HH3, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH3, any(grepl("PD1\\-CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1-CD8"))
 
 c<-with(HH3[which(with(HH3, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH3, any(grepl("PD1\\+CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1+CD4"))
 d<-with(HH3[which(with(HH3, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH3, any(grepl("PD1\\-CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1-CD4"))

 e<-with(HH3[which(with(HH3, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH3, any(grepl("PD1\\+CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1+CD8"))
 
 f<-with(HH3[which(with(HH3, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH3, any(grepl("PD1\\-CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1-CD8"))
 
 g<-with(HH3[which(with(HH3, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH3, any(grepl("PD1\\+CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1+CD4"))
 
 h<-with(HH3[which(with(HH3, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH3, any(grepl("PD1\\-CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1-CD4"))
 
 
 ## HL
  a2<-with(HH4[which(with(HH4, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH4, any(grepl("PD1\\+CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1+CD8"))
  
  b2<-with(HH4[which(with(HH4, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH4, any(grepl("PD1\\-CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1-CD8")) 


  c2<-with(HH4[which(with(HH4, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH4, any(grepl("PD1\\+CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1+CD4"))
  
  d2<-with(HH4[which(with(HH4, any(grepl("PD1\\+CCR4\\+TREG",unique(marks(point)))))& with(HH4, any(grepl("PD1\\-CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PD1+CCR4+TREG",j="PD1-CD4")) 

 e2<-with(HH4[which(with(HH4, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH4, any(grepl("PD1\\+CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1+CD8"))
 
 f2<-with(HH4[which(with(HH4, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH4, any(grepl("PD1\\-CD8",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1-CD8"))
 
 g2<-with(HH4[which(with(HH4, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH4, any(grepl("PD1\\+CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1+CD4"))
 
 h2<-with(HH4[which(with(HH4, any(grepl("PDL1\\+Macrophage",unique(marks(point)))))& with(HH4, any(grepl("PD1\\-CD4",unique(marks(point)))))  )  ,],
    findNND(point=point,i="PDL1+Macrophage",j="PD1-CD4"))
 
 
  ## gathering the data
  
  asum<-sapply(a,function(x) mean(x$dist))
  bsum<-sapply(b,function(x) mean(x$dist))
  csum<-sapply(c,function(x) mean(x$dist))
  dsum<-sapply(d,function(x) mean(x$dist))
  esum<-sapply(e,function(x) mean(x$dist))
  fsum<-sapply(f,function(x) mean(x$dist))
  gsum<-sapply(g,function(x) mean(x$dist))
  hsum<-sapply(h,function(x) mean(x$dist))

  
 nnD.dl<-data.frame(nnd=c(asum,bsum,csum,
                   dsum,esum,fsum,gsum,hsum),
            to=c( rep("PD1+CD8",length(asum)),
                         rep("PD1-CD8",length(bsum)),
                         rep("PD1+CD4",length(csum)),
                          rep("PD1-CD4",length(dsum)),
                  
                          rep("PD1+CD8",length(esum)),
                          rep("PD1-CD8",length(fsum)),
                         rep("PD1+CD4",length(gsum)),
                       rep("PD1-CD4",length(hsum))
                          ),
            from=c( rep("PD1+CCR4+TREG",length(asum)),
                         rep("PD1+CCR4+TREG",length(bsum)),
                         rep("PD1+CCR4+TREG",length(csum)),
                          rep("PD1+CCR4+TREG",length(dsum)),
                  
                          rep("PDL1+MAC",length(esum)),
                          rep("PDL1+MAC",length(fsum)),
                         rep("PDL1+MAC",length(gsum)),
                       rep("PDL1+MAC",length(hsum))
                          )
              )
 
            
 nnD.dl$cohort<-"DLBCL"
 nnD.dl$id<-rownames(nnD.dl)
 
 
 ggplot(nnD.dl,aes(x=to,y=nnd))+geom_boxplot()+facet_wrap(~from)

 
 ## HL
  asum2<-sapply(a2,function(x) mean(x$dist))
  bsum2<-sapply(b2,function(x) mean(x$dist))
  csum2<-sapply(c2,function(x) mean(x$dist))
  dsum2<-sapply(d2,function(x) mean(x$dist))
  esum2<-sapply(e2,function(x) mean(x$dist))
  fsum2<-sapply(f2,function(x) mean(x$dist))
  gsum2<-sapply(g2,function(x) mean(x$dist))
  hsum2<-sapply(h2,function(x) mean(x$dist))

  
 nnD.hl<-data.frame(nnd=c(asum2,bsum2,
                          csum2,dsum2,
                          esum2,fsum2,gsum2,hsum2),
            to=c( rep("PD1+CD8",length(asum2)),
                         rep("PD1-CD8",length(bsum2)),
                         rep("PD1+CD4",length(csum2)),
                          rep("PD1-CD4",length(dsum2)),
                  
                          rep("PD1+CD8",length(esum2)),
                          rep("PD1-CD8",length(fsum2)),
                         rep("PD1+CD4",length(gsum2)),
                       rep("PD1-CD4",length(hsum2))
                          ),
            from=c( rep("PD1+CCR4+TREG",length(asum2)),
                         rep("PD1+CCR4+TREG",length(bsum2)),
                         rep("PD1+CCR4+TREG",length(csum2)),
                          rep("PD1+CCR4+TREG",length(dsum2)),
                  
                          rep("PDL1+MAC",length(esum2)),
                          rep("PDL1+MAC",length(fsum2)),
                         rep("PDL1+MAC",length(gsum2)),
                       rep("PDL1+MAC",length(hsum2))
                          )
              )
 
            
 nnD.hl$cohort<-"HL"
 nnD.hl$id<-rownames(nnD.hl)
 
 total.nnd<-rbind(nnD.dl,nnD.hl)
total.nnd$id<-rownames(total.nnd) 


  ggplot(total.nnd[which(total.nnd$nnd<150),],aes(x=from,y=nnd,fill=to))+geom_boxplot()+facet_wrap(~cohort)+ggpubr::stat_compare_means(method='anova')
 
   lm(nnd~cohort,total.nnd[which(total.nnd$phenotype=="PD1Pos.KiNeg"),])%>%summary

  
 
 total.nnd$state<-"Exhausted"
 total.nnd$state[total.nnd$phenotype%in%c("PD1Mid.KiPos",
                                        "PD1Neg.KiMid",
                                       "PD1Neg.KiPos",
                                       "PD1Pos.KiPos")]<-"Activated"
 ## fit the contrast
  nndfit<-aov.car(nnd~cohort*to+Error(id),data=total.nnd[which(total.nnd$from=="PD1+CCR4+TREG"& total.nnd$nnd<150),],return="aov")
  nndref1<-lsmeans(nndfit,c("cohort","to"))  
  

  nnd_list <- list(CD8PvsCD8N = c(0,0, -1/2,-1/2, 0,0, 1/2,1/2),
               CD4P.vs.CD4N = c(-1/2,-1/2, 0,0, 1/2,1/2, 0,0),
                CD8P.vs.CD8N.DLBCL = c(0,0, -1,0, 0,0, 1,0),
               CD8P.vs.CD8N.HL = c(0,0, 0,-1, 0,0, 0,1),
               CD4P.vs.CD4N.DLBCL = c(-1,0, 0,0, 1,0, 0,0),
               CD4P.vs.CD4N.HL = c(0,-1, 0,0, 0,1, 0,0),
               
               CD8P.vs.CD8N.DLBCL.vs.HL = c(0,0, -1,1, 0,0, 1,-1),
               CD4P.vs.CD4N.DLBCL.vs.HL = c(-1,1, 0,0, 1,-1, 0,0),
               
               CD8P.DLBCL = c(0,0, 0,0, 0,0, 1,0),
               CD8N.DLBCL = c(0,0, 1,0, 0,0, 0,0),
               CD8P.HL = c(0,0, 0,0, 0,0, 0,1),
               CD8N.HL = c(0,0, 0,1, 0,0, 0,0),
               CD4P.DLBCL = c(0,0, 0,0, 1,0, 0,0),
               CD4N.DLBCL = c(1,0, 0,0, 0,0, 0,0),
               CD4P.HL = c(0,0, 0,0, 0,1, 0,0),
               CD4N.HL = c(0,1, 0,0, 0,0, 0,0)
               )
   
 
   
  summary(contrast(nndref1, nnd_list), adjust = "BH")
 nnd.rez.data<- confint(contrast(nndref1, nnd_list), adjust = "BH")%>%data.frame
  nnd.rez.data$p.value<-summary(contrast(nndref1, nnd_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(nndref1, nnd_list)), test = adjusted("BH"))

  
  nnd.rez.data$cohort<-c(rep("contrast",8),rep(c("DLBCL","DLBCL","HL","HL"),2))  
  nnd.rez.data$phenotype<-c(rep("contrast",8),rep(c("PD1+CD8+","PD1-CD8"),2),
                            rep(c("PD1+CD4","PD1-CD4"),2))  
 
 
## macs
   nndfit<-aov.car(nnd~cohort*to+Error(id),data=total.nnd[which(total.nnd$from=="PDL1+MAC"& total.nnd$nnd<150),],return="aov")
  nndref1<-lsmeans(nndfit,c("cohort","to"))  
  summary(contrast(nndref1, nnd_list), adjust = "BH")
   
```

## PD1 TREG spatial interaction
- we don't observe the same pattern from CCR4+PD1+
```{r}
    zz$PD1.anno<-as.character(zz$unsup.subcommunity)
   zz$PD1.anno[grepl("TREG",zz$PD1.anno)]<-paste0(zz$PD1.thresh[grepl("TREG",zz$PD1.anno)],zz$unsup.subcommunity[grepl("TREG",zz$PD1.anno)])
  zz$PD1.anno[grepl("CD4",zz$PD1.anno)]<-paste0(zz$PD1.thresh[grepl("CD4",zz$PD1.anno)],zz$unsup.subcommunity[grepl("CD4",zz$PD1.anno)])
    zz$PD1.anno[grepl("CD8",zz$PD1.anno)]<-paste0(zz$PD1.thresh[grepl("CD8",zz$PD1.anno)],zz$unsup.subcommunity[grepl("CD8",zz$PD1.anno)])
 zz$PD1.anno[grepl("Macrophage",zz$PD1.anno)]<-paste0(zz$PDL1.thresh[grepl("Macrophage",zz$PD1.anno)],zz$unsup.subcommunity[grepl("Macrophage",zz$PD1.anno)])
 
 table(zz$PD1.anno)
 
 zz.tme<-zz[which(zz$subType.correction!="Tumor"),]

 dat_cells<-data.frame(ImageNumber=zz.tme$ROIID,
	ObjectNumber=zz.tme$CellId,
	label=as.character(zz.tme$PD1.anno),
	group=zz.tme$ROIID)
  dat_cells<-data.table(dat_cells)
  
  dat_relation_15<-createRelationTable(dn5=zz.tme,
	myPheno="PD1.anno",
	minDistance=15)
  
save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_PD1_PD1Tcells.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_PD1_PD1Tcells.RData")

  
 ## using the p=0.01 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

  save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_PD1_PD1Tcells.01.RData")

 dat_cells<-NULL
 dat_relation_15<-NULL
 zz.tme<-NULL
  
  ### HL
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/steidl.hl.CD4.CD8.MAC.TREG.subpop.RData")
 # hd.z2$ICOS.Ki67.anno<-hd.z2$subpopulation_meta
   #hd.z2$ICOS.Ki67.anno[grepl("TREG",hd.z2$ICOS.Ki67.anno)]<-paste0(hd.z2$ICOS.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],                                                     #       hd.z2$Ki67.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],
   #     hd.z2$PD1.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],
  #     hd.z2$subpopulation_meta[grepl("TREG",hd.z2$ICOS.Ki67.anno)])
  
   hd.z2$PD1.anno<-hd.z2$subpopulation_meta
   hd.z2$PD1.anno[grepl("TREG",hd.z2$PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("TREG",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("TREG",hd.z2$PD1.anno)])
 hd.z2$PD1.anno[grepl("CD4",hd.z2$PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("CD4",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("CD4",hd.z2$PD1.anno)])
hd.z2$PD1.anno[grepl("CD8",hd.z2$PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("CD8",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("CD8",hd.z2$PD1.anno)])
  hd.z2$PD1.anno[grepl("Macrophage",hd.z2$PD1.anno)]<-paste0(hd.z2$PDL1.thresh[grepl("Macrophage",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("Macrophage",hd.z2$PD1.anno)])

   
  
  
  table(hd.z2$PD1.anno)
  toFilt<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0",
           "Placenta_1_3_a0",
           "Placenta_2_4_a0",
           "Tonsil_1_1_a0",
           "Tonsil_2_2_a0")

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[which(!hd.z.tme$ROIID%in%toFilt),]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$PD1.anno)
  
 dat_cells_hl<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_hl<-data.table(dat_cells_hl)
  
  dat_relation_15_hl<-createRelationTable(dn5=hd.z.tme,
	myPheno="PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_PD1_PD1Tcells_hl.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_PD1_PD1Tcells_hl.RData")

  #filter tonsil and RLN.
 
 ## using the p=0.01 threshold.
  p1_hl<-makePermutationNBHD(dat_cells=dat_cells_hl,
	dat_relation=dat_relation_15_hl,
	n_perm=1000,pthreshold=0.01)

  save(p1_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_PD1_PD1Tcells.01.RData")

  
  ### RLN analysis.
  toRLN<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0"
          )

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[hd.z.tme$ROIID%in%toRLN,]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$PD1.anno)
  
 
 dat_cells_rln<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_rln<-data.table(dat_cells_rln)
  
  dat_relation_15_rln<-createRelationTable(dn5=hd.z.tme,
	myPheno="PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_PD1_PD1Tcells_rln.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_PD1_PD1Tcells_rln.RData")

  #filter tonsil and RLN.
 dat_cells_rln$label<-factor(dat_cells_rln$label)
 ## using the p=0.01 threshold.
  p1_rln<-makePermutationNBHD(dat_cells=dat_cells_rln,
	dat_relation=dat_relation_15_rln,
	n_perm=1000,pthreshold=0.01)

  save(p1_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_PD1_PD1Tcells.01.RData")

  
  dat_relation_15_hl<-NULL
 dat_relation_15<-NULL
 hd.z.tme<-NULL
 
 
# load the permutation results and model the hypothesis. 
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_PD1_PD1Tcells.01.RData")
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_PD1_PD1Tcells.01.RData")
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_PD1_PD1Tcells.01.RData")

 ######################## gather ###################

 p1$pmat[grepl("PD1\\+TREG",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]
 sum1<-100*p1$pmat[grepl("PD1\\+TREG",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)

 
 p1_hl$pmat[grepl("PD1\\+TREG",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]
  sum2<-100*p1_hl$pmat[grepl("PD1\\+TREG",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)


  p1_rln$pmat[grepl("PD1\\+TREG",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]
  sum3<-100*p1_rln$pmat[grepl("PD1\\+TREG",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)


  
 cd4.sum<-data.frame(CD4=c(sum1[grepl("PD1\\-CD4",names(sum1))],
                           sum2[grepl("PD1\\-CD4",names(sum2))],
                           sum3[grepl("PD1\\-CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD4",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PD1-CD4"
                    )  
 cd4.sum2<-data.frame(CD4=c(sum1[grepl("PD1\\+CD4",names(sum1))],
                           sum2[grepl("PD1\\+CD4",names(sum2))],
                           sum3[grepl("PD1\\+CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD4",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PD1+CD4"
                    )
 
 cd4.sum3<-data.frame(CD4=c(sum1[grepl("PD1.midCD4",names(sum1))],
                           sum2[grepl("PD1.midCD4",names(sum2))],
                           sum3[grepl("PD1.midCD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD4",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PD1.midCD4"
                    )
 cd4.sum.final<-rbind(cd4.sum,cd4.sum2,cd4.sum3)
# cd8
 cd8.sum<-data.frame(CD8=c(sum1[grepl("PD1\\-CD8",names(sum1))],
                           sum2[grepl("PD1\\-CD8",names(sum2))],
                           sum3[grepl("PD1\\-CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD8",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PD1-CD8"
                    )  
 cd8.sum2<-data.frame(CD8=c(sum1[grepl("PD1\\+CD8",names(sum1))],
                           sum2[grepl("PD1\\+CD8",names(sum2))],
                           sum3[grepl("PD1\\+CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD8",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PD1+CD8"
                    )
 
 cd8.sum3<-data.frame(CD8=c(sum1[grepl("PD1.midCD8",names(sum1))],
                           sum2[grepl("PD1.midCD8",names(sum2))],
                           sum3[grepl("PD1.midCD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD8",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PD1.midCD8"
                    )
 cd8.sum.final<-rbind(cd8.sum,cd8.sum2,cd8.sum3)
lm(CD4~cohort*interaction,cd4.sum.final)%>%summary
lm(CD8~cohort*interaction,cd8.sum.final)%>%summary
 
 ## PDL1+ MAC

 m.sum<-data.frame(MAC=c(sum1[grepl("PDL1\\-Macrophage",names(sum1))],
                           sum2[grepl("PDL1\\-Macrophage",names(sum2))],
                           sum3[grepl("PDL1\\-Macrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1\\-Macrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1\\-Macrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1\\-Macrophage",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PDL1-Macrophage"
                    )  
 m.sum2<-data.frame(MAC=c(sum1[grepl("PDL1\\+Macrophage",names(sum1))],
                           sum2[grepl("PDL1\\+Macrophage",names(sum2))],
                           sum3[grepl("PDL1\\+Macrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1\\+Macrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1\\+Macrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1\\+Macrophage",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PDL1+Macrophage"
                    )
 
 m.sum3<-data.frame(MAC=c(sum1[grepl("PDL1.midMacrophage",names(sum1))],
                           sum2[grepl("PDL1.midMacrophage",names(sum2))],
                           sum3[grepl("PDL1.midMacrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1.midMacrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1.midMacrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1.midMacrophage",names(sum3))]))),
                     pheno="PD1+TREG",
                     interaction="PDL1.midMacrophage"
                    )
 m.sum.final<-rbind(m.sum,m.sum2,m.sum3)


 
 
 ## we test for interactions checking to see if stratified analysis is justified
############### MODELING ############################
##########################
 ########################
######################### 
 
 ## no interactions between phenotypes across cohorts.
  cd4.global<-cd4.sum.final
  aov(CD4~cohort*interaction,cd4.global)%>%summary
  
  
   cd4.global$id<-rownames(cd4.global)
   cd4.global$id<-factor(cd4.global$id)
  
  
  ## marginal interaction p=0.09, but not significant
  cd8.global<-cd8.sum.final
  ## fixing id
  cd8.global$id<-rownames(cd8.global)
  cd8.global$id<-factor(cd8.global$id)
    
  aov(CD8~cohort*interaction,cd8.global)%>%summary
  
  
  ## macrphage
  m.global<-m.sum.final
  ## fixing id
  m.global$id<-rownames(m.global)
  m.global$id<-factor(m.global$id)
    
  aov(MAC~cohort*interaction,m.global)%>%summary
  
  
  
  ## we will not stratify but form a single LM with contrasting.
  ## checking residuals
  colnames(cd8.global)[1]<-"Proportion"
  colnames(cd4.global)[1]<-"Proportion"
    colnames(m.global)[1]<-"Proportion"

    ## trend test
    tester<-cd8.global[which(cd8.global$cohort=="DLBCL"),]
    tester$class<-0
    tester$class[which(tester$interaction=="PD1.midCD8")]<-1
    tester$class[which(tester$interaction=="PD1+CD8")]<-2
####    
    
    
 # tcells<-rbind(cd8.global,cd4.global)
  ### checking fit residuals.
  anova2<-lm(Proportion~cohort*pheno,cd8.global)
  qqnorm(anova2$residuals)
  qqline(anova2$residuals, datax=FALSE, distribution=qnorm)
  plot(anova2$residuals~anova2$fitted.values)
  plot(density(anova2$residuals))
 
   #library(multcomp);library(lme4)
#	spatial.global <-  lmer(Proportion~0+pheno+cohort+(1|interaction),tcells)
#	  alpha_noint <- rbind(
	#HL.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_1",36,-1)/37,
	#RLN.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_2",36,-1)/37,
  library(lsmeans);library(afex)
  fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd8.global,return="aov")
  ref1<-lsmeans(fit,c("interaction","cohort"))  
  
  ## dropping PD1-CCR4+
 # fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd8.global[which(cd8.global$pheno!="PD1-CCR4+"),],return="aov")
  #ref1<-lsmeans(fit,c("pheno","cohort")) 

  c_list <- list(PD1CD8.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0, 0, -1/2,0,0, -1/2,0,0),
                PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1/2,0, 0,-1/2,0),
               PD1CD8.DLBCL=c(0,0, 1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0,0,0,0,0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
               
 library(multcomp)
  

  summary(contrast(ref1, c_list), adjust = "BH")
 cd8.rez.data<- confint(contrast(ref1, c_list), adjust = "BH")%>%data.frame
  cd8.rez.data$p.value<-summary(contrast(ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(ref1, c_list)), test = adjusted("BH"))

  
cd8.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
cd8.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CD8+",3),rep("PD1midCD8+",3),rep("PD1-CD8+",3))  

  
  ## CD4 anova
  cd4fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd4.global,return="aov")
  cd4ref1<-lsmeans(cd4fit,c("interaction","cohort"))  
  

  c_list2 <- list(PD1CD8.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0, 0, -1,0,0, -1,0,0),
                PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1,0, 0,-1,0),
               
                PD1CD8.DLBCL=c(0,0, 1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0,0,0,0,0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
   
 
   
  summary(contrast(cd4ref1, c_list2), adjust = "BH")
 cd4.rez.data<- confint(contrast(cd4ref1, c_list2), adjust = "BH")%>%data.frame
  cd4.rez.data$p.value<-summary(contrast(cd4ref1, c_list2), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd4ref1, c_list2)), test = adjusted("BH"))

  
cd4.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
cd4.rez.data$phenotype<-c(rep("contrast",3),rep("PD1+CD4+",3),rep("PD1midCD4+",3),rep("PD1-CD4+",3))  

## macrophage
 mfit<-aov.car(Proportion~interaction*cohort+Error(id),data=m.global,return="aov")
 mref1<-lsmeans(mfit,c("interaction","cohort"))  
  

  c_list3 <- list(PDL1Mac.DLBCL.vOther = c(0,0, 1, 0, 0, -1/2,0,0,-1/2),
               PDL1nMac.DLBCL.vOther = c(1,0, 0, -1,0,0, -1,0,0),
                PDL1midMac.DLBCL.vOther = c(0,1,0, 0,-1,0, 0,-1,0),
               
                PDL1Mac.DLBCL=c(0,0, 1, 0,0,0, 0,0,0),
               PDL1Mac.HL=c(0,0,0, 0,0,1, 0,0,0),
               PDL1Mac.RLN=c(0,0,0,0,0,0,0,0,1),
               PDL1midMac.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PDL1midMac.HL=c(0,0,0, 0,1,0, 0,0,0),
               PDL1midMac.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PDL1nMac.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PDL1nMac.HL=c(0,0,0, 1,0,0, 0,0,0),
               PDL1nMac.RLN=c(0,0,0, 0,0,0, 1,0,0))
   
 
   
  summary(contrast(mref1, c_list3), adjust = "BH")
 m.rez.data<- confint(contrast(mref1, c_list3), adjust = "BH")%>%data.frame
  m.rez.data$p.value<-summary(contrast(mref1, c_list3), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(mref1, c_list3)), test = adjusted("BH"))

  
m.rez.data$cohort<-c(rep("contrast",3),rep(c("DLBCL","HL","RLN"),3))  
m.rez.data$phenotype<-c(rep("contrast",3),rep("PDL1+MAC+",3),rep("PDL1midMAC+",3),rep("PDL1-MAC+",3))  

################################
  
  
### PD1-CD4
 lm(CD4~0+cohort,cd4.sum)%>%summary
# PD1+CD4
 lm(CD4~0+cohort,cd4.sum2)%>%summary
 #PD1 mid CD4
  lm(CD4~0+cohort,cd4.sum3)%>%summary
  cd4.rez.data
  
  lm(CD8~0+cohort,cd8.sum)%>%summary
# PD1+CD4
 lm(CD8~0+cohort,cd8.sum2)%>%summary
 #PD1 mid CD4
  lm(CD8~0+cohort,cd8.sum3)%>%summary
  cd8.rez.data
 #######################################
 
################ 
 # figure making of the ANOVA for CD8
   cd8.d<-cd8.rez.data[which(cd8.rez.data$cohort!="contrast"),]
   cd8.d$phenotype<-factor(cd8.d$phenotype,levels=c("PD1+CD8+","PD1midCD8+","PD1-CD8+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
  
  cohortColors<-threeColors
  names(cohortColors)[2]<-"HL"
  

  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                      estimate = c(3.5,3.8,4, 4.25,4.55,4.85, 1.25,1.55,1.85),
                     cohort=NA,
                    label=c("*",rep("",8))
                  )
  
  

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p2<- ggplot(cd8.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2.7),lty=2)+ylab("Spatial interaction with CD8 proportion (p<0.01)")+xlab("CCR4+PD1+TREG phenotype")+theme(axis.text=element_text(color='black'))+geom_hline(yintercept = 0,linetype="dashed",color='red')
  
## CD4
 
  cd4.d<-cd4.rez.data[which(cd4.rez.data$cohort!="contrast"),]
   cd4.d$phenotype<-factor(cd4.d$phenotype,levels=c("PD1+CD4+","PD1midCD4+","PD1-CD4+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.3, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("*",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p3<- ggplot(cd4.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+ylab("Spatial interaction with CD4 proportion (p<0.01)")+xlab("CCR4+PD1+TREG phenotype")+theme(axis.text=element_text(color='black'))+geom_hline(yintercept = 0,linetype="dashed",color='red')
  
 
 ### macrophages
 
  m.d<-m.rez.data[which(m.rez.data$cohort!="contrast"),]
  m.d$phenotype<-factor(m.d$phenotype,levels=c("PDL1+MAC+","PDL1midMAC+","PDL1-MAC+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.2, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("*",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p4<- ggplot(m.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+ylab("Spatial interaction with MAC proportion (p<0.01)")+xlab("PD1+CCR4+TREG phenotype")+theme(axis.text=element_text(color='black'))+geom_hline(yintercept = 0,linetype="dashed",color='red')
  
 
  grid.arrange(p2,p3,p4,nrow=2)

    grid.arrange(p2,p3,nrow=2)

  ## save ggplot for Tucker
    save(p2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/p2.CD8.Figure5D.RData")
    save(p3,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/p3.CD4.Figure5D.RData")
  ## as a positive control  examine the PDL1+MAC and the PD1+ Tcell interactions. checking Shipp model

```

# Generalizing
```{r}
  ## custom labeling by gating.
   zz$PD1.anno<-as.character(zz$unsup.subcommunity)
   zz$PD1.anno[grepl("TREG",zz$PD1.anno)]<-paste0(zz$Ki67.thresh[grepl("TREG",zz$PD1.anno)],zz$unsup.subcommunity[grepl("TREG",zz$PD1.anno)])
   
  zz$PD1.anno[grepl("CD4",zz$PD1.anno)]<-paste0(zz$PD1.thresh[grepl("CD4",zz$PD1.anno)],zz$unsup.subcommunity[grepl("CD4",zz$PD1.anno)])
    zz$PD1.anno[grepl("CD8",zz$PD1.anno)]<-paste0(zz$PD1.thresh[grepl("CD8",zz$PD1.anno)],zz$unsup.subcommunity[grepl("CD8",zz$PD1.anno)])
 zz$PD1.anno[grepl("Macrophage",zz$PD1.anno)]<-paste0(zz$PDL1.thresh[grepl("Macrophage",zz$PD1.anno)],zz$unsup.subcommunity[grepl("Macrophage",zz$PD1.anno)])
 
 table(zz$PD1.anno)
 
 
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/steidl.hl.CD4.CD8.MAC.TREG.subpop.RData")
 # hd.z2$ICOS.Ki67.anno<-hd.z2$subpopulation_meta
   #hd.z2$ICOS.Ki67.anno[grepl("TREG",hd.z2$ICOS.Ki67.anno)]<-paste0(hd.z2$ICOS.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],                                                     #       hd.z2$Ki67.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],
   #     hd.z2$PD1.thresh[grepl("TREG",hd.z2$ICOS.Ki67.anno)],
  #     hd.z2$subpopulation_meta[grepl("TREG",hd.z2$ICOS.Ki67.anno)])
  
   hd.z2$PD1.anno<-hd.z2$subpopulation_meta
   hd.z2$PD1.anno[grepl("TREG",hd.z2$PD1.anno)]<-paste0(hd.z2$Ki67.thresh[grepl("TREG",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("TREG",hd.z2$PD1.anno)])
   
 hd.z2$PD1.anno[grepl("CD4",hd.z2$PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("CD4",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("CD4",hd.z2$PD1.anno)])
hd.z2$PD1.anno[grepl("CD8",hd.z2$PD1.anno)]<-paste0(hd.z2$PD1.thresh[grepl("CD8",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("CD8",hd.z2$PD1.anno)])
  hd.z2$PD1.anno[grepl("Macrophage",hd.z2$PD1.anno)]<-paste0(hd.z2$PDL1.thresh[grepl("Macrophage",hd.z2$PD1.anno)],hd.z2$subpopulation_meta[grepl("Macrophage",hd.z2$PD1.anno)])
  
 
 
 
 zz.tme<-zz[which(zz$subType.correction!="Tumor"),]

 dat_cells<-data.frame(ImageNumber=zz.tme$ROIID,
	ObjectNumber=zz.tme$CellId,
	label=as.character(zz.tme$PD1.anno),
	group=zz.tme$ROIID)
  dat_cells<-data.table(dat_cells)
  
  dat_relation_15<-createRelationTable(dn5=zz.tme,
	myPheno="PD1.anno",
	minDistance=15)
  
save(dat_relation_15,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_Ki67_PD1Tcells.RData")
  

  
 ## using the p=0.01 threshold.
  p1<-makePermutationNBHD(dat_cells=dat_cells,
	dat_relation=dat_relation_15,
	n_perm=1000,pthreshold=0.01)

  save(p1,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_Ki67_PD1Tcells.01.RData")

 dat_cells<-NULL
 dat_relation_15<-NULL
 zz.tme<-NULL
  
  ### HL
 

   
  
  
  table(hd.z2$PD1.anno)
  toFilt<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0",
           "Placenta_1_3_a0",
           "Placenta_2_4_a0",
           "Tonsil_1_1_a0",
           "Tonsil_2_2_a0")

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[which(!hd.z.tme$ROIID%in%toFilt),]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$PD1.anno)
  
  dat_cells_hl<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_hl<-data.table(dat_cells_hl)
  
  dat_relation_15_hl<-createRelationTable(dn5=hd.z.tme,
	myPheno="PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_Ki67_PD1Tcells_hl.RData")
  
  

  #filter tonsil and RLN.
 
 ## using the p=0.01 threshold.
  p1_hl<-makePermutationNBHD(dat_cells=dat_cells_hl,
	dat_relation=dat_relation_15_hl,
	n_perm=1000,pthreshold=0.01)

  save(p1_hl,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_Ki67_PD1Tcells.01.RData")

  
  ### RLN analysis.
  toRLN<-c("Case1_ROI_01_5_a0",
           "Case1_ROI_02_6_a0",
           "Case2_ROI_01_7_a0",
           "Case5_ROI_01_10_a0",
           "Case4_ROI_02_9_a0",
           "Case3_ROI_01_8_a0"
          )

   hd.z.tme<-hd.z2[which(hd.z2$annotation!="BCell"),]
   hd.z.tme<-hd.z.tme[hd.z.tme$ROIID%in%toRLN,]
   hd.z.tme$ROIID<-as.character(hd.z.tme$ROIID)
   
  table(hd.z.tme$PD1.anno)
  
 
 dat_cells_rln<-data.frame(ImageNumber=hd.z.tme$ROIID,
	ObjectNumber=hd.z.tme$CellId,
	label=as.character(hd.z.tme$PD1.anno),
	group=hd.z.tme$ROIID)
  dat_cells_rln<-data.table(dat_cells_rln)
  
  dat_relation_15_rln<-createRelationTable(dn5=hd.z.tme,
	myPheno="PD1.anno",
	minDistance=15)
  
  save(dat_relation_15_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_Ki67_PD1Tcells_rln.RData")
  
  
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/dat_relation_15_Ki67_PD1Tcells_rln.RData")

  #filter tonsil and RLN.
 dat_cells_rln$label<-factor(dat_cells_rln$label)
 ## using the p=0.01 threshold.
  p1_rln<-makePermutationNBHD(dat_cells=dat_cells_rln,
	dat_relation=dat_relation_15_rln,
	n_perm=1000,pthreshold=0.01)

  save(p1_rln,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_Ki67_PD1Tcells.01.RData")

  
  dat_relation_15_hl<-NULL
 dat_relation_15<-NULL
 hd.z.tme<-NULL
 
 
# load the permutation results and model the hypothesis. 
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_Ki67_PD1Tcells.01.RData")
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_Ki67_PD1Tcells.01.RData")
 load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_Ki67_PD1Tcells.01.RData")

 #load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_15_PD1Ki67.01.RData")
 #load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_hl_15_PD1Ki67.01.RData")
 #load(file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/p1_rln_15_PD1Ki67.01.RData")

 
 ### anova model
 
 p1$pmat[grepl("Ki67\\+TREG",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]
 sum1<-100*p1$pmat[grepl("Ki67\\+TREG",rownames(p1$pmat)),!grepl("TREG",colnames(p1$pmat))]%>%colSums/(p1$dat_p$group%>%unique%>%length)

 
 p1_hl$pmat[grepl("Ki67\\+TREG",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]
  sum2<-100*p1_hl$pmat[grepl("Ki67\\+TREG",rownames(p1_hl$pmat)),!grepl("TREG",colnames(p1_hl$pmat))]%>%colSums/(p1_hl$dat_p$group%>%unique%>%length)


  p1_rln$pmat[grepl("Ki67\\+TREG",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]
  sum3<-100*p1_rln$pmat[grepl("Ki67\\+TREG",rownames(p1_rln$pmat)),!grepl("TREG",colnames(p1_rln$pmat))]%>%colSums/(p1_rln$dat_p$group%>%unique%>%length)


  
 cd4.sum<-data.frame(CD4=c(sum1[grepl("PD1\\-CD4",names(sum1))],
                           sum2[grepl("PD1\\-CD4",names(sum2))],
                           sum3[grepl("PD1\\-CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD4",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PD1-CD4"
                    )  
 cd4.sum2<-data.frame(CD4=c(sum1[grepl("PD1\\+CD4",names(sum1))],
                           sum2[grepl("PD1\\+CD4",names(sum2))],
                           sum3[grepl("PD1\\+CD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD4",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PD1+CD4"
                    )
 
 cd4.sum3<-data.frame(CD4=c(sum1[grepl("PD1.midCD4",names(sum1))],
                           sum2[grepl("PD1.midCD4",names(sum2))],
                           sum3[grepl("PD1.midCD4",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD4",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD4",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD4",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PD1.midCD4"
                    )
 cd4.sum.final<-rbind(cd4.sum,cd4.sum2,cd4.sum3)
# cd8
 cd8.sum<-data.frame(CD8=c(sum1[grepl("PD1\\-CD8",names(sum1))],
                           sum2[grepl("PD1\\-CD8",names(sum2))],
                           sum3[grepl("PD1\\-CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\-CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\-CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\-CD8",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PD1-CD8"
                    )  
 cd8.sum2<-data.frame(CD8=c(sum1[grepl("PD1\\+CD8",names(sum1))],
                           sum2[grepl("PD1\\+CD8",names(sum2))],
                           sum3[grepl("PD1\\+CD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1\\+CD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1\\+CD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1\\+CD8",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PD1+CD8"
                    )
 
 cd8.sum3<-data.frame(CD8=c(sum1[grepl("PD1.midCD8",names(sum1))],
                           sum2[grepl("PD1.midCD8",names(sum2))],
                           sum3[grepl("PD1.midCD8",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PD1.midCD8",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PD1.midCD8",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PD1.midCD8",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PD1.midCD8"
                    )
 cd8.sum.final<-rbind(cd8.sum,cd8.sum2,cd8.sum3)
lm(CD4~cohort*interaction,cd4.sum.final)%>%summary
lm(CD8~cohort*interaction,cd8.sum.final)%>%summary
 
 ## PDL1+ MAC

 m.sum<-data.frame(MAC=c(sum1[grepl("PDL1\\-Macrophage",names(sum1))],
                           sum2[grepl("PDL1\\-Macrophage",names(sum2))],
                           sum3[grepl("PDL1\\-Macrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1\\-Macrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1\\-Macrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1\\-Macrophage",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PDL1-Macrophage"
                    )  
 m.sum2<-data.frame(MAC=c(sum1[grepl("PDL1\\+Macrophage",names(sum1))],
                           sum2[grepl("PDL1\\+Macrophage",names(sum2))],
                           sum3[grepl("PDL1\\+Macrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1\\+Macrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1\\+Macrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1\\+Macrophage",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PDL1+Macrophage"
                    )
 
 m.sum3<-data.frame(MAC=c(sum1[grepl("PDL1.midMacrophage",names(sum1))],
                           sum2[grepl("PDL1.midMacrophage",names(sum2))],
                           sum3[grepl("PDL1.midMacrophage",names(sum3))]),
                     cohort=c(rep("DLBCL",length((sum1[grepl("PDL1.midMacrophage",names(sum1))]))),
                              rep("HL",length(sum2[grepl("PDL1.midMacrophage",names(sum2))])),
                      rep("RLN",length(sum3[grepl("PDL1.midMacrophage",names(sum3))]))),
                     pheno="Ki67+TREG",
                     interaction="PDL1.midMacrophage"
                    )
 m.sum.final<-rbind(m.sum,m.sum2,m.sum3)


 
 
 ## we test for interactions checking to see if stratified analysis is justified
############### MODELING ############################
##########################
 ########################
######################### 
 
 ## no interactions between phenotypes across cohorts.
  cd4.global<-cd4.sum.final
  aov(CD4~cohort*interaction,cd4.global)%>%summary
  
  
   cd4.global$id<-rownames(cd4.global)
   cd4.global$id<-factor(cd4.global$id)
  
  
  ## marginal interaction p=0.09, but not significant
  cd8.global<-cd8.sum.final
  ## fixing id
  cd8.global$id<-rownames(cd8.global)
  cd8.global$id<-factor(cd8.global$id)
  cd8.global$cohort<-factor(cd8.global$cohort,levels=c("DLBCL","HL","RLN")) 
  aov(CD8~cohort*interaction,cd8.global)%>%summary
  
  
  ## macrphage
  m.global<-m.sum.final
  ## fixing id
  m.global$id<-rownames(m.global)
  m.global$id<-factor(m.global$id)
    
  aov(MAC~cohort*interaction,m.global)%>%summary
  
  
  
  ## we will not stratify but form a single LM with contrasting.
  ## checking residuals
  colnames(cd8.global)[1]<-"Proportion"
  colnames(cd4.global)[1]<-"Proportion"
    colnames(m.global)[1]<-"Proportion"

    ## trend test
    tester<-cd8.global[which(cd8.global$cohort=="DLBCL"),]
    tester$class<-0
    tester$class[which(tester$interaction=="PD1.midCD8")]<-1
    tester$class[which(tester$interaction=="PD1+CD8")]<-2
####    
    
    
 # tcells<-rbind(cd8.global,cd4.global)
  ### checking fit residuals.
  anova2<-lm(Proportion~cohort*interaction,cd8.global)
  qqnorm(anova2$residuals)
  qqline(anova2$residuals, datax=FALSE, distribution=qnorm)
  plot(anova2$residuals~anova2$fitted.values)
  plot(density(anova2$residuals))
 
   #library(multcomp);library(lme4)
#	spatial.global <-  lmer(Proportion~0+pheno+cohort+(1|interaction),tcells)
#	  alpha_noint <- rbind(
	#HL.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_1",36,-1)/37,
	#RLN.vs.DLBCL_PD1CCR4=ifelse(subExpr[,2]%>%levels=="CD4_2",36,-1)/37,
  library(lsmeans);library(afex)
  fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd8.global,return="aov")
  ref1<-lsmeans(fit,c("interaction","cohort"))  
  
  ## dropping PD1-CCR4+
 # fit<-aov.car(Proportion~pheno*cohort+Error(id),data=cd8.global[which(cd8.global$pheno!="PD1-CCR4+"),],return="aov")
  #ref1<-lsmeans(fit,c("pheno","cohort")) 

  c_list <- list(PD1CD8.DLBCL.vOther = c(0,0,1, 0,0,-1/2, 0,0,-1/2),
               PD1nCD8.DLBCL.vOther = c(1,0, 0, -1/2,0,0, -1/2,0,0),
               PD1midCD8.DLBCL.vOther = c(0,1,0, 0,-1/2,0, 0,-1/2,0),
               CD8.DLBCL.vOther = c(1/3,1/3,1/3, -1/6,-1/6,-1/6, -1/6,-1/6,-1/6),
               PD1CD8.DLBCL=c(0,0,1, 0,0,0, 0,0,0),
               PD1CD8.HL=c(0,0,0, 0,0,1, 0,0,0),
               PD1CD8.RLN=c(0,0,0,0,0,0,0,0,1),
               PD1midCD8.DLBCL=c(0,1,0, 0,0,0, 0,0,0),
               PD1midCD8.HL=c(0,0,0, 0,1,0, 0,0,0),
               PD1midCD8.RLN=c(0,0,0, 0,0,0, 0,1,0),
               PD1nCD8.DLBCL=c(1,0,0, 0,0,0, 0,0,0),
               PD1nCD8.HL=c(0,0,0, 1,0,0, 0,0,0),
               PD1nCD8.RLN=c(0,0,0, 0,0,0, 1,0,0))
               
 library(multcomp)
  

  summary(contrast(ref1, c_list), adjust = "BH")
 cd8.rez.data<- confint(contrast(ref1, c_list), adjust = "BH")%>%data.frame
  cd8.rez.data$p.value<-summary(contrast(ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(ref1, c_list)), test = adjusted("BH"))

  
cd8.rez.data$cohort<-c(rep("contrast",4),rep(c("DLBCL","HL","RLN"),3))  
cd8.rez.data$phenotype<-c(rep("contrast",4),rep("PD1+CD8+",3),rep("PD1midCD8+",3),rep("PD1-CD8+",3))  

  
  ## CD4 anova
  cd4fit<-aov.car(Proportion~interaction*cohort+Error(id),data=cd4.global,return="aov")
  cd4ref1<-lsmeans(cd4fit,c("interaction","cohort"))  
  

  summary(contrast(cd4ref1, c_list), adjust = "BH")
 cd4.rez.data<- confint(contrast(cd4ref1, c_list), adjust = "BH")%>%data.frame
  cd4.rez.data$p.value<-summary(contrast(cd4ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd4ref1, c_list)), test = adjusted("BH"))

  
cd4.rez.data$cohort<-c(rep("contrast",4),rep(c("DLBCL","HL","RLN"),3))  
cd4.rez.data$phenotype<-c(rep("contrast",4),rep("PD1+CD4+",3),rep("PD1midCD4+",3),rep("PD1-CD4+",3))  

## macrophage
 mfit<-aov.car(Proportion~interaction*cohort+Error(id),data=m.global,return="aov")
 mref1<-lsmeans(mfit,c("interaction","cohort"))  
  

 
   
 
   
  summary(contrast(mref1, c_list), adjust = "BH")
 m.rez.data<- confint(contrast(mref1, c_list), adjust = "BH")%>%data.frame
  m.rez.data$p.value<-summary(contrast(mref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(mref1, c_list)), test = adjusted("BH"))

  
m.rez.data$cohort<-c(rep("contrast",4),rep(c("DLBCL","HL","RLN"),3))  
m.rez.data$phenotype<-c(rep("contrast",4),rep("PDL1+MAC+",3),rep("PDL1midMAC+",3),rep("PDL1-MAC+",3))  

################################
  
  
### PD1-CD4
 lm(CD4~0+cohort,cd4.sum)%>%summary
# PD1+CD4
 lm(CD4~0+cohort,cd4.sum2)%>%summary
 #PD1 mid CD4
  lm(CD4~0+cohort,cd4.sum3)%>%summary
  cd4.rez.data
  
  lm(CD8~0+cohort,cd8.sum)%>%summary
# PD1+CD4
 lm(CD8~0+cohort,cd8.sum2)%>%summary
 #PD1 mid CD4
  lm(CD8~0+cohort,cd8.sum3)%>%summary
  cd8.rez.data
 #######################################
 
################ 
 # figure making of the ANOVA for CD8
   cd8.d<-cd8.rez.data[which(cd8.rez.data$cohort!="contrast"),]
   cd8.d$phenotype<-factor(cd8.d$phenotype,levels=c("PD1+CD8+","PD1midCD8+","PD1-CD8+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
  
  cohortColors<-threeColors
  names(cohortColors)[2]<-"HL"
  

  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                      estimate = c(3.5,3.8,4, 4.25,4.55,4.85, 1.25,1.55,1.85),
                     cohort=NA,
                    label=c("",rep("",8))
                  )
  
  

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 kip2<- ggplot(cd8.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      #geom_line(data=arc.df,aes(phenotype+1,Estimate+2.7),lty=2)+
   ylab("Spatial interaction proportion (p<0.01)")+xlab("Ki67+TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y=element_text(size=11))+geom_hline(yintercept = 0,linetype="dashed",color='red')+ggtitle("CD8")
  
## CD4
 
  cd4.d<-cd4.rez.data[which(cd4.rez.data$cohort!="contrast"),]
   cd4.d$phenotype<-factor(cd4.d$phenotype,levels=c("PD1+CD4+","PD1midCD4+","PD1-CD4+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.3, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 kip3<- ggplot(cd4.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      #geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+
   ylab("Spatial interaction proportion (p<0.01)")+xlab("Ki67+TREG phenotype")+theme(axis.text=element_text(color='black'),axis.title.y = element_text(size=11))+geom_hline(yintercept = 0,linetype="dashed",color='red')+ggtitle("CD4")
  
 
 ### macrophages
 
  m.d<-m.rez.data[which(m.rez.data$cohort!="contrast"),]
  m.d$phenotype<-factor(m.d$phenotype,levels=c("PDL1+MAC+","PDL1midMAC+","PDL1-MAC+"))
 
 library(ggpubr) 
  
  my_comparisons <- list( c("0.5", "1"), c("1", "2"), c("0.5", "2") )
 
  label.df <- data.frame(phenotype = c(1,1,1, 2,2,2, 3,3,3),
                       estimate = c(2.8,3,3.2, 3.25,3.55,3.85, 1.25,1.55,1.85),
                       cohort=NA,
                       label=c("*",rep("",8))
                      )

# Define arc coordinates
r <- 0.15
t <- seq(0, 180, by = 1) * pi / 180
x <- r * cos(t)
y <- r*5 * sin(t)
  arc.df <- data.frame(phenotype = x, Estimate = y,cohort=NA)
  
 p4<- ggplot(m.d,aes(x=phenotype,y=estimate,fill=cohort))+
                 geom_point(shape=21,stroke=1,size=5.5,position=position_dodge(width=0.25))+
   geom_errorbar(aes(ymin=lower.CL,ymax=upper.CL),position=position_dodge(width=0.25),width=0.15,color='gray54')+
   scale_fill_manual(values=cohortColors)+theme_ipsum(axis_text_size = 16,axis_col = "black")+
    geom_text(data=label.df,size=7,aes(label=label))+
      geom_line(data=arc.df,aes(phenotype+1,Estimate+2),lty=2)+ylab("Spatial interaction with MAC proportion (p<0.01)")+xlab("Ki67+TREG phenotype")+theme(axis.text=element_text(color='black'))+geom_hline(yintercept = 0,linetype="dashed",color='red')
  
 
  grid.arrange(p2,p3,p4,nrow=2)

    grid.arrange(p2,p3,nrow=2)

  ## save ggplot for Tucker
    save(p2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/p2.CD8.Figure5D.RData")
    save(p3,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/p3.CD4.Figure5D.RData")
  ## as a positive control  examine the PDL1+MAC and the PD1+ Tcell interactions. checking Shipp model


```



### Revision 12-17 comparing relative proportions
- 1-10: regenerates supplementary figure 15
```{r}

 ## TME relative proportion.
 ta4<-as.data.frame.matrix(table(joint$ROIID,joint$phenotype))
 ta4<-ta4[,c('CD4','CD8','Macrophage','TREG','Endothelial')]
 ta4<-100*ta4/rowSums(ta4)
  ta4$feature<-joint$experiment[match(rownames(ta4),joint$ROIID)]
  ta4$feature<-factor(ta4$feature,levels=c("RLN","DLBCL","HL"))
  summary(lm(Macrophage~feature,ta4))
    summary(lm(CD4~feature,ta4))
  summary(lm(CD8~feature,ta4))
  summary(lm(TREG~feature,ta4))



 joint$tim3.pdl1<-paste0(joint$TIM3,"_",joint$PDL1,"_",joint$phenotype)
 joint$tim3.pdl1.simple<-as.character(joint$tim3.pdl1)
  joint$tim3.pdl1.simple<-gsub("\\-","Neg.",joint$tim3.pdl1.simple)
 joint$tim3.pdl1.simple<-gsub("\\+","Pos.",joint$tim3.pdl1.simple)
  joint$tim3.pdl1.simple<-gsub(" ","",joint$tim3.pdl1.simple) 
  
  
 joint$tim3.simple<-as.character(paste0(joint$TIM3,joint$phenotype))
  joint$tim3.simple<-gsub("\\-","Neg.",joint$tim3.simple)
 joint$tim3.simple<-gsub("\\+","Pos.",joint$tim3.simple)
  joint$tim3.simple<-gsub(" ","",joint$tim3.simple)
  
  
  joint$pd1.simple<-as.character(paste0(joint$PD1,joint$phenotype))
  joint$pd1.simple<-gsub("\\-","Neg.",joint$pd1.simple)
 joint$pd1.simple<-gsub("\\+","Pos.",joint$pd1.simple)
  joint$pd1.simple<-gsub(" ","",joint$pd1.simple)
  
   joint$tim3.pd1<-paste0(joint$TIM3,"_",joint$PD1,"_",joint$phenotype)
 joint$tim3.pd1.simple<-as.character(joint$tim3.pd1)
  joint$tim3.pd1.simple<-gsub("\\-","Neg.",joint$tim3.pd1.simple)
 joint$tim3.pd1.simple<-gsub("\\+","Pos.",joint$tim3.pd1.simple)
  joint$tim3.pd1.simple<-gsub(" ","",joint$tim3.pd1.simple) 
  
   joint$ccr4.pd1<-paste0(joint$CCR4,"_",joint$PD1,"_",joint$phenotype)
 joint$ccr4.pd1.simple<-as.character(joint$ccr4.pd1)
  joint$ccr4.pd1.simple<-gsub("\\-","Neg.",joint$ccr4.pd1.simple)
 joint$ccr4.pd1.simple<-gsub("\\+","Pos.",joint$ccr4.pd1.simple)
  joint$ccr4.pd1.simple<-gsub(" ","",joint$ccr4.pd1.simple)
  
   joint$ki67<-paste0(joint$Ki67,"_",joint$phenotype)
 joint$ki67.simple<-as.character(joint$ki67)
  joint$ki67.simple<-gsub("\\-","Neg.",joint$ki67.simple)
 joint$ki67.simple<-gsub("\\+","Pos.",joint$ki67.simple)
  joint$ki67.simple<-gsub(" ","",joint$ki67.simple)
  
  
  
  ### JOINT TME ANALYSIS
  ## JOINT TME phenotypes gathered.
    tme<-subset(joint,joint$phenotype%in%c("CD4","CD8","Macrophage","TREG",'Endothelial'))
    ## ROI proportions (%)
  ta4<-as.data.frame.matrix(table(tme$ROIID,tme$tim3.pdl1.simple))
 ta4<-100*ta4/rowSums(ta4)
  ta4$feature<-joint$experiment[match(rownames(ta4),joint$ROIID)]
  ta4$feature<-factor(ta4$feature,levels=c("RLN","DLBCL","HL"))
  ## gathering only the macrophages
  M<-ta4[,grepl("Macrophage",colnames(ta4))]
  ## macrophage relative proportion.
  M<-100*M/rowSums(M)
  M$feature<-ta4$feature
  fit1a<-lm(TIM3Pos._PDL1Pos._Macrophage ~0+feature,M)
  fit1b<-lm(TIM3Pos._PDL1Pos._Macrophage ~feature,M)

  fit2a<-lm(TIM3Pos._PDL1Neg._Macrophage ~0+feature,M)
  fit2b<-lm(TIM3Pos._PDL1Neg._Macrophage ~feature,M)

  fit3a<-lm(TIM3.mid_PDL1Pos._Macrophage ~0+feature,M)
  fit3b<-lm(TIM3.mid_PDL1Pos._Macrophage ~feature,M)

  fit4a<-lm(TIM3.mid_PDL1Neg._Macrophage ~0+feature,M)
  fit4b<-lm(TIM3.mid_PDL1Neg._Macrophage ~feature,M)
  
  


  prepLinear<-function(fit=NULL,label='TIM3+PDL1+Macrophage',comparisonFit=NULL){
    sum<-as.data.frame(coef(summary(fit)))
    fit2<-comparisonFit
    diffp<-as.data.frame(coef(summary(fit2)))
    ci<-confint(fit)
    ci<-data.frame(Estimate=sum$Estimate,ci,label=label)
    rownames(ci)<-gsub("feature","",rownames(ci))
    ci$experiment<-rownames(ci)
    colnames(ci)[2:3]<-c("lower2.5","upper97.5")
    ci$p<-round(diffp[,'Pr(>|t|)'],3)
    return(ci)
  }
  
  a<-prepLinear(fit=fit1a,label='TIM-3+PD-L1+',comparisonFit=fit1b)
  b<-prepLinear(fit=fit2a,label='TIM-3+PD-L1-',comparisonFit=fit2b)
  c<-prepLinear(fit=fit3a,label='TIM-3 mid PD-L1+',comparisonFit=fit3b)
  d<-prepLinear(fit=fit4a,label='TIM-3 mid PD-L1-',comparisonFit=fit4b)
  dat<-rbind(a,b,c,d)
  dat$annotation<-""
  dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p>=0.1)]<-"NS"
  dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p<0.1)]<-"."
    dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p<0.05)]<-"*"
    dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p<0.01)]<-"**"


  threeColors2<-threeColors
  names(threeColors2)<-c("DLBCL","HL","RLN")
 ## add stars of significance
  #sigDat<-data.frame(x=c(0.5,1.5,2.5,3.5), xend=c(1.75, 2.75,3.75,4.75),
   #                           y=c(11, 20,15,35), annotation=c("*","**",".","."))

  library(ggsignif)
  ggplot(dat,aes(x=label,y=Estimate,fill=experiment))+geom_bar(stat='identity',position='dodge')+geom_errorbar(position=position_dodge(0.9),aes(ymin=lower2.5,ymax=upper97.5),width=0.5)+scale_fill_manual('Cohort',values=threeColors2)+
  geom_signif(stat="identity",
              data=data.frame(x=c(0.7,1.7,2.7,3.7), xend=c(1.25, 2.25,3.25,4.25),
                              y=c(13, 20,18,35),experiment="RLN", annotation=c("p=0.03","p<0.01","p=0.06","NS")),
              aes(x=x,xend=xend, y=y, yend=y, annotation=annotation))+theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme(axis.text.y=element_text(size=10,colour='black'))+ylab("Within macrophage relative TME proportion (%)")+ggtitle("Macrophage comparison to reactive lymph node")+xlab("Phenotype")

  
  ##TREG Ki+/- and PD1+/- CCR4+
   
  ## TREG
  pd1<-as.data.frame.matrix(table(tme$ROIID,tme$ccr4.pd1.simple))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   G<-pd1[,grepl("TREG",colnames(pd1))]
  ## TREG relative proportion.
  G<-100*G/rowSums(G)
  G$feature<-pd1$feature
 
  fit1a<-lm(CCR4Pos._PD1Pos._TREG ~0+feature,G)
  fit1b<-lm(CCR4Pos._PD1Pos._TREG ~feature,G)

   fit2a<-lm(CCR4Pos._PD1Neg._TREG ~0+feature,G)
  fit2b<-lm(CCR4Pos._PD1Neg._TREG ~feature,G)
  
  fit3a<-lm(CCR4Neg._PD1Pos._TREG ~0+feature,G)
  fit3b<-lm(CCR4Neg._PD1Pos._TREG ~feature,G)
  fit4a<-lm(CCR4Neg._PD1Neg._TREG ~0+feature,G)
  fit4b<-lm(CCR4Neg._PD1Neg._TREG ~feature,G)
  
  ki<-as.data.frame.matrix(table(tme$ROIID,tme$ki67.simple))
  ki<-100*ki/rowSums(ki)
   ki$feature<-joint$experiment[match(rownames(ki),joint$ROIID)]
  ki$feature<-factor(ki$feature,levels=c("RLN","DLBCL","HL"))
   G<-ki[,grepl("TREG",colnames(ki))]
  ## TREG relative proportion.
  G<-100*G/rowSums(G)
  G$feature<-ki$feature
  fit5a<-lm(Ki67Pos._TREG ~0+feature,G)
  fit5b<-lm(Ki67Pos._TREG ~feature,G)

  fit6a<-lm(Ki67Neg._TREG ~0+feature,G)
  fit6b<-lm(Ki67Neg._TREG ~feature,G)
  
  
  a<-prepLinear(fit=fit1a,label='CCR4+PD-1+',comparisonFit=fit1b)
  b<-prepLinear(fit=fit2a,label='CCR4+PD-1-',comparisonFit=fit2b)
  c<-prepLinear(fit=fit3a,label='CCR4- PD-1+',comparisonFit=fit3b)
  d<-prepLinear(fit=fit4a,label='CCR4- PD-1-',comparisonFit=fit4b)
  e<-prepLinear(fit=fit5a,label='Ki67+',comparisonFit=fit5b)
  f<-prepLinear(fit=fit6a,label='Ki67-',comparisonFit=fit6b)

  dat<-rbind(a,b,c,d,e,f)
  dat$annotation<-""
  dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p>=0.1)]<-"NS"
  dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p<0.1)]<-"."
    dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p<0.05)]<-"*"
    dat$annotation[which(dat$experiment%in%c("DLBCL","HL")&dat$p<0.01)]<-"**"

  library(ggsignif)
  ggplot(dat,aes(x=label,y=Estimate,fill=experiment))+geom_bar(stat='identity',position='dodge')+geom_errorbar(position=position_dodge(0.9),aes(ymin=lower2.5,ymax=upper97.5),width=0.5)+scale_fill_manual('Cohort',values=threeColors2)+
  geom_signif(stat="identity",
              data=data.frame(x=c(4,4.7,5.7), xend=c(4.35,5.25,6.25),
                              y=c(30,60,60),experiment="RLN", annotation=c("p=0.047","p=0.06","p<0.01")),
              aes(x=x,xend=xend, y=y, yend=y, annotation=annotation))+theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme(axis.text.y=element_text(size=10,colour='black'))+ylab("Within T-REG relative TME proportion (%)")+ggtitle("T-REG comparison to reactive lymph node")+xlab("Phenotype")+geom_vline(xintercept=4.5)

  
  
 
  ## CD4
  ta4<-as.data.frame.matrix(table(tme$ROIID,tme$tim3.pd1.simple))
 ta4<-100*ta4/rowSums(ta4)
  ta4$feature<-joint$experiment[match(rownames(ta4),joint$ROIID)]
  ta4$feature<-factor(ta4$feature,levels=c("RLN","DLBCL","HL"))
  ## gathering only the macrophages
  M<-ta4[,grepl("CD4",colnames(ta4))]
  ## macrophage relative proportion.
  M<-100*M/rowSums(M)
  M$feature<-ta4$feature
  summary(lm(TIM3Pos._PD1Pos._CD4 ~feature,M))
  summary(lm(TIM3Pos._PD1Neg._CD4 ~feature,M))
  summary(lm(TIM3.mid_PD1Pos._CD4 ~feature,M))
  summary(lm(TIM3.mid_PD1Neg._CD4 ~feature,M))

  
    
    mac<-subset(tme,tme$phenotype=='Macrophage')
     ta2<-as.data.frame.matrix(table(tme$ROIID,tme$tim3.simple))
  # ta2<-as.data.frame.matrix(table(joint$ROIID,joint$tim3.simple))
 ta2<-100*ta2/rowSums(ta2)
 ta2$feature<-joint$experiment[match(rownames(ta2),joint$ROIID)]
 ta2$feature<-factor(ta2$feature,levels=c("RLN","DLBCL","HL"))
 summary(lm(TIM3Neg.Macrophage~feature,ta2))
 summary(lm(TIM3Pos.Macrophage~feature,ta2))
 summary(lm(TIM3.midMacrophage~feature,ta2))
 
 summary(lm(TIM3Neg.CD4~feature,ta2))
 summary(lm(TIM3Pos.CD4~feature,ta2))
 summary(lm(TIM3.midCD4~feature,ta2))
 
 ##relative to only macs
  mac<-subset(joint,joint$phenotype=='Macrophage')
     ta5<-as.data.frame.matrix(table(mac$ROIID,mac$tim3.simple))
  # ta2<-as.data.frame.matrix(table(joint$ROIID,joint$tim3.simple))
 ta5<-100*ta5/rowSums(ta5)
 ta5$feature<-joint$experiment[match(rownames(ta5),joint$ROIID)]
 ta5$feature<-factor(ta5$feature,levels=c("RLN","DLBCL","HL"))
 
 summary(lm(TIM3Neg.Macrophage~feature,ta5))
 summary(lm(TIM3Pos.Macrophage~feature,ta5))
 summary(lm(TIM3.midMacrophage~feature,ta5))
 
 summary(lm(TIM3Neg.CD4~feature,ta2))
 summary(lm(TIM3Pos.CD4~feature,ta2))
 summary(lm(TIM3.midCD4~feature,ta2))
  
 joint$pdl1.simple<-as.character(paste0(joint$PDL1,joint$phenotype))
  joint$pdl1.simple<-gsub("\\-","Neg.",joint$pdl1.simple)
 joint$pdl1.simple<-gsub("\\+","Pos.",joint$pdl1.simple)
  joint$pdl1.simple<-gsub(" ","",joint$pdl1.simple)
  
   ta3<-as.data.frame.matrix(table(joint$ROIID,joint$pdl1.simple))
 ta3<-100*ta3/rowSums(ta3)
 ta3$feature<-joint$experiment[match(rownames(ta3),joint$ROIID)]
 ta3$feature<-factor(ta3$feature,levels=c("RLN","DLBCL","HL"))
  summary(lm(PDL1Neg.Macrophage~feature,ta3))
 summary(lm(PDL1Pos.Macrophage~feature,ta3))
 summary(lm(PDL1.midMacrophage~feature,ta3))
 
 ta<-as.data.frame.matrix(table(tme$ROIID,tme$tim3.pdl1.simple))
 ta<-100*ta/rowSums(ta)
 ta$feature<-joint$experiment[match(rownames(ta),joint$ROIID)]
 ta$feature<-factor(ta$feature,levels=c("RLN","DLBCL","HL"))
 
 summary(lm(TIM3Pos._PDL1Pos._Macrophage~feature,ta))
  summary(lm(TIM3Pos._PDL1.mid_Macrophage~feature,ta))
 summary(lm(TIM3Pos._PDL1Neg._Macrophage~feature,ta))

 summary(lm(TIM3.mid_PDL1Neg._Macrophage~feature,ta))
 summary(lm(TIM3Pos._PDL1.mid_Macrophage~feature,ta))

 
 ta<-as.data.frame.matrix(table(mac$ROIID,mac$tim3.pdl1.simple))
 ta<-100*ta/rowSums(ta)
 ta$feature<-joint$experiment[match(rownames(ta),joint$ROIID)]
 ta$feature<-factor(ta$feature,levels=c("RLN","DLBCL","HL"))
 
 
 
 ## TREG
  treg<-subset(joint,joint$phenotype=='TREG')
     ta6<-as.data.frame.matrix(table(treg$ROIID,treg$tim3.simple))
  # ta2<-as.data.frame.matrix(table(joint$ROIID,joint$tim3.simple))
 #ta6<-100*ta5/rowSums(ta6)
 #ta6$feature<-joint$experiment[match(rownames(ta6),joint$ROIID)]
 #ta6$feature<-factor(ta6$feature,levels=c("RLN","DLBCL","HL"))
 
# summary(lm(TIM3Neg.Macrophage~feature,ta5))
# summary(lm(TIM3Pos.Macrophage~feature,ta5))
# summary(lm(TIM3.midMacrophage~feature,ta5))
```

### DLBCL pre-integration processing
- Shows the standardized values and performs PCA on both TME at single cell level.
- 12-4 revision path update for windows local reproduction regenerates supplementary figure 13
```{r dlbclPreIntegrate,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 12, fig.width = 15}

  # load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/Box-RData/Figure4/original-paper-data/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")
  ## DLBCL Z-score.
     zz<- zScorePatientExpression(full.dn4=full.dn4,markers=c(1:31,104:111),ROIID.column=32)

   myMarkerList<-setUpMarkers()

  expr<-full.dn4[,c(myMarkerList[["lineage_markers"]],myMarkerList[["induc"]],"response","treatment.status","uniqueLabel","ROIID")]
  expr<-expr[which(expr$treatment.status!="ND"),]
  expr.ref<-expr[which(expr$response=="Primary refract"),]
  expr.cr<-expr[which(expr$response=="CR"),]
  labels<-expr[,"uniqueLabel"]

   ##new way of sampling.
  tsne_inds<-sapply(unique(expr.cr$ROIID),function(x) randomSampleUniqueLabel(which.ROIID=x,N=1250,full.dn4=expr.cr))
   tsne_inds<-as.vector(tsne_inds)
  ## refractory samples
  tsne_inds_ref<-sapply(unique(expr.ref$ROIID),function(x) randomSampleUniqueLabel(which.ROIID=x,N=4532,full.dn4=expr.ref))
  tsne_inds_ref<-as.vector(tsne_inds_ref)
  randomSamps<-c(tsne_inds,tsne_inds_ref)



  ## single cell integration. 
dlbcl.sub<-zz[match(randomSamps,zz$uniqueLabel),c("Cell_CCR4",
	"Cell_CD206","Cell_CD20","Cell_CD30","Cell_CD3",
	"Cell_CD45RA","Cell_CD45RO","Cell_CD4","Cell_CD68","Cell_CD8",
	"Cell_CXCR3","Cell_DNA","Cell_DNA2","Cell_EphrinB2","Cell_FOXP3","Cell_HLADR",
	"Cell_ICOS","Cell_Ki67","Cell_Lag3","Cell_PD1","Cell_PDL1","Cell_Tim3",
	"Cell_Tbet","ROIID","uniqueLabel")]
  colnames(dlbcl.sub)<-gsub("Cell_","",colnames(dlbcl.sub))
  colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="Lag3")]<-"LAG3"
  colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="Tim3")]<-"TIM3"
   colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="DNA")]<-"DNA1"
   colnames(dlbcl.sub)[which(colnames(dlbcl.sub)=="CD8")]<-"CD8a"
   dlbcl.sub$experiment<-zz$response[match(dlbcl.sub$uniqueLabel,zz$uniqueLabel)]
  ## merging.
  ## Z-transforms both complete experiments
   integ<-integrateHD.DLBCL(hd=NULL,dlbcl=NULL)
   integ<-integ[which(!is.nan(integ$DNA1)),]
   integ<-integ[which(!is.nan(integ$DNA2)),]
   integ<-left_join(integ,full.dn4[,c("treatment.status","response","uniqueLabel")],by="uniqueLabel")
   integ$batch<-"HD"
   integ$batch[which(integ$response=="CR" | integ$response=="Primary refract")]<-"DLBCL"
   

  ## plot the distributions.

   plotIntegrationPlots(integ=integ)


  ### perform the tsne
   ## sample balanced in each experiment
  ## sample balanced 
   ## both experiments together.
  ## after pooling them together.
   hd.rln<-integ[which(integ$experiment!="DLBCL"),]
   hd.rln<-hd.rln[grepl("Case",hd.rln$ROIID),]
  ## sample 4,200 cells per ROIID per experiment.
  # then sample 140,000 cells balanced across DLBCL response.
   hd_ids<-sapply(unique(hd.rln$ROIID),function(x) randomSampleUniqueLabel(which.ROIID=x,N=3500,full.dn4=hd.rln))
   hd_ids<-as.vector(hd_ids)
 

     
  
   hd.rln2<-hd.rln[match(hd_ids,hd.rln$uniqueLabel),c("CCR4",
	"CD206","CD20","CD30","CD3","CD45RA","CD45RO","CD4","CD68",
	"CD8a","CXCR3","DNA1","DNA2","EphrinB2","FOXP3","HLADR",
	"ICOS","Ki67","LAG3","PD1","PDL1","TIM3","Tbet","ROIID","uniqueLabel")]
	hd.rln2$experiment<-dn4$diagnosis[match(hd.rln2$uniqueLabel,dn4$uniqueLabel)]

 all(colnames(dlbcl.sub)==colnames(hd.rln2))
     
     integ.sub<-rbind(dlbcl.sub,hd.rln2)
      integ.sub$batch<-"HD"
      integ.sub$batch[which(integ.sub$response=="CR" | integ.sub$response=="Primary refract")]<-"DLBCL"
   
  ## use Seurat to plot single cell tSNE.
  library(Seurat)
  pbmc<- seuratDownSampleIntegrationEvaluation(downSample=TRUE)
  ##this shows 
  p1 <- DimPlot(pbmc, reduction = "pca", group.by = "tech")
  p1

```


### Combine both disease types
 The joint TME experiment contains 1,397,176 cells from both experiment.
  REVISION: 11-1 
```{r limma, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 16}

  require(limma)
  dim(integ)
  annos<-data.frame(annotation=hd.z$annotation,label=hd.z$uniqueLabel)
  annos2<-data.frame(annotation=full.dn4$subType.correction,label=full.dn4$uniqueLabel)
  annos<-rbind(annos,annos2)
  annos2<-NULL
  annos<-annos[match(integ$uniqueLabel,annos$label),]

  integ$annotation='none'
  integ$annotation[match(annos$label,integ$uniqueLabel)]<-as.character(annos$annotation)
 library(kableExtra)
 dim(integ)%>%
    kable() %>%
  kable_styling()
```
 
### Meta-cluster the joint TME experiment
Slow, runs the meta-clustering.
```{r, eval=FALSE}
 ## meta-cluster the experiment.
 ### meta-cluster both experiments.
 ##slow.
 library(igraph)
  set.seed(1234)
 ## this performs case independent phenograph
  ### includes meta-cluster with k=15
  integ2<-metaClusterExperiment(dn4=integ,
	markers=c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3"),
	plotPCA=FALSE)
    #save(integ2,file="./integ2.casePheno.k45.metaClust.k45.RData")
```

### Batch correct the joint TME 
- Revision 11-4: for the expression boxplots after normalization, we include the RLN as the 3rd factor of comparing expression.
- 12-4 local windows path reproduction
```{r batchAdjust, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 16}
 library(Rphenograph)
library(igraph)
 #load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/back-up-data-reproduced/integ2.casePheno.k45.metaClust.k45.RData")
 #  load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/back-up-data-reproduced/integ.metaclustered.jointTME.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/back-up-data-reproduced/integ.metaclustered.jointTME.RData") 	
integ2<-integ 
 
     integ2<-plotMeta(dn4=integ2[,which(colnames(integ2)!="phenograph_cluster_meta")])
    ## the original time I metaclustered used k2=45 for 10 MCs.
  ErikMeta=integ2[,
	c(c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3"),
	"ROIID",
	"phenograph_cluster_meta")] %>% group_by(ROIID,phenograph_cluster_meta) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
  
   ## plot the original cluster numbers.   #plot_clustering_heatmap_wrapper(expr=ErikMeta[,c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3")], 
 # cell_clustering=ErikMeta[,"phenograph_cluster_meta"],
 # useMedians=FALSE,
 # useQuantiles=TRUE,
 # color_clusters=NULL)

 ### match the reproduced analysis to the original.
  ## the cluster numbers in the reproduction are randomly assigned.
	integ2$phenograph_cluster_meta2<-integ2$phenograph_cluster_meta
     integ2[which(integ2$phenograph_cluster_meta==2),"phenograph_cluster_meta2"]<-3
     integ2[which(integ2$phenograph_cluster_meta==3),"phenograph_cluster_meta2"]<-4
     integ2[which(integ2$phenograph_cluster_meta==1),"phenograph_cluster_meta2"]<-2
     integ2[which(integ2$phenograph_cluster_meta==4),"phenograph_cluster_meta2"]<-9
     integ2[which(integ2$phenograph_cluster_meta==9),"phenograph_cluster_meta2"]<-6
     integ2[which(integ2$phenograph_cluster_meta==5),"phenograph_cluster_meta2"]<-1
     integ2[which(integ2$phenograph_cluster_meta==6),"phenograph_cluster_meta2"]<-7
     integ2[which(integ2$phenograph_cluster_meta==7),"phenograph_cluster_meta2"]<-5

 ## filtering
  ## filter dim CD68 from mc 7
  id7<-which(integ2$phenograph_cluster_meta2==7 & integ2$CD68<(0.15))
  integ2<-integ2[-id7,]
    id9<-which(integ2$phenograph_cluster_meta2==9 & integ2$CD20<(-0.3))
 integ2<-integ2[-id9,]
  id6<-which(integ2$phenograph_cluster_meta2==6 & integ2$CD20<(-0.3))
 integ2<-integ2[-id9,]

   integ2[which(integ2$phenograph_cluster_meta2==7),"phenograph_cluster_meta2"]<-1


  ErikMeta=integ2[,
	c(c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3"),
	"ROIID",
	"phenograph_cluster_meta2")] %>% group_by(ROIID,phenograph_cluster_meta2) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
   	  plot_clustering_heatmap_wrapper(expr=ErikMeta[,c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3")], 
  cell_clustering=ErikMeta[,"phenograph_cluster_meta2"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)

 ## annotate
   anno<-c("Macrophage","HRS","TREG","BCell","CD4","BCell","CD8","BCell","BCell")
   anno<-data.frame(anno)
   anno$cluster.id<-c(1,2,3,4,5,6,8,9,10)
  integ2$annotation<-anno[match(integ2$phenograph_cluster_meta2,anno$cluster.id),"anno"]

  ##filter out ND from DLBCL
   nd.id<-integ2$uniqueLabel[which(integ2$treatment.status=="ND")]
    ltf.id<-integ2$uniqueLabel[which(integ2$response=='LTF')]

    integ2<-integ2[which(!integ2$uniqueLabel%in%c(nd.id,ltf.id)),]
    integ2<-integ2[which(!integ2$ROIID %in%c("Placenta_1_3_a0","Placenta_2_4_a0","Tonsil_1_1_a0","Tonsil_2_2_a0")),]


 ## INTEG2 has the RLN, but drops ND and tonsil, and placenta.
    
  ### ANOVA comparing cluster differences.
   immune<-integ2[integ2$annotation%in%c("CD4","CD8","TREG","Macrophage"),]
   rln.id<-c("Case1_ROI_01_5_a0",
	      "Case1_ROI_02_6_a0",
	"Case2_ROI_01_7_a0",
	     "Case3_ROI_01_8_a0",
	"Case4_ROI_02_9_a0",
	    "Case5_ROI_01_10_a0")


  ## plot the TME only.
 plot_clustering_heatmap_wrapper(expr=immune[,c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3")], 
  cell_clustering=immune[,"annotation"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 


 # load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/integ2.metacluster.HD.DLBCL.RLN.RData")

  ## omit ambiguous cluster from HL
 ## Cluster 7 was an ambiguous cluster. 
  ## so we filter out the lowly expressed cells from the analysis.
  ### this will increase the phenotype accuracy, but add errors in the spatial analysis.
  ## the alternative is to keep phenotype ambiguity, with accurate spatial results.
  ## either way, there is a trade off. 
  hd.omit.id<-hd.z$uniqueLabel[which(hd.z[which(hd.z$phenograph_cluster_meta==7),"CD4"]<=(-0.7) |hd.z[which(hd.z$phenograph_cluster_meta==7),"CD3"]<=(-0.7))]
   table(hd.omit.id%in%integ2$ROIID)
 #  integ2<-integ2[!integ2$uniqueLabel%in%hd.omit.id,]

  

 ### use ComBat to batch correct
  library(sva)
   mod<-model.matrix(~annotation,integ2)
   batch<-integ2$experiment
   test<-integ2[,1:24]
	rownames(test)<-integ2$uniqueLabel
  test<-t(test)
   cleandat <- ComBat(test,batch,mod)
    cleandat<-t(cleandat)
   cleandat<-as.data.frame(cleandat)
    cleandat$experiment<-integ2$experiment
     cleandat$ROIID<-integ2$ROIID
      cleandat$annotation<-integ2$annotation
         cleandat$uniqueLabel<-integ2$uniqueLabel
     test<-cleandat%>%group_by(ROIID,annotation)%>%summarise_all(funs(mean))%>%data.frame
	test$experiment<-integ2$experiment[match(test$ROIID,integ2$ROIID)]
	##append the RLN
	test$experiment<-as.character(test$experiment)
	 test$experiment[which(cleandat$normalLymphnode[match(test$ROIID,cleandat$ROIID)]=='RLN')]<-'RLN'
  
	  
	 
	 
	test%>%reshape2::melt()%>%ggplot(aes(x=annotation,y=asinh(value),fill=experiment))+geom_boxplot()+facet_wrap(~variable)+theme_bw(base_size=14)+theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_fill_manual(values=threeColors)

  ## append clinical data.
  cleandat$response<-integ2$response[match(cleandat$uniqueLabel,integ2$uniqueLabel)]
  cleandat$clinical.type<-"NA"
  cleandat$clinical.type<-as.character(dn4$diagnosis[match(cleandat$uniqueLabel,dn4$uniqueLabel)])
    cleandat$clinical.type[which(is.na(cleandat$clinical.type))]<-as.character(full.dn4$COO[match(cleandat$uniqueLabel[which(is.na(cleandat$clinical.type))],full.dn4$uniqueLabel)])
 cleandat$clinical.type[which(cleandat$clinical.type=="RLN1" |cleandat$clinical.type=="RLN2" |  cleandat$clinical.type=="RLN3" | cleandat$clinical.type=="RLN4" | cleandat$clinical.type=="RLN5" )] <-"RLN" 

  ### create a column for RLN.
   cleandat$normalLymphnode<-"disease"
   cleandat$normalLymphnode[cleandat$ROIID%in%c("Case1_ROI_01_5_a0","Case1_ROI_02_6_a0","Case2_ROI_01_7_a0","Case4_ROI_02_9_a0","Case5_ROI_01_10_a0","Case3_ROI_01_8_a0")]<-"RLN"
 

    ##Append the caes information per each ROIID.
 cleandat$normalLymphnode<-factor(cleandat$normalLymphnode,levels=c("RLN","disease"))
 cleandat$case<-"NA"
 id<-which(cleandat$experiment=='DLBCL')
 cleandat$case[id]<-as.character(full.dn4$case[match(cleandat$uniqueLabel[id],full.dn4$uniqueLabel)])
  id<-which(cleandat$experiment!='DLBCL')

  cleandat$case[id]<-as.character(dn4$case[match(cleandat$uniqueLabel[id],dn4$uniqueLabel)])

 
```



### Test for batch effect evaluation
using K-nearest neighbor batch effect test we can test for batch effects quantitatively.  The p-value >0.05 from the kBET test indicate homogeneity of expression.  The ANOVA p-value <0.05 indicate significant heterogeneity of expression explained by the phenotype level (MAC,TREG,CD4,CD8) compared to the disease explained variance (DLBCL, or HL). 
- 1-10: reproduces Figure S12 for Kbet result, the heatmap is a little noisy.
```{r kBET,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 16}
   require(limma)
   require(kBET)

 #load the original clean dat data.
  #load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.annotated.July29.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.annotated.July29.RData")
  ## loads the original data
 ## using cleandat.limma object
  immune<-cleandat[cleandat$annotation%in%c("CD4","CD8","Macrophage","TREG"),]
    ### create a column for RLN.
   immune$normalLymphnode<-"disease"
     immune$normalLymphnode[immune$ROIID%in%c("Case1_ROI_01_5_a0",
	"Case1_ROI_02_6_a0",
	"Case2_ROI_01_7_a0","Case4_ROI_02_9_a0","Case5_ROI_01_10_a0","Case3_ROI_01_8_a0")]<-"RLN"
  immune$normalLymphnode<-factor(immune$normalLymphnode,levels=c("RLN","disease"))
  immune$annotation<-factor(immune$annotation)
 immune$annotation<-droplevels(immune$annotation)
  
  plot_clustering_heatmap_wrapper(expr=immune[,c("CD20","CD30","CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","EphrinB2","FOXP3","DNA1","DNA2","Area")], 
  cell_clustering=immune[,"annotation"],
  useMedians=TRUE,
  useQuantiles=TRUE,
  color_clusters=NULL)

  X<-asinh(immune[,c(1:22,24)])

   X<-t(X)
  rv<-removeBatchEffect(X,batch=as.character(immune$experiment),batch2=immune$ROIID)
  rv<-as.data.frame(t(rv))
  rv$ROIID<-immune$ROIID
  rv$experiment<-immune$experiment
   rv$annotation<-immune$annotation
  all(rownames(rv)==rownames(immune))
  immune[rownames(rv),colnames(rv)]<-rv
  immune$annotation<-as.character(immune$annotation)
  table(immune$experiment)%>%print


 #############################################
 
#### TEST FOR BATCH EFFECTS KBET
 ## perform kBET to check integration quality.
  library(kBET)
  # subset_size <- 0.01 #subsample to 10% of the data
    ##new way of sampling.
  tsne_inds<-sapply(unique(immune$ROIID[which(immune$experiment=="HD")]),
	function(x) randomSampleUniqueLabel(which.ROIID=x,N=400,full.dn4=immune[which(immune$experiment=="HD"),]))
   tsne_inds<-as.vector(tsne_inds)
  ## refractory samples
  tsne_inds_ref<-sapply(unique(immune$ROIID[which(immune$experiment!="HD")]),
	function(x) randomSampleUniqueLabel(which.ROIID=x,N=432,full.dn4=immune[which(immune$experiment!="HD"),]))
  tsne_inds_ref<-as.vector(tsne_inds_ref)
  randomSamps<-c(tsne_inds,tsne_inds_ref)

  data<-immune[match(randomSamps,immune$uniqueLabel),]
  
  
 
      batch<-ifelse(data$experiment=="HD",1,0)
   pca.data <- prcomp(data[,c("CCR4","CD3","CD45RA","CD45RO","CD4",
	"CD8a","CXCR3","PDL1","PD1","TIM3","CD68","FOXP3","LAG3")], center=TRUE) 
  
   #compute PCA representation of the data
  # batch.silhouette <- batch_sil(pca.data, batch)
  # batch.pca <- pcRegression(pca.data, batch)
  #   batch.data<-data.frame(maxSigPC=batch.pca$msigPC,
#			maxExplained.Var=batch.pca$maxVar,
#			maxCorr=batch.pca$maxCorr,
#			samplePercent=nrow(data)/nrow(immune),
#			batch.pca$r2)
 #  #write.csv(batch.data,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/integration-QC/immune-stratified-limma-norm/kBET.results.0.05.downsample.csv")



### TEST FOR BATCH EFFECTS USING ANOVA
  data<-immune%>%group_by(ROIID,experiment,annotation)%>%summarise(
	CCR4=mean(CCR4),
	CD3=mean(CD3),
	CD45RA=mean(CD45RA),
	CD45RO=mean(CD45RO),
	CD4=mean(CD4),	
	CD8=mean(CD8a),
	CXCR3=mean(CXCR3),
	DNA1=mean(DNA1),
	PDL1=mean(PDL1),
	PD1=mean(PD1),
	TIM3=mean(TIM3),
	CD68=mean(CD68),
	#LAG3=mean(LAG3),
	#Ki67=mean(Ki67),
	### 1-3
	CD206=mean(CD206),
	FOXP3=mean(FOXP3))%>%data.frame
 
 # library(xlsx)
 colnames(data)[which(colnames(data)=="annotation")]<-"phenotype"

 ## all values are constant.
  data2<-immune%>%group_by(ROIID,experiment)%>%summarise(
	CCR4=mean(CCR4),
	CD3=mean(CD3),
	CD45RA=mean(CD45RA),
	CD45RO=mean(CD45RO),
	CD4=mean(CD4),	
	CD8=mean(CD8a),
	CXCR3=mean(CXCR3),
	DNA1=mean(DNA1),
	PDL1=mean(PDL1),
	PD1=mean(PD1),
	TIM3=mean(TIM3),
	CD68=mean(CD68),
	#LAG3=mean(LAG3),
	#Ki67=mean(Ki67),
		CD206=mean(CD206),
	FOXP3=mean(FOXP3))%>%data.frame

 for(i in colnames(data)[c(-1,-2,-3)]){
  data[,i]<-data[,i]-data2[1,i] 
  }


  batch<-ifelse(data$experiment=="HD",1,2)
  clusters<-unique(data$phenotype)
  kBET_result_list <- list()
  sum_kBET <- 0
  for (cluster_level in unique(clusters)){
   batch_tmp <- batch[clusters == cluster_level]
   data_tmp <- data[clusters == cluster_level,c(-1,-2,-3)]
   kBET_tmp <- kBET(df=data_tmp, batch=batch_tmp, plot=FALSE)
   kBET_result_list[[cluster_level]] <- kBET_tmp
   sum_kBET <- sum_kBET + kBET_tmp$summary$kBET.observed[1]
}
  #averaging
  mean_kBET = sum_kBET/length(unique(clusters))
  means<-sapply(kBET_result_list,function(x) x$average.pval)
  sds<-sapply(kBET_result_list,function(x) sd(x$results[,2]))
  error <- qt(0.975,df=75-1)*sds/sqrt(75)
  left <- means-error
  right <- means+error
  kbetResults<-data.frame(rejection.rate=means,min=left,max=right)
  kbetResults$phenotype<-rownames(kbetResults)

  
   library(hrbrthemes)
   ggplot(kbetResults,aes(x=phenotype,y=rejection.rate))+geom_point(size=5)+geom_errorbar(aes(ymin=min,ymax=max),width=0.2)+scale_y_continuous(limits=c(0,1))+geom_hline(yintercept=0.05,col='red',linetype='dashed')+geom_rect( aes(xmin=0,xmax=5,ymin=0,ymax=0.05),fill='grey',alpha=0.5)+theme_ipsum(base_family="Arial")+ylab("p-value")+ggtitle("Quantification of batch effect")
 
kbet1<-ggplot(kbetResults,aes(x=phenotype,y=rejection.rate))+geom_point(size=5)+geom_errorbar(aes(ymin=min,ymax=max),width=0.2)+scale_y_continuous(limits=c(0,1))+geom_hline(yintercept=0.05,col='red',linetype='dashed')+geom_rect( aes(xmin=0,xmax=5,ymin=0,ymax=0.05),fill='grey',alpha=0.5)+theme_ipsum(base_family="Arial")+ylab("p-value")+ggtitle("Quantification of batch effect")
 
#   ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/integration-QC/immune-stratified-limma-norm/kBET.batchEffect.test.png")


  means<-sapply(kBET_result_list,function(x) mean(x$summary$kBET.observed))
  sds<-sapply(kBET_result_list,function(x) sd(x$summary$kBET.observed))
  error <- qt(0.975,df=75-1)*sds/sqrt(4)
  left <- means-error
  right <- means+error
  kbetResults<-data.frame(kBET=means,min=left,max=right)
  kbetResults$phenotype<-rownames(kbetResults)


 

   ggplot(kbetResults,aes(x=phenotype,y=kBET))+geom_point(size=5)+geom_errorbar(aes(ymin=min,ymax=max),width=0.2)+scale_y_continuous(limits=c(0,1))+theme_ipsum(base_family="Arial")+ylab("kBET score")+ggtitle("Quantification of batch effect")

   # ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/integration-QC/immune-stratified-limma-norm/kBET.scores.test.png")

 ## load the ANOVA workbook and write the KBET results there.
 



  results<-data.frame(CD4=c(kBET_result_list[[1]]$result[,2],
	CD8=kBET_result_list[[2]]$result[,2],
	MAC=kBET_result_list[[3]]$result[,2],
	TREG=kBET_result_list[[4]]$result[,2]))
  #write.csv(results,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/integration-QC/immune-stratified-limma-norm/kBET.results.test.csv")



 ## explained variance.
  ##REVISION: adding RLN to the factors
   immune$experiment_RLN<-as.character(immune$experiment)
   immune$experiment_RLN[which(immune$normalLymphnode=='RLN')]<-'RLN'
   
  
   data<-immune%>%group_by(experiment_RLN,annotation)%>%summarise(
	CCR4=mean(CCR4),
	CD3=mean(CD3),
	CD45RA=mean(CD45RA),
	CD45RO=mean(CD45RO),
	CD4=mean(CD4),	
	CD8=mean(CD8a),
	CXCR3=mean(CXCR3),
	DNA1=mean(DNA1),
	PDL1=mean(PDL1),
	PD1=mean(PD1),
	TIM3=mean(TIM3),
	CD68=mean(CD68),
	#LAG3=mean(LAG3),
	Ki67=mean(Ki67),
	FOXP3=mean(FOXP3))%>%data.frame
  library(xlsx)
 colnames(data)[which(colnames(data)=="annotation")]<-"phenotype"
 colnames(data)[1]<-'experiment'
  

  anovas<-data.frame(p.values= c(anova(lm(CCR4~experiment+phenotype,data))[1:2,5],
	anova(lm(CXCR3~experiment+phenotype,data))[1:2,5],
	anova(lm(PDL1~experiment+phenotype,data))[1:2,5],
	anova(lm(TIM3~experiment+phenotype,data))[1:2,5],
	anova(lm(PD1~experiment+phenotype,data))[1:2,5]),
	#anova(lm(Ki67~experiment+phenotype,data))[1:2,5]),
	feature=rep(c("experiment","phenotype"),5),
	phenotype=c(rep("CCR4",2),
	rep("CXCR3",2),
	rep("PDL1",2),
	rep("TIM3",2),
	rep("PD1",2)))
	#rep("Ki67",2))
  ggplot(anovas,aes(x=phenotype,y=p.values,color=feature))+geom_point(size=4)+geom_hline(yintercept=0.05,col='pink2',linetype='dashed')+theme_ipsum(base_family="Arial")+scale_color_manual(values=c("black","red"))+ylab("ANOVA p-value")+ggtitle("Expression variability explained by experiment/batch or phenotype")
  # ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/integration-QC/immune-stratified-limma-norm/ANOVA.average.expression.phenotype.experiment.png")





```

### Write kBET to table
```{r,eval=FALSE}
library("xlsx")
   write.xlsx(kbetResults, file = "kbet.rejection.xlsx",
      sheetName = "average.pval", append = FALSE)
 write.xlsx(kbetResults, file = "kbet.rejection.xlsx",
      sheetName = "kBET.scores", append = TRUE)
 

```


### Bar plot comparisons of the TME
 Now we compare the abundances of each TME component.  The PCA plots the ROI, whereas the Figure4 depicts all the sub-communities.
 - Revision 1-3: saving the barplot ggplot file for final figure
 - 1-10: reproduces final 5A,  and the TME barplot (supplementary).
```{r abundance,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 16}
 ## let RLN be the baseline reference.
  ## load immune normalized
 load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.annotated.July29.RData")
  ## loads the original data
 ## using cleandat.limma object
  immune<-cleandat.limma[cleandat.limma$annotation%in%c("CD4","CD8","Macrophage","TREG"),]
    ### create a column for RLN.
   immune$normalLymphnode<-"disease"
     immune$normalLymphnode[immune$ROIID%in%c("Case1_ROI_01_5_a0",
	"Case1_ROI_02_6_a0",
	"Case2_ROI_01_7_a0","Case4_ROI_02_9_a0","Case5_ROI_01_10_a0","Case3_ROI_01_8_a0")]<-"RLN"
  immune$normalLymphnode<-factor(immune$normalLymphnode,levels=c("RLN","disease"))
  immune$annotation<-factor(immune$annotation)
 immune$annotation<-droplevels(immune$annotation)
 immune$experiment_RLN<-as.character(immune$experiment)
 immune$experiment_RLN[which(immune$normalLymphnode=="RLN")]<-"RLN"
 immune$experiment_RLN<-factor(immune$experiment_RLN,levels=c("RLN","DLBCL","HD"))
 ### barplot
 ## barplot of primary phenotypes comparing HD vs DLBCL.
  integratedPrimaryCommunityBarPlot(cleandat=immune,experiment='experiment_RLN')
  immune$subannotation2<-immune$annotation
  
  ## saving the image for Tucker and final print out
  tmeBar<-integratedPrimaryCommunityBarPlot(cleandat=immune,experiment='experiment_RLN')
  
  save(tmeBar,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/tmeBar_Figure5.RData")
 ######################################################
  
 ## REVISION 11-1: we include the RLN
# community<-immune[which(immune$normalLymphnode!='RLN'),]%>%group_by(ROIID,subannotation2)%>%dplyr::summarise(TIM3=mean(TIM3),
  community<-immune%>%group_by(ROIID,subannotation2)%>%dplyr::summarise(TIM3=mean(TIM3),
  	CD45RO=mean(CD45RO),
	CD45RA=mean(CD45RA),
	DNA1=mean(DNA1),
	DNA2=mean(DNA2),
	Area=mean(Area),
	HLADR=mean(HLADR),
	PD1=mean(PD1),
	PDL1=mean(PDL1),
	CCR4=mean(CCR4),
	CXCR3=mean(CXCR3))%>%data.frame

    PC<-prcomp(community[,c(-1,-2)],scale=TRUE,center=TRUE)
    PCi<-data.frame(PC$x,ROIID=community$ROIID,annotation=community$subannotation2)
  #  PCi$experiment<-immune$experiment[match(PCi$ROIID,immune$ROIID)]
      PCi$experiment<-immune$experiment_RLN[match(PCi$ROIID,immune$ROIID)]
 pca.perc<-100*round(PC$sd/sum(PC$sd),3)
   threeColors<-getExperimentColors()
    threeColors<-c(threeColors,'lightcoral')
    names(threeColors)[3]<-'RLN'
 
 experiment.states.pca<- ggplot(PCi,aes(x=PC1,y=PC2,col=experiment,label=ROIID))+geom_point(color="black",size=3.5,alpha=0.95)+geom_point(aes(fill=experiment,col=experiment),size=2.8)+xlim(c( min(PCi$PC1)-0.5,  max(PCi$PC1)+0.05))+scale_color_manual(values=threeColors)+theme_classic()+ xlab(paste0("PC1 (",pca.perc[1],"%)"))+ylab(paste0("PC2 (",pca.perc[2],"%)"))
 print(experiment.states.pca)

  integratedColors<-getIntegratedColors(immune)
 #### PCA with the color coded sub-community.
   experiment.states.pca2<- ggplot(PCi,aes(x=PC1,y=PC2,col=annotation))+geom_point(color="black",size=3.5,alpha=0.95)+geom_point(aes(fill=annotation,col=annotation),size=2.8)+xlim(c( min(PCi$PC1)-0.5,  max(PCi$PC1)+0.05))+
	theme_classic()+ xlab(paste0("PC1 (",pca.perc[1],"%)"))+ylab(paste0("PC2 (",pca.perc[2],"%)"))+scale_color_manual(values=c("blue","green","magenta","purple"))
 print(experiment.states.pca2)
  library(gridExtra)
 grid.arrange(experiment.states.pca,experiment.states.pca2,ncol=2)
  #  savePlot(file="immune.states.integration.PCA.subcommunity.png")
  
#save(experiment.states.pca,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/experiment.stats.pca_Figure5.RData")
 
#save(experiment.states.pca2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/experiment.stats.pca2_Figure5.RData")
```


### Expression analysis of integrated data (Final 5B)
 The distributions show that TIM3, PD-1, and PD-L1 have uniform distribution across disease type.  The integrated boxplots show that Ki67, and LAG-3 did not integrate well, whereas TIM-3, DNA, and PD1/PDL1 did at the patient level.   Across phenotypes the boxplots show differential expression.4
 - 1-3: reproduced figure 5B
```{r exprAnalysis, eval=TRUE,warning=FALSE,message=FALSE,fig.height = 16, fig.width = 18}
 ## MakeJoy Plots only of the TME.

 #load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/immune.integrated.ComBat.limma.BatchCorrected.immuneStratified.RData")

#### cleand dat limma object
   load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.annotated.July29.RData")
  ## loads the original data
 ## using cleandat.limma object
  immune<-cleandat.limma[cleandat.limma$annotation%in%c("CD4","CD8","Macrophage","TREG"),]
    ### create a column for RLN.
   immune$normalLymphnode<-"disease"
     immune$normalLymphnode[immune$ROIID%in%c("Case1_ROI_01_5_a0",
	"Case1_ROI_02_6_a0",
	"Case2_ROI_01_7_a0","Case4_ROI_02_9_a0","Case5_ROI_01_10_a0","Case3_ROI_01_8_a0")]<-"RLN"
  immune$normalLymphnode<-factor(immune$normalLymphnode,levels=c("RLN","disease"))
  immune$annotation<-factor(immune$annotation)
 immune$annotation<-droplevels(immune$annotation)
   immune$experiment_RLN<-as.character(immune$experiment)
   immune$experiment_RLN[which(immune$normalLymphnode=='RLN')]<-'RLN'
 


  require(ComplexHeatmap)
  plotIntegrationPlots.IMMUNE(integ=immune)
  
  immune$Area.norm<-immune$Area
  integratedBoxPlotsBatchEvaluation(cleandat=immune)
  ## normalize by Area
 
 



  ### Integrated comparision of PD1/PDL1/TIM3
    test<-immune[,c("PDL1","PD1","TIM3","ROIID","annotation")]%>%group_by(ROIID,annotation)%>%summarise_all(funs(mean))%>%data.frame
    exCol<-getExperimentColors()
      threeColors<-getExperimentColors()
    threeColors<-c(threeColors,'indianred3')
    names(threeColors)[3]<-'RLN'
     names(threeColors)[which(names(threeColors)=="HD")]<-'HL'

 immune$experiment_RLN<-as.character(immune$experiment_RLN)
 immune$experiment_RLN[which(immune$experiment_RLN=="HD")]<-"HL"
 immune$experiment_RLN<-factor(immune$experiment_RLN,levels=c("DLBCL","HL","RLN"))

	test$experiment<-immune$experiment_RLN[match(test$ROIID,immune$ROIID)]
	test%>%reshape2::melt()%>%ggplot(aes(x=annotation,y=value,fill=experiment))+geom_boxplot()+facet_wrap(~variable)+theme_bw(base_size=14)+theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_fill_manual(values=threeColors)+ylab("Measurement mean")

  ## save final
 integrateBoxFinal<-	test%>%reshape2::melt()%>%ggplot(aes(x=annotation,y=value,fill=experiment))+geom_boxplot()+facet_wrap(~variable)+theme_bw(base_size=14)+theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_fill_manual(values=threeColors)+ylab("Measurement mean")
  save(integrateBoxFinal,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/integrateBoxFinal_Figure5.RData")

 cleandat$subannotation<-as.character(cleandat$annotation)
  cleandat$subannotation[which(cleandat$annotation=="Macrophage")]<-paste0("Macrophage_",cleandat$Macrophage_subphenograph_cluster_meta[which(cleandat$annotation=="Macrophage")])
  cleandat$subannotation[which(cleandat$annotation=="CD4")]<-paste0("CD4_",cleandat$CD4_subphenograph_cluster_meta[which(cleandat$annotation=="CD4")])
  cleandat$subannotation[which(cleandat$annotation=="CD8")]<-paste0("CD8_",cleandat$CD8_subphenograph_cluster_meta[which(cleandat$annotation=="CD8")])
  cleandat$subannotation[which(cleandat$annotation=="TREG")]<-paste0("TREG_",cleandat$TREG_subphenograph_cluster_meta[which(cleandat$annotation=="TREG")])


 ##HL vs DLBCL
  mac<-empiricalFit(immune,phenotype="Macrophage")
  cd4<-empiricalFit(immune,phenotype="CD4")
  cd8<-empiricalFit(immune,phenotype="CD8")
   treg<-empiricalFit(immune,phenotype="TREG")

 
 

```


### 4B revision Boxplots
- 1-10: reproduces the boxplots across cohorts and performs simple linear regression.
- 1-11: compares CD8 Tim3/PD1 relative CD8 proportions using stratified linear models, because there was an interaction term between phenotypes and cohorts.  creates a supplementary figure 15B comparing CD8 Tim3/PD1 proportions.
- 1-11: Uses an ANOVA for TIM3/PD1/PDL1 expression differences because we did not see significant interaction terms between the cohort/phenotypes.  The ANOVA uses contrasting to report the PD1/PDL1/TIM3 differences in cohorts.
- 1-11: the results section uses ANOVA results and not the empirical design model, and averages each patient level.
- 1-12: proportions initially modeled with SLR, but included beta regression for proportions.
- 1-18: remember that betareg is on the log scale so "among those with X, the proportion of Y increase exp(beta) times that of those not in X".
```{r boxplots}
  immune$experiment_RLN<-"DLBCL"
  immune$experiment_RLN[grepl("Case",immune$ROIID)]<-"HL"
#immune$experiment_RLN[immune$case%in%c("Case1_ROI_01_5_a0","Case1_ROI_02_6_a0","Case2_ROI_01_7_a0","Case3_ROI_01_8_a0","Case4_ROI_02_9_a0","Case5_ROI_01_10_a0")]<-"RLN"
 immune$experiment_RLN[immune$ROIID%in%c("Case1_ROI_01_5_a0",
                                         "Case1_ROI_02_6_a0",
                                         "Case2_ROI_01_7_a0",
                                         "Case3_ROI_01_8_a0",
                                         "Case4_ROI_02_9_a0",
                                         "Case5_ROI_01_10_a0")]<-"RLN"
  immune$experiment_RLN<-factor(immune$experiment_RLN,levels=c("DLBCL","HL","RLN"))

  test<-immune[,c("PDL1","PD1","TIM3","ROIID","annotation")]%>%group_by(ROIID,annotation)%>%summarise_all(funs(mean))%>%data.frame
    exCol<-getExperimentColors()
      threeColors<-getExperimentColors()
    threeColors<-c(threeColors,'indianred3')
    
    names(threeColors)[3]<-'RLN'
     names(threeColors)[2]<-'HL'

  ## reproduces final figure 5B  
	test$experiment_RLN<-immune$experiment_RLN[match(test$ROIID,immune$ROIID)]
	test%>%reshape2::melt()%>%ggplot(aes(x=annotation,y=value,fill=experiment_RLN))+geom_boxplot()+facet_wrap(~variable)+theme_bw(base_size=14)+theme(axis.text.x = element_text(angle = 90, hjust = 1))+scale_fill_manual(values=threeColors)+ylab("Measurement mean")

	
 treg.pd1.fit<- lm(PD1~experiment_RLN,test[which(test$annotation=="TREG"),])
  cd8.pd1.fit<- lm(PD1~experiment_RLN,test[which(test$annotation=="CD8"),])

 mac.pdl1<-lm(PDL1~experiment_RLN,test[which(test$annotation=="Macrophage"),])
  cd4.tim3<-lm(TIM3~experiment_RLN,test[which(test$annotation=="CD4"),])%>%summary
  cd8.tim3<-lm(TIM3~experiment_RLN,test[which(test$annotation=="CD8"),])%>%summary

  
## library
  library(sjPlot)
  tab_model(treg.pd1.fit,
            mac.pdl1,
            cd4.tim3,
            cd8.tim3,
             pred.labels = c("Intercept", 
                  "HL.vs.DLBCL",
                  "RLN.vs.DLBCL"
                    ),
  dv.labels = c("T-REG PD1","MAC PD-L1","CD4 TIM-3","CD8 TIM-3"),
  string.pred = "Average mean intensity",
  string.ci = "Conf. Int (95%)",
  string.p = "P-Value",
  show.aic=TRUE,
    file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/CaseExpression-BoxPlot.intensity.html"
)
  test$experiment_RLN<-relevel(test$experiment_RLN,ref="RLN")
  ##lincom
  # library(biostat3)
 #lm(PD1~experiment_RLN+annotation,test[which(test$annotation!="Macrophage"),])%>%summary
 #regfit<-lm(PD1~experiment_RLN*annotation,test[which(test$annotation!="Macrophage"),])
 #lincom(regfit,c("experiment_RLNDLBCL-experiment_RLNHL"),eform=TRUE)
#  treg.pd1.fit<- lm(PD1~experiment_RLN,test[which(test$annotation=="TREG"),])
#  lincom(treg.pd1.fit,c("experiment_RLNDLBCL-experiment_RLNHL"),eform=TRUE)

  ##we model based on cell type that can express PD1/TIM3/PDL1
  ## anova on the expression 
  test$experiment_RLN<-factor(test$experiment_RLN,levels=c("DLBCL","HL","RLN"))
  tme.tim3.exp<-test
  tme.tim3.exp$id<-seq(1,nrow(tme.tim3.exp))
  
  ## pd1 (we omit MAC because they express PDL1, and exclude PD1 on CD4, to focus on Cytotoxic hypothesis)
  tme.pd1.exp<-test[which(test$annotation!="Macrophage" & test$annotation!="CD4"),]
  tme.pd1.exp$id<-seq(1,nrow(tme.pd1.exp))
  
  #pdl1  (all celltypes i guess can express PDL1?)
  tme.pdl1.exp<-test
  tme.pdl1.exp$id<-seq(1,nrow(tme.pdl1.exp))
  

  library(lsmeans);library(afex);library(multcomp)
  
  ## should match cd4.prop means
  #cd4
 # cd4tim3aov<-aov.car(TIM3~experiment_RLN+annotation+Error(id),data=tme.exp,return="aov")
#  cd4tim3ref1<-lsmeans(cd4tim3aov,c("experiment_RLN","annotation")) 
  ## all cell types can expres TIM3
  #cd8
   tim3aov<-aov.car(TIM3~experiment_RLN+annotation+Error(id),data=tme.tim3.exp,return="aov")
  tim3ref1<-lsmeans(tim3aov,c("experiment_RLN","annotation"))

    ## PD1 cd8  ##
  # CD4, CD8 and TREG can express PD1  (omit mac)
  pd1aov<-aov.car(PD1~experiment_RLN+annotation+Error(id),data=tme.pd1.exp,return="aov")
  pd1ref1<-lsmeans(pd1aov,c("experiment_RLN","annotation")) 
  
  
  #mac  
  ## MACs, and others have been reported to express PDL1
   macpdl1aov<-aov.car(PDL1~experiment_RLN+annotation+Error(id),data=tme.pdl1.exp,return="aov")
   macpdl1ref1<-lsmeans(macpdl1aov,c("experiment_RLN","annotation")) 
   
    
  ## contrast for all cell types 
  c_list <- list(macDLBCL.v.RLN = c( 0,0,0, 0,0,0, 1,0,-1, 0,0,0),
               macHL.v.RLN = c( 0,0,0, 0,0,0, 0,1,-1, 0,0,0),
              macDLBCLRLN.v.HLRLN=c( 0,0,0, 0,0,0,  1,-1,0, 0,0,0),
              
              cd8DLBCL.v.RLN = c(0,0,0, 1,0,-1, 0,0,0, 0,0,0),
              cd8HL.v.RLN = c(0,0,0, 0,1,-1, 0,0,0, 0,0,0),
              cd8DLBCLRLN.v.HLRLN=c(0,0,0, 1,-1,0, 0,0,0, 0,0,0),
              
              cd4DLBCL.v.RLN = c(1,0,-1, 0,0,0, 0,0,0, 0,0,0),
              cd4HL.v.RLN = c(0,1,-1, 0,0,0, 0,0,0, 0,0,0),
              cd4DLBCLRLN.v.HLRLN=c(1,-1/2,-1/2, 0,0,0, 0,0,0, 0,0,0),
              
              regDLBCL.v.RLN = c(0,0,0, 0,0,0, 0,0,0, 1,0,-1),
               regHL.v.RLN = c(0,0,0, 0,0,0, 0,0,0, 0,1,-1),
              regDLBCLRLN.v.HLRLN=c(0,0,0, 0,0,0, 0,0,0, 1,-1,0)
             )
  ## omit mac contrast.
  treg_list <- list(
              cd8DLBCL.v.RLN = c( 1,0,-1,  0,0,0),
              cd8HL.v.RLN = c( 0,1,-1,  0,0,0),
              cd8DLBCLRLN.v.HLRLN=c( 1,-1,0, 0,0,0),
              
              regDLBCL.v.RLN = c( 0,0,0,  1,0,-1),
               regHL.v.RLN = c( 0,0,0, 0,1,-1),
              regDLBCLRLN.v.HLRLN=c( 0,0,0,  1,-1,0)
             )
  
 ### CD8 Tim3
 summary(contrast(tim3ref1, c_list), adjust = "BH")
 tim3.rez.data<- confint(contrast(tim3ref1, c_list), adjust = "BH")%>%data.frame
  tim3.rez.data$p.value<-summary(contrast(tim3ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(tim3ref1, c_list)), test = adjusted("BH"))
 tim3.rez.data$bonf.p<-tim3.rez.data$p.value*3
 ## logfold
log2(tim3.rez.data[which(tim3.rez.data$contrast=="cd8DLBCLRLN.v.HLRLN"),"estimate"]/tim3.rez.data[which(tim3.rez.data$contrast=="cd8DLBCL.v.RLN"),"estimate"])

### CD4 Tim3

 ## logfold
log2(tim3.rez.data[which(tim3.rez.data$contrast=="cd4DLBCLRLN.v.HLRLN"),"estimate"]/tim3.rez.data[which(tim3.rez.data$contrast=="cd4DLBCL.v.RLN"),"estimate"])


 ## cd8 PD1 (NULL)
summary(contrast(pd1ref1, treg_list), adjust = "BH")
 pd1.rez.data2<- confint(contrast(pd1ref1, treg_list), adjust = "BH")%>%data.frame
  pd1.rez.data2$p.value<-summary(contrast(pd1ref1, treg_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(pd1ref1, treg_list)), test = adjusted("BH"))
 pd1.rez.data2$bonf.p<-pd1.rez.data2$p.value*3
 ## logfold
log2(pd1.rez.data2[which(pd1.rez.data2$contrast=="cd8DLBCLRLN.v.HLRLN"),"estimate"]/pd1.rez.data2[which(pd1.rez.data2$contrast=="cd8DLBCL.v.RLN"),"estimate"])
log2(pd1.rez.data2[which(pd1.rez.data2$contrast=="regDLBCLRLN.v.HLRLN"),"estimate"]/pd1.rez.data2[which(pd1.rez.data2$contrast=="regDLBCL.v.RLN"),"estimate"])


   summary(contrast(macpdl1ref1, c_list), adjust = "BH")
 mac.rez.data<- confint(contrast(macpdl1ref1, c_list), adjust = "BH")%>%data.frame
  mac.rez.data$p.value<-summary(contrast(macpdl1ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(macpdl1ref1, c_list)), test = adjusted("BH"))
 ### multiple testing bonferroni by 4 for each TME model.
  mac.rez.data$bonf.p<-mac.rez.data$p.value*3
   ## logfold
log2(mac.rez.data[which(mac.rez.data$contrast=="macDLBCLRLN.v.HLRLN"),"estimate"]/mac.rez.data[which(mac.rez.data$contrast=="macDLBCL.v.RLN"),"estimate"])


####### CD8 Proportions TIm3/PD1 comparison stratified analysis.
 ## 1-12: should use beta-regression for proportions.
  ## looking at PD1/TIM3 on CD8 relative proportions across cohorts using within gating.
tot<-as.data.frame.matrix(table(joint$experiment,joint$phenotype))
  
 # nchl<-tot["HL",c("CD4","CD8","Macrophage","TREG")]%>%sum
#  ndlbcl<-tot["DLBCL",c("CD4","CD8","Macrophage","TREG")]%>%sum
#  nrln<-tot["RLN",c("CD4","CD8","Macrophage","TREG")]%>%sum

   nchl<-tot["HL",c("CD8")]%>%sum
  ndlbcl<-tot["DLBCL",c("CD8")]%>%sum
  nrln<-tot["RLN",c("CD8")]%>%sum
  tregs<-joint[grepl("CD8",joint$phenotype),]
  ## CHL
  chl.tim3pd1P<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$tim3.pd1=="TIM3+_PD1+_CD8"),])/nchl
     chl.tim3pd1N<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$tim3.pd1=="TIM3+_PD1-_CD8"),])/nchl
     chl.tim3Npd1P<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$tim3.pd1=="TIM3-_PD1+_CD8"),])/nchl

   ## DLBCL
    dl.tim3pd1P<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$tim3.pd1=="TIM3+_PD1+_CD8"),])/ndlbcl
     dl.tim3pd1N<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$tim3.pd1=="TIM3+_PD1-_CD8"),])/ndlbcl
     dl.tim3Npd1P<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$tim3.pd1=="TIM3-_PD1+_CD8"),])/ndlbcl

  
   ## RLN
 rln.tim3pd1P<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$tim3.pd1=="TIM3+_PD1+_CD8"),])/nrln
  rln.tim3pd1N<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$tim3.pd1=="TIM3+_PD1-_CD8"),])/nrln
  rln.tim3Npd1P<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$tim3.pd1=="TIM3-_PD1+_CD8"),])/nrln

  ## testing proportions.
  tim3.pd1<-as.data.frame.matrix(table(tregs$ROIID,tregs$tim3.pd1.simple))%>%data.frame
  tim3.pd1<-tim3.pd1/rowSums(tim3.pd1)
  tim3.pd1$experiment<-tregs$experiment[match(rownames(tim3.pd1),tregs$ROIID)]
  ## 
  tim3.pd1<-tim3.pd1[,c("TIM3Pos._PD1Pos._CD8","TIM3Pos._PD1Neg._CD8","TIM3Neg._PD1Pos._CD8","experiment")]
  tim3.pd1$id<-seq(1,nrow(tim3.pd1))
  ## MACs, and others have been reported to express PDL1
  colnames(tim3.pd1)<-c("TIM3PD1P","TIM3PPD1N","TIM3NPD1P","experiment","id")
   tim3.pd12<-melt(tim3.pd1,id.vars=c("experiment","id"))
   tim3.pd12$id<-rownames(tim3.pd12)

   tim3pd1aov<-aov.car(value~experiment*variable+Error(id),data=tim3.pd12,return="aov")
   summary(tim3pd1aov)
   ## we see significant interaction between experiment and variable, so we do a stratified analysis and stratified modeling of each phenotype.
   
   
 #  tim3pd1ref1<-lsmeans(tim3pd1aov,c("experiment","variable")) 
  #     cd8_list <- list(TIM3PD1DLBCL.v.RLN = c(1,0,-1, 0,0,0, 0,0,0),
   #            TIM3PD1HL.v.RLN = c( 0,1,-1, 0,0,0, 0,0,0),
    #          TIM3PD1DLBCLRLN.v.HLRLN=c( 1,-1,0, 0,0,0,  0,0,0),
             
    #          TIM3PPD1NDLBCL.v.RLN = c(0,0,0, 1,0,-1, 0,0,0),
    #           TIM3PPD1NHL.v.RLN = c( 0,0,0, 0,1,-1, 0,0,0),
    #          TIM3PPD1NDLBCLRLN.v.HLRLN=c( 0,0,0, 1,-1,0,  0,0,0),
              
     #        TIM3NPD1PDLBCL.v.RLN = c(0,0,0, 0,0,0, 1,0,-1),
      #         TIM3NPD1PHL.v.RLN = c( 0,0,0, 0,0,0, 0,1,-1),
        #      TIM3NPD1PDLBCLRLN.v.HLRLN=c( 0,0,0, 0,0,0,  1,-1,0)
       #      )
  
 ### CD8 Tim3
# summary(contrast(tim3pd1ref1, cd8_list), adjust = "BH")
# tim3.rez.data<- confint(contrast(tim3pd1ref1, cd8_list), adjust = "BH")%>%data.frame
#  tim3.rez.data$p.value<-summary(contrast(tim3pd1ref1, cd8_list), adjust = "BH")[,"p.value"]
#  summary(as.glht(contrast(tim3pd1ref1, cd8_list)), test = adjusted("BH"))
# tim3.rez.data$bonf.p<-tim3.rez.data$p.value*3
   tim3.pd1$experiment<-factor(tim3.pd1$experiment,levels=c("HL","DLBCL","RLN"))
   
   library(betareg)
   ### beta reg for proportions estimates
   ## TIM3+PD1+
   betareg(TIM3PD1P~experiment,tim3.pd1)%>%summary
      betareg(TIM3PD1P~experiment,tim3.pd1)%>%coef%>%exp
      betareg(TIM3PD1P~experiment,tim3.pd1)%>%confint%>%exp

   
   betareg(TIM3PPD1N~experiment,tim3.pd1)%>%summary
      betareg(TIM3PPD1N~experiment,tim3.pd1)%>%coef%>%exp
         betareg(TIM3PPD1N~experiment,tim3.pd1)%>%confint%>%exp

   
   betareg(TIM3NPD1P~experiment,tim3.pd1)%>%summary
      betareg(TIM3NPD1P~experiment,tim3.pd1)%>%confint

  
   
  ### figure generation uses the estimated means (the means are similar between beta-reg and linear model, but estimates are not )
   ## okay to use the mean estimates.
    a<-lm(TIM3PD1P~0+experiment,tim3.pd1)
    b<-lm(TIM3PPD1N~0+experiment,tim3.pd1)
    c<-lm(TIM3NPD1P~0+experiment,tim3.pd1)
   
   # a<-betareg(TIM3PD1P~0+experiment,tim3.pd1)
  #  b<-betareg(TIM3PPD1N~0+experiment,tim3.pd1)
  #  c<-betareg(TIM3NPD1P~0+experiment,tim3.pd1)
   
  #  lm(TIM3PD1P~experiment,tim3.pd1)%>%summary
  #  lm(TIM3PPD1N~experiment,tim3.pd1)%>%summary
  # lm(TIM3NPD1P~experiment,tim3.pd1)%>%summary
   
   A<-100*cbind(coef(a),confint(a))
   B<-100*cbind(coef(b),confint(b))
   C<-100*cbind(coef(c),confint(c))
  cd8.props<-rbind(A,B,C)
  cd8.props<-cd8.props[!grepl("phi",rownames(cd8.props)),]
  colnames(cd8.props)<-c("estimate","lower","upper")
  cd8.props<-as.data.frame(cd8.props)
  cd8.props$cohort<-rep(c("DLBCL","HL","RLN"),3)
  cd8.props$phenotype<-c( rep("TIM3+PD1+",3),rep("TIM3+PD1-",3),rep("TIM3-PD1+",3))
  
   library(ggsignif)
  #ggplot(cd8.props,aes(x=cohort,y=estimate))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.2)+facet_wrap(~phenotype)
  
  ggplot(cd8.props,aes(x=phenotype,y=estimate,fill=cohort))+geom_bar(stat='identity',position='dodge')+
    geom_errorbar(position=position_dodge(0.9),aes(ymin=lower,ymax=upper),width=0.5)+
    scale_fill_manual(values=threeColors2)+
    geom_signif(stat="identity",
              data=data.frame(x=c(0.7,1.7,2.7), xend=c(1.25, 2.25,3.25),
                              y=c(13, 20,33),cohort=c("DLBCL","RLN","HL"), annotation=c("p<0.001","p<0.001","p<0.001")),
              aes(x=x,xend=xend, y=y, yend=y, annotation=annotation))+theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme(axis.text.y=element_text(size=10,colour='black'))+
    ylab("Within CD8 relative TME proportion (%)")+ggtitle("CD8 stratified comparison")+xlab("CD8 Phenotype")
  
  
  ############################################
    ############################################
    ############################################
    ############################################
    ############################################
    ############################################
  ## looking at PD1/Ki67 within stratified gating proportions.
  ###### TREG
    joint$ki67.pd1<-paste0(joint$Ki67,joint$PD1,joint$phenotype)
      joint$tim3.pdl1<-paste0(joint$TIM3,joint$PDL1,joint$phenotype)
      
    joint$ki67.pd1.simple<-gsub("\\+",".Pos.",joint$ki67.pd1)  
    joint$ki67.pd1.simple<-gsub("\\-",".Neg.",joint$ki67.pd1.simple)  

       joint$tim3.pdl1.simple<-gsub("\\+",".Pos.",joint$tim3.pdl1)  
    joint$tim3.pdl1.simple<-gsub("\\-",".Neg.",joint$tim3.pdl1.simple)  

   tregs<-joint[grepl("TREG",joint$phenotype),]
  

  
  ## Ki67+PD1+ testing proportions.
  tim3.pd1<-as.data.frame.matrix(table(tregs$ROIID,tregs$ki67.pd1.simple))%>%data.frame
  tim3.pd1<-tim3.pd1/rowSums(tim3.pd1)
  tim3.pd1$experiment<-tregs$experiment[match(rownames(tim3.pd1),tregs$ROIID)]
  ## 
  tim3.pd1<-tim3.pd1[,c("Ki67.Pos.PD1.Pos.TREG","Ki67.Pos.PD1.Neg.TREG","Ki67.Neg.PD1.Pos.TREG","experiment")]
  tim3.pd1$id<-seq(1,nrow(tim3.pd1))
  ## MACs, and others have been reported to express PDL1
  colnames(tim3.pd1)<-c("Ki67PD1P","Ki67PPD1N","Ki67NPD1P","experiment","id")
   tim3.pd12<-melt(tim3.pd1,id.vars=c("experiment","id"))
   tim3.pd12$id<-rownames(tim3.pd12)

   tim3pd1aov<-aov.car(value~experiment*variable+Error(id),data=tim3.pd12,return="aov")
   summary(tim3pd1aov)
   ## we see significant interaction between experiment and variable, so we do a stratified analysis and stratified modeling of each phenotype.
      tim3.pd1$experiment<-factor(tim3.pd1$experiment,levels=c("HL","DLBCL","RLN"))

 
    betareg(Ki67PD1P~experiment,dplyr::filter(tim3.pd1, Ki67PD1P>0))%>%summary
        betareg(Ki67PD1P~experiment,dplyr::filter(tim3.pd1, Ki67PD1P>0))%>%coef%>%exp
    betareg(Ki67PD1P~experiment,dplyr::filter(tim3.pd1, Ki67PD1P>0))%>%confint

    
    
   betareg(Ki67PPD1N~experiment,dplyr::filter(tim3.pd1, Ki67PPD1N>0))%>%summary
      betareg(Ki67PPD1N~experiment,dplyr::filter(tim3.pd1, Ki67PPD1N>0))%>%coef%>%exp
      
     betareg(Ki67PPD1N~experiment,dplyr::filter(tim3.pd1, Ki67PPD1N>0))%>%confint%>%exp
     
     

   betareg(Ki67NPD1P~experiment,dplyr::filter(tim3.pd1, Ki67NPD1P>0))%>%summary
      betareg(Ki67NPD1P~experiment,dplyr::filter(tim3.pd1, Ki67NPD1P>0))%>%confint
      
      
 
   tim3.pd1$experiment<-factor(tim3.pd1$experiment,levels=c("DLBCL","HL","RLN"))
  a<-lm(Ki67PD1P~0+experiment,tim3.pd1)
    b<-lm(Ki67PPD1N~0+experiment,tim3.pd1)
   c<-lm(Ki67NPD1P~0+experiment,tim3.pd1)
   
    lm(Ki67PD1P~experiment,tim3.pd1)%>%summary
    lm(Ki67PPD1N~experiment,tim3.pd1)%>%summary
   lm(Ki67NPD1P~experiment,tim3.pd1)%>%summary
   
    A<-100*cbind(coef(a),confint(a))
   B<-100*cbind(coef(b),confint(b))
   C<-100*cbind(coef(c),confint(c))
  cd8.props<-rbind(A,B,C)
  colnames(cd8.props)<-c("estimate","lower","upper")
  cd8.props<-as.data.frame(cd8.props)
  cd8.props$cohort<-rep(c("DLBCL","HL","RLN"),3)
  cd8.props$phenotype<-c( rep("Ki67+PD1+",3),rep("Ki67+PD1-",3),rep("Ki67-PD1+",3))
  
   library(ggsignif)
  #ggplot(cd8.props,aes(x=cohort,y=estimate))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.2)+facet_wrap(~phenotype)
  
  ggplot(cd8.props,aes(x=phenotype,y=estimate,fill=cohort))+geom_bar(stat='identity',position='dodge')+
    geom_errorbar(position=position_dodge(0.9),aes(ymin=lower,ymax=upper),width=0.5)+
    scale_fill_manual(values=threeColors2)+
    geom_signif(stat="identity",
              data=data.frame(x=c(0.7,1.7,2.7), xend=c(1.25, 2.25,3.25),
                              y=c(23, 20,20),cohort=c("DLBCL","RLN","HL"), annotation=c("p<0.001","p<0.001","p<0.001")),
              aes(x=x,xend=xend, y=y, yend=y, annotation=annotation))+theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme(axis.text.y=element_text(size=10,colour='black'))+
    ylab("Within TREG relative TME proportion (%)")+ggtitle("TREG stratified comparison")+xlab("TREG Phenotype")
  
  
  ## CCR4+PD1+ TREG comparison
  
   tregs<-joint[grepl("TREG",joint$phenotype),]
   tregs$experiment<-factor(tregs$experiment,level=c("HL","DLBCL","RLN"))
   

  
  ## CCR4 PD1 testing proportions.
  tim3.pd1<-as.data.frame.matrix(table(tregs$ROIID,tregs$ccr4.pd1.simple))%>%data.frame
  tim3.pd1<-tim3.pd1/rowSums(tim3.pd1)
  tim3.pd1$experiment<-tregs$experiment[match(rownames(tim3.pd1),tregs$ROIID)]
  ## 
  tim3.pd1<-tim3.pd1[,c("CCR4Pos._PD1Pos._TREG","CCR4Pos._PD1Neg._TREG","CCR4Neg._PD1Pos._TREG","experiment")]
  tim3.pd1$id<-seq(1,nrow(tim3.pd1))
  ## MACs, and others have been reported to express PDL1
  colnames(tim3.pd1)<-c("CCR4PD1P","CCR4PPD1N","CCR4NPD1P","experiment","id")
   tim3.pd12<-melt(tim3.pd1,id.vars=c("experiment","id"))
   tim3.pd12$id<-rownames(tim3.pd12)

      betareg(CCR4PD1P~experiment,dplyr::filter(tim3.pd1, CCR4PD1P>0))%>%summary
            betareg(CCR4PD1P~experiment,dplyr::filter(tim3.pd1, CCR4PD1P>0))%>%coef%>%exp
    betareg(CCR4PD1P~experiment,dplyr::filter(tim3.pd1, CCR4PD1P>0))%>%confint%>%exp

    
    
   betareg(CCR4PPD1N~experiment,dplyr::filter(tim3.pd1, CCR4PPD1N>0))%>%summary
      betareg(CCR4PPD1N~experiment,dplyr::filter(tim3.pd1, CCR4PPD1N>0))%>%confint

   betareg(CCR4NPD1P~experiment,filter(tim3.pd1, CCR4NPD1P>0))%>%summary
      betareg(CCR4NPD1P~experiment,filter(tim3.pd1, CCR4NPD1P>0))%>%confint
      
      
 
   tim3.pd1$experiment<-factor(tim3.pd1$experiment,levels=c("DLBCL","HL","RLN"))
  a<-lm(CCR4PD1P~0+experiment,tim3.pd1)
    b<-lm(CCR4PPD1N~0+experiment,tim3.pd1)
   c<-lm(CCR4NPD1P~0+experiment,tim3.pd1)
   
   
   
    A<-100*cbind(coef(a),confint(a))
   B<-100*cbind(coef(b),confint(b))
   C<-100*cbind(coef(c),confint(c))
  cd8.props<-rbind(A,B,C)
  colnames(cd8.props)<-c("estimate","lower","upper")
  cd8.props<-as.data.frame(cd8.props)
  cd8.props$cohort<-rep(c("DLBCL","HL","RLN"),3)
  cd8.props$phenotype<-c( rep("CCR4+PD1+",3),rep("CCR4+PD1-",3),rep("CCR4-PD1+",3))
  
   library(ggsignif)
  #ggplot(cd8.props,aes(x=cohort,y=estimate))+geom_bar(stat='identity')+geom_errorbar(aes(ymin=lower,ymax=upper),width=0.2)+facet_wrap(~phenotype)
  
  ggplot(cd8.props,aes(x=phenotype,y=estimate,fill=cohort))+geom_bar(stat='identity',position='dodge')+
    geom_errorbar(position=position_dodge(0.9),aes(ymin=lower,ymax=upper),width=0.5)+
    scale_fill_manual(values=threeColors2)+
    geom_signif(stat="identity",
              data=data.frame(x=c(0.7,1.7,2.7), xend=c(1.25, 2.25,3.25),
                              y=c(13, 18,28),cohort=c("DLBCL","RLN","HL"), annotation=c("N.S.","N.S.","p<0.001")),
              aes(x=x,xend=xend, y=y, yend=y, annotation=annotation))+theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme_bw()+theme(axis.text.x=element_text(size=11,colour='black'))+
    theme(axis.text.y=element_text(size=10,colour='black'))+
    ylab("Within TREG relative TME proportion (%)")+ggtitle("TREG stratified comparison")+xlab("TREG Phenotype")
  
```


### PCA plot with the cluster level final Figure 5A (final and reproduced)
- Selects panels used for annotating, with lineage markers to match original 4A/4B that includes the lymphnode. the PCA will change depending on the factors to include.  this matches the first final draft.
- 12-4: final PCA plot of the phenotypes
- 1-3: save the ggplot object reproduces exactly 5A
```{r clusterPCA}
 ## REVISION 11-1: we include the RLN
# community<-immune[which(immune$normalLymphnode!='RLN'),]%>%group_by(ROIID,subannotation2)%>%dplyr::summarise(TIM3=mean(TIM3),
  community<-immune%>%group_by(ROIID,subannotation2)%>%dplyr::summarise(TIM3=mean(TIM3),
             CD4=mean(CD4),
	CD68=mean(CD68),
	CD8a=mean(CD8a),
	FOXP3=mean(FOXP3),                                                         
  	CD45RO=mean(CD45RO),
	CD45RA=mean(CD45RA),
		#HLADR=mean(HLADR),
	PD1=mean(PD1),
	#ICOS=mean(ICOS),
	PDL1=mean(PDL1),
	#CCR4=mean(CCR4),
	#CXCR3=mean(CXCR3),
	CD206=mean(CD206),
	Ki67=mean(Ki67))%>%data.frame

   PC<-prcomp(community[,c(-1,-2)],scale=FALSE,center=FALSE)
    PCi<-data.frame(PC$x,ROIID=community$ROIID,annotation=community$subannotation2)
   # PCi$experiment<-immune$experiment_RLN[match(PCi$ROIID,immune$ROIID)]
     PCi$experiment<-immune$experiment[match(PCi$ROIID,immune$ROIID)]
  PCi$experiment[PCi$ROIID%in%c("Case1_ROI_01_5_a0","Case1_ROI_02_6_a0","Case2_ROI_01_7_a0","Case3_ROI_01_8_a0","Case4_ROI_02_9_a0","Case5_ROI_01_10_a0")]<-"RLN"

 pca.perc<-100*round(PC$sd/sum(PC$sd),3)
   threeColors<-getExperimentColors()
    threeColors<-c(threeColors,'indianred3')
    names(threeColors)[3]<-'RLN'
 
 experiment.states.pca<- ggplot(PCi,aes(x=PC1,y=PC2,col=experiment,label=ROIID))+geom_point(color="black",size=3.5,alpha=0.95)+geom_point(aes(fill=experiment,col=experiment),size=2.8)+xlim(c( min(PCi$PC1)-0.5,  max(PCi$PC1)+0.05))+scale_color_manual(values=threeColors)+theme_classic()+ xlab(paste0("PC1 (",pca.perc[1],"%)"))+ylab(paste0("PC2 (",pca.perc[2],"%)"))
 print(experiment.states.pca)

  integratedColors<-getIntegratedColors(immune)
 #### PCA with the color coded sub-community.
   experiment.states.pca2<- ggplot(PCi,aes(x=PC1,y=PC2,col=annotation))+geom_point(color="black",size=3.5,alpha=0.95)+geom_point(aes(fill=annotation,col=annotation),size=2.8)+xlim(c( min(PCi$PC1)-0.5,  max(PCi$PC1)+0.05))+
	theme_classic()+ xlab(paste0("PC1 (",pca.perc[1],"%)"))+ylab(paste0("PC2 (",pca.perc[2],"%)"))+scale_color_manual(values=c(integratedColors))
 print(experiment.states.pca2)
  library(gridExtra)
 grid.arrange(experiment.states.pca,experiment.states.pca2,ncol=2)
  #  savePlot(file="immune.states.integration.PCA.subcommunity.png")
  

 save(experiment.states.pca,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/experiment.stats.pca_Figure5.RData")
 
  save(experiment.states.pca2,file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/ggplot-data-final-figures-Tucker/experiment.stats.pca2_Figure5.RData")

   immune$experiment_RLN2<-as.character(immune$experiment_RLN)
   immune$experiment_RLN2[which(immune$experiment_RLN2=="HL")]<-"HD"
   immune$experiment_RLN2<-factor(immune$experiment_RLN2,levels=c("RLN","HD","DLBCL"))
   
   
  tmeBar<-integratedPrimaryCommunityBarPlot(cleandat=immune,experiment='experiment_RLN2')
  #### TME ANOVA compare DLBCL.vs.RLN  and HL.vs.RLN  in proportions abundances and expression.
  
  
   tums<-immune[immune$annotation%in%c("CD4","CD8","TREG","Macrophage"),]
   tums$response<-factor(as.character(tums$response))
  ## filter prevalance
  x<-as.data.frame.matrix(table(tums[,'experiment_RLN2'],tums$annotation))
  x<-x/rowSums(x)
  toFilter<-which(apply(apply(x,2,function(x) x==0),2,function(y) any(y==TRUE)))
  tums<-tums[!tums$annotation%in%names(toFilter),]

 ## need to plot all the tumor proportions CR vs. REF
  #response vars=c("darkslateblue","firebrick2"))
 
 computeImmuneProportions<-function(full.dn4=NULL,primary=NULL,clinicalFactor=NULL){
  tumor.abund<-as.data.frame.matrix(table(tums$ROIID,tums$annotation))
  ## immune proportions
  tumor.abund<-100*(tumor.abund/(rowSums(tumor.abund)))

  tumor.abund$response<-full.dn4[,clinicalFactor][match(rownames(tumor.abund),full.dn4$ROIID)]
  tumor.means<-tumor.abund%>%group_by(response)%>%summarise_all(funs(mean))%>%data.frame
  rownames(tumor.means)<-tumor.means[,1]
  tumor.means<-tumor.means[,-1]
  tumor.means<-t(tumor.means)
  tumor.means<-as.data.frame(tumor.means)
  tumor.means$pheno<-rownames(tumor.means)
  #tumor.means<-melt(tumor.means)
  tumors<-reshape(tumor.means,varying=1:3,v.names='response',timevar='pheno',direction='long')
  tumors$pheno<-colnames(tumor.means)[tumors$pheno]
  tumors$id<-tumor.means$pheno[tumors$id]
  tumor.means<-tumors
  #tumor.means<-data.frame(means=c(tumor.means[,1],tumor.means[,2]),response=c(rep(colnames(tumor.means)[1],nrow(tumor.means)),rep(colnames(tumor.means)[2],nrow(tumor.means))),names=rep(rownames(tumor.means,2)))
  tumor.sd<-tumor.abund%>%group_by(response)%>%summarise_all(funs(sd))%>%data.frame
  tumor.se<-tumor.sd[,-1]/sqrt(table(tumor.abund$response))
   tumor.se<-t(tumor.se)
   colnames(tumor.se)<-tumor.sd$response
   tumor.se<-as.data.frame(tumor.se)
     tumor.se$pheno<-rownames(tumor.se)
   tumorSE<-reshape(tumor.se,varying=1:3,v.names='response',timevar='pheno',direction='long')
 tumorSE$pheno<-colnames(tumor.se)[tumorSE$pheno]
  tumorSE$id<-tumor.se$pheno[tumorSE$id]
  tumor.se<-tumorSE
     # tumor.means$se<-c(tumor.se[,1],tumor.se[,2])
  ## need to plot all the tumor proportions CR vs. REF
  tumor.means<- left_join(tumor.means,tumor.se,by=c("pheno","id"))
  colnames(tumor.means)[which(colnames(tumor.means)=='response.x')]<-'means'
    colnames(tumor.means)[which(colnames(tumor.means)=='response.y')]<-'se'
  return(tumor.means)
 }



 ## add the immune populations.
  ## MAC, TREG, CD4, CD8.
   cd4.prop<-computeImmuneProportions(full.dn4=tums,primary="CD4",clinicalFactor='experiment_RLN2')
  cd4.prop<-cd4.prop[which(cd4.prop$means>0),]
    cd4s<-getExperimentColors()
    cd4s<-c(cd4s,'lightcoral')
    names(cd4s)[3]<-'RLN'
    cd4.exp<-ggplot(cd4.prop,aes(x=id,y=means,fill=pheno))+geom_bar(stat='identity',position=position_dodge(),colour='black')+geom_errorbar(aes(ymin=abs(means-2*se), ymax=means+2*se), width=.2,
                 position=position_dodge(.9))+theme_bw(base_size=14)+ theme(axis.text.x = element_text(angle = 45, hjust = 1,color='black',size=17),
	axis.text.y=element_text(size=15,color='black'))+ylab("Relative proportion (%)")+xlab("TME communities")+scale_fill_manual("Cohort",values=cd4s)
  cd4.exp
 ### immune proportions
  
     ## FIX ME:  anova and contrast with RLN.  differences.
 tumor.abund<-as.data.frame.matrix(table(tums$ROIID,tums$annotation))
  ## immune proportions
  tumor.abund<-100*(tumor.abund/(rowSums(tumor.abund)))
   tumor.abund$response<-tums[,"experiment_RLN2"][match(rownames(tumor.abund),tums$ROIID)]
   tumor.abund$id<-seq(1,nrow(tumor.abund))
   
  library(lsmeans);library(afex);library(multcomp)
  
  ## should match cd4.prop means
  #cd4
  cd4aov<-aov.car(CD4~response+Error(id),data=tumor.abund,return="aov")
  cd4ref1<-lsmeans(cd4aov,c("response"))  
  #cd8
   cd8aov<-aov.car(CD8~response+Error(id),data=tumor.abund,return="aov")
  cd8ref1<-lsmeans(cd8aov,c("response"))  
  #mac
   macaov<-aov.car(Macrophage~response+Error(id),data=tumor.abund,return="aov")
   macref1<-lsmeans(macaov,c("response"))  
    # treg
   regaov<-aov.car(TREG~response+Error(id),data=tumor.abund,return="aov")
  regref1<-lsmeans(regaov,c("response"))  
  
  ## dropping PD1-CCR4+
  c_list <- list(DLBCL.v.RLN = c(-1,0, 1),
               HL.v.RLN = c(-1,1,0),
              DLBCLRLN.v.HLRLN=c(0,-1,1)
             )
  

  summary(contrast(cd4ref1, c_list), adjust = "BH")
 cd4.rez.data<- confint(contrast(cd4ref1, c_list), adjust = "BH")%>%data.frame
  cd4.rez.data$p.value<-summary(contrast(cd4ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd4ref1, c_list)), test = adjusted("BH"))
cd4.rez.data$bonf.p<-cd4.rez.data$p.value*4
  
 summary(contrast(cd8ref1, c_list), adjust = "BH")
 cd8.rez.data<- confint(contrast(cd8ref1, c_list), adjust = "BH")%>%data.frame
  cd8.rez.data$p.value<-summary(contrast(cd8ref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(cd8ref1, c_list)), test = adjusted("BH"))
cd8.rez.data$bonf.p<-cd8.rez.data$p.value*4

   summary(contrast(macref1, c_list), adjust = "BH")
 mac.rez.data<- confint(contrast(macref1, c_list), adjust = "BH")%>%data.frame
  mac.rez.data$p.value<-summary(contrast(macref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(macref1, c_list)), test = adjusted("BH"))
 ### multiple testing bonferroni by 4 for each TME model.
  mac.rez.data$bonf.p<-mac.rez.data$p.value*4
  
   summary(contrast(regref1, c_list), adjust = "BH")
 reg.rez.data<- confint(contrast(regref1, c_list), adjust = "BH")%>%data.frame
  reg.rez.data$p.value<-summary(contrast(regref1, c_list), adjust = "BH")[,"p.value"]
  summary(as.glht(contrast(regref1, c_list)), test = adjusted("BH"))
reg.rez.data$bonf.p<-reg.rez.data$p.value*4

 ### anova on the expression of given markers comparing cohorts.


  
 
  
 
```


### Write integration comparison to a table
```{r, eval=FALSE}

  library("xlsx")
  write.xlsx(topTable(mac,coef=2), file = "empiricalBayes.integrated.unsupervised.community.immune.proportions.HD.vs.DLBCL.xlsx",
    sheetName = "MAC", append = FALSE)
 write.xlsx(topTable(cd4,coef=2), file = "empiricalBayes.integrated.unsupervised.community.immune.proportions.HD.vs.DLBCL.xlsx",
  sheetName = "CD4", append = TRUE)
 write.xlsx(topTable(cd8,coef=2), file = "empiricalBayes.integrated.unsupervised.community.immune.proportions.HD.vs.DLBCL.xlsx",
    sheetName = "CD8", append = TRUE)
 write.xlsx(topTable(treg,coef=2), file = "empiricalBayes.integrated.unsupervised.community.immune.proportions.HD.vs.DLBCL.xlsx",
    sheetName = "TREG", append = TRUE)


```


### Examine the Heatmap for Figure 4A (deprecated)
Generate the heatmap in Figure 4A, and append on it the odds ratio for enrichment.  Also shows the cluster level joint TME expression heatmap, and the labeled annotation construction.
- 1-3:  the heatmap is omitted.
```{r heatmap4A,eval=TRUE,warning=FALSE,message=FALSE,fig.height = 13, fig.width = 16}

  #load("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.RData")
  load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.RData")
 
    immune<-cleandat.limma[which(cleandat.limma$annotation!="BCell" & cleandat.limma$annotation!="HRS"),]
 
 ##plot clustering heatmap.
   immune$subannotation2<-as.character(immune$subannotation)
   immune$subannotation2<-sapply( strsplit(immune$subannotation2,"_"),function(x) paste0(x[1],"_",letters[as.numeric(x[2])]))
   immune$subannotation2<-factor(immune$subannotation2,levels=rownames(table(immune$subannotation2,immune$subannotation)))


  ErikMeta=immune[,
	c(c("CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","FOXP3","CXCR3","CCR4","HLADR","Ki67","PD1","PDL1","TIM3"),
	#"ROIID",
	"subannotation2")] %>% group_by(subannotation2) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame  

    integratedColors<-getIntegratedColors(immune)
   plot_clustering_heatmap_wrapper(expr=immune[,c("CD206","CD3","CD45RA","CD45RO","CD4","CD68","CD8a","FOXP3","CXCR3","CCR4","HLADR","Ki67","PD1","PDL1","TIM3")], 
  cell_clustering=immune[,"subannotation2"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=integratedColors)

  #savePlot(file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/TME.immune.HD.DLBCL.subannotation.singleCellLevel.png")


  immune$annotated<-as.character(immune$subannotation2)
 immune$experiment.binary<-ifelse(immune$experiment=='DLBCL',1,0)
 

  #PDL1+TIM3- MAC
  immune$annotated[immune$annotated%in%c("Macrophage_h",
	"Macrophage_c",	
	"Macrophage_d",
	"Macrophage_e")]<-"TIM3-PDL1+MAC"
 #TIM3+PDL1+MAC
  immune$annotated[immune$annotated%in%c("Macrophage_g",
	"Macrophage_f",	
	"Macrophage_i")]<-"TIM3+PDL1+MAC" 
 #TIM3-PDL1-
 immune$annotated[immune$annotated%in%c("Macrophage_a",
	"Macrophage_b")]<-"TIM3-PDL1-MAC"

 ##PD1+ HS TREG
  immune$annotated[immune$annotated%in%c("TREG_a",
	'TREG_f',
	"TREG_d")]<-"PD1+Highlysuppressive TREG"

 ##PD1- HS TREG

 ##PD1- TREG
 immune$annotated[immune$annotated%in%c("TREG_b",
	"TREG_g",'TREG_c',
	"TREG_h",
	"TREG_e")]<-"PD1-TREG"
 
 ## CD8 TIM3-PD1-
 immune$annotated[immune$annotated%in%c("CD8_d")]<-"TIM3-PD1-CD8"
 #CD8 TIM3+PD1+
 immune$annotated[immune$annotated%in%c("CD8_b",
	"CD8_c",
	"CD8_e")]<-"TIM3+PD1+CD8"

#CD8  TIM3+PD1-
 immune$annotated[immune$annotated%in%c("CD8_f")]<-"TIM3+PD1-CD8"
 ## TIM3-PD1+
 immune$annotated[immune$annotated%in%c("CD8_a")]<-"TIM3-PD1+CD8"


 ## CD4 TIM3+PD1+
 immune$annotated[immune$annotated%in%c("CD4_a",
	"CD4_h",
	"CD4_e",
	"CD4_c",
	"CD4_b")]<-"TIM3+PD1+CD4"
 ## CD4 TIM3+PD1-
 immune$annotated[immune$annotated%in%c("CD4_a")]<-"TIM3+PD1-CD4"

 ## CD4 TIM3-PD1+
immune$annotated[immune$annotated%in%c("CD4_g",
	"CD4_d",
	"CD4_f")]<-"TIM3-PD1+ CD4"

 immune$annotated.simple<-as.character(immune$annotated) 
  immune$annotated.simple<-gsub("\\-","Neg.",immune$annotated.simple)
 immune$annotated.simple<-gsub("\\+","Pos.",immune$annotated.simple)
 immune$annotated.simple<-gsub(" ","",immune$annotated.simple)
 immune$case<-immune$ROIID

 ## REVISION 11-1 PERFORM THE LINEAR MODEL TO SHOW UP-EXPRESSION PER CLUSTER.
 
 
 
 ## save the immune data.
 # save(immune,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/immune.subset.cleandat.limma.RData")
 # append to cleandat.limma
  cleandat.limma$annotated.simple<-cleandat.limma$annotated<-cleandat.limma$subannotation2<-NA
 
 cleandat.limma$annotated[match(immune$uniqueLabel,cleandat.limma$uniqueLabel)]<-immune$annotated


cleandat.limma$annotated.simple[match(immune$uniqueLabel,cleandat.limma$uniqueLabel)]<-immune$annotated.simple


 cleandat.limma$subannotation2[match(immune$uniqueLabel,cleandat.limma$uniqueLabel)]<-as.character(immune$subannotation2)
 
#save(cleandat.limma,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/##cleandat.limma.integrated.ComBat.limma.BatchCorrected.annotated.July29.RData")

 
 
 ## chi square Regression.
 library(dendextend);library(dendsort)
 markers<-c("CD206","CD3",
	"CD45RA","CD45RO",
	"CD4","CD68","CD8a",
	"FOXP3","CXCR3","CCR4","HLADR","Ki67","PD1","PDL1",
	"TIM3")
   zz<-zScorePatientExpression(full.dn4=immune,
	markers=which(colnames(immune)%in%markers),
	ROIID.column=which(colnames(immune)=='ROIID'))
  X<-zz[,c("CD206",
	"CD3",
	"CD45RA",
	"CD45RO",
	"CD4","CD68","CD8a","FOXP3","CXCR3","CCR4","HLADR","Ki67","PD1","PDL1","TIM3",
	"annotated")] %>% group_by(annotated) %>% 
       dplyr::summarise_all(funs(mean)) %>%data.frame  
  rownames(X)<-X[,1]
   library(ComplexHeatmap);library(dendextend)
  library(pvclust)
    rpt.pv<-pvclust(log(1+X[,-1]),nboot=100,method.hclust="average")
  rpt.dend<-dendsort(hclust(dist(log(1+X[,c(-1)]))),isReverse=TRUE)
 ## uses the full panel and the unsupervised community
  rownames(X)<-gsub("TIM3","TIM-3",rownames(X))
   rownames(X)<-gsub("PD1","PD-1",rownames(X))
    rownames(X)<-gsub("PDL1","PD-L1",rownames(X))
     rownames(X)<-gsub("Highlysuppressive TREG","highly suppressive TREG",rownames(X))
  colnames(X)<-gsub("TIM3","TIM-3",colnames(X))
   colnames(X)<-gsub("PD1","PD-1",colnames(X))
   colnames(X)<-gsub("PDL1","PD-L1",colnames(X))
    colnames(X)<-gsub("CD8a","CD8",colnames(X))

  heats<-Heatmap(X[,c(-1)],
                  cluster_rows=rpt.dend,
                  cluster_columns=rpt.pv$hclust,
		name="Measurement mean",
	show_row_names=TRUE,
		rect_gp = gpar(col = "white", lwd = 1))
  heats

  ## DLBCL>1
    exp.Odds<-lapply(as.character(unique(immune$annotated.simple)),function(x) logisticRegressionClinicalOdds(pheno=x,full.dn4=immune,
	phenotypeColumn="annotated.simple",
	mycFeature="experiment.binary",
	groupA="1"))

 
  glmOdds<-t(sapply(exp.Odds,function(x) x[3:4]))%>%data.frame
  glmOdds[,2]<-unlist(glmOdds[,2])
  glmOdds[,1]<-unlist(glmOdds[,1])

 rownames(glmOdds)<-sapply(exp.Odds,function(x) rownames(x))
 rownames(glmOdds)<-c("TIM-3+PD-1+CD4",
	"TIM-3+PD-1+CD8",
	"TIM-3+PD-L1+MAC",
	"TIM-3-PD-L1+MAC",
	"TIM-3-PD-L1-MAC",
	"PD-1+highly suppressive TREG",
	"TIM-3-PD-1+CD8",
	"PD-1-TREG",
	"TIM-3+PD-1-CD8",
	"TIM-3-PD-1+ CD4",
	"TIM-3-PD-1-CD8")              
 glmOdds<-glmOdds[match(rownames(X),rownames(glmOdds)),]
 glmOdds$logOdds<-log2(glmOdds[,1])
   ors<-glmOdds[,'logOdds']
  or.p<-glmOdds[,2]
 ors<-data.frame(OR=ors,p=or.p,row.names=rownames(X))
  ORS<-data.frame(OR=rep(0,nrow(X)),row.names=rownames(X))
   ORS[,1]<-ors[match(rownames(ORS),rownames(ors)),1]

	ORS$p<-ors[match(rownames(ORS),rownames(ors)),"p"]


   ors<-ORS
 or.cols<-rep("bisque1",nrow(ors))
  or.cols[which(ors[,1]>0 & ors[,"p"]<0.01)]<-"lightgoldenrod"
  or.cols[which(ors[,1]<0 & ors[,"p"]<0.01 )]<-"plum2"
      ORS<-cbind(ORS,glmOdds[rownames(ORS),])
  colnames(ORS)<-c("OR","p","LR.Estimate","LR.p")


   hans = rowAnnotation(
   Cohort  = anno_points(ors[,1], 
        gp = gpar(col = or.cols),
	height = unit(3, "cm"),
	width = unit(1.5, "cm"),	
	size=unit(ifelse(or.cols=="bisque1",0.5,3.5)
,"mm"),
	axis_param=list(side="top",
	labels_rot=45) ))

 lgd.or = Legend(at = c("DLBCL enriched (p<0.01)","HL enriched (p<0.01)"),ncol=1,
	 title = "Cohort log-odds Ratio",
	type="points", 
	title_position="topcenter",
	size=unit(5,"mm"),
	legend_gp = gpar(col =  c("khaki2",
	"plum2")))

  ht<-draw(hans+heats,annotation_legend_list=list(lgd.or))
  decorate_annotation("Cohort",{
	grid.lines(0, c(0.5,11.5), gp = gpar(lty = 2),
        default.units = "native")
	popViewport()
	}) 

#  savePlot(file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/annotated.jointTME.heatmap.png")


 
  
  

```

### Figure 4 REVISION-11-1 PERFORM THE LINEAR REGRESSION (CLUSTER LEVEL)
The reviewer wanted to know how we determined positive/negative status per cluster, so we model it.
We use a linear model to measure differences of expression across phenotypes, and include a fixed effects model that has the TME (DLBCL/HL/RLN) as an explantory variable.

## FIXME: include the tissue type as an explanatoy variable.
## change the logistic regression to handle more complex estimates (GLMER).
```{r clusterLevelRegression}

subExpr<-immune%>%group_by(ROIID,subannotation2)%>%summarise(CXCR3=mean(CXCR3),
	CCR4=mean(CCR4),
	Ki67=mean(Ki67),
	PDL1=mean(PDL1),
	PD1=mean(PD1),
	TIM3=mean(TIM3),
	LAG3=mean(LAG3),
	CD3=mean(CD3),
	CD4=mean(CD4),
	CD8=mean(CD8a),
	Ki67=mean(Ki67),
	FOXP3=mean(FOXP3),
	CD68=mean(CD68),
	CD206=mean(CD206))%>%data.frame

 subExpr$subannotation2<-factor(subExpr$subannotation2)
  subExpr$subannotation2<-droplevels(subExpr$subannotation2)
 ##append clinical info
  	subExpr$case<-immune$case[match(subExpr$ROIID,immune$ROIID)]
	subExpr$experiment<-immune$experiment[match(subExpr$case,immune$case)]
		subExpr$normalLymphnode<-immune$normalLymphnode[match(subExpr$case,immune$case)]
  subExpr$experiment_RLN<-as.character(subExpr$experiment)
 subExpr$experiment_RLN[which(subExpr$normalLymphnode=='RLN')]<-'RLN'
 
 colnames(subExpr)[2]<-'unsup.subcommunity'
 subExpr$primaryPheno<-sapply(strsplit(as.character(subExpr$unsup.subcommunity),"_"),function(x) x[1])
 
  ## create the contrasts.
  ## we've create a model that has a random intercept for each primary phenotype (CD4/CD8,.../MAC), and compare the sub-community mean to the mean of that group.
 ## we include a factor for the fixed effect of DLBCL/HL/RLN as a fixed effect and average those effects into the estimates.
 alpha_noint <- rbind(
	grandMean= c(rep(1,31)/31,c(1,1)/2))
 colnames(alpha_noint)<-c(levels(subExpr[,2]),"DLBCL.vs.HL","DLBCL.vs.RLN")
 
  ## FIXME: need to add the factor for DLBCL.vs.HL or DLBLC.vs.RLN.
 
 alpha_noint <- rbind(
alpha_noint,
	cd4a.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_a",30,-1)/31,
	##primary group averages
	CD4mean=ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	CD8mean=ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),
	MACmean=ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
	TREGmean=ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),

	cd4a.vs.cd4=ifelse(grepl("CD4_a",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4b.vs.cd4=ifelse(grepl("CD4_b",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4c.vs.cd4=ifelse(grepl("CD4_c",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4d.vs.cd4=ifelse(grepl("CD4_d",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4e.vs.cd4=ifelse(grepl("CD4_e",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4f.vs.cd4=ifelse(grepl("CD4_f",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4g.vs.cd4=ifelse(grepl("CD4_g",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	cd4h.vs.cd4=ifelse(grepl("CD4_h",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD4",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD4",colnames(alpha_noint)),1,0)),
	
#	cd4b.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_b",30,-1)/31,
#	cd4c.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_c",30,-1)/31,
#	cd4d.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_d",30,-1)/31,
#	cd4e.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_e",30,-1)/31,
#	cd4f.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_f",30,-1)/31,
#	cd4g.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_g",30,-1)/31,	
#	cd4h.vs.mean=ifelse(subExpr[,2]%>%levels=="CD4_h",30,-1)/31,	


	#cd8a.vs.mean=ifelse(subExpr[,2]%>%levels=="CD8_a",30,-1)/31,
	#cd8b.vs.mean=ifelse(subExpr[,2]%>%levels=="CD8_b",30,-1)/31,
	#cd8c.vs.mean=ifelse(subExpr[,2]%>%levels=="CD8_c",30,-1)/31,
	#cd8d.vs.mean=ifelse(subExpr[,2]%>%levels=="CD8_d",30,-1)/31,
	#cd8e.vs.mean=ifelse(subExpr[,2]%>%levels=="CD8_e",30,-1)/31,
	#cd8f.vs.mean=ifelse(subExpr[,2]%>%levels=="CD8_f",30,-1)/31,
	cd8a.vs.cd8=ifelse(grepl("CD8_a",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),
	cd8b.vs.cd8=ifelse(grepl("CD8_b",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),
	cd8c.vs.cd8=ifelse(grepl("CD8_c",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),
	cd8d.vs.cd8=ifelse(grepl("CD8_d",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),
	cd8e.vs.cd8=ifelse(grepl("CD8_e",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),
	cd8f.vs.cd8=ifelse(grepl("CD8_f",colnames(alpha_noint)),8,0)/8-ifelse(grepl("CD8",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("CD8",colnames(alpha_noint)),1,0)),

	## 

#	trega.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_a",30,-1)/31,
#	tregb.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_b",30,-1)/31,
#	tregc.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_c",30,-1)/31,
#	tregd.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_d",30,-1)/31,
#	trege.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_e",30,-1)/31,
#	tregf.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_f",30,-1)/31,
#		tregg.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_g",30,-1)/31,
#			tregh.vs.mean=ifelse(subExpr[,2]%>%levels=="TREG_h",30,-1)/31,
trega.vs.treg=ifelse(grepl("TREG_a",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
tregb.vs.treg=ifelse(grepl("TREG_b",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
tregc.vs.treg=ifelse(grepl("TREG_c",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
tregd.vs.treg=ifelse(grepl("TREG_d",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
trege.vs.treg=ifelse(grepl("TREG_e",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
tregf.vs.treg=ifelse(grepl("TREG_f",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
tregg.vs.treg=ifelse(grepl("TREG_g",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),
tregh.vs.treg=ifelse(grepl("TREG_h",colnames(alpha_noint)),8,0)/8-ifelse(grepl("TREG",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("TREG",colnames(alpha_noint)),1,0)),


	#maca.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_a",30,-1)/31,
	#macb.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_b",30,-1)/31,
	#macc.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_c",30,-1)/31,
	#macd.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_d",30,-1)/31,
	#mace.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_e",30,-1)/31,
	#macf.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_f",30,-1)/31,
#		macg.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_g",30,-1)/31,
#	mach.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_h",30,-1)/31,
#	maci.vs.mean=ifelse(subExpr[,2]%>%levels=="Macrophage_i",30,-1)/31
maca.vs.mac=ifelse(grepl("Macrophage_a",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
macb.vs.mac=ifelse(grepl("Macrophage_b",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
macc.vs.mac=ifelse(grepl("Macrophage_c",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
macd.vs.mac=ifelse(grepl("Macrophage_d",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
mace.vs.mac=ifelse(grepl("Macrophage_e",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
macf.vs.mac=ifelse(grepl("Macrophage_f",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
macg.vs.mac=ifelse(grepl("Macrophage_g",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
mach.vs.mac=ifelse(grepl("Macrophage_h",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)),
maci.vs.mac=ifelse(grepl("Macrophage_i",colnames(alpha_noint)),8,0)/8-ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0)/sum(ifelse(grepl("Macrophage",colnames(alpha_noint)),1,0))
	   )

 
  colnames(subExpr)[which(colnames(subExpr)=='subannotation2')]<-'unsup.subcommunity'
  library(multcomp);library(lme4)
	tim3.global <- lmer(TIM3 ~ 0+unsup.subcommunity+experiment_RLN+(1|case), data = subExpr)
	#lag3.global <- lmer(LAG3 ~ 0+unsup.subcommunity+(1|case), data = subExpr)
	pd1.global <- lmer(PD1 ~ 0+unsup.subcommunity +experiment_RLN+(1|case), data = subExpr)
  ccr4.global <- lmer(CCR4 ~ 0+unsup.subcommunity+experiment_RLN+(1|case), data = subExpr)
	pdl1.global <- lmer(PDL1 ~ 0+unsup.subcommunity+experiment_RLN+(1|case), data = subExpr)
	ki67.global <- lmer(Ki67 ~ 0+unsup.subcommunity+experiment_RLN+(1|case), data = subExpr)

	# cd68.global <- lmer(CD68 ~ 0+unsup.subcommunity +(1|case), data = subExpr)
 #cd3.global <- lmer(CD3 ~ 0+unsup.subcommunity +(1|case), data = subExpr)
#  cd4.global <- lmer(CD4 ~ 0+unsup.subcommunity +(1|case), data = subExpr)
 #  cd8.global <- lmer(CD8 ~ 0+unsup.subcommunity +(1|case), data = subExpr)
#    foxp3.global <- lmer(FOXP3 ~ 0+unsup.subcommunity +(1|case), data = subExpr)

  table(immune$subannotation2,immune$annotated)


	##TIM3
  tim3.global <- glht(tim3.global, linfct=alpha_noint)
  tim3.global.ci<-confint(tim3.global, calpha=univariate_calpha())
  tim3.global.ci<- tim3.global.ci$confint%>%data.frame
	tim3.global.ci$marker<-'TIM3'
	 xx<-tim3.global%>%summary
  tim3.global.ci$pvalues<-xx$test$pvalues
  ##add the annotations
  tim3.global.ci$label<-'TIM3-mod'
 tim3.global.ci[which(tim3.global.ci$Estimate<0 & tim3.global.ci$pvalues<0.05),'label']<-'TIM3-'  
tim3.global.ci[which(tim3.global.ci$Estimate>0 & tim3.global.ci$pvalues<0.05),'label']<-'TIM3+'
  

	# PD1
	 pd1.global <- glht(pd1.global, linfct=alpha_noint)
 	 pd1.global.ci<-confint(pd1.global, calpha=univariate_calpha())
  	pd1.global.ci<- pd1.global.ci$confint%>%data.frame
	pd1.global.ci$marker<-'PD1'
 	 xx<-pd1.global%>%summary
	pd1.global.ci$pvalues<-xx$test$pvalues
	
	 pd1.global.ci$label<-'PD1-mod'
 pd1.global.ci[which(pd1.global.ci$Estimate<0 & pd1.global.ci$pvalues<0.05),'label']<-'PD1-'  
pd1.global.ci[which(pd1.global.ci$Estimate>0 & pd1.global.ci$pvalues<0.05),'label']<-'PD1+'
  

		## CCR4
	 ccr4.global <- glht(ccr4.global, linfct=alpha_noint)
 	 ccr4.global.ci<-confint(ccr4.global, calpha=univariate_calpha())
  	ccr4.global.ci<- ccr4.global.ci$confint%>%data.frame
	ccr4.global.ci$marker<-'CCR4'
  	xx<-ccr4.global%>%summary
	ccr4.global.ci$pvalues<-xx$test$pvalues
	
	 ccr4.global.ci$label<-'CCR4-mod'
 ccr4.global.ci[which(ccr4.global.ci$Estimate<0 & ccr4.global.ci$pvalues<0.05),'label']<-'CCR4-'  
  ccr4.global.ci[which(ccr4.global.ci$Estimate>0 & ccr4.global.ci$pvalues<0.05),'label']<-'CCR4+'
  


	
	## PDL1
	 pdl1.global <- glht(pdl1.global, linfct=alpha_noint)
 	 pdl1.global.ci<-confint(pdl1.global, calpha=univariate_calpha())
  	pdl1.global.ci<- pdl1.global.ci$confint%>%data.frame
	pdl1.global.ci$marker<-'PDL1'
	 xx<-pdl1.global%>%summary
	pdl1.global.ci$pvalues<-xx$test$pvalues
	
	 pdl1.global.ci$label<-'PDL1-mod'
 pdl1.global.ci[which(pdl1.global.ci$Estimate<0 & pdl1.global.ci$pvalues<0.05),'label']<-'PDL1-'  
  pdl1.global.ci[which(pdl1.global.ci$Estimate>0 & pdl1.global.ci$pvalues<0.05),'label']<-'PDL1+'
  

	
		##ki67
	 ki.global <- glht(ki67.global, linfct=alpha_noint)
 	 ki.global.ci<-confint(ki.global, calpha=univariate_calpha())
  	ki.global.ci<- ki.global.ci$confint%>%data.frame
	ki.global.ci$marker<-'Ki67'
 	  xx<-ki.global%>%summary
	ki.global.ci$pvalues<-xx$test$pvalues
	 ki.global.ci$label<-'Ki67-mod'
 ki.global.ci[which(ki.global.ci$Estimate<0 & ki.global.ci$pvalues<0.05),'label']<-'Ki67-'  
  ki.global.ci[which(ki.global.ci$Estimate>0 & ki.global.ci$pvalues<0.05),'label']<-'Ki67+'

	output<-rbind(tim3.global.ci,pd1.global.ci,ccr4.global.ci,pdl1.global.ci,ki.global.ci)

	  
	 
	

```

### Write the regression effect model to table
* Not needed to overwrite
```{r revisedTable, eval=FALSE}
 	library(openxlsx)
 	fig2<-loadWorkbook("~/Documents/imageAnalysis/DLBCL/revision-rebuttal-10-1/proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes.xlsx")
	addWorksheet(fig2,"global.estimates.IntegrModel")
	writeData(fig2,'global.estimates.IntegrModel',
	output,
	rowNames=TRUE,withFilter=FALSE,colNames=TRUE,startCol=1,startRow=1)
	openxlsx::saveWorkbook(fig2,file="~/Documents/imageAnalysis/DLBCL/revision-rebuttal-10-1/proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes.xlsx",overwrite=TRUE)
		
	

```

### The linear model annotations revision
The mixed-effect model identified heterogeneity of expression and we can annotate based on the cluster expression. We use the linear model to annotate labels.
```{r figure4ARevised}
#MAC has TIM3+/- PDL1+/-
 output[which(output$label=='TIM3-'& grepl("mac",rownames(output)) | output$label=='PDL1-' & grepl("mac",rownames(output))),]
 ### MAC_a TIM3-,PDL1-
## MAC_b: TIM3-,PDL1-
#MAC_c: TIM3-,PDL1-mod
##MAC_d: TIM3-mod, PDL1-mod
## MAC_e: TIM3+
 ## MAC_f:  TIM3+ , PDL1-mod
##MAC_g:  TIM3+,
 ##MAC_h: TIM3-,PDL1-mod
##MAC_i : TIM3+, 
 output[which(output$label=='TIM3-'& grepl("mac",rownames(output)) | output$label=='PDL1-mod' & grepl("mac",rownames(output))),]

  output[which(output$label=='TIM3-mod'& grepl("mac",rownames(output)) | output$label=='PDL1-mod' & grepl("mac",rownames(output))),]

     output[which(output$label=='TIM3+'& grepl("mac",rownames(output)) | output$label=='PDL1-mod' & grepl("mac",rownames(output))),]


tt<-output[grepl(".vs.",rownames(output)),]
tt$phenos<-sapply(strsplit(rownames(tt),".vs."),function(x) x[1])
tt$subPheno<-sapply(tt$phenos,function(x) paste0(toupper(substring(x,1,nchar(x)-1)),"_",substring(x,nchar(x))))
tt$subPheno<-gsub("MAC",'Macrophage',tt$subPheno)

 immune$lmer.annotation<-as.character(immune$annotation)
 immune$lmer.tim3<-'NA'
  immune$lmer.pd1<-'NA'
 immune$lmer.ccr4<-'NA'
 immune$lmer.pdl1<-'NA'
 immune$lmer.ki67<-'NA'

   d<-tt[which(tt$marker=='TIM3'),]
  immune$lmer.tim3<-d$label[match(immune$subannotation2,d$subPheno)]
  d<-tt[which(tt$marker=='PD1'),]
  immune$lmer.pd1<-d$label[match(immune$subannotation2,d$subPheno)]
 d<-tt[which(tt$marker=='CCR4'),]
  immune$lmer.ccr4<-d$label[match(immune$subannotation2,d$subPheno)]
 d<-tt[which(tt$marker=='PDL1'),]
  immune$lmer.pdl1<-d$label[match(immune$subannotation2,d$subPheno)]
 d<-tt[which(tt$marker=='Ki67'),]
  immune$lmer.ki67<-d$label[match(immune$subannotation2,d$subPheno)]


  immune$lmer.annotation[which(immune$annotation=='Macrophage')]<-paste0(immune[which(immune$annotation=='Macrophage'),'lmer.tim3'],"_",immune[which(immune$annotation=='Macrophage'),'lmer.pdl1'],"_MAC")
  immune$lmer.annotation[which(immune$annotation=='CD4')]<-paste0(immune[which(immune$annotation=='CD4'),'lmer.tim3'],"_",immune[which(immune$annotation=='CD4'),'lmer.pd1'],"_CD4")
  immune$lmer.annotation[which(immune$annotation=='CD8')]<-paste0(immune[which(immune$annotation=='CD8'),'lmer.tim3'],"_",immune[which(immune$annotation=='CD8'),'lmer.pd1'],"_CD8")
  immune$lmer.annotation[which(immune$annotation=='TREG')]<-paste0(immune[which(immune$annotation=='TREG'),'lmer.ccr4'],"_",immune[which(immune$annotation=='TREG'),'lmer.pd1'],"_TREG")

  #factors to combine moderate and positive groups.
   immune$lmer.final<-"NA"
   immune$lmer.final[immune$lmer.annotation%in%c("TIM3+_PDL1-mod_MAC","TIM3+_PDL1+_MAC")]<-"TIM-3+PD-L1+ MAC"
   immune$lmer.final[immune$lmer.annotation%in%c("TIM3-_PDL1-mod_MAC")]<-"TIM-3-PD-L1 mid MAC"
   immune$lmer.final[immune$lmer.annotation%in%c("TIM3-_PDL1-_MAC")]<-"TIM-3-PD-L1- MAC"
   immune$lmer.final[immune$lmer.annotation%in%c("TIM3-mod_PDL1-mod_MAC")]<-"TIM-3 mid PD-L1 mid MAC"
   
   ##CD8
  immune$lmer.final[immune$lmer.annotation%in%c("TIM3-_PD1-_CD8")]<-"TIM-3-PD-1- CD8"
    immune$lmer.final[immune$lmer.annotation%in%c("TIM3-_PD1-mod_CD8")]<-"TIM-3-PD-1 mid CD8"
  immune$lmer.final[immune$lmer.annotation%in%c("TIM3-mod_PD1+_CD8","TIM3+_PD1-mod_CD8","TIM3+_PD1+_CD8")]<-"TIM-3+PD-1+ CD8"
    immune$lmer.final[immune$lmer.annotation%in%c("TIM3+_PD1-_CD8")]<-"TIM-3+PD-1- CD8"

    ##TREG
  immune$lmer.final[immune$lmer.annotation%in%c("CCR4-_PD1-_TREG")]<-"PD-1- TREG"
    immune$lmer.final[immune$lmer.annotation%in%c("CCR4-mod_PD1-mod_TREG")]<-"Proliferative TREG"  
    immune$lmer.final[immune$lmer.annotation%in%c("CCR4+_PD1-mod_TREG","CCR4+_PD1+_TREG")]<-"PD-1+ Highly Suppressive TREG"
    
  
  ## CD4
  immune$lmer.final[immune$lmer.annotation%in%c("TIM3+_PD1+_CD4","TIM3-mod_PD1+_CD4")]<-"TIM-3+PD-1+ CD4"
  immune$lmer.final[immune$lmer.annotation%in%c("TIM3-mod_PD1-mod_CD4")]<-"TIM-3-PD-1 mid CD4"


 ## add immune case
 ##FIX ME: add case information.
     

 ## add RLN to immune feature
   immune$experiment_RLN<-as.character(immune$experiment)
 immune$experiment_RLN[which(immune$normalLymphnode=='RLN')]<-'RLN'
   
  
```

### Heatmap of the linear model annotations
 This used ordinary logistic regression and included the RLN into hte multivariate model to control for RLN as a nuisance variable in multiple regression, which modeled the odds ratio comparing DLBCL to HL after controlling for the RLN heterogeneity.
```{r heatmapLMER}

 
 ## chi square Regression.
 library(dendextend);library(dendsort)
 markers<-c("CD206","CD3",
	"CD45RA","CD45RO",
	"CD4","CD68","CD8a",
	"FOXP3","CXCR3","CCR4","HLADR","Ki67","PD1","PDL1",
	"TIM3")
   zz<-zScorePatientExpression(full.dn4=immune,
	markers=which(colnames(immune)%in%markers),
	ROIID.column=which(colnames(immune)=='ROIID'))
  X<-zz[,c("CD206",
	"CD3",
	"CD45RA",
	"CD45RO",
	"CD4","CD68","CD8a","FOXP3","CXCR3","CCR4","HLADR","Ki67","PD1","PDL1","TIM3",
	"lmer.final")] %>% group_by(lmer.final) %>% 
       dplyr::summarise_all(funs(mean)) %>%data.frame  
  rownames(X)<-X[,1]
   library(ComplexHeatmap);library(dendextend)
  library(pvclust)
    rpt.pv<-pvclust(log(1+X[,-1]),nboot=100,method.hclust="average")
  rpt.dend<-dendsort(hclust(dist(log(1+X[,c(-1)]))),isReverse=TRUE)
 ## uses the full panel and the unsupervised community
  rownames(X)<-gsub("TIM3","TIM-3",rownames(X))
   rownames(X)<-gsub("PD1","PD-1",rownames(X))
    rownames(X)<-gsub("PDL1","PD-L1",rownames(X))
     rownames(X)<-gsub("Highlysuppressive TREG","highly suppressive TREG",rownames(X))
  colnames(X)<-gsub("TIM3","TIM-3",colnames(X))
   colnames(X)<-gsub("PD1","PD-1",colnames(X))
   colnames(X)<-gsub("PDL1","PD-L1",colnames(X))
    colnames(X)<-gsub("CD8a","CD8",colnames(X))

  heats<-Heatmap(X[,c(-1)],
                  cluster_rows=rpt.dend,
                  cluster_columns=rpt.pv$hclust,
		name="Measurement mean",
	show_row_names=TRUE,
		rect_gp = gpar(col = "white", lwd = 1))
  heats

  ## logistic regression that includes RLN as a nuisance variable.
  logisticRegressionClinicalOdds_RLN<-function(pheno=NULL,full.dn4=NULL,phenotypeColumn=NULL,mycFeature=NULL,groupA=1){
    message(pheno)
    ta<-as.data.frame.matrix(table(full.dn4$case,full.dn4[,phenotypeColumn]))
    ta<-ta/rowSums(ta)
    ## direct standardization
    ta<-round(mean(table(full.dn4$case))*ta)
	ta<-as.data.frame(scale(ta))
    ta$feature<-full.dn4[match(rownames(ta),full.dn4$case),mycFeature]
    ta$RLN<-immune$normalLymphnode[match(rownames(ta),immune$case)]
    ta$feature<-ifelse(ta$feature==groupA,1,0)
  colnames(ta)[which(colnames(ta)=="feature")]<-mycFeature
    myForm<-formula(paste0(mycFeature,"~",pheno,"+RLN"))
	fit<-glm(myForm,ta,family='binomial')
	 ## can add a mixed effects model
	  # myForm2<-formula(paste0(mycFeature,"~",pheno,"+(1|RLN)"))
	#fit.re<-glmer(myForm,data=ta,family=binomial(link='logit'),nAGQ=25,control=glmerControl(optimizer='bobyqa'))

	ci<-exp(confint(fit))
  dd<-exp(coef(glm(myForm,ta,family='binomial')))
   dd<-cbind(ci,dd)%>%data.frame
  colnames(dd)<-c("lower.2.5%","upper.97.5%",paste0("OddsRatio.",groupA))
  pv<-summary(fit)%>%coef%>%data.frame
  dd$p<-pv[,4]
	return(dd[pheno,])
  }
  
  immune$lmer.simple<-as.character(immune$lmer.final)
  immune$lmer.simple<-gsub(" ","",immune$lmer.simple)
  immune$lmer.simple<-gsub("\\+","pos.",immune$lmer.simple)
  immune$lmer.simple<-gsub("TIM-3","TIM3",immune$lmer.simple)
  immune$lmer.simple<-gsub("PD-1","PD1",immune$lmer.simple)
  immune$lmer.simple<-gsub("PD-L1","PDL1",immune$lmer.simple)
  immune$lmer.simple<-gsub("\\-","neg.",immune$lmer.simple)
  ## DLBCL>1
    exp.Odds<-lapply(as.character(unique(immune$lmer.simple)),function(x) logisticRegressionClinicalOdds_RLN(pheno=x,full.dn4=immune,
	phenotypeColumn="lmer.simple",
	mycFeature="experiment.binary",
	groupA="1"))

      mult.Odds<-lapply(as.character(unique(immune$lmer.simple)),function(x) 
      multinomialRegressionClinicalOdds.RLNRatio(pheno=x,
                                                 full.dn4=immune,
	phenotypeColumn="lmer.simple",
	mycFeature="experiment_RLN"))
      
 names(mult.Odds)<-unique(immune$lmer.simple)
 
  glmOdds<-t(sapply(exp.Odds,function(x) x[3:4]))%>%data.frame
  glmOdds[,2]<-unlist(glmOdds[,2])
  glmOdds[,1]<-unlist(glmOdds[,1])

 rownames(glmOdds)<-sapply(exp.Odds,function(x) rownames(x))
 rownames(glmOdds)<- immune$lmer.final[match(rownames(glmOdds),immune$lmer.simple)]

 glmOdds<-glmOdds[match(rownames(X),rownames(glmOdds)),]
 glmOdds$logOdds<-log2(glmOdds[,1])
 
 ## table data to save.
  glmOdds.out<-t(sapply(exp.Odds,function(x) x[1:4]))%>%data.frame
  glmOdds.out[,2]<-unlist(glmOdds.out[,2])
  glmOdds.out[,1]<-unlist(glmOdds.out[,1])
    glmOdds.out[,3]<-unlist(glmOdds.out[,3])
  glmOdds.out[,4]<-unlist(glmOdds.out[,4])
 rownames(glmOdds.out)<-sapply(exp.Odds,function(x) rownames(x))
 rownames(glmOdds.out)<- immune$lmer.final[match(rownames(glmOdds.out),immune$lmer.simple)]
 glmOdds.out<-glmOdds.out[match(rownames(X),rownames(glmOdds.out)),]
 glmOdds.out$logOdds<-log2(glmOdds.out[,1])
 
   ors<-glmOdds[,'logOdds']
  or.p<-glmOdds[,2]
 ors<-data.frame(OR=ors,p=or.p,row.names=rownames(X))
  ORS<-data.frame(OR=rep(0,nrow(X)),row.names=rownames(X))
   ORS[,1]<-ors[match(rownames(ORS),rownames(ors)),1]

	ORS$p<-ors[match(rownames(ORS),rownames(ors)),"p"]


   ors<-ORS
 or.cols<-rep("bisque1",nrow(ors))
  or.cols[which(ors[,1]>0 & ors[,"p"]<0.01)]<-"lightgoldenrod"
  or.cols[which(ors[,1]<0 & ors[,"p"]<0.01 )]<-"plum2"
      ORS<-cbind(ORS,glmOdds[rownames(ORS),])
  colnames(ORS)<-c("OR","p","LR.Estimate","LR.p")


   hans = rowAnnotation(
   Cohort  = anno_points(ors[,1], 
        gp = gpar(col = or.cols),
	height = unit(3, "cm"),
	width = unit(1.5, "cm"),	
	size=unit(ifelse(or.cols=="bisque1",0.5,3.5)
,"mm"),
	axis_param=list(side="top",
	labels_rot=45) ))

 lgd.or = Legend(at = c("DLBCL enriched (p<0.01)","HL enriched (p<0.01)"),ncol=1,
	 title = "Cohort log-odds Ratio",
	type="points", 
	title_position="topcenter",
	size=unit(5,"mm"),
	legend_gp = gpar(col =  c("khaki2",
	"plum2")))

  ht<-draw(hans+heats,annotation_legend_list=list(lgd.or))
  decorate_annotation("Cohort",{
	grid.lines(0, c(0.5,13.5), gp = gpar(lty = 2),
        default.units = "native")
	popViewport()
	}) 


```

### write the updated threshold to table
```{r}

chap<-loadWorkbook("~/Documents/imageAnalysis/DLBCL/revision-rebuttal-10-1/Figure4-revision/proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes_RLNREFERENCE.xlsx")
addWorksheet(chap,"Fig4.labeled.GLMmodel")
 	writeData(chap,'Fig4.labeled.GLMmodel',
	glmOdds.out,
	rowNames=TRUE,withFilter=FALSE,colNames=TRUE,startCol=1,startRow=1)
 	
openxlsx::saveWorkbook(chap,file="~/Documents/imageAnalysis/DLBCL/revision-rebuttal-10-1/Figure4-revision/proportion.LM.DLBCL.vs.HD.RLN.primaryPhenotypes_RLNREFERENCE.xlsx",overwrite=TRUE)

```


### TIM-3 relative proportions
```{r}
 
 
	
    ta<-as.data.frame.matrix((table(immune$case,immune$lmer.simple)))
    ta<-100*ta/rowSums(ta)
       ## direct standardization
  #  ta<-round(mean(table(full.dn4$case))*ta)
	#ta<-as.data.frame(scale(ta))
    ta$feature<-immune[match(rownames(ta),immune$case),'experiment_RLN']
    ##model forward selection.
    mac.fit<- nnet::multinom('feature~TIM3midPDL1midMAC+TIM3neg.PDL1midMAC+TIM3neg.PDL1neg.MAC+TIM3pos.PDL1pos.MAC' , data = ta)
    z <- summary(mac.fit)$coefficients/summary(mac.fit)$standard.errors
    p <- (1 - pnorm(abs(z), 0, 1)) * 2
   
    ##cd4
    cd4.fit<- nnet::multinom('feature~TIM3neg.PD1midCD4+TIM3pos.PD1pos.CD4' , data = ta)
    z <- summary(cd4.fit)$coefficients/summary(cd4.fit)$standard.errors
    cd4.p <- (1 - pnorm(abs(z), 0, 1)) * 2
    
     ## CD8
     cd8.fit<- nnet::multinom('feature~TIM3neg.PD1midCD8+TIM3neg.PD1neg.CD8+TIM3pos.PD1neg.CD8+TIM3pos.PD1pos.CD8' , data = ta)
    z <- summary(cd8.fit)$coefficients/summary(cd8.fit)$standard.errors
    cd8.p <- (1 - pnorm(abs(z), 0, 1)) * 2
    
    
    ## TREG
    reg.fit<- nnet::multinom('feature~PD1neg.TREG+PD1pos.HighlySuppressiveTREG+ProliferativeTREG' , data = ta)
    z <- summary(reg.fit)$coefficients/summary(reg.fit)$standard.errors
    reg.p <- (1 - pnorm(abs(z), 0, 1)) * 2

     ##final
     final.fit<- nnet::multinom('feature~TIM3midPDL1midMAC+TIM3neg.PDL1neg.MAC+PD1neg.TREG +TIM3pos.PD1neg.CD8+ProliferativeTREG' , data = ta)
    final.z <- summary(final.fit)$coefficients/summary(final.fit)$standard.errors
    final.p <- (1 - pnorm(abs(final.z), 0, 1)) * 2
    final.p<-final.p*3
    summary(final.fit);final.p
    
    y<-as.numeric(ta$feature)
    x<-as.matrix(ta[,-14])
    fit<-glmnet::cv.glmnet(x,y,family='multinomial')
    
 immune$experiment_RLN<-factor(immune$experiment_RLN,levels=c("RLN","DLBCL","HD"))
 immune$tim3.phenotypes<-paste0(immune$lmer.tim3,"_",immune$annotation)
 immune$tim3.phenotypes<-gsub("\\-mod",".mid",immune$tim3.phenotypes)
 immune$tim3.phenotypes<-gsub("\\-",".neg",immune$tim3.phenotypes)
 immune$tim3.phenotypes<-gsub("\\+",".pos",immune$tim3.phenotypes)
                             
## DLBCL>1
    tim3.Odds<-lapply(as.character(unique(immune$tim3.phenotypes)),function(x) 
      multinomialRegressionClinicalOdds.RLNRatio(pheno=x,full.dn4=immune,
	phenotypeColumn="tim3.phenotypes",
	mycFeature="experiment_RLN",
	groupA="1"))
 names(tim3.Odds)<-unique(immune$tim3.phenotypes)
     ## table data to save.
  glmOdds.out<-t(sapply(tim3.Odds,function(x) x[1:4]))%>%data.frame
  glmOdds.out[,2]<-unlist(glmOdds.out[,2])
  glmOdds.out[,1]<-unlist(glmOdds.out[,1])
    glmOdds.out[,3]<-unlist(glmOdds.out[,3])
  glmOdds.out[,4]<-unlist(glmOdds.out[,4])
 rownames(glmOdds.out)<-sapply(tim3.Odds,function(x) rownames(x))
 
 

```

### Specific barplot for TIM3+PDL1+ TME
```{r}
 x<-as.data.frame.matrix(table(immune$case,immune$lmer.simple))
 x<-100*x/rowSums(x)
 x$ROIID<-rownames(x)
 x$experiment_RLN<-'NA'
  for(i in x$ROIID){
   x$experiment_RLN[which(x$ROIID==i)]<-unique(immune$experiment_RLN[which(immune$ROIID==i)]) 
  }
 x$experiment_RLN<-factor(x$experiment_RLN,levels=c('RLN',"DLBCL","HD"))
 
  ##macrophage panel quadrants
  a<-lm(TIM3pos.PDL1pos.MAC~0+experiment_RLN,x)
  A<- data.frame(confint(a),Estimate=coef(a),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3+PD-L1+ Macrophage",row.names=c("RLN","DLBCL","HL"))
  
  a1<-lm(TIM3midPDL1midMAC~0+experiment_RLN,x)
  A1<- data.frame(confint(a1),Estimate=coef(a1),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3 mid PD-L1 mid Macrophage",row.names=c("RLN","DLBCL","HL"))
  a2<-lm(TIM3neg.PDL1midMAC~0+experiment_RLN,x)
  A2<- data.frame(confint(a2),Estimate=coef(a2),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3- PD-L1 mid Macrophage",row.names=c("RLN","DLBCL","HL"))
  a3<-lm(TIM3neg.PDL1neg.MAC~0+experiment_RLN,x)
  A3<- data.frame(confint(a3),Estimate=coef(a3),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3- PD-L1- Macrophage",row.names=c("RLN","DLBCL","HL"))
  MAC<-rbind(A,A1,A2,A3)
 colnames(MAC)[1:2]<-c("lower","upper")

 MAC[which(MAC$lower<0),'lower']<-0
  MAC[which(MAC$Estimate==0),'Estimate']<-1

 threeColors2<-threeColors
 names(threeColors2)<-c("DLBCL","HL","RLN")
mac.plot<-ggplot(MAC,aes(x=phenotype,y=Estimate,fill=experiment))+geom_bar(stat='identity',position=position_dodge(),color='black')+geom_errorbar(aes(ymin=lower,ymax=upper),position=position_dodge())+scale_fill_manual('Cohort',values=threeColors2)+ylab("Relative proportion (%)")+xlab("Phenotype")+theme_ipsum(base_family='Arial',base_size=9)+theme(axis.text.x=element_text(angle=90,hjust=1,color='black'))

  ##CD4
  
  b<-lm(TIM3pos.PD1pos.CD4~0+experiment_RLN,x)
  B<- data.frame(confint(b),Estimate=coef(b),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3+PD-1+ CD4",row.names=c("RLN","DLBCL","HL"))
  b1<-lm(TIM3neg.PD1midCD4~0+experiment_RLN,x)
  B1<- data.frame(confint(b1),Estimate=coef(b1),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3-PD-1 mid CD4",row.names=c("RLN","DLBCL","HL"))
  
  CD4<-rbind(B,B1)
 colnames(CD4)[1:2]<-c("lower","upper")

 CD4[which(CD4$lower<0),'lower']<-0
  CD4[which(CD4$Estimate==0),'Estimate']<-1

cd4.plot<-ggplot(CD4,aes(x=phenotype,y=Estimate,fill=experiment))+geom_bar(stat='identity',position=position_dodge(),color='black')+geom_errorbar(aes(ymin=lower,ymax=upper),position=position_dodge())+scale_fill_manual('Cohort',values=threeColors2)+ylab("Relative proportion (%)")+xlab("Phenotype")+theme_ipsum(base_family='Arial',base_size=9)+theme(axis.text.x=element_text(angle=90,hjust=1,color='black'))
  ##CD8
  
  

  c<-lm(TIM3pos.PD1pos.CD8~0+experiment_RLN,x)
  C<- data.frame(confint(c),Estimate=coef(c),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3+PD-1+ CD8",row.names=c("RLN","DLBCL","HL"))
   c1<-lm(TIM3neg.PD1midCD8~0+experiment_RLN,x)
  C1<- data.frame(confint(c1),Estimate=coef(c1),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3-PD-1 mid CD8",row.names=c("RLN","DLBCL","HL"))
  
   c2<-lm(TIM3neg.PD1neg.CD8~0+experiment_RLN,x)
  C2<- data.frame(confint(c2),Estimate=coef(c2),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3-PD-1- CD8",row.names=c("RLN","DLBCL","HL"))
   c3<-lm(TIM3pos.PD1neg.CD8~0+experiment_RLN,x)
  C3<- data.frame(confint(c3),Estimate=coef(c3),experiment=c("RLN","DLBCL","HL"),phenotype="TIM-3+PD-1- CD8",row.names=c("RLN","DLBCL","HL"))
  
    CD8<-rbind(C,C1,C2,C3)
 colnames(CD8)[1:2]<-c("lower","upper")

 CD8[which(CD8$lower<0),'lower']<-0

cd8.plot<-ggplot(CD8,aes(x=phenotype,y=Estimate,fill=experiment))+geom_bar(stat='identity',position=position_dodge(),color='black')+geom_errorbar(aes(ymin=lower,ymax=upper),position=position_dodge())+scale_fill_manual('Cohort',values=threeColors2)+ylab("Relative proportion (%)")+xlab("Phenotype")+theme_ipsum(base_family='Arial',base_size=9)+theme(axis.text.x=element_text(angle=90,hjust=1,color='black'))

  
##TREG  
  d<-lm(PD1pos.HighlySuppressiveTREG~0+experiment_RLN,x)
  D<- data.frame(confint(d),Estimate=coef(d),experiment=c("RLN","DLBCL","HL"),phenotype="PD-1+ Highly suppressive TREG",row.names=c("RLN","DLBCL","HL"))
  e<-lm(ProliferativeTREG~0+experiment_RLN,x)
  E<- data.frame(confint(e),Estimate=coef(e),experiment=c("RLN","DLBCL","HL"),phenotype="Proliferative TREG",row.names=c("RLN","DLBCL","HL"))
 X<-rbind(D,E)
 colnames(X)[1:2]<-c("lower","upper")

 X[which(X$lower<0),'lower']<-0
 
treg.plot<-ggplot(X,aes(x=phenotype,y=Estimate,fill=experiment))+geom_bar(stat='identity',position=position_dodge(),color='black')+geom_errorbar(aes(ymin=lower,ymax=upper),position=position_dodge())+scale_fill_manual('Cohort',values=threeColors2)+ylab("Relative proportion (%)")+xlab("Phenotype")+theme_ipsum(base_family='Arial',base_size=9)+theme(axis.text.x=element_text(angle=90,hjust=1,color='black'))

 grid.arrange(mac.plot,cd4.plot,cd8.plot,treg.plot,nrow=1,ncol=4)


```



### update tables for figure 5
Need to print the odds ratio, annotation label assignments, proportions, and expression
```{r writeTable,eval=FALSE}
 library(openxlsx)

fig4<-loadWorkbook("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/FinalFigure4-table.xlsx")
 
 ## GLMODDS
    addWorksheet(fig4,"logOR.Cohort.labeled")
	writeData(fig4,'logOR.Cohort.labeled',	
	glmOdds,
	rowNames=TRUE,withFilter=FALSE,colNames=TRUE,startCol=1,startRow=1)
	
	

 ## label assignments
  lab.assign<- table(immune$subannotation2,immune$annotated)%>%as.data.frame.matrix
addWorksheet(fig4,"cluster.label.assign")
	writeData(fig4,'cluster.label.assign',	
	lab.assign,
	rowNames=TRUE,withFilter=FALSE,colNames=TRUE,startCol=1,startRow=1)
	



 ### proportions.
 addWorksheet(fig4,"clusterProportions")
	writeData(fig4,'clusterProportions',	
	cd4.prop,
	rowNames=TRUE,withFilter=FALSE,colNames=TRUE,startCol=1,startRow=1)
	




 openxlsx::saveWorkbook(fig4,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/final-submission-folder/figure4-final/FinalFigure4-table2.xlsx",overwrite=TRUE)

```


# Radar plot of the T-reg ICOS, CCR4, PD1
- requires joint to be created from stratified gating in the early code
- hypothesis: to gate TREGs by TIM3/CCR4/PD1/ICOS
```{r}
  library(fmsb)
  tot<-as.data.frame.matrix(table(joint$experiment,joint$phenotype))
  
 # nchl<-tot["HL",c("CD4","CD8","Macrophage","TREG")]%>%sum
#  ndlbcl<-tot["DLBCL",c("CD4","CD8","Macrophage","TREG")]%>%sum
#  nrln<-tot["RLN",c("CD4","CD8","Macrophage","TREG")]%>%sum

   nchl<-tot["HL",c("TREG")]%>%sum
  ndlbcl<-tot["DLBCL",c("TREG")]%>%sum
  nrln<-tot["RLN",c("TREG")]%>%sum
  tregs<-joint[grepl("TREG",joint$phenotype),]
  ## CHL
  chl.ccr4<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$CCR4=="CCR4+"),])/nchl
    chl.ccr4.pd1<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$CCR4=="CCR4+" & tregs$PD1=="PD1+"),])/nchl
  chl.pd1<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$PD1=="PD1+"),])/nchl
  chl.pd1.ki<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$PD1=="PD1+" & tregs$Ki67=="Ki67+"),])/nchl
  chl.ki<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$Ki67=="Ki67+"),])/nchl
  chl.ki.icos<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$Ki67=="Ki67+" & tregs$ICOS=="ICOS+"),])/nchl
  chl.icos<-100*nrow(tregs[which(tregs$experiment=="HL" &  tregs$ICOS=="ICOS+"),])/nchl
  chl.ccr4.icos<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$CCR4=="CCR4+" & tregs$ICOS=="ICOS+"),])/nchl

   ## DLBCL
  dl.ccr4<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$CCR4=="CCR4+"),])/ndlbcl
  dl.ccr4.pd1<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+" & tregs$CCR4=="CCR4+"),])/ndlbcl
  dl.pd1<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+"),])/ndlbcl
  dl.pd1.ki<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+" & tregs$Ki67=="Ki67+"),])/ndlbcl
  dl.ki<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$Ki67=="Ki67+"),])/ndlbcl
  dl.ki.icos<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$Ki67=="Ki67+" & tregs$ICOS=="ICOS+"),])/ndlbcl
  dl.icos<-100*nrow(tregs[which(tregs$experiment=="DLBCL" &  tregs$ICOS=="ICOS+"),])/ndlbcl
  dl.ccr4.icos<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$CCR4=="CCR4+" & tregs$ICOS=="ICOS+"),])/ndlbcl
#  dl.icos.pd1<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+" &  tregs$ICOS=="ICOS+"),])/ndlbcl
  
   ## RLN
  rl.ccr4<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$CCR4=="CCR4+"),])/nrln
  rl.ccr4.pd1<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$PD1=="PD1+" & tregs$CCR4=="CCR4+"),])/nrln
  rl.pd1<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$PD1=="PD1+"),])/nrln
  rl.pd1.ki<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$PD1=="PD1+" & tregs$Ki67=="Ki67+"),])/nrln
  rl.ki<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$Ki67=="Ki67+"),])/nrln
  rl.ki.icos<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$Ki67=="Ki67+" & tregs$ICOS=="ICOS+"),])/nrln
  rl.icos<-100*nrow(tregs[which(tregs$experiment=="RLN" &  tregs$ICOS=="ICOS+"),])/nrln
  rl.ccr4.icos<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$CCR4=="CCR4+" & tregs$ICOS=="ICOS+"),])/nrln
  
  data<- matrix(as.numeric(c(chl.ccr4,
                             chl.ccr4.pd1,
                             chl.pd1,
                             chl.pd1.ki,
                             chl.ki,
                             chl.ki.icos,
                             chl.icos,
                             chl.ccr4.icos)),nrow=1)%>%as.data.frame.matrix()
  colnames(data) <- c("CCR4+" , "CCR4+PD1+" , "PD1+" , "PD1+Ki67+" , "Ki67+", "Ki67+ICOS+","ICOS+","ICOS+CCR4+")
 
  data<-rbind(data,as.numeric(c(dl.ccr4,
                             dl.ccr4.pd1,
                             dl.pd1,
                             dl.pd1.ki,
                             dl.ki,
                             dl.ki.icos,
                             dl.icos,
                             dl.ccr4.icos)))
    data<-rbind(data,as.numeric(c(rl.ccr4,
                             rl.ccr4.pd1,
                             rl.pd1,
                             rl.pd1.ki,
                             rl.ki,
                             rl.ki.icos,
                             rl.icos,
                             rl.ccr4.icos)))
  rownames(data)<-c("HL","DLBCL","RLN")
  
 data<-data[1:2,]
# To use the fmsb package, I have to add 2 lines to the dataframe: the max and min of each topic to show on the plot!
data <- rbind(rep(max(data),ncol(data)) , rep(min(data),ncol(data)) , data)
 
# Check your data, it has to look like this!
# head(data)
# colors_border=c( rgb(0.8,0.08,147/255,0.9), rgb(1,1,51/255,0.9) )
 colors_border=c( rgb(0.8,0.08,147/255,0.9), rgb(153/255,153/255,0,0.7) )

# colors_border=c("lightpink3", "gold")
 #colors_in=c( "#EEAEEE", "#FFF68F")
colors_in=c( rgb(1,182/255,193/255,0.4), rgb(1,1,153/255,0.4)  )
# Custom the radarChart !
  radarchart( data  , axistype=1 , 
 
    #custom polygon
    pcol=colors_border ,
    pfcol=colors_in ,
    plwd=4, 
    plty=1,
 
    #custom the grid
    cglcol="grey", cglty=1, axislabcol="grey", caxislabels=c("10%","20%","35%","50%","70%"), cglwd=0.8,
 
    #custom labels
    vlcex=1 
    )
  
  legend(x=0.9, y=0.6, legend = rownames(data[-c(1,2),]), bty = "n", pch=20 , col=colors_in , text.col = "grey", cex=1.2, pt.cex=3)

  ## output PDF for tucker.
radarchart( data,
                                   axistype=1 , 
    #custom polygon
    pcol=colors_border ,
    pfcol=colors_in ,
    plwd=4, 
    plty=1,
    #custom the grid
    cglcol="grey", cglty=1, axislabcol="grey", caxislabels=c("10%","20%","35%","50%","70%"), cglwd=0.8,
    #custom labels
    vlcex=2 
    )
  
  

  prepLinear<-function(fit=NULL,label='TIM3+PDL1+Macrophage',comparisonFit=NULL){
    sum<-as.data.frame(coef(summary(fit)))
    fit2<-comparisonFit
    diffp<-as.data.frame(coef(summary(fit2)))
    ci<-confint(fit)
    ci<-data.frame(Estimate=sum$Estimate,ci,label=label)
    rownames(ci)<-gsub("feature","",rownames(ci))
    ci$experiment<-rownames(ci)
    colnames(ci)[2:3]<-c("lower2.5","upper97.5")
    ci$p<-round(diffp[,'Pr(>|t|)'],3)
    return(ci)
  }
  
  ## need to customize joint + expression.
  ## t-test
   tme<-subset(joint,joint$phenotype%in%c("TREG"))
    ## ROI proportions (%)
     tme$ccr4.pd1.simple<-paste0(tme$CCR4,tme$PD1)


  ## ICOS+ CCR4+
  tme$icos.ccr4<-paste0(tme$ICOS,tme$CCR4)
  tme$icos.ccr4<-gsub("\\+",".Pos.",tme$icos.ccr4)
  
  ## Ki67 + ics+
  tme$icos.ki67<-paste0(tme$ICOS,tme$Ki67)
  tme$icos.ki67<-gsub("\\+",".Pos.",tme$icos.ki67)
  
  ## Ki + PD1+
   tme$ki67.pd1<-paste0(tme$Ki67,tme$PD1)
  tme$ki67.pd1<-gsub("\\+",".Pos.",tme$ki67.pd1)
  
  ## CCR4+PD1+
   tme$ccr4.pd1<-paste0(tme$PD1,tme$CCR4)
  tme$ccr4.pd1<-gsub("\\+",".Pos.",tme$ccr4.pd1)
  
  # totals
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  ## ICOS+CCR4+
  icos.ccr4<-as.data.frame.matrix(table(tme$ROIID,tme$icos.ccr4))
  tme.total<-tme.total[match(rownames(icos.ccr4),names(tme.total))]

    
 # icos.ccr4<-100*icos.ccr4/rowSums(icos.ccr4)
  icos.ccr4$total<-tme.total
  icos.ccr4$prop<-100*icos.ccr4$ICOS.Pos.CCR4.Pos./icos.ccr4$total

     icos.ccr4$feature<-joint$experiment[match(rownames(icos.ccr4),joint$ROIID)]
  icos.ccr4$feature<-factor(icos.ccr4$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ccr4[which(icos.ccr4$feature!="RLN"),])%>%summary
 lm(prop~feature,icos.ccr4[which(icos.ccr4$feature!="RLN"),])%>%confint()
 
 icos.ccr4.fit<- lm(prop~feature,icos.ccr4[which(icos.ccr4$feature!="RLN"),])

 ###
  
 ## ICOS+
    # totals
   tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$ICOS))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$ICOS.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%confint()
 
  icos.fit<-  lm(prop~feature,icos[which(icos$feature!="RLN"),])


 
 ### Ki67+ICOS+
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  ## ICOS+Ki+
  icos.ki<-as.data.frame.matrix(table(tme$ROIID,tme$icos.ki67))
  tme.total<-tme.total[match(rownames(icos.ki),names(tme.total))]

    
  icos.ki$total<-tme.total
  icos.ki$prop<-100*icos.ki$ICOS.Pos.Ki67.Pos./icos.ki$total

     icos.ki$feature<-joint$experiment[match(rownames(icos.ki),joint$ROIID)]
  icos.ki$feature<-factor(icos.ki$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%summary
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%confint()
 
    icos.ki.fit<-  lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])

 #### Ki67
   tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$Ki67))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$Ki67.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%confint()
    ki.fit<- lm(prop~feature,icos[which(icos$feature!="RLN"),])

 
  ###  PD1+Ki67+
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos.ki<-as.data.frame.matrix(table(tme$ROIID,tme$ki67.pd1))
  tme.total<-tme.total[match(rownames(icos.ki),names(tme.total))]

    
  icos.ki$total<-tme.total
  icos.ki$prop<-100*icos.ki$Ki67.Pos.PD1.Pos./icos.ki$total

     icos.ki$feature<-joint$experiment[match(rownames(icos.ki),joint$ROIID)]
  icos.ki$feature<-factor(icos.ki$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%summary
 pd1.ki.fit<- lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])

 ############
  
  ## pd1 
 ## PD1
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$PD1))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$PD1.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 pd1.fit<- lm(prop~feature,icos[which(icos$feature!="RLN"),])

 ###
 ## CCR4+PD1+
 tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  ## CCR4+PD1++
  icos.ki<-as.data.frame.matrix(table(tme$ROIID,tme$ccr4.pd1))
  tme.total<-tme.total[match(rownames(icos.ki),names(tme.total))]

    
  icos.ki$total<-tme.total
  icos.ki$prop<-100*(icos.ki$PD1.Pos.CCR4.Pos.)/icos.ki$total

     icos.ki$feature<-joint$experiment[match(rownames(icos.ki),joint$ROIID)]
  icos.ki$feature<-factor(icos.ki$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%summary
  pd1.ccr4.fit<-  lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])

 ## CCR4 
    tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$CCR4))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$CCR4.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 ccr4.fit<- lm(prop~feature,icos[which(icos$feature!="RLN"),])
  ############
  library(sjPlot)
 
  tab_model(ccr4.fit,
            icos.ccr4.fit,
            icos.fit,
            icos.ki.fit,
            ki.fit,
            pd1.ki.fit,
            pd1.fit,
            pd1.ccr4.fit,
  pred.labels = c("Intercept", 
                  "HL.vs.DLBCL"
                    ),
  dv.labels = c("CCR4","ICOS+CCR4+","ICOS+","ICOS+Ki67+","Ki67+","PD1+Ki67+","PD1+","PD1+CCR4+"),
  string.pred = "TME rel. proportion (%)",
  string.ci = "Conf. Int (95%)",
  string.p = "P-Value",
  show.aic=TRUE,
    file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/TME.relprop.html"
)

 
 ## relative T-regs.
 
 
  ## CCR4+
 pd1<-as.data.frame.matrix(table(tme$ROIID,tme$CCR4))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   ccr4.fit<-  lm(CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])

   ## CCR4+PD1+
   pd1<-as.data.frame.matrix(table(tme$ROIID,tme$ccr4.pd1))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(PD1.Pos.CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   ccr4.pd1.fit<-lm(PD1.Pos.CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])

  ### ICOS
     pd1<-as.data.frame.matrix(table(tme$ROIID,tme$ICOS))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(ICOS.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   icos.fit<-  lm(ICOS.Pos.~feature,pd1[which(pd1$feature!="RLN"),])
 
 ## ICOS+ Ki+
    pd1<-as.data.frame.matrix(table(tme$ROIID,tme$icos.ki67))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(ICOS.Pos.Ki67.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   icos.ki.fit<-  lm(ICOS.Pos.Ki67.Pos.~feature,pd1[which(pd1$feature!="RLN"),])
   
 
  tab_model(ccr4.fit,
            ccr4.pd1.fit,
            icos.fit,
            icos.ki.fit,
            
  pred.labels = c("Intercept", 
                  "HL.vs.DLBCL"
                    ),
  dv.labels = c("CCR4+","CCR4+PD1+","ICOS+","ICOS+Ki67+"),
  string.pred = "Case relative T-reg proportion (%)",
  string.ci = "Conf. Int (95%)",
  string.p = "P-Value",
  show.aic=TRUE,
    file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/TregRelativeProp.html"
)
 
  ## Nearest neighbor distance comparisons  CCR4/ CCR4.PD1 / PD1  comparing to CD8/CD4 PD1+/PD1-
    load("C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/simpleModel-1-14/Figure3-HD.vs.DLBCL/cleandat.limma.integrated.ComBat.limma.BatchCorrected.RData")

  
                           
```


# Radar plot for MAC
- requires joint to be created from stratified gating in the early code
- hypothesis: to gate TREGs by TIM3/CCR4/PD1/ICOS
```{r}

  library(fmsb)
  tot<-as.data.frame.matrix(table(joint$experiment,joint$phenotype))
  
 # nchl<-tot["HL",c("CD4","CD8","Macrophage","TREG")]%>%sum
#  ndlbcl<-tot["DLBCL",c("CD4","CD8","Macrophage","TREG")]%>%sum
#  nrln<-tot["RLN",c("CD4","CD8","Macrophage","TREG")]%>%sum

   nchl<-tot["HL",c("Macrophage")]%>%sum
  ndlbcl<-tot["DLBCL",c("Macrophage")]%>%sum
  nrln<-tot["RLN",c("Macrophage")]%>%sum
  tregs<-joint[grepl("Macrophage",joint$phenotype),]
  ## CHL
  chl.ccr4<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$CCR4=="CCR4+"),])/nchl
    chl.ccr4.pd1<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$CCR4=="CCR4+" & tregs$PD1=="PD1+"),])/nchl
  chl.pd1<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$PD1=="PD1+"),])/nchl
  chl.pd1.ki<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$PD1=="PD1+" & tregs$Ki67=="Ki67+"),])/nchl
  chl.ki<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$Ki67=="Ki67+"),])/nchl
  chl.ki.icos<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$Ki67=="Ki67+" & tregs$ICOS=="ICOS+"),])/nchl
  chl.icos<-100*nrow(tregs[which(tregs$experiment=="HL" &  tregs$ICOS=="ICOS+"),])/nchl
  chl.ccr4.icos<-100*nrow(tregs[which(tregs$experiment=="HL" & tregs$CCR4=="CCR4+" & tregs$ICOS=="ICOS+"),])/nchl

   ## DLBCL
  dl.ccr4<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$CCR4=="CCR4+"),])/ndlbcl
  dl.ccr4.pd1<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+" & tregs$CCR4=="CCR4+"),])/ndlbcl
  dl.pd1<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+"),])/ndlbcl
  dl.pd1.ki<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+" & tregs$Ki67=="Ki67+"),])/ndlbcl
  dl.ki<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$Ki67=="Ki67+"),])/ndlbcl
  dl.ki.icos<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$Ki67=="Ki67+" & tregs$ICOS=="ICOS+"),])/ndlbcl
  dl.icos<-100*nrow(tregs[which(tregs$experiment=="DLBCL" &  tregs$ICOS=="ICOS+"),])/ndlbcl
  dl.ccr4.icos<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$CCR4=="CCR4+" & tregs$ICOS=="ICOS+"),])/ndlbcl
#  dl.icos.pd1<-100*nrow(tregs[which(tregs$experiment=="DLBCL" & tregs$PD1=="PD1+" &  tregs$ICOS=="ICOS+"),])/ndlbcl
  
   ## RLN
  rl.ccr4<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$CCR4=="CCR4+"),])/nrln
  rl.ccr4.pd1<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$PD1=="PD1+" & tregs$CCR4=="CCR4+"),])/nrln
  rl.pd1<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$PD1=="PD1+"),])/nrln
  rl.pd1.ki<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$PD1=="PD1+" & tregs$Ki67=="Ki67+"),])/nrln
  rl.ki<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$Ki67=="Ki67+"),])/nrln
  rl.ki.icos<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$Ki67=="Ki67+" & tregs$ICOS=="ICOS+"),])/nrln
  rl.icos<-100*nrow(tregs[which(tregs$experiment=="RLN" &  tregs$ICOS=="ICOS+"),])/nrln
  rl.ccr4.icos<-100*nrow(tregs[which(tregs$experiment=="RLN" & tregs$CCR4=="CCR4+" & tregs$ICOS=="ICOS+"),])/nrln
  
  data<- matrix(as.numeric(c(chl.ccr4,
                             chl.ccr4.pd1,
                             chl.pd1,
                             chl.pd1.ki,
                             chl.ki,
                             chl.ki.icos,
                             chl.icos,
                             chl.ccr4.icos)),nrow=1)%>%as.data.frame.matrix()
  colnames(data) <- c("CCR4+" , "CCR4+PD1+" , "PD1+" , "PD1+Ki67+" , "Ki67+", "Ki67+ICOS+","ICOS+","ICOS+CCR4+")
 
  data<-rbind(data,as.numeric(c(dl.ccr4,
                             dl.ccr4.pd1,
                             dl.pd1,
                             dl.pd1.ki,
                             dl.ki,
                             dl.ki.icos,
                             dl.icos,
                             dl.ccr4.icos)))
    data<-rbind(data,as.numeric(c(rl.ccr4,
                             rl.ccr4.pd1,
                             rl.pd1,
                             rl.pd1.ki,
                             rl.ki,
                             rl.ki.icos,
                             rl.icos,
                             rl.ccr4.icos)))
  rownames(data)<-c("HL","DLBCL","RLN")
  
 data<-data[1:2,]
# To use the fmsb package, I have to add 2 lines to the dataframe: the max and min of each topic to show on the plot!
data <- rbind(rep(max(data),ncol(data)) , rep(min(data),ncol(data)) , data)
 
# Check your data, it has to look like this!
# head(data)
 colors_border=c( rgb(0.8,0.08,147/255,0.9), rgb(1,1,51/255,0.9) )
# colors_border=c("lightpink3", "gold")
 #colors_in=c( "#EEAEEE", "#FFF68F")
colors_in=c( rgb(1,182/255,193/255,0.4), rgb(1,1,153/255,0.4)  )
# Custom the radarChart !
  radarchart( data  , axistype=1 , 
 
    #custom polygon
    pcol=colors_border ,
    pfcol=colors_in ,
    plwd=4, 
    plty=1,
 
    #custom the grid
    cglcol="grey", cglty=1, axislabcol="grey", caxislabels=c("10%","20%","30%","50%","70%"), cglwd=0.8,
 
    #custom labels
    vlcex=0.8 
    )
  
  legend(x=0.9, y=0.6, legend = rownames(data[-c(1,2),]), bty = "n", pch=20 , col=colors_in , text.col = "grey", cex=1.2, pt.cex=3)

  
  

  prepLinear<-function(fit=NULL,label='TIM3+PDL1+Macrophage',comparisonFit=NULL){
    sum<-as.data.frame(coef(summary(fit)))
    fit2<-comparisonFit
    diffp<-as.data.frame(coef(summary(fit2)))
    ci<-confint(fit)
    ci<-data.frame(Estimate=sum$Estimate,ci,label=label)
    rownames(ci)<-gsub("feature","",rownames(ci))
    ci$experiment<-rownames(ci)
    colnames(ci)[2:3]<-c("lower2.5","upper97.5")
    ci$p<-round(diffp[,'Pr(>|t|)'],3)
    return(ci)
  }
  
  ## need to customize joint + expression.
  ## t-test
   tme<-subset(joint,joint$phenotype%in%c("TREG"))
    ## ROI proportions (%)
     tme$ccr4.pd1.simple<-paste0(tme$CCR4,tme$PD1)


  ## ICOS+ CCR4+
  tme$icos.ccr4<-paste0(tme$ICOS,tme$CCR4)
  tme$icos.ccr4<-gsub("\\+",".Pos.",tme$icos.ccr4)
  
  ## Ki67 + ics+
  tme$icos.ki67<-paste0(tme$ICOS,tme$Ki67)
  tme$icos.ki67<-gsub("\\+",".Pos.",tme$icos.ki67)
  
  ## Ki + PD1+
   tme$ki67.pd1<-paste0(tme$Ki67,tme$PD1)
  tme$ki67.pd1<-gsub("\\+",".Pos.",tme$ki67.pd1)
  
  ## CCR4+PD1+
   tme$ccr4.pd1<-paste0(tme$PD1,tme$CCR4)
  tme$ccr4.pd1<-gsub("\\+",".Pos.",tme$ccr4.pd1)
  
  # totals
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  ## ICOS+CCR4+
  icos.ccr4<-as.data.frame.matrix(table(tme$ROIID,tme$icos.ccr4))
  tme.total<-tme.total[match(rownames(icos.ccr4),names(tme.total))]

    
 # icos.ccr4<-100*icos.ccr4/rowSums(icos.ccr4)
  icos.ccr4$total<-tme.total
  icos.ccr4$prop<-100*icos.ccr4$ICOS.Pos.CCR4.Pos./icos.ccr4$total

     icos.ccr4$feature<-joint$experiment[match(rownames(icos.ccr4),joint$ROIID)]
  icos.ccr4$feature<-factor(icos.ccr4$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ccr4[which(icos.ccr4$feature!="RLN"),])%>%summary
 lm(prop~feature,icos.ccr4[which(icos.ccr4$feature!="RLN"),])%>%confint()
 
 icos.ccr4.fit<- lm(prop~feature,icos.ccr4[which(icos.ccr4$feature!="RLN"),])

 ###
  
 ## ICOS+
    # totals
   tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$ICOS))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$ICOS.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%confint()
 
  icos.fit<-  lm(prop~feature,icos[which(icos$feature!="RLN"),])


 
 ### Ki67+ICOS+
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  ## ICOS+Ki+
  icos.ki<-as.data.frame.matrix(table(tme$ROIID,tme$icos.ki67))
  tme.total<-tme.total[match(rownames(icos.ki),names(tme.total))]

    
  icos.ki$total<-tme.total
  icos.ki$prop<-100*icos.ki$ICOS.Pos.Ki67.Pos./icos.ki$total

     icos.ki$feature<-joint$experiment[match(rownames(icos.ki),joint$ROIID)]
  icos.ki$feature<-factor(icos.ki$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%summary
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%confint()
 
    icos.ki.fit<-  lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])

 #### Ki67
   tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$Ki67))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$Ki67.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%confint()
    ki.fit<- lm(prop~feature,icos[which(icos$feature!="RLN"),])

 
  ###  PD1+Ki67+
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos.ki<-as.data.frame.matrix(table(tme$ROIID,tme$ki67.pd1))
  tme.total<-tme.total[match(rownames(icos.ki),names(tme.total))]

    
  icos.ki$total<-tme.total
  icos.ki$prop<-100*icos.ki$Ki67.Pos.PD1.Pos./icos.ki$total

     icos.ki$feature<-joint$experiment[match(rownames(icos.ki),joint$ROIID)]
  icos.ki$feature<-factor(icos.ki$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%summary
 pd1.ki.fit<- lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])

 ############
  
  ## pd1 
 ## PD1
  tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$PD1))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$PD1.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 pd1.fit<- lm(prop~feature,icos[which(icos$feature!="RLN"),])

 ###
 ## CCR4+PD1+
 tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  ## CCR4+PD1++
  icos.ki<-as.data.frame.matrix(table(tme$ROIID,tme$ccr4.pd1))
  tme.total<-tme.total[match(rownames(icos.ki),names(tme.total))]

    
  icos.ki$total<-tme.total
  icos.ki$prop<-100*(icos.ki$PD1.Pos.CCR4.Pos.)/icos.ki$total

     icos.ki$feature<-joint$experiment[match(rownames(icos.ki),joint$ROIID)]
  icos.ki$feature<-factor(icos.ki$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])%>%summary
  pd1.ccr4.fit<-  lm(prop~feature,icos.ki[which(icos.ki$feature!="RLN"),])

 ## CCR4 
    tj<-as.data.frame.matrix( table(joint$ROIID,joint$phenotype))
  tme.total<-tj[,c("CD4","CD8","Macrophage","TREG")]%>%rowSums
  
  icos<-as.data.frame.matrix(table(tme$ROIID,tme$CCR4))
    tme.total<-tme.total[match(rownames(icos),names(tme.total))]

   colnames(icos)<-gsub("\\+",".Pos.",colnames(icos))

  icos$total<-tme.total
  icos$prop<-100*icos$CCR4.Pos./icos$total

  icos$feature<-joint$experiment[match(rownames(icos),joint$ROIID)]
  icos$feature<-factor(icos$feature,levels=c("RLN","DLBCL","HL"))
   
 lm(prop~feature,icos[which(icos$feature!="RLN"),])%>%summary
 ccr4.fit<- lm(prop~feature,icos[which(icos$feature!="RLN"),])
  ############
  library(sjPlot)
 
  tab_model(ccr4.fit,
            icos.ccr4.fit,
            icos.fit,
            icos.ki.fit,
            ki.fit,
            pd1.ki.fit,
            pd1.fit,
            pd1.ccr4.fit,
  pred.labels = c("Intercept", 
                  "HL.vs.DLBCL"
                    ),
  dv.labels = c("CCR4","ICOS+CCR4+","ICOS+","ICOS+Ki67+","Ki67+","PD1+Ki67+","PD1+","PD1+CCR4+"),
  string.pred = "TME rel. proportion (%)",
  string.ci = "Conf. Int (95%)",
  string.p = "P-Value",
  show.aic=TRUE,
    file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/TME.relprop.html"
)

 
 ## relative T-regs.
 
 
  ## CCR4+
 pd1<-as.data.frame.matrix(table(tme$ROIID,tme$CCR4))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   ccr4.fit<-  lm(CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])

   ## CCR4+PD1+
   pd1<-as.data.frame.matrix(table(tme$ROIID,tme$ccr4.pd1))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(PD1.Pos.CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   ccr4.pd1.fit<-lm(PD1.Pos.CCR4.Pos.~feature,pd1[which(pd1$feature!="RLN"),])

  ### ICOS
     pd1<-as.data.frame.matrix(table(tme$ROIID,tme$ICOS))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(ICOS.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   icos.fit<-  lm(ICOS.Pos.~feature,pd1[which(pd1$feature!="RLN"),])
 
 ## ICOS+ Ki+
    pd1<-as.data.frame.matrix(table(tme$ROIID,tme$icos.ki67))
  pd1<-100*pd1/rowSums(pd1)
   pd1$feature<-joint$experiment[match(rownames(pd1),joint$ROIID)]
  pd1$feature<-factor(pd1$feature,levels=c("RLN","DLBCL","HL"))
   
  colnames(pd1)<-gsub("\\+",".Pos.",colnames(pd1))
  lm(ICOS.Pos.Ki67.Pos.~feature,pd1[which(pd1$feature!="RLN"),])%>%summary
 
   icos.ki.fit<-  lm(ICOS.Pos.Ki67.Pos.~feature,pd1[which(pd1$feature!="RLN"),])
   
 
  tab_model(ccr4.fit,
            ccr4.pd1.fit,
            icos.fit,
            icos.ki.fit,
            
  pred.labels = c("Intercept", 
                  "HL.vs.DLBCL"
                    ),
  dv.labels = c("CCR4+","CCR4+PD1+","ICOS+","ICOS+Ki67+"),
  string.pred = "Case relative T-reg proportion (%)",
  string.ci = "Conf. Int (95%)",
  string.p = "P-Value",
  show.aic=TRUE,
    file="C:/Users/UOSC/Documents/IMC-Ranalysis/DLBCL/blood-revisions-5-1-21/final-figure-revision/Figure5/TregRelativeProp.html"
)
 
                           
```

