---
title: "Imaging mass cytometry reveals tumor and immune spatial and phenotypic clusters associated with clinical outcomes in diffuse large B cell lymphoma (DLBCL)"
author: "Anthony Colombo+,Monirath Hav+, Erik Gerdtsson"
date: "`r Sys.Date()`"
output: html_document
---


# {.tabset}  
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Setup {.tabset}

### Helper Functions

```{r, eval=FALSE}


metaClusterPhenotype<-function(dn4=NULL,phenotype=NULL,myPheno=NULL,markers=NULL,plotPCA=FALSE,k2=15){
 #myPheno is the column for the phenotype.
    
   erik<-subset(dn4,dn4[,myPheno]==phenotype)
  ## subMeta holds the casePheno 
  dn4[,paste0(phenotype,"_subMeta")]<-NA
    erik[,paste0(phenotype,"_subMeta")]<-NA
 require(Rphenograph)
   roiData<-as.data.frame.matrix(table(erik$ROIID,erik[,myPheno]))

   for(roi in rownames(roiData)[which(roiData[,phenotype]>45)]){
     myroi<-subset(erik,erik$ROIID==roi)
     graph<-Rphenograph(myroi[,markers],k=45)
     myroi[,paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
    dn4[match(myroi$uniqueLabel,dn4$uniqueLabel),paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
     erik[match(myroi$uniqueLabel,erik$uniqueLabel),paste0(phenotype,"_subMeta")]<-membership(graph[[2]])
   }

  ErikMeta=erik[,
	c(markers,
	"ROIID",
	paste0(phenotype,"_subMeta"))] %>% group_by(ROIID,.dots=paste0(phenotype,"_subMeta")) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame
     
          ##Levine used k=15 for meta.
     erikPheno<- Rphenograph(ErikMeta[,c(markers)], k = k2)
     subphenograph_cluster_meta<- factor(membership(erikPheno[[2]]))
     table(subphenograph_cluster_meta)
     ErikMeta=cbind(ErikMeta, subphenograph_cluster_meta)
     ### plot PCA, label each case.  COO,  status.  ## 
   if( any(colnames(erik)== paste0(phenotype,"_subphenograph_cluster_meta"))){
     erik<-erik[,which(colnames(erik)!= paste0(phenotype,"_subphenograph_cluster_meta"))]
 } 

   myLocal=left_join(erik, ErikMeta[,c("ROIID",paste0(phenotype,"_subMeta"),"subphenograph_cluster_meta")], by = c("ROIID"))

  colnames(myLocal)[which(colnames(myLocal)=="subphenograph_cluster_meta")]<-paste0(phenotype,"_subphenograph_cluster_meta") 
  dn4[,paste0(phenotype,"_subphenograph_cluster_meta")]<-NA
  dn4[match(myLocal$uniqueLabel,dn4$uniqueLabel),paste0(phenotype,"_subphenograph_cluster_meta")]<-myLocal[,which(colnames(myLocal)==paste0(phenotype,"_subphenograph_cluster_meta"))]

  plot_clustering_heatmap_wrapper(expr=ErikMeta[,c(markers)], 
  cell_clustering=ErikMeta[,"subphenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
 return(dn4)
}



plot_clustering_heatmap_wrapper<-function(expr=NULL, 
  cell_clustering=NULL, cluster_merging = NULL,useMedians=TRUE,useQuantiles=FALSE,
color_clusters=NULL){
  ## cluster_merging should be character merging vector of 29 or so clusters. this is input from the ANNOTATION object.  not a matched object level.
   expr<-as.matrix(expr)
  if(useQuantiles==TRUE){
  ##max min normalization
  library(matrixStats)
  rng <- colQuantiles(expr, probs = c(0.001, 0.999))
  expr01 <- t((t(expr) - rng[, 1]) / (rng[, 2] - rng[, 1]))
  expr01[expr01 < 0] <- 0
  expr01[expr01 > 1] <- 1
  }else{
   expr01<-expr 

  }

 if(is.null(color_clusters)==TRUE){
  color_clusters <- c("#DC050C", "#FB8072", "#1965B0", "#7BAFDE", "#882E72", 
  "#B17BA6", "#FF7F00", "#FDB462", "#E7298A", "#E78AC3", 
  "#33A02C", "#B2DF8A", "#55A1B1", "#8DD3C7", "#A6761D", 
  "#E6AB02", "#7570B3", "#BEAED4", "#666666", "#999999", 
  "#aa8282", "#d4b7b7", "#8600bf", "#ba5ce3", "#808000", 
  "#aeae5c", "#1e90ff", "#00bfff", "#56ff0d", "#ffff00")
 }

  # Calculate the median expression 
  library(dplyr)
  expr_median <- data.frame(expr, cell_clustering = cell_clustering) %>%
   dplyr::group_by(cell_clustering) %>% 
    dplyr::summarize_all(funs(median))
 if(useMedians==TRUE){
  expr01_median <- data.frame(expr01, cell_clustering = cell_clustering) %>%
    group_by(cell_clustering) %>% 
    dplyr::summarize_all(funs(median))
   }else{
   expr01_median <- data.frame(expr01, cell_clustering = cell_clustering) %>%
    dplyr::group_by(cell_clustering) %>% 
   dplyr::summarize_all(funs(mean))
 }
  # Calculate cluster frequencies
  clustering_table <- as.numeric(table(cell_clustering))
   print(dim(expr_median))
  # This clustering is based on the markers that were used for the main clustering
  d <- dist(expr_median[, colnames(expr)], method = "euclidean")
  cluster_rows <- hclust(d, method = "average")
   expr_heat <- as.matrix(expr01_median[, colnames(expr01)])
  rownames(expr_heat) <- expr01_median$cell_clustering
 
  labels_row <- paste0(rownames(expr_heat), " (", 
    round(clustering_table / sum(clustering_table) * 100, 2), "%)")
  labels_col <- colnames(expr_heat)
  
  # Row annotation for the heatmap
  annotation_row <- data.frame(cluster = factor(expr01_median$cell_clustering))
  rownames(annotation_row) <- rownames(expr_heat)
  
  color_clusters <- color_clusters[1:nlevels(annotation_row$cluster)]
  names(color_clusters) <- levels(annotation_row$cluster)
  annotation_colors <- list(cluster = color_clusters)
  annotation_legend <- FALSE
  
  if(!is.null(cluster_merging)){
    
    annotation_row$cluster_merging <- factor(cluster_merging)
    color_clusters <- color_clusters[1:nlevels(factor(cluster_merging))]
    names(color_clusters) <- levels(factor(cluster_merging))
    annotation_colors$cluster_merging <- color_clusters
    annotation_legend <- TRUE
  }
  
  # Colors for the heatmap
  library(RColorBrewer);library(pheatmap)
  color <- colorRampPalette(rev(brewer.pal(n = 9, name = "RdYlBu")))(100)
  
  pheatmap::pheatmap(expr_heat, color = color, 
    cluster_cols = FALSE, cluster_rows = cluster_rows, 
    labels_col = labels_col, labels_row = labels_row, 
    display_numbers = TRUE, number_color = "black", 
    fontsize = 12, fontsize_number = 12,
    annotation_row = annotation_row, annotation_colors = annotation_colors,
    annotation_legend = annotation_legend)
  
}



mynormalize<-function(data=NULL,percentile=NULL){
   if(percentile>1){
     percentile<-percentile/100
   }
    if(is.null(percentile)==TRUE){
    minValues<-apply(data,2,min)
    maxValues<-apply(data,2,max) 
   }else{
    minValues<-apply(data,2,function(x) quantile(x,1-percentile)  )
    maxValues<-apply(data,2,function(x) quantile(x,percentile)  )
   }
    hv2<-maxValues-minValues
    dataXR<-data
    stopifnot(all( colnames(data)==names(minValues)))
    stopifnot(all(colnames(data)==names(maxValues)))
    stopifnot(all(colnames(data)==names(hv2)))
    for(j in 1:ncol(data)){
    dataXR[,j]<-(data[,j]-minValues[j])/hv2[j]
    }
    dataXR[dataXR<0]<-0
    dataXR[dataXR>1]<-1

  stopifnot(all(apply(dataXR,2,max)==1))
  stopifnot(all(apply(dataXR,2,min)==0))
   return(dataXR)
}



mutationClinicalSheetAssembly<-function(mutationsToGather=c("TP53","BCL2","BCL6","CD79B","SGK1","MYD88","MYC","NOTCH1","NOTCH2","EZH2","CREBBP"),full.dn4=NULL){
 ## the mutation of MYD88 is specifically controlled for L265P region.
 ## the column header for the non-syn SNP annotation for MYD88 is AA.Mutation.Cosmic_v70

   mutations<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/mutational-genomic-table/TMA CGI genomic data.csv")
   colnames(mutations)<-gsub("USC076","USC76",colnames(mutations))
  #  clinic<-read.csv("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/patient.clinical.table.csv",header=T)
     library(XLConnect) 
   wb<-XLConnect::loadWorkbook("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/clinicalDataRevisedFinal_final.xlsx")
   clinic <- readWorksheet(wb, 1);  
  clinic$MB<-gsub(" ","",clinic$End..MB)
 mut.tag<-colnames(mutations)[9:67]
  mutt<-data.frame(mut.tag,id=sapply(strsplit(gsub("\\.","_",mut.tag),"_"),function(x) x[1])
) 

  mutt$id[which(mutt[,"id"]=="USC076")]<-"USC76"
  tag<- as.character(clinic[,"MB"])
  tag<-gsub(" ","",tag)
  tag<-data.frame(tag=tag,stringsAsFactors=FALSE)
  inters<-intersect(tag[,1],mutt$id)
   key<-data.frame(MB=tag[match(inters,tag[,1]),],stringsAsFactors=FALSE)
  key$id<-mutt[match(inters,mutt[,2]),1]
 ##fix a column header
 ## USC76 has a funcy header.
 ##
  casemuts<-mutations[,match(key$id,colnames(mutations))]
  casemuts$gene.name<-mutations[,"gene_name..Homo_sapiens_ensembl_v74_mRNA."]
  casemuts$chromosome<-as.character(mutations$Chromosome)
 casemuts$aminoAcid<-as.character(mutations$AA.Mutation.Cosmic_v70)
 casemuts$aminoAcid<-gsub( "\"","",mutations$AA.Mutation.Cosmic_v70)
  casemuts$gene.name<-as.character(casemuts$gene.name)
 ##for MYD88, the gene name will have the aminoAcid appended.
  casemuts$gene.name[which(casemuts$gene.name=="MYD88" & casemuts$aminoAcid=="L265P, L265P")]<-"MYD88.L265P"
  casemuts<-casemuts[which(casemuts$aminoAcid!="S219C, S219C, S219C, S219C" & casemuts$aminoAcid!="M232T, M232T" & casemuts$aminoAcid!="0.999, 0.999" & casemuts$aminoAcid!="S251N, S243N, S251N, S243N"),]
  casemuts[which(casemuts$gene.name=="MYD88" & casemuts$chromosome==""),"gene.name"]<-"MYD88.L265P"
 casemuts[which(casemuts$gene.name=="MYD88.L265P" & casemuts$chromosome==""),1:20]<-ifelse(casemuts[which(casemuts$gene.name=="MYD88.L265P" & casemuts$chromosome!=""),1:20]!=0,1,0)
  casem<- casemuts[which(casemuts$chromosome==""),]
 mutationsToGather[which(mutationsToGather=="MYD88")]<-"MYD88.L265P"
  dat<-casem[casem$gene.name%in%mutationsToGather,]
   ### 
   for(i in mutationsToGather){
  key[,i]<-0
 key[match(colnames(dat)[which(dat[dat$gene.name==i,]==1)],key$id),i]<-1
   }
  write.csv(key,file="~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/DLBCL-mutation-analysis/Mutational.CNV.categorical.table.csv")
  ## append to the clinical sheet
  clinic2<-clinic[,!colnames(clinic)%in%mutationsToGather]
 clinic2<-left_join(clinic2,key[,c("MB",mutationsToGather)],by="MB")
  for(i in mutationsToGather){
  clinic2[which(is.na(clinic2[,i])), i]<-0
  }
 ##add the factor to control for replicates
  replicate<-model.matrix(~0+clinic2$MB)
  replicate<-replicate[,which(colSums(replicate)>1)]  
 colnames(replicate)<-substring(colnames(replicate),11)
  clinic2<-data.frame(clinic2,replicate)
 clinic2$WES.avail<-FALSE
  clinic2$WES.avail[match(key$MB,clinic2$MB)]<-TRUE
 clinic2$Case_ID<-paste0("Case_",clinic2$Col1)
  return(clinic2)

}


```
### Library
```{r,eval=FALSE}

library(magrittr)
  library(matrixStats)
```

### Figure 1
This shows the original clustering.
```{r,eval=FALSE}
  load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")

 set.seed(1234)
  table(full.dn4$phenograph_cluster_meta,full.dn4$cell.type.new)
    

  lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
	"Area.norm",
	"Eccentricity.norm",
	"Solidity.norm",
	"Perimeter.norm",
	"Percent_Touching.norm",
	"Number_Neighbors.norm")


  plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers)], 
  cell_clustering=full.dn4[,"phenograph_cluster_meta"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)

### identify the major components of the TME
  anno<-data.frame("1"="Endothelial",
		"2"="CD8",
		"3"="CD4",
		"4"="Tumor",
		"5"="CD8",
		"6"="CD4",
		"7"="Treg",
		"8"="Macrophages",
		"9"="Tumor",
		"10"="Tumor",
		"11"="CD4",
		"12"="Tumor",
		"13"="Tumor",
		"14"="Tumor")
  anno<-as.data.frame(t(anno))
  anno$cluster<-rownames(anno)


```




### Re-assigning Tumor and TME populations
We cluster each major TME component (k=50) and re-assign labels based on lineage expression.
```{r DLBCLsubType,eval=FALSE}
   full.dn4$subType<-as.character(full.dn4$cell.type.new)
     full.dn4$subType.correction<-as.character(full.dn4$cell.type.new)
 full.dn4$subType.reassignedLabel<-as.character(full.dn4$subType)


   for (i in unique(as.character(full.dn4$cell.type.new))[-1]){
    ROI=data.frame(subset(full.dn4,full.dn4$subType==i))
      set.seed(42)
    graph<- Rphenograph(ROI[,c(lineage_markers,induc)], k = 50)
    phenograph_cluster<- factor(membership(graph[[2]]))
   full.dn4$subType[match(ROI$uniqueLabel,full.dn4$uniqueLabel)]=paste0(i,"_",phenograph_cluster)
  
 }
 
  ##plot the sub-phenographs co-expression.
  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tumor"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
 # savePlot(file="full.allROI.Tumor.subTypes.png")
  
  ## CD8 reassign
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_28")]<-"Tcyt"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_28")]<-"Tumor_28.Reassigned.CD8"
   ##MAC re-assign
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_1")]<-"Macrophages"
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_27")]<-"Macrophages"
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_34")]<-"Macrophages"
 
  full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_1")]<-"Tumor_1.Reassigned.MAC"

 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_27")]<-"Tumor_27.Reassigned.MAC"

 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_34")]<-"Tumor_34.Reassigned.MAC"



  ##CD4 re-assign
    full.dn4$subType.correction[which(full.dn4$subType=="Tumor_10")]<-"Th"
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_41")]<-"Th"
  full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_10")]<-"Tumor_10.Reassigned.CD4"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_41")]<-"Tumor_41.Reassigned.CD4"

    #Endo re-assign
   full.dn4$subType.correction[which(full.dn4$subType=="Tumor_25")]<-"Endothelial"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_25")]<-"Tumor_25.Reassigned.Endo"

 ##Th re-assign
  full.dn4$subType.correction[which(full.dn4$subType=="Tumor_20")]<-"Th"
 full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Tumor_20")]<-"Tumor_20.Reassigned.CD4"

 #### 
 
 



 ##plot the tumor correction
    plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tumor"),"subType.correction"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  #savePlot(file="full.allROI.Tumor.subType.correction.png")



  
   plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tumor"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tumor"),"subType.reassignedLabel"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  #savePlot(file="full.allROI.Tumor.subType.reassignedLabel.png")


 ## re-assigning TME populations.

  ## Tcyt
 plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Tcyt"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Tcyt"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
 # savePlot(file="full.allROI.CD8.subTypes.png")
  

 ## CD4 needed re-assignment.
 ##Th
 plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Th"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Th"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.CD4.subTypes.png")

 ##Th_12 has FOXP3+ 
     full.dn4$subType.correction[which(full.dn4$subType=="Th_12")]<-"Treg"
  full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Th_12")]<-"Th_12.Reassigned.Treg"

      full.dn4$subType.correction[which(full.dn4$subType=="Th_21")]<-"Treg"
full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Th_21")]<-"Th_21.Reassigned.Treg"

     full.dn4$subType.correction[which(full.dn4$subType=="Th_9")]<-"Tcyt"
full.dn4$subType.reassignedLabel[which(full.dn4$subType=="Th_9")]<-"Th_9.Reassigned.CD8"

 ##plot the corrected
   plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Th"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Th"),"subType.reassignedLabel"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Th.subType.reassignedLabel.png")
  




 #MAC
  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Macrophages"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Macrophages"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.MAC.subTypes.png")


 ##Endo
   plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Endothelial"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Endothelial"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.Endo.subTypes.png")
##Treg
  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$cell.type.new=="Treg"),c(lineage_markers,induc)], 
  cell_clustering=full.dn4[which(full.dn4$cell.type.new=="Treg"),"subType"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
  savePlot(file="full.allROI.treg.subTypes.png")

 ##plot the heatmap of corrected
  plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers,induc)], 
  cell_clustering=full.dn4[,"subType.correction"],
  useMedians=TRUE,
  useQuantiles=FALSE,
  color_clusters=NULL)
 # savePlot(file="full.allROI.corrected.subTypes.png")




```



### Meta-cluster heatmap of primary phenotypes
```{r, eval=FALSE}

 ErikMeta=full.dn4[,
	c("Cell_CD20","Cell_CD3","Cell_CD4","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3",
	"ROIID",
	"phenograph_cluster_meta")] %>% group_by(ROIID,phenograph_cluster_meta) %>% 
       dplyr::summarise_all(funs(median)) %>%data.frame

   plot_clustering_heatmap_wrapper(expr=full.dn4[,c("Cell_CD20","Cell_CD3","Cell_CD4","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3")], 
  cell_clustering=full.dn4[,"subType.correction"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 #savePlot(file="allCells.corrected.png")

  plot_clustering_heatmap_wrapper(expr=full.dn4[which(full.dn4$immune.status=="immune"),c("Cell_CD20","Cell_CD3","Cell_CD4","Cell_CD8","Cell_CD68","Cell_CD31","Cell_FOXP3")], 
  cell_clustering=full.dn4[which(full.dn4$immune.status=="immune"),"subType.correction"],
  useMedians=FALSE,
  useQuantiles=TRUE,
  color_clusters=NULL)
 #savePlot(file="TME.corrected.png")


```



```{r,eval=FALSE}
# meta-cluster each TME population.

######## unsupervised community detection.
 ## subPhenograph the CD4.
  full.dn4$CD4.subPhenograph<-NA
   morphs<-mynormalize(data=full.dn4[,c("Area",
	"Eccentricity",
	"Solidity",
	"Perimeter",
	"Percent_Touching",
	"Number_Neighbors")],percentile=0.99)
  colnames(morphs)<-paste0(colnames(morphs),".norm")
  full.dn4<-cbind(full.dn4,morphs)

  lineage_markers<-c( "Cell_BCL2",
                       "Cell_BCL6",
                       "Cell_CD20",
                       "Cell_CD206",
                       "Cell_CD3",
                       "Cell_CD30",
                       "Cell_CD31",
                       "Cell_CD4",
                       "Cell_CD45RA",
                       "Cell_CD45RO",
                       "Cell_CD68" ,
                       "Cell_CD8",
                       "Cell_EphrinB2",
                       "Cell_FOXP3",
                       "Cell_HLADR" )
   induc<-c("Cell_C",
            "Cell_CCR4",
            "Cell_CD134",
            "Cell_CXCR3",
            "Cell_Granzym",
            "Cell_ICOS",
            "Cell_Ki67",
            "Cell_Lag3",
            "Cell_PD1",
            "Cell_PDL1",
            "Cell_PDL2",
            "Cell_Tbet",
            "Cell_Tim3",
            "Cell_Vimentin",
            "Cell_Vista",
            "Cell_p",
	"Area.norm",
	"Eccentricity.norm",
	"Solidity.norm",
	"Perimeter.norm",
	"Percent_Touching.norm",
	"Number_Neighbors.norm")
 ##meta cluster the CD4.
  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="CD4",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
  savePlot(file="CD4_subMeta.cluster.png")
  ## CD8
  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="CD8",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
   savePlot(file="CD8_subMeta.cluster.png")
  table(full.dn4$subType.correction,full.dn4$CD8_subphenograph_cluster_meta)
 ## TREG
  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="TREG",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
 savePlot(file="TREG_subMeta.cluster.png")

 table(full.dn4$subType.correction,full.dn4$TREG_subphenograph_cluster_meta)
 ##MAC
 full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="Macrophage",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
 savePlot(file="MAC_subMeta.cluster.png")
  table(full.dn4$subType.correction,full.dn4$Macrophage_subphenograph_cluster_meta)

  full.dn4<-metaClusterPhenotype(dn4=full.dn4,
	phenotype="Tumor",
	myPheno="subType.correction",
	markers=c(lineage_markers,induc))
 savePlot(file="Tumor_subMeta.cluster.png")

 


 ##compile into 1 column unsupervised.community  
  full.dn4$unsup.subcommunity<-as.character(full.dn4$subType.correction)
 full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="CD4")]<-paste0("CD4_",full.dn4$CD4_subphenograph_cluster_meta[which(full.dn4$subType.correction=="CD4")])
 ## CD8
  full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="CD8")]<-paste0("CD8_",full.dn4$CD8_subphenograph_cluster_meta[which(full.dn4$subType.correction=="CD8")])
 ## TREG
 full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="TREG")]<-paste0("TREG_",full.dn4$TREG_subphenograph_cluster_meta[which(full.dn4$subType.correction=="TREG")])
 ## MAC
 full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="Macrophage")]<-paste0("Macrophage_",full.dn4$Macrophage_subphenograph_cluster_meta[which(full.dn4$subType.correction=="Macrophage")])
 ## Tumor
  full.dn4$unsup.subcommunity[which(full.dn4$subType.correction=="Tumor")]<-paste0("Tumor_",full.dn4$Tumor_subphenograph_cluster_meta[which(full.dn4$subType.correction=="Tumor")])


 # save(full.dn4,file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")



   plot_clustering_heatmap_wrapper(expr=full.dn4[,c(lineage_markers,induc)], 
  cell_clustering=full.dn4[,"unsup.subcommunity"], useMedians=TRUE,useQuantiles=FALSE,
color_clusters=NULL)
 #  savePlot(file="subcommunities.png")
 ##### end of the meta-clustering of the sub-types.



```



```{r,eval=FALSE}

 ################ FIGURE 1C,D,E. ################################
   library(dplyr)
   library(magrittr)
   library(ComplexHeatmap)
 #  load("~/DLBCL/expr2.subPhenotyped.revised.RData")
 load("~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/full.dn4.subType.corrected.mutation.tsne.CD20dim.filtered.RData")

   
  source("~/Documents/imageAnalysis/Ansell/cluster-phenograph-analysis/clusterFunctions.R")
  #enum<-readRDS("~/DLBCL/enumeration.large.april8.rds")
  options(stringsAsFacotrs=FALSE)

   #clinic<-read.csv("~/Documents/imageAnalysis/DLBCL/monirath/myData/USC_CGI_DLBCL_Data_Final.csv",header=T)
 clinic<-mutationClinicalSheetAssembly(mutationsToGather=c("TP53","BCL2","BCL6","CD79B","SGK1","MYD88","MYC","NOTCH1","NOTCH2","EZH2"))
  clinic$Case_ID<-factor(clinic$Case_ID,levels=unique(clinic$Case_ID))

 ##total proportions.
  100*table(full.dn4$subType.correction)%>%prop.table
  
####
 ratio<- full.dn4%>%group_by(case,immune.status)%>%summarise(count=n()) %>%data.frame
    rat<-c()
  for(roi in unique(ratio$case)){
  x<-round(100*ratio[which(ratio$case==roi & ratio$immune.status=="immune"),"count"]/sum(ratio[which(ratio$case==roi) ,"count"]),2)
  rat<-c(rat, x)
  }
  rat<-data.frame(ratio=rat,case=unique(ratio$case),row.names=unique(ratio$case))

  rat$label<-gsub("_"," ",rownames(rat))
  rat<-rat[order(rat$ratio,decreasing=TRUE),]
   rat$label<-factor(rat$label,levels=rat$label)
   write.csv(rat,file="~/Documents/imageAnalysis/DLBCL/monirath/myData/revision-comments-11-6/survival-analysis-morphological-variants/PDL1-neighborhood-analysis/workingDir/re-phenotype/ratioPlot.immune.infiltrate.csv")

   ratioPlot<-ggplot(rat,aes(y=ratio,x=label))+geom_bar(stat='identity')+coord_flip()+ylab("Immune absolute proportion (%)")+xlab("Case")+theme_bw(base_size=14)
   print( ratioPlot)

#####



    immune<-full.dn4
   immune<-immune[which(immune$subType.correction!="Tumor" & immune$subType.correction!="Dendritic.Cell" & immune$subType.correction!="Endothelial"),]
   immune$cell.type.new<-immune$subType.correction
   immune$rename<-as.character(immune$subType.correction)
   immune$rename<-gsub("CD8","T cytotoxic",immune$rename)
    immune$rename<-gsub("CD4","T Helper",immune$rename)
    immune$rename<-gsub("TREG","FOXP3 T REG",immune$rename)
    immune$rename<-gsub("Macrophage","CD68 Macrophage",immune$rename)     

  
   cellType<-data.frame(type=c("CD8","CD4","TREG","Macrophage"),stringsAsFactors=FALSE)
   cellType$color<-c("green","blue","purple","magenta")
   cellType$name<-c("T cytotoxic","T Helper","FOXP3 T REG","CD68 Macrophage")

  
     roiSum<-immune%>%group_by(case,rename)%>%summarise(count=n())%>%mutate(freq=count/sum(count))%>%data.frame
      roiSum$color<-cellType$color[match(roiSum$rename,cellType$name)]
      roiSum$rename<-factor(roiSum$rename,levels=cellType$name)
      rownames(roiSum)<-seq(1,nrow(roiSum))

  roiSum$case<-as.character(roiSum$case)
  roiSum$case<-gsub("_"," ",roiSum$case)

  roiSum$case<-factor(roiSum$case, levels=gsub("_"," ",rownames(rat)))

    colnames(roiSum)<-c("case","Phenotype","count","proportion","color")
  finalStack<-ggplot(roiSum,aes(y=proportion,x=case,fill=Phenotype))+geom_bar(stat='identity')+coord_flip()+scale_fill_manual(values=cellType$color )+theme_bw()
  ##FIGURE 1D 
  print(finalStack)
 ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure1-lowCD20-revision/immune.proportions.adjusted.finalRevised.png")

     
   ###multivariate association
   rat$case<-gsub("_"," ",rownames(rat))
   fullRat<-left_join(rat,roiSum[which(roiSum$Phenotype=='CD68 Macrophage'),c('case','proportion')],by='case')
   fullRat<-left_join(fullRat,roiSum[which(roiSum$Phenotype=='FOXP3 T REG'),c('case','proportion')],by='case')
   fullRat<-left_join(fullRat,roiSum[which(roiSum$Phenotype=='T Helper'),c('case','proportion')],by='case')
   fullRat<-left_join(fullRat,roiSum[which(roiSum$Phenotype=='T cytotoxic'),c('case','proportion')],by='case')
 colnames(fullRat)<-c("ratio","case","label","CD68","TREG","CD4","CD8")

 ## comparing multivariate to MACs
 lm(ratio~CD4+TREG+CD8,fullRat)%>%summary

 mult<-lm(ratio~CD4+TREG+CD8,fullRat)%>%confint%>%data.frame
 mult<-cbind(mult, lm(ratio~CD4+TREG+CD8,fullRat)%>%coef)
 colnames(mult)<-c("X2.5","X97.5","Estimate")
  mult$TME<-c("MAC","CD4","TREG","CD8")
 library(hrbrthemes)
 ggplot(mult,aes(x=TME,y=Estimate))+geom_errorbar(aes(ymin=X2.5,ymax=X97.5))+geom_point()+geom_hline(yintercept=0,linetype='dashed',color='red')+theme_ipsum(base_family='Arial')+ylab("Immune absolute proportion rate of change (%)")+ggtitle("% explained by major TME")+theme(axis.text.x=element_text(size=18),axis.text.y=element_text(size=18)) 
 ggsave("~/Documents/imageAnalysis/DLBCL/simpleModel-1-14/Figure1-lowCD20-revision/immune.proportions.Multivariate.regression.png")


 ### Figure 1G
 ggplot(fullRat,aes(x=CD68,y=CD4))+geom_point()+geom_smooth(method='lm')



```



### PCA and Batch/Entropy analysis

```{r,eval=FALSE}
 library(gridExtra)
 data<-as.data.frame.matrix( table(full.dn4$ROIID,full.dn4$subType.correction2))
 data<- asinh(data)
  pcs<-prcomp((data))
  
   PCi<-data.frame(pcs$x,ROIID=rownames(pcs$x))
  PCi$TMA<-full.dn4$TMA[match(PCi$ROIID,full.dn4$ROIID)]
   PCi$case<-full.dn4$case[match(PCi$ROIID,full.dn4$ROIID)]
  varz<-100*round(pcs$sdev/sum(pcs$sdev),3)
 EBV.pca<- ggplot(PCi,aes(x=PC1,y=PC2,col=case,label=case))+
	geom_point(size=6)+xlim(c( min(PCi$PC1)-0.05,  max(PCi$PC1)+0.05))+
	theme_classic()+
	xlab(paste0("PC1 (",varz[1],"%)"))+ylab(paste0("PC2 (",varz[2],"%)"))+
  geom_text(aes(label=case),hjust=1.2, vjust=0)+ggtitle("99th Percentile Norm")
 print(EBV.pca)

  EBV.pca2<- ggplot(PCi,aes(x=PC1,y=PC2,col=TMA,label=case))+
	geom_point(size=6)+xlim(c( min(PCi$PC1)-0.05,  max(PCi$PC1)+0.05))+
	theme_classic()+
	xlab(paste0("PC1 (",varz[1],"%)"))+ylab(paste0("PC2 (",varz[2],"%)"))+
  geom_text(aes(label=case),hjust=1.2, vjust=0)+ggtitle("99th Percentile Norm")
 print(EBV.pca2)
 grid.arrange(EBV.pca,EBV.pca2,nrow=2)
 EBV.pca+facet_wrap(~case)
   EBV.pca+facet_wrap(~case)
  # ggsave("PCA.case.ROI.level.png")
   EBV.pca2
 # ggsave("PCA.TMA.heterogeneity.png")


 library(entropy)
  ### Compute Entropy.
  ## for inter-core heterogeneity, compute the Kullback-Leibler divergence of given ROI to the case average.
   data<-as.data.frame.matrix( table(full.dn4$ROIID,full.dn4$subType.correction2))
  data<-data/rowSums(data) 
  data$case<-full.dn4$case[match(rownames(data),full.dn4$ROIID)]

  KLdata<-c()
  for(i in c("Case_17","Case_18","Case_21","Case_26","Case_27","Case_30","Case_31")){
   d<-subset(data,data$case==i)
  y1<-d[1,which(colnames(d)!="case")]
  y2<-d[2,which(colnames(d)!="case")]
  y3<-y2
  if(nrow(d)==3){
  y3<-d[3,which(colnames(d)!="case")]
  }
  yavg<-d%>%group_by(case)%>%summarise_all(funs(mean))%>%data.frame
  yavg<-yavg[,-1]
  kl1<- KL.empirical(y1, yavg, unit=c("log2"))
  hp.avg<-entropy.plugin(as.numeric(yavg))
  kl2<-KL.empirical(y2, yavg, unit=c("log2"))
  kl3<-kl2
    if(nrow(d)==3){
   kl3<-KL.empirical(y3, yavg, unit=c("log2"))
  }
  kldata<-data.frame(KL1=kl1,KL2=kl2,KL3=kl3,HP=hp.avg,
	entropy1=kl1/hp.avg,
	entropy2=kl2/hp.avg,
	entropy3=kl3/hp.avg,row.names=i)
  KLdata<-rbind(KLdata,kldata)
  }

  klmeans<-data.frame(KL=rowMeans(KLdata[,c(1:3)]))
  klmeans$case<-rownames(klmeans)
 ggplot(klmeans,aes(x=case,y=KL))+geom_point()+theme_bw(base_size=14)
 #  ggsave("KL-divergence.scores.png")

```





